// devexpress-richedit/lib/base/rich-utils/debug/chrome-dev-tools-custom-formatters/model/table/table-row.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var table_row_1 = require("../../../../../../core/model/tables/main-structures/table-row");
var table_base_structures_1 = require("../../../../../../core/model/tables/secondary-structures/table-base-structures");
var log_1 = require("../../../../../../core/rich-utils/debug/logger/base-logger/log");
var pair_1 = require("@devexpress/utils/lib/class/pair");
var number_1 = require("@devexpress/utils/lib/utils/map/number");
var base_formatter_1 = require("../../base-formatter");
var CCF_TableRow = (function (_super) {
    tslib_1.__extends(CCF_TableRow, _super);
    function CCF_TableRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CCF_TableRow.prototype.isHandleObject = function (obj) {
        return obj instanceof table_row_1.TableRow;
    };
    CCF_TableRow.prototype.getShortDescription = function (_config) {
        var list = [
            new pair_1.Pair("", this.getIndex() + ":"),
            new pair_1.Pair("cells", this.curr.cells.length),
            new pair_1.Pair("height", this.curr.height),
        ];
        return this.stdShow(new base_formatter_1.StdProps(list).showAsLine());
    };
    CCF_TableRow.prototype.availableFullDescription = function (_config) {
        return true;
    };
    CCF_TableRow.prototype.getFullDescription = function (_config) {
        var list = [
            new pair_1.Pair("index", this.getIndex()),
            new pair_1.Pair("gridBefore", this.curr.gridBefore),
            new pair_1.Pair("gridAfter", this.curr.gridAfter),
            new pair_1.Pair("cells", this.curr.cells),
            new pair_1.Pair("height", this.curr.height),
            new pair_1.Pair("text", this.getRawText(this.getSubDocument().getText(this.curr.interval)).setLen(30)),
            new pair_1.Pair("runsOnCell", this.getSubDocument().getRunsByInterval(this.curr.interval)),
            new pair_1.Pair("conditionalFormatting", log_1.Log.mask(table_base_structures_1.ConditionalTableStyleFormatting, this.curr.conditionalFormatting)),
            new pair_1.Pair("properties", this.curr.properties),
            new pair_1.Pair("tablePropertiesException", this.curr.tablePropertiesException),
        ];
        if (this.curr.gridBefore)
            list.push(new pair_1.Pair("widthBefore", this.curr.widthBefore));
        if (this.curr.gridAfter)
            list.push(new pair_1.Pair("widthAfter", this.curr.widthAfter));
        list.push(new pair_1.Pair("parentTable", this.curr.parentTable));
        return this.stdShow(new base_formatter_1.StdProps(list).showAsColumn());
    };
    CCF_TableRow.prototype.getIndex = function () {
        return this.curr.parentTable.rows.indexOf(this.curr);
    };
    CCF_TableRow.prototype.getSubDocument = function () {
        var _this = this;
        return number_1.NumberMapUtils.elementBy(this.model.subDocuments, function (subDoc) {
            return subDoc.tables[_this.curr.parentTable.index] == _this.curr.parentTable;
        });
    };
    CCF_TableRow._foo = base_formatter_1.BaseFormatter.addToFormattersList(new CCF_TableRow());
    return CCF_TableRow;
}(base_formatter_1.BaseFormatter));
exports.CCF_TableRow = CCF_TableRow;
