// devexpress-dashboard/common/docking-layout/drag-and-drop/_layout-drag-over-state.js
/**
* DevExpress Dashboard (_layout-drag-over-state.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _layout_item_1 = require("../core/_layout-item");
var DragOverStateController = (function () {
    function DragOverStateController() {
    }
    DragOverStateController.prototype.update = function (targetItem, currentHoverLocation) {
        var _this = this;
        if (!this._inTimer && (this._dragOverItem !== targetItem || currentHoverLocation !== this._hoverLocation)) {
            this.reset();
            this._hoverLocation = currentHoverLocation;
            this._dragOverItem = targetItem;
            _layout_item_1.setHoverLocation(this._dragOverItem, this._hoverLocation);
            this._currentTargetItem = this._dragOverItem;
            this._currentTargetItemBehavior = 'InsertBesideGroup';
            var selectionList = this._dragOverItem.getSelectionParentsList(this._hoverLocation);
            if (selectionList.length > 1) {
                var currentIndex = 1;
                this._timer && clearInterval(this._timer);
                this._timer = window.setInterval(function () {
                    _this._inTimer = true;
                    try {
                        _layout_item_1.setHoverLocation(_this._currentTargetItem, null, _this._currentTargetItemBehavior);
                        _this._currentTargetItem = selectionList[currentIndex].node;
                        _this._currentTargetItemBehavior = selectionList[currentIndex].dropBehavior;
                        _layout_item_1.setHoverLocation(_this._currentTargetItem, _this._hoverLocation, _this._currentTargetItemBehavior);
                        currentIndex++;
                        if (currentIndex >= selectionList.length) {
                            currentIndex = 0;
                        }
                    }
                    finally {
                        _this._inTimer = false;
                    }
                }, 1000);
            }
        }
    };
    DragOverStateController.prototype.getState = function () {
        if (this._currentTargetItem) {
            return {
                targetItem: this._currentTargetItem,
                targetItemBehavior: this._currentTargetItemBehavior,
                hoverLocation: this._hoverLocation
            };
        }
        return null;
    };
    DragOverStateController.prototype.reset = function () {
        this._timer && clearInterval(this._timer);
        _layout_item_1.setHoverLocation(this._dragOverItem);
        _layout_item_1.setHoverLocation(this._currentTargetItem);
        this._currentTargetItem = null;
        this._currentTargetItemBehavior = 'InsertBesideGroup';
        this._dragOverItem = null;
        this._timer = null;
        this._inTimer = null;
    };
    return DragOverStateController;
}());
exports.DragOverStateController = DragOverStateController;
