// devexpress-richedit/lib/rtf/import/destination/shape/shape-properties-destination-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var chunk_1 = require("../../../../core/model/chunk");
var dx_color_1 = require("../../../../core/model/color/dx-color");
var enums_1 = require("../../../../core/model/enums");
var enums_2 = require("../../../../core/model/floating-objects/enums");
var number_1 = require("@devexpress/utils/lib/utils/map/number");
var rtf_drawing_keywords_1 = require("../../../translation-table/rtf-drawing-keywords");
var shape_properties_info_1 = require("../../model/shape/shape-properties-info");
var destination_1 = require("../base/destination");
var skip_destination_1 = require("../base/skip-destination");
var destination_type_1 = require("../utils/destination-type");
var enums_3 = require("./enums");
var shape_property_destination_1 = require("./shape-property-destination");
var shape_text_destination_1 = require("./shape-text-destination");
var ShapePropertiesDestinationBase = (function (_super) {
    tslib_1.__extends(ShapePropertiesDestinationBase, _super);
    function ShapePropertiesDestinationBase(importer, shapeProperties) {
        if (shapeProperties === void 0) { shapeProperties = new shape_properties_info_1.RtfShapePropertiesInfo(); }
        var _this = _super.call(this, importer) || this;
        _this.shapeProperties = shapeProperties;
        return _this;
    }
    Object.defineProperty(ShapePropertiesDestinationBase.prototype, "destinationType", {
        get: function () { return destination_type_1.DestinationType.ShapePropertiesDestinationBase; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShapePropertiesDestinationBase.prototype, "controlCharHT", {
        get: function () { return null; },
        enumerable: true,
        configurable: true
    });
    ShapePropertiesDestinationBase.getThis = function (rtfImporter) {
        return rtfImporter.destination;
    };
    ShapePropertiesDestinationBase.onShapeId = function (_importer, _parameterValue, _hasParameter) {
    };
    ShapePropertiesDestinationBase.onShapeTextKeyword = function (importer, _parameterValue, _hasParameter) {
        var destination = ShapePropertiesDestinationBase.getThis(importer);
        var textBox = importer.documentModel.createSubDocument(enums_1.SubDocumentInfoType.TextBox, importer.subDocument.id, true);
        textBox.chunks = [new chunk_1.Chunk(textBox.positionManager.registerPosition(0), '', true)];
        destination.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeText, textBox.id);
        destination.markAsShape();
        importer.destination = new shape_text_destination_1.ShapeTextDestination(importer, textBox);
    };
    ShapePropertiesDestinationBase.onShapePropertyKeyword = function (importer, _parameterValue, _hasParameter) {
        importer.destination = new shape_property_destination_1.ShapePropertyDestination(importer, ShapePropertiesDestinationBase.getThis(importer).shapeProperties);
    };
    ShapePropertiesDestinationBase.onShapeResultKeyword = function (importer, _parameterValue, _hasParameter) {
        importer.destination = new skip_destination_1.SkipDestination(importer);
    };
    ShapePropertiesDestinationBase.onShapeLockedKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeLocked, ShapePropertiesDestinationBase.trueIntValue);
    };
    ShapePropertiesDestinationBase.onShapeWrapTextSideKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeWrapTextSide, parameterValue);
    };
    ShapePropertiesDestinationBase.onShapeWrapTextTypeZOrderKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter) {
            var destination = ShapePropertiesDestinationBase.getThis(importer);
            var rawWrapType = destination.shapeProperties.getPropertyOrNull(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeWrapTextType);
            if (rawWrapType !== null && rawWrapType == enums_2.AnchorObjectTextWrapType.None)
                destination.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeWrapTextTypeZOrder, parameterValue);
        }
    };
    ShapePropertiesDestinationBase.onShapeWrapTextTypeKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter) {
            var newWrapType = ShapePropertiesDestinationBase.convertDrawingObjectTextWrapType(parameterValue);
            if (newWrapType == enums_2.AnchorObjectTextWrapType.None)
                return;
            ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeWrapTextType, newWrapType);
        }
    };
    ShapePropertiesDestinationBase.convertDrawingObjectTextWrapType = function (parameterValue) {
        var textWrapType = enums_2.AnchorObjectTextWrapType.None;
        number_1.NumberMapUtils.forEach(rtf_drawing_keywords_1.RtfDrawingKeywords.DrawingObjectTextWrapTypeTable, function (obj, key) {
            if (parameterValue == obj)
                textWrapType = key;
        });
        return textWrapType;
    };
    ShapePropertiesDestinationBase.onShapeHorizontalPositionTypePageKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectHorizontalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectHorizontalPositionType.Page);
    };
    ShapePropertiesDestinationBase.onShapeHorizontalPositionTypeMarginKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectHorizontalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectHorizontalPositionType.Margin);
    };
    ShapePropertiesDestinationBase.onShapeHorizontalPositionTypeColumnKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectHorizontalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectHorizontalPositionType.Column);
    };
    ShapePropertiesDestinationBase.onShapeHorizontalPositionTypeIgnoreKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeIgnoreLegacyHorizontalPositionType, ShapePropertiesDestinationBase.trueIntValue);
    };
    ShapePropertiesDestinationBase.onShapeVerticalPositionTypePageKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectVerticalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectVerticalPositionType.Page);
    };
    ShapePropertiesDestinationBase.onShapeVerticalPositionTypeMarginKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectVerticalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectVerticalPositionType.Margin);
    };
    ShapePropertiesDestinationBase.onShapeVerticalPositionTypeParagraphKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.setDrawingObjectVerticalPositionType(ShapePropertiesDestinationBase.getThis(importer).shapeProperties, enums_3.DrawingObjectVerticalPositionType.Paragraph);
    };
    ShapePropertiesDestinationBase.onShapeVerticalPositionTypeIgnoreKeyword = function (importer, _parameterValue, _hasParameter) {
        ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeIgnoreLegacyVerticalPositionType, ShapePropertiesDestinationBase.trueIntValue);
    };
    ShapePropertiesDestinationBase.onShapeLeftKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).onShapeLeftKeyword(parameterValue);
    };
    ShapePropertiesDestinationBase.onShapeTopKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).onShapeTopKeyword(parameterValue);
    };
    ShapePropertiesDestinationBase.onShapeRightKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).onShapeRightKeyword(parameterValue);
    };
    ShapePropertiesDestinationBase.onShapeBottomKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).onShapeBottomKeyword(parameterValue);
    };
    ShapePropertiesDestinationBase.onShapeZOrderKeyword = function (importer, parameterValue, hasParameter) {
        if (hasParameter)
            ShapePropertiesDestinationBase.getThis(importer).shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeZOrder, Math.max(0, parameterValue));
    };
    ShapePropertiesDestinationBase.setDrawingObjectVerticalPositionType = function (shapePropertiesInfo, drawingObjectVerticalPositionType) {
        shapePropertiesInfo.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.DXInternalLegacyVerticalPositionAlignment, drawingObjectVerticalPositionType);
    };
    ShapePropertiesDestinationBase.setDrawingObjectHorizontalPositionType = function (shapePropertiesInfo, drawingObjectHorizontalPositionType) {
        shapePropertiesInfo.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.DXInternalLegacyHorizontalPositionAlignment, drawingObjectHorizontalPositionType);
    };
    ShapePropertiesDestinationBase.prototype.processControlCharCore = function (_ch) { };
    ShapePropertiesDestinationBase.prototype.processCharCore = function (_ch) { };
    ShapePropertiesDestinationBase.prototype.processKeywordCore = function (keyword, parameterValue, hasParameter) {
        var translator = this.keywordHT[keyword];
        if (translator !== undefined) {
            translator(this.importer, parameterValue, hasParameter);
            return true;
        }
        return false;
    };
    ShapePropertiesDestinationBase.prototype.hasColorProperty = function (name) {
        return this.shapeProperties.getPropertyOrNull(name) != null;
    };
    ShapePropertiesDestinationBase.prototype.getColorPropertyValue = function (name) {
        var r = this.shapeProperties.getNullableColorProperty(name);
        return r != null ? r : dx_color_1.DXColor.empty;
    };
    ShapePropertiesDestinationBase.prototype.onShapeLeftKeyword = function (parameterValue) {
        this.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeLeft, parameterValue);
        this.markAsShape();
    };
    ShapePropertiesDestinationBase.prototype.onShapeTopKeyword = function (parameterValue) {
        this.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeTop, parameterValue);
        this.markAsShape();
    };
    ShapePropertiesDestinationBase.prototype.onShapeRightKeyword = function (parameterValue) {
        this.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeRight, parameterValue);
        this.markAsShape();
    };
    ShapePropertiesDestinationBase.prototype.onShapeBottomKeyword = function (parameterValue) {
        this.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeBottom, parameterValue);
        this.markAsShape();
    };
    ShapePropertiesDestinationBase.prototype.markAsShape = function () {
        this.shapeProperties.addProperty(rtf_drawing_keywords_1.RtfDrawingKeywords.DXInternalIsShape, ShapePropertiesDestinationBase.trueIntValue);
    };
    ShapePropertiesDestinationBase.distanceFromText = 114305;
    ShapePropertiesDestinationBase.trueIntValue = 1;
    return ShapePropertiesDestinationBase;
}(destination_1.DestinationBase));
exports.ShapePropertiesDestinationBase = ShapePropertiesDestinationBase;
