// devexpress-richedit/lib/base-utils/direction-adaptor/y-direction-adaptor.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var YDirectionAdaptor = (function () {
    function YDirectionAdaptor() {
    }
    YDirectionAdaptor.prototype.isXDirection = function () {
        return false;
    };
    return YDirectionAdaptor;
}());
exports.YDirectionAdaptor = YDirectionAdaptor;
var PointDirectionAdaptorY = (function (_super) {
    __extends(PointDirectionAdaptorY, _super);
    function PointDirectionAdaptorY() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PointDirectionAdaptorY.prototype.init = function (obj) {
        this.obj = obj;
        return this;
    };
    Object.defineProperty(PointDirectionAdaptorY.prototype, "position", {
        get: function () {
            return this.obj.y;
        },
        set: function (y) {
            this.obj.y = y;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PointDirectionAdaptorY.prototype, "anotherPosition", {
        get: function () {
            return this.obj.x;
        },
        set: function (x) {
            this.obj.x = x;
        },
        enumerable: true,
        configurable: true
    });
    return PointDirectionAdaptorY;
}(YDirectionAdaptor));
exports.PointDirectionAdaptorY = PointDirectionAdaptorY;
var SizeDirectionAdaptorY = (function (_super) {
    __extends(SizeDirectionAdaptorY, _super);
    function SizeDirectionAdaptorY() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SizeDirectionAdaptorY.prototype.init = function (obj) {
        this.obj = obj;
        return this;
    };
    Object.defineProperty(SizeDirectionAdaptorY.prototype, "length", {
        get: function () {
            return this.obj.height;
        },
        set: function (height) {
            this.obj.height = height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SizeDirectionAdaptorY.prototype, "anotherLength", {
        get: function () {
            return this.obj.width;
        },
        set: function (width) {
            this.obj.width = width;
        },
        enumerable: true,
        configurable: true
    });
    return SizeDirectionAdaptorY;
}(YDirectionAdaptor));
exports.SizeDirectionAdaptorY = SizeDirectionAdaptorY;
var RectangleDirectionAdaptorY = (function (_super) {
    __extends(RectangleDirectionAdaptorY, _super);
    function RectangleDirectionAdaptorY() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RectangleDirectionAdaptorY.prototype.init = function (obj) {
        this.obj = obj;
        return this;
    };
    Object.defineProperty(RectangleDirectionAdaptorY.prototype, "length", {
        get: function () {
            return this.obj.height;
        },
        set: function (height) {
            this.obj.height = height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RectangleDirectionAdaptorY.prototype, "anotherLength", {
        get: function () {
            return this.obj.width;
        },
        set: function (width) {
            this.obj.width = width;
        },
        enumerable: true,
        configurable: true
    });
    return RectangleDirectionAdaptorY;
}(PointDirectionAdaptorY));
exports.RectangleDirectionAdaptorY = RectangleDirectionAdaptorY;
