// devexpress-richedit/lib/base-utils/web-utils/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_1 = require("@devexpress/utils/lib/browser");
var attr_1 = require("@devexpress/utils/lib/utils/attr");
var common_1 = require("@devexpress/utils/lib/utils/common");
function SetStyles(element, styles, makeImportant) {
    if (common_1.isDefined(styles.cssText))
        element.style.cssText = styles.cssText;
    if (common_1.isDefined(styles.className))
        element.className = styles.className;
    for (var property in styles) {
        if (!Object.prototype.hasOwnProperty.call(styles, property))
            continue;
        var value = styles[property];
        switch (property) {
            case "cssText":
            case "className":
                break;
            case "float":
                SetElementFloat(element, value);
                break;
            case "opacity":
                element.style.opacity = value.toString();
                break;
            case "zIndex":
                SetStylesCore(element, property, value, makeImportant);
                break;
            default:
                SetStylesCore(element, property, value + (typeof (value) == "number" ? "px" : ""), makeImportant);
        }
    }
}
exports.SetStyles = SetStyles;
;
function SetElementFloat(element, value) {
    if (common_1.isDefined(element.style.cssFloat))
        element.style.cssFloat = value;
    else if (common_1.isDefined(element.style.styleFloat))
        element.style.styleFloat = value;
    else
        attr_1.AttrUtils.setElementAttribute(element, "float", value);
}
;
function SetStylesCore(element, property, value, makeImportant) {
    if (makeImportant) {
        var index = property.search("[A-Z]");
        if (index != -1)
            property = property.replace(property.charAt(index), "-" + property.charAt(index).toLowerCase());
        element.style.setProperty(property, value, "important");
    }
    else
        element.style[property] = value;
}
function IsExistsElement(element) {
    return element && IsValidElement(element);
}
exports.IsExistsElement = IsExistsElement;
function IsValidElement(element) {
    if (!element)
        return false;
    if (!(browser_1.Browser.Firefox && browser_1.Browser.Version < 4)) {
        if (element.ownerDocument && element.ownerDocument.body && element.ownerDocument.body.compareDocumentPosition)
            return element.ownerDocument.body.compareDocumentPosition(element) % 2 === 0;
    }
    if (!browser_1.Browser.Opera && !(browser_1.Browser.IE && browser_1.Browser.Version < 9) && element.offsetParent && element.parentNode.tagName)
        return true;
    while (element != null) {
        if (element.tagName == "BODY")
            return true;
        element = element.parentNode;
    }
    return false;
}
exports.IsValidElement = IsValidElement;
function IsPercentageSize(size) {
    return size && size.indexOf('%') != -1;
}
exports.IsPercentageSize = IsPercentageSize;
