// devexpress-richedit/lib/core/model/caches/images.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var unit_converter_1 = require("@devexpress/utils/lib/class/unit-converter");
var size_1 = require("@devexpress/utils/lib/geometry/size");
var base64_1 = require("@devexpress/utils/lib/utils/base64");
var number_1 = require("@devexpress/utils/lib/utils/map/number");
var CacheImageInfo = (function () {
    function CacheImageInfo(base64, actualId, tmpId, imageUrl, file, referenceInfo, size, isLoaded) {
        this._base64 = base64 !== undefined ? base64_1.Base64Utils.normalizeToDataUrl(base64, "image/png") : undefined;
        this.actualId = actualId;
        this.tmpId = tmpId;
        this._referenceInfo = referenceInfo;
        this._size = size ? size : CacheImageInfo.emptyPictureSize;
        this._isLoaded = isLoaded !== undefined ? isLoaded : false;
        this.imageUrl = imageUrl;
        this.file = file;
    }
    Object.defineProperty(CacheImageInfo, "emptyPictureSize", {
        get: function () { return new size_1.Size(CacheImageInfo.emptyPicDimension, CacheImageInfo.emptyPicDimension); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CacheImageInfo.prototype, "isLoaded", {
        get: function () { return this._referenceInfo ? this._referenceInfo._isLoaded : this._isLoaded; },
        set: function (val) { this._isLoaded = val; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CacheImageInfo.prototype, "size", {
        get: function () { return this._referenceInfo ? this._referenceInfo._size : this._size; },
        set: function (val) { this._size = val; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CacheImageInfo.prototype, "currId", {
        get: function () { return this.actualId !== undefined ? this.actualId : this.tmpId; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CacheImageInfo.prototype, "base64", {
        get: function () { return this._referenceInfo ? this._referenceInfo._base64 : this._base64; },
        set: function (val) { this._base64 = base64_1.Base64Utils.normalizeToDataUrl(val, "image/png"); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CacheImageInfo.prototype, "referenceInfo", {
        get: function () { return this._referenceInfo; },
        set: function (val) {
            this._referenceInfo = val;
            this._base64 = undefined;
            this._size = undefined;
            this._isLoaded = undefined;
            this.file = undefined;
        },
        enumerable: true,
        configurable: true
    });
    CacheImageInfo.prototype.equals = function (obj) {
        return this._referenceInfo === obj._referenceInfo ||
            this.actualId == obj.actualId &&
                this.tmpId == obj.tmpId &&
                this.base64 === obj.base64 &&
                this.imageUrl === obj.imageUrl &&
                this.file === obj.file &&
                this.size.equals(obj.size);
    };
    CacheImageInfo.prototype.clone = function () {
        return new CacheImageInfo(this._base64, this.actualId, this.tmpId, this.imageUrl, this.file, this._referenceInfo, this._size, this._isLoaded);
    };
    CacheImageInfo.emptyPicDimension = unit_converter_1.UnitConverter.pixelsToTwips(32);
    return CacheImageInfo;
}());
exports.CacheImageInfo = CacheImageInfo;
var ImageCache = (function () {
    function ImageCache() {
        this.emptyImageId = 0;
        this.lastTmpId = 0;
        this.lastActualId = 1;
        this.cache = {};
        var onePixelSize = unit_converter_1.UnitConverter.pixelsToTwips(1);
        var emptyImage = this.createUnloadedInfoByBase64(ImageCache.transparentWhiteImage1_1, new size_1.Size(onePixelSize, onePixelSize));
        emptyImage.isLoaded = false;
    }
    Object.defineProperty(ImageCache.prototype, "emptyImage", {
        get: function () { return this.cache[this.emptyImageId]; },
        enumerable: true,
        configurable: true
    });
    ImageCache.prototype.getPictureData = function (id) {
        return this.cache[id];
    };
    ImageCache.prototype.createUnloadedInfoByUrl = function (imageUrl, size) {
        var info = this.findInfoByUrl(imageUrl);
        if (info)
            return info;
        return this.registerPictureData(new CacheImageInfo(this.cache[this.emptyImageId].base64, undefined, this.lastTmpId--, imageUrl, undefined, undefined, this.origSizeCorrect(size) ? size : undefined));
    };
    ImageCache.prototype.createUnloadedInfoByFile = function (file) {
        return this.registerPictureData(new CacheImageInfo(this.cache[this.emptyImageId].base64, undefined, this.lastTmpId--, undefined, file));
    };
    ImageCache.prototype.registerFromAnotherModel = function (imageInfo) {
        imageInfo.isLoaded = false;
        imageInfo.actualId = undefined;
        imageInfo.base64 = this.cache[this.emptyImageId].base64;
        imageInfo.tmpId = this.lastTmpId--;
    };
    ImageCache.prototype.createUnloadedInfoByBase64 = function (base64, size) {
        var info = this.findInfoByBase64(base64);
        if (info)
            return info;
        var origSizeCorrect = this.origSizeCorrect(size);
        return this.registerPictureData(new CacheImageInfo(base64, undefined, this.lastTmpId--, undefined, undefined, undefined, origSizeCorrect ? size : undefined));
    };
    ImageCache.prototype.createLoadedInfo = function (base64, size, id) {
        var info = this.findInfoByBase64(base64);
        if (info) {
            return id === undefined || id === info.actualId ?
                info :
                this.registerPictureData(new CacheImageInfo(undefined, id, undefined, undefined, undefined, info, undefined));
        }
        if (id === undefined)
            id = this.getNextActualId();
        return this.registerPictureData(new CacheImageInfo(base64, id, undefined, undefined, undefined, undefined, size, !!size));
    };
    ImageCache.prototype.createUnloadedByBase64OrUrl = function (data, size) {
        return base64_1.Base64Utils.checkPrependDataUrl(data) ?
            this.createUnloadedInfoByBase64(data, size) :
            this.createUnloadedInfoByUrl(data, size);
    };
    ImageCache.prototype.finalizeLoading = function (existingInfo, loadedInfo) {
        existingInfo.isLoaded = true;
        existingInfo.size = loadedInfo.size;
        existingInfo.actualId = loadedInfo.actualId;
        if (existingInfo.actualId === undefined)
            existingInfo.actualId = this.getNextActualId();
        this.registerPictureData(existingInfo);
        if (existingInfo.referenceInfo)
            return;
        var base64 = base64_1.Base64Utils.normalizeToDataUrl(loadedInfo.base64, "image/png");
        if (!number_1.NumberMapUtils.containsBy(this.cache, function (cacheElem) {
            var isReference = cacheElem.base64 == base64 && cacheElem !== existingInfo && cacheElem.isLoaded;
            if (isReference)
                existingInfo.referenceInfo = cacheElem.referenceInfo ? cacheElem.referenceInfo : cacheElem;
            return isReference;
        }))
            existingInfo.base64 = base64;
    };
    ImageCache.prototype.getNextActualId = function () {
        return this.lastActualId++;
    };
    ImageCache.prototype.isDataRegistered = function (data) {
        return !!this.cache[data.actualId] || !!this.cache[data.tmpId];
    };
    ImageCache.prototype.registerPictureData = function (data) {
        var existingData = this.cache[data.actualId];
        if (!existingData)
            existingData = this.cache[data.tmpId];
        if (!existingData)
            existingData = data;
        if (data.actualId !== undefined)
            this.cache[data.actualId] = existingData;
        if (data.tmpId !== undefined)
            this.cache[data.tmpId] = existingData;
        return existingData;
    };
    ImageCache.prototype.loadAllPictures = function (picture) {
        var _this = this;
        number_1.NumberMapUtils.forEach(this.cache, function (cacheInfo) {
            if (_this.emptyImageId != cacheInfo.actualId)
                picture.loader.load(cacheInfo);
        });
    };
    ImageCache.prototype.findInfoByBase64 = function (base64) {
        base64 = base64_1.Base64Utils.normalizeToDataUrl(base64, "image/png");
        return number_1.NumberMapUtils.elementBy(this.cache, function (info) { return info.base64 == base64; });
    };
    ImageCache.prototype.findInfoByUrl = function (imageUrl) {
        return number_1.NumberMapUtils.elementBy(this.cache, function (info) { return info.imageUrl == imageUrl; });
    };
    ImageCache.prototype.origSizeCorrect = function (size) {
        return size && !!size.width && !!size.height;
    };
    ImageCache.prototype.clone = function () {
        var result = new ImageCache();
        result.lastTmpId = this.lastTmpId;
        result.lastActualId = this.lastActualId;
        result.cache = number_1.NumberMapUtils.deepCopy(this.cache);
        number_1.NumberMapUtils.forEach(result.cache, function (el) {
            if (el.referenceInfo)
                el.referenceInfo = result.cache[el.currId];
        });
        return result;
    };
    ImageCache.transparentWhiteImage1_1 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAANSURBVBhXY/j///9/AAn7A/0FQ0XKAAAAAElFTkSuQmCC';
    return ImageCache;
}());
exports.ImageCache = ImageCache;
