// devexpress-richedit/lib/core/model/character/character-property-descriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var comparers_1 = require("@devexpress/utils/lib/utils/comparers");
var color_model_info_1 = require("../color/color-model-info");
var character_properties_history_items_1 = require("../history/items/character-properties-history-items");
var json_character_enums_1 = require("../json/enums/json-character-enums");
var shading_info_1 = require("../shadings/shading-info");
var composite_font_info_1 = require("./composite-font-info");
var enums_1 = require("./enums");
var lang_info_1 = require("./lang-info");
var CharacterPropertiesAllCapsDescriptor = (function () {
    function CharacterPropertiesAllCapsDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesAllCapsDescriptor.prototype.setProp = function (props, newValue) {
        props.allCaps = newValue;
    };
    CharacterPropertiesAllCapsDescriptor.prototype.getProp = function (props) {
        return props.allCaps;
    };
    CharacterPropertiesAllCapsDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseAllCaps;
    };
    CharacterPropertiesAllCapsDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontCapsHistoryItem;
    };
    CharacterPropertiesAllCapsDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.AllCaps;
    };
    return CharacterPropertiesAllCapsDescriptor;
}());
exports.CharacterPropertiesAllCapsDescriptor = CharacterPropertiesAllCapsDescriptor;
var CharacterPropertiesFontSizeDescriptor = (function () {
    function CharacterPropertiesFontSizeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 11;
    }
    CharacterPropertiesFontSizeDescriptor.prototype.setProp = function (props, newValue) {
        props.fontSize = newValue;
    };
    CharacterPropertiesFontSizeDescriptor.prototype.getProp = function (props) {
        return props.fontSize;
    };
    CharacterPropertiesFontSizeDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseDoubleFontSize;
    };
    CharacterPropertiesFontSizeDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontSizeHistoryItem;
    };
    CharacterPropertiesFontSizeDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.Size;
    };
    return CharacterPropertiesFontSizeDescriptor;
}());
exports.CharacterPropertiesFontSizeDescriptor = CharacterPropertiesFontSizeDescriptor;
var CharacterPropertiesFontBoldDescriptor = (function () {
    function CharacterPropertiesFontBoldDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesFontBoldDescriptor.prototype.setProp = function (props, newValue) {
        props.fontBold = newValue;
    };
    CharacterPropertiesFontBoldDescriptor.prototype.getProp = function (props) {
        return props.fontBold;
    };
    CharacterPropertiesFontBoldDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseFontBold;
    };
    CharacterPropertiesFontBoldDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontBoldHistoryItem;
    };
    CharacterPropertiesFontBoldDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.Bold;
    };
    return CharacterPropertiesFontBoldDescriptor;
}());
exports.CharacterPropertiesFontBoldDescriptor = CharacterPropertiesFontBoldDescriptor;
var CharacterPropertiesFontItalicDescriptor = (function () {
    function CharacterPropertiesFontItalicDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesFontItalicDescriptor.prototype.setProp = function (props, newValue) {
        props.fontItalic = newValue;
    };
    CharacterPropertiesFontItalicDescriptor.prototype.getProp = function (props) {
        return props.fontItalic;
    };
    CharacterPropertiesFontItalicDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseFontItalic;
    };
    CharacterPropertiesFontItalicDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontItalicHistoryItem;
    };
    CharacterPropertiesFontItalicDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.Italic;
    };
    return CharacterPropertiesFontItalicDescriptor;
}());
exports.CharacterPropertiesFontItalicDescriptor = CharacterPropertiesFontItalicDescriptor;
var CharacterPropertiesFontInfoDescriptor = (function () {
    function CharacterPropertiesFontInfoDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = undefined;
    }
    CharacterPropertiesFontInfoDescriptor.prototype.setProp = function (props, newValue) {
        props.fontInfo = newValue;
    };
    CharacterPropertiesFontInfoDescriptor.prototype.getProp = function (props) {
        return props.fontInfo;
    };
    CharacterPropertiesFontInfoDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseFontName;
    };
    CharacterPropertiesFontInfoDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontNameHistoryItem;
    };
    CharacterPropertiesFontInfoDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.FontName;
    };
    return CharacterPropertiesFontInfoDescriptor;
}());
exports.CharacterPropertiesFontInfoDescriptor = CharacterPropertiesFontInfoDescriptor;
var CharacterPropertiesScriptDescriptor = (function () {
    function CharacterPropertiesScriptDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = enums_1.CharacterFormattingScript.Normal;
    }
    CharacterPropertiesScriptDescriptor.prototype.setProp = function (props, newValue) {
        props.script = newValue;
    };
    CharacterPropertiesScriptDescriptor.prototype.getProp = function (props) {
        return props.script;
    };
    CharacterPropertiesScriptDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseScript;
    };
    CharacterPropertiesScriptDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontScriptHistoryItem;
    };
    CharacterPropertiesScriptDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.Script;
    };
    return CharacterPropertiesScriptDescriptor;
}());
exports.CharacterPropertiesScriptDescriptor = CharacterPropertiesScriptDescriptor;
var CharacterPropertiesStrikeoutTypeDescriptor = (function () {
    function CharacterPropertiesStrikeoutTypeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = enums_1.StrikeoutType.None;
    }
    CharacterPropertiesStrikeoutTypeDescriptor.prototype.setProp = function (props, newValue) {
        props.fontStrikeoutType = newValue;
    };
    CharacterPropertiesStrikeoutTypeDescriptor.prototype.getProp = function (props) {
        return props.fontStrikeoutType;
    };
    CharacterPropertiesStrikeoutTypeDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseFontStrikeoutType;
    };
    CharacterPropertiesStrikeoutTypeDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontStrikeoutTypeHistoryItem;
    };
    CharacterPropertiesStrikeoutTypeDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.StrikeoutType;
    };
    return CharacterPropertiesStrikeoutTypeDescriptor;
}());
exports.CharacterPropertiesStrikeoutTypeDescriptor = CharacterPropertiesStrikeoutTypeDescriptor;
var CharacterPropertiesUnderlineTypeDescriptor = (function () {
    function CharacterPropertiesUnderlineTypeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = enums_1.UnderlineType.None;
    }
    CharacterPropertiesUnderlineTypeDescriptor.prototype.setProp = function (props, newValue) {
        props.fontUnderlineType = newValue;
    };
    CharacterPropertiesUnderlineTypeDescriptor.prototype.getProp = function (props) {
        return props.fontUnderlineType;
    };
    CharacterPropertiesUnderlineTypeDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseFontUnderlineType;
    };
    CharacterPropertiesUnderlineTypeDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontUnderlineTypeHistoryItem;
    };
    CharacterPropertiesUnderlineTypeDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.UnderlineType;
    };
    return CharacterPropertiesUnderlineTypeDescriptor;
}());
exports.CharacterPropertiesUnderlineTypeDescriptor = CharacterPropertiesUnderlineTypeDescriptor;
var CharacterPropertiesUnderlineWordsOnlyDescriptor = (function () {
    function CharacterPropertiesUnderlineWordsOnlyDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesUnderlineWordsOnlyDescriptor.prototype.setProp = function (props, newValue) {
        props.underlineWordsOnly = newValue;
    };
    CharacterPropertiesUnderlineWordsOnlyDescriptor.prototype.getProp = function (props) {
        return props.underlineWordsOnly;
    };
    CharacterPropertiesUnderlineWordsOnlyDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseUnderlineWordsOnly;
    };
    CharacterPropertiesUnderlineWordsOnlyDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontUnderlineWordsOnlyHistoryItem;
    };
    CharacterPropertiesUnderlineWordsOnlyDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.UnderlineWordsOnly;
    };
    return CharacterPropertiesUnderlineWordsOnlyDescriptor;
}());
exports.CharacterPropertiesUnderlineWordsOnlyDescriptor = CharacterPropertiesUnderlineWordsOnlyDescriptor;
var CharacterPropertiesStrikeoutWordsOnlyDescriptor = (function () {
    function CharacterPropertiesStrikeoutWordsOnlyDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesStrikeoutWordsOnlyDescriptor.prototype.setProp = function (props, newValue) {
        props.strikeoutWordsOnly = newValue;
    };
    CharacterPropertiesStrikeoutWordsOnlyDescriptor.prototype.getProp = function (props) {
        return props.strikeoutWordsOnly;
    };
    CharacterPropertiesStrikeoutWordsOnlyDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseStrikeoutWordsOnly;
    };
    CharacterPropertiesStrikeoutWordsOnlyDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontStrikeoutWordsOnlyHistoryItem;
    };
    CharacterPropertiesStrikeoutWordsOnlyDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.StrikeoutWordsOnly;
    };
    return CharacterPropertiesStrikeoutWordsOnlyDescriptor;
}());
exports.CharacterPropertiesStrikeoutWordsOnlyDescriptor = CharacterPropertiesStrikeoutWordsOnlyDescriptor;
var CharacterPropertiesNoProofDescriptor = (function () {
    function CharacterPropertiesNoProofDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesNoProofDescriptor.prototype.setProp = function (props, newValue) {
        props.noProof = newValue;
    };
    CharacterPropertiesNoProofDescriptor.prototype.getProp = function (props) {
        return props.noProof;
    };
    CharacterPropertiesNoProofDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseNoProof;
    };
    CharacterPropertiesNoProofDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontNoProofHistoryItem;
    };
    CharacterPropertiesNoProofDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.NoProof;
    };
    return CharacterPropertiesNoProofDescriptor;
}());
exports.CharacterPropertiesNoProofDescriptor = CharacterPropertiesNoProofDescriptor;
var CharacterPropertiesHiddenDescriptor = (function () {
    function CharacterPropertiesHiddenDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    CharacterPropertiesHiddenDescriptor.prototype.setProp = function (props, newValue) {
        props.hidden = newValue;
    };
    CharacterPropertiesHiddenDescriptor.prototype.getProp = function (props) {
        return props.hidden;
    };
    CharacterPropertiesHiddenDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseHidden;
    };
    CharacterPropertiesHiddenDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontHiddenHistoryItem;
    };
    CharacterPropertiesHiddenDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.Hidden;
    };
    return CharacterPropertiesHiddenDescriptor;
}());
exports.CharacterPropertiesHiddenDescriptor = CharacterPropertiesHiddenDescriptor;
var CharacterPropertiesTextColorDescriptor = (function () {
    function CharacterPropertiesTextColorDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = color_model_info_1.ColorModelInfo.autoColor;
    }
    CharacterPropertiesTextColorDescriptor.prototype.setProp = function (props, newValue) {
        props.textColor = newValue;
    };
    CharacterPropertiesTextColorDescriptor.prototype.getProp = function (props) {
        return props.textColor;
    };
    CharacterPropertiesTextColorDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseForeColorIndex;
    };
    CharacterPropertiesTextColorDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontTextColorHistoryItem;
    };
    CharacterPropertiesTextColorDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.TextColor;
    };
    return CharacterPropertiesTextColorDescriptor;
}());
exports.CharacterPropertiesTextColorDescriptor = CharacterPropertiesTextColorDescriptor;
var CharacterPropertiesShadingInfoColorDescriptor = (function () {
    function CharacterPropertiesShadingInfoColorDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = shading_info_1.ShadingInfo.noColor;
    }
    CharacterPropertiesShadingInfoColorDescriptor.prototype.setProp = function (props, newValue) {
        props.shadingInfo = newValue;
    };
    CharacterPropertiesShadingInfoColorDescriptor.prototype.getProp = function (props) {
        return props.shadingInfo;
    };
    CharacterPropertiesShadingInfoColorDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseShadingInfoIndex;
    };
    CharacterPropertiesShadingInfoColorDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontShadingInfoHistoryItem;
    };
    CharacterPropertiesShadingInfoColorDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.ShadingInfo;
    };
    return CharacterPropertiesShadingInfoColorDescriptor;
}());
exports.CharacterPropertiesShadingInfoColorDescriptor = CharacterPropertiesShadingInfoColorDescriptor;
var CharacterPropertiesStrikeoutColorDescriptor = (function () {
    function CharacterPropertiesStrikeoutColorDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = color_model_info_1.ColorModelInfo.noColor;
    }
    CharacterPropertiesStrikeoutColorDescriptor.prototype.setProp = function (props, newValue) {
        props.strikeoutColor = newValue;
    };
    CharacterPropertiesStrikeoutColorDescriptor.prototype.getProp = function (props) {
        return props.strikeoutColor;
    };
    CharacterPropertiesStrikeoutColorDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseStrikeoutColorIndex;
    };
    CharacterPropertiesStrikeoutColorDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontStrikeoutColorHistoryItem;
    };
    CharacterPropertiesStrikeoutColorDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.StrikeoutColor;
    };
    return CharacterPropertiesStrikeoutColorDescriptor;
}());
exports.CharacterPropertiesStrikeoutColorDescriptor = CharacterPropertiesStrikeoutColorDescriptor;
var CharacterPropertiesUnderlineColorDescriptor = (function () {
    function CharacterPropertiesUnderlineColorDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = color_model_info_1.ColorModelInfo.noColor;
    }
    CharacterPropertiesUnderlineColorDescriptor.prototype.setProp = function (props, newValue) {
        props.underlineColor = newValue;
    };
    CharacterPropertiesUnderlineColorDescriptor.prototype.getProp = function (props) {
        return props.underlineColor;
    };
    CharacterPropertiesUnderlineColorDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseUnderlineColorIndex;
    };
    CharacterPropertiesUnderlineColorDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontUnderlineColorHistoryItem;
    };
    CharacterPropertiesUnderlineColorDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.UnderlineColor;
    };
    return CharacterPropertiesUnderlineColorDescriptor;
}());
exports.CharacterPropertiesUnderlineColorDescriptor = CharacterPropertiesUnderlineColorDescriptor;
var CharacterPropertiesHighlightColorDescriptor = (function () {
    function CharacterPropertiesHighlightColorDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = color_model_info_1.ColorModelInfo.noColor;
    }
    CharacterPropertiesHighlightColorDescriptor.prototype.setProp = function (props, newValue) {
        props.highlightColor = newValue;
    };
    CharacterPropertiesHighlightColorDescriptor.prototype.getProp = function (props) {
        return props.highlightColor;
    };
    CharacterPropertiesHighlightColorDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseHighlightColorIndex;
    };
    CharacterPropertiesHighlightColorDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontHighlightColorHistoryItem;
    };
    CharacterPropertiesHighlightColorDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.HighlightColor;
    };
    return CharacterPropertiesHighlightColorDescriptor;
}());
exports.CharacterPropertiesHighlightColorDescriptor = CharacterPropertiesHighlightColorDescriptor;
var CharacterPropertiesLangInfoDescriptor = (function () {
    function CharacterPropertiesLangInfoDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new lang_info_1.LangInfo();
    }
    CharacterPropertiesLangInfoDescriptor.prototype.setProp = function (props, newValue) {
        props.langInfo = newValue;
    };
    CharacterPropertiesLangInfoDescriptor.prototype.getProp = function (props) {
        return props.langInfo;
    };
    CharacterPropertiesLangInfoDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseLangInfo;
    };
    CharacterPropertiesLangInfoDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontLangInfoHistoryItem;
    };
    CharacterPropertiesLangInfoDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.LangInfo;
    };
    return CharacterPropertiesLangInfoDescriptor;
}());
exports.CharacterPropertiesLangInfoDescriptor = CharacterPropertiesLangInfoDescriptor;
var CharacterPropertiesCompositeFontInfoDescriptor = (function () {
    function CharacterPropertiesCompositeFontInfoDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new composite_font_info_1.CompositeFontInfo();
    }
    CharacterPropertiesCompositeFontInfoDescriptor.prototype.setProp = function (props, newValue) {
        props.compositeFontInfo = newValue;
    };
    CharacterPropertiesCompositeFontInfoDescriptor.prototype.getProp = function (props) {
        return props.compositeFontInfo;
    };
    CharacterPropertiesCompositeFontInfoDescriptor.prototype.maskValue = function () {
        return enums_1.CharacterPropertiesMask.UseCompositeFontInfo;
    };
    CharacterPropertiesCompositeFontInfoDescriptor.prototype.getHistoryItemConstructor = function () {
        return character_properties_history_items_1.FontCompositeFontInfoHistoryItem;
    };
    CharacterPropertiesCompositeFontInfoDescriptor.prototype.getJSONProperty = function () {
        return json_character_enums_1.JSONCharacterFormattingProperty.CompositeFontInfo;
    };
    return CharacterPropertiesCompositeFontInfoDescriptor;
}());
exports.CharacterPropertiesCompositeFontInfoDescriptor = CharacterPropertiesCompositeFontInfoDescriptor;
var CharacterPropertyDescriptor = (function () {
    function CharacterPropertyDescriptor() {
    }
    CharacterPropertyDescriptor.allCaps = new CharacterPropertiesAllCapsDescriptor();
    CharacterPropertyDescriptor.size = new CharacterPropertiesFontSizeDescriptor();
    CharacterPropertyDescriptor.bold = new CharacterPropertiesFontBoldDescriptor();
    CharacterPropertyDescriptor.italic = new CharacterPropertiesFontItalicDescriptor();
    CharacterPropertyDescriptor.fontInfo = new CharacterPropertiesFontInfoDescriptor();
    CharacterPropertyDescriptor.script = new CharacterPropertiesScriptDescriptor();
    CharacterPropertyDescriptor.strikeoutType = new CharacterPropertiesStrikeoutTypeDescriptor();
    CharacterPropertyDescriptor.underlineType = new CharacterPropertiesUnderlineTypeDescriptor();
    CharacterPropertyDescriptor.underlineWordsOnly = new CharacterPropertiesUnderlineWordsOnlyDescriptor();
    CharacterPropertyDescriptor.strikeoutWordsOnly = new CharacterPropertiesStrikeoutWordsOnlyDescriptor();
    CharacterPropertyDescriptor.noProof = new CharacterPropertiesNoProofDescriptor();
    CharacterPropertyDescriptor.hidden = new CharacterPropertiesHiddenDescriptor();
    CharacterPropertyDescriptor.langInfo = new CharacterPropertiesLangInfoDescriptor();
    CharacterPropertyDescriptor.compositeFontInfo = new CharacterPropertiesCompositeFontInfoDescriptor();
    CharacterPropertyDescriptor.textColor = new CharacterPropertiesTextColorDescriptor();
    CharacterPropertyDescriptor.shadingInfo = new CharacterPropertiesShadingInfoColorDescriptor();
    CharacterPropertyDescriptor.highlightColor = new CharacterPropertiesHighlightColorDescriptor();
    CharacterPropertyDescriptor.strikeoutColor = new CharacterPropertiesStrikeoutColorDescriptor();
    CharacterPropertyDescriptor.underlineColor = new CharacterPropertiesUnderlineColorDescriptor();
    CharacterPropertyDescriptor.ALL_FIELDS = [
        CharacterPropertyDescriptor.allCaps,
        CharacterPropertyDescriptor.size,
        CharacterPropertyDescriptor.bold,
        CharacterPropertyDescriptor.italic,
        CharacterPropertyDescriptor.fontInfo,
        CharacterPropertyDescriptor.script,
        CharacterPropertyDescriptor.strikeoutType,
        CharacterPropertyDescriptor.underlineType,
        CharacterPropertyDescriptor.underlineWordsOnly,
        CharacterPropertyDescriptor.strikeoutWordsOnly,
        CharacterPropertyDescriptor.noProof,
        CharacterPropertyDescriptor.hidden,
        CharacterPropertyDescriptor.langInfo,
        CharacterPropertyDescriptor.compositeFontInfo,
        CharacterPropertyDescriptor.textColor,
        CharacterPropertyDescriptor.shadingInfo,
        CharacterPropertyDescriptor.highlightColor,
        CharacterPropertyDescriptor.strikeoutColor,
        CharacterPropertyDescriptor.underlineColor,
    ];
    CharacterPropertyDescriptor.whatNeedSetWhenCreateHyperlinkField = [
        CharacterPropertyDescriptor.size,
        CharacterPropertyDescriptor.bold,
        CharacterPropertyDescriptor.italic,
        CharacterPropertyDescriptor.fontInfo,
        CharacterPropertyDescriptor.script,
        CharacterPropertyDescriptor.strikeoutType,
        CharacterPropertyDescriptor.allCaps,
        CharacterPropertyDescriptor.underlineWordsOnly,
        CharacterPropertyDescriptor.strikeoutWordsOnly,
        CharacterPropertyDescriptor.noProof,
        CharacterPropertyDescriptor.hidden,
        CharacterPropertyDescriptor.shadingInfo,
        CharacterPropertyDescriptor.strikeoutColor,
        CharacterPropertyDescriptor.underlineColor,
        CharacterPropertyDescriptor.langInfo,
    ];
    return CharacterPropertyDescriptor;
}());
exports.CharacterPropertyDescriptor = CharacterPropertyDescriptor;
