// devexpress-richedit/lib/rtf/export/exporters/rtf-anchored-text-box-run-exporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var unit_converter_1 = require("@devexpress/utils/lib/class/unit-converter");
var rtf_drawing_keywords_1 = require("../../translation-table/rtf-drawing-keywords");
var rtf_export_sr_1 = require("../../translation-table/rtf-export-sr");
var rtf_anchored_run_exporter_1 = require("./rtf-anchored-run-exporter");
var common_1 = require("@devexpress/utils/lib/utils/common");
var RtfAnchoredTextBoxRunExporter = (function (_super) {
    tslib_1.__extends(RtfAnchoredTextBoxRunExporter, _super);
    function RtfAnchoredTextBoxRunExporter(rtfBuilder, textBoxRun, exportTextBoxContent) {
        var _this = _super.call(this, rtfBuilder, textBoxRun.anchorInfo, textBoxRun.shape, textBoxRun.size) || this;
        _this.textBoxRun = textBoxRun;
        _this.exportTextBoxContent = exportTextBoxContent;
        return _this;
    }
    RtfAnchoredTextBoxRunExporter.prototype.getWidth = function () {
        return this.textBoxRun.size.absoluteSize.width;
    };
    RtfAnchoredTextBoxRunExporter.prototype.getHeight = function () {
        return this.textBoxRun.size.absoluteSize.height;
    };
    RtfAnchoredTextBoxRunExporter.prototype.exportContent = function () {
        this.exportTextBoxProperties(this.textBoxRun.textBoxProperties);
        this.rtfBuilder.openGroup();
        this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.ShapeText);
        this.exportTextBoxContent();
        this.rtfBuilder.closeGroup();
    };
    RtfAnchoredTextBoxRunExporter.prototype.exportTextBoxProperties = function (properties) {
        if (common_1.isDefined(properties.leftMargin))
            this.rtfBuilder.writeShapeIntegerProperty("dxTextLeft", unit_converter_1.UnitConverter.twipsToEmu(properties.leftMargin));
        if (common_1.isDefined(properties.rightMargin))
            this.rtfBuilder.writeShapeIntegerProperty("dxTextRight", unit_converter_1.UnitConverter.twipsToEmu(properties.rightMargin));
        if (common_1.isDefined(properties.topMargin))
            this.rtfBuilder.writeShapeIntegerProperty("dyTextTop", unit_converter_1.UnitConverter.twipsToEmu(properties.topMargin));
        if (common_1.isDefined(properties.bottomMargin))
            this.rtfBuilder.writeShapeIntegerProperty("dyTextBottom", unit_converter_1.UnitConverter.twipsToEmu(properties.bottomMargin));
        if (common_1.isDefined(properties.resizeShapeToFitText))
            this.rtfBuilder.writeShapeBoolProperty("fFitShapeToText", properties.resizeShapeToFitText);
        if (common_1.isDefined(properties.wrapText) && properties.wrapText == false)
            this.rtfBuilder.writeShapeIntegerProperty("WrapText", 2);
    };
    RtfAnchoredTextBoxRunExporter.prototype.exportFloatingObjectRelativeSize = function () {
        var textBoxSize = this.textBoxRun.size;
        if (!textBoxSize.useAbsoluteWidth()) {
            this.rtfBuilder.writeShapeIntegerProperty("sizerelh", rtf_drawing_keywords_1.RtfDrawingKeywords.DrawingObjectRelativeWidthTypeTable[textBoxSize.relativeWidthType]);
            this.rtfBuilder.writeShapeIntegerProperty("pctHoriz", textBoxSize.relativeSize.width / 100);
        }
        if (!textBoxSize.useAbsoluteHeight()) {
            this.rtfBuilder.writeShapeIntegerProperty("sizerelv", rtf_drawing_keywords_1.RtfDrawingKeywords.DrawingObjectRelativeHeightTypeTable[textBoxSize.relativeHeightType]);
            this.rtfBuilder.writeShapeIntegerProperty("pctVert", textBoxSize.relativeSize.height / 100);
        }
        _super.prototype.exportFloatingObjectRelativeSize.call(this);
    };
    return RtfAnchoredTextBoxRunExporter;
}(rtf_anchored_run_exporter_1.RtfAnchoredRunExporter));
exports.RtfAnchoredTextBoxRunExporter = RtfAnchoredTextBoxRunExporter;
