// devexpress-dashboard/model/items/map/metadata/_chorolpeth-map-item.js
/**
* DevExpress Dashboard (_chorolpeth-map-item.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _data_item_1 = require("../../../data-item/metadata/_data-item");
var _base_metadata_1 = require("../../../metadata/_base-metadata");
var _map_item_1 = require("./_map-item");
exports.maps = {
    propertyName: 'maps',
    modelName: 'Maps',
    displayName: 'DashboardWebStringId.Binding.Maps',
    array: true
};
exports.includeSummaryValueToShapeTitle = {
    propertyName: 'includeSummaryValueToShapeTitle',
    modelName: '@IncludeSummaryValueToShapeTitle',
    displayName: 'DashboardWebStringId.Map.IncludeSummaryValueToShapeTitle',
    defaultVal: false,
    simpleFormAdapterItem: 'yesNoButtonGroupEditor',
    from: _base_metadata_1.parseBool
};
exports.attributeDimension = {
    propertyName: '__attributeDimension',
    modelName: 'AttributeDimension',
    displayName: 'DashboardWebStringId.Map.AttributeDimension',
    info: _data_item_1.dataItemLinkSerializationsInfo
};
exports.attributeName = {
    propertyName: 'attributeName',
    modelName: '@AttributeName',
    displayName: 'DashboardWebStringId.Map.AttributeName',
    defaultVal: '',
    category: _base_metadata_1.PropertyCategory.Map
};
exports.tooltipAttributeName = {
    propertyName: 'tooltipAttributeName',
    modelName: '@ToolTopAttributeName',
    displayName: 'DashboardWebStringId.Map.TooltipAttribute',
    defaultVal: '',
    category: _base_metadata_1.PropertyCategory.Map
};
exports.choroplethMapDashboardItemSerializationsInfo = _map_item_1.mapDashboardItemSerializationsInfo.concat([exports.maps, exports.attributeDimension, exports.attributeName, exports.includeSummaryValueToShapeTitle, exports.tooltipAttributeName, _map_item_1.colorLegend]);
