// devexpress-dashboard/model/items/map/metadata/_map-legend.js
/**
* DevExpress Dashboard (_map-legend.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var _base_metadata_1 = require("../../../metadata/_base-metadata");
exports.legendVisible = { propertyName: 'visible', modelName: '@Visible', displayName: 'DashboardWebStringId.Map.ShowLegend', defaultVal: false, simpleFormAdapterItem: 'yesNoButtonGroupEditor', from: _base_metadata_1.parseBool };
exports.weightedLegendVisible = __assign({}, exports.legendVisible, { displayName: 'DashboardWebStringId.Map.ShowWeightedLegend' });
exports.legendPosition = {
    propertyName: 'position', modelName: '@Position', displayName: 'DashboardWebStringId.WeightedLegendGalleryGroup', defaultVal: 'TopLeft', simpleFormAdapterItem: 'listEditor', values: {
        'TopLeft': 'DashboardWebStringId.Map.Position.TopLeft',
        'TopCenter': 'DashboardWebStringId.Map.Position.TopCenter',
        'TopRight': 'DashboardWebStringId.Map.Position.TopRight',
        'BottomLeft': 'DashboardWebStringId.Map.Position.BottomLeft',
        'BottomCenter': 'DashboardWebStringId.Map.Position.BottomCenter',
        'BottomRight': 'DashboardWebStringId.Map.Position.BottomRight'
    }
};
exports.legendType = {
    propertyName: 'type', modelName: '@WeightedLegendType', displayName: 'DashboardWebStringId.Map.WeightedLegendType', defaultVal: 'Linear', values: {
        'Linear': 'DashboardWebStringId.Map.WeightedLegendType.Linear',
        'Nested': 'DashboardWebStringId.Map.WeightedLegendType.Nested'
    }
};
exports.weightedLegendSerializationsInfo = [exports.weightedLegendVisible, exports.legendPosition, exports.legendType];
exports.legendOrientation = {
    propertyName: 'orientation', modelName: '@Orientation', displayName: 'DashboardWebStringId.Chart.Orientation', defaultVal: 'Vertical', simpleFormAdapterItem: 'buttonGroupEditor', values: {
        'Vertical': 'DashboardWebStringId.Map.Orientation.Vertical',
        'Horizontal': 'DashboardWebStringId.Map.Orientation.Horizontal'
    }
};
exports.mapLegendSerializationsInfo = [exports.legendVisible, exports.legendPosition, exports.legendOrientation];
