// devexpress-richedit/lib/base/rich-utils/debug/chrome-dev-tools-custom-formatters/layout/anchor-objects-position-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var document_layout_1 = require("../../../../../core/layout/document-layout");
var pair_1 = require("@devexpress/utils/lib/class/pair");
var number_1 = require("@devexpress/utils/lib/utils/map/number");
var base_formatter_1 = require("../base-formatter");
var elements_1 = require("../elements");
var CCF_AnchorObjectsPositionInfo = (function (_super) {
    tslib_1.__extends(CCF_AnchorObjectsPositionInfo, _super);
    function CCF_AnchorObjectsPositionInfo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CCF_AnchorObjectsPositionInfo.prototype.isHandleObject = function (obj) {
        return obj instanceof document_layout_1.AnchorObjectsPositionInfo;
    };
    CCF_AnchorObjectsPositionInfo.prototype.getShortDescription = function (_config) {
        if (number_1.NumberMapUtils.isEmpty(this.curr.cache))
            return new elements_1.SpanElement().setText("Empty");
        return this.stdShow(new base_formatter_1.StdProps(this.getList()).showAsLine().setStdNumElements());
    };
    CCF_AnchorObjectsPositionInfo.prototype.hasBody = function (_, _config) {
        return !number_1.NumberMapUtils.isEmpty(this.curr.cache);
    };
    CCF_AnchorObjectsPositionInfo.prototype.getFullDescription = function (_config) {
        return this.stdShow(new base_formatter_1.StdProps(this.getList()).showAsColumn());
    };
    CCF_AnchorObjectsPositionInfo.prototype.getList = function () {
        return this.mapNumberAsKeyToListOfPairs(this.curr.cache, function (_, obj) { return obj.modelPosition; }, function (key, obj) { return new pair_1.Pair(key.toString(), obj.modelPosition); });
    };
    CCF_AnchorObjectsPositionInfo._foo = base_formatter_1.BaseFormatter.addToFormattersList(new CCF_AnchorObjectsPositionInfo());
    return CCF_AnchorObjectsPositionInfo;
}(base_formatter_1.BaseFormatter));
exports.CCF_AnchorObjectsPositionInfo = CCF_AnchorObjectsPositionInfo;
