// devexpress-richedit/lib/client/ui/client-quick-search-panel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var toolbar_1 = require("devextreme/ui/toolbar");
var quick_search_panel_1 = require("../../base/ui/quick-search-panel");
var dom_1 = require("@devexpress/utils/lib/utils/dom");
var dxt_theme_css_classes_1 = require("../../dxt-utils/dxt-utils/dxt-theme-css-classes");
var dxt_utils_1 = require("../../dxt-utils/dxt-utils/dxt-utils");
var ClientQuickSearchPanel = (function (_super) {
    tslib_1.__extends(ClientQuickSearchPanel, _super);
    function ClientQuickSearchPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ClientQuickSearchPanel.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.getQuickSearchPanelWrapper().parentElement.style.display = 'none';
    };
    ClientQuickSearchPanel.prototype.show = function () {
        this.getQuickSearchPanelWrapper().parentElement.style.display = 'block';
        _super.prototype.show.call(this);
    };
    ClientQuickSearchPanel.prototype.hide = function (resetSearch) {
        _super.prototype.hide.call(this, resetSearch);
        this.getQuickSearchPanelWrapper().parentElement.style.display = 'none';
    };
    ClientQuickSearchPanel.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this.searchTextBox.dispose();
        this.searchTextBox = null;
        dom_1.DomUtils.hideNode(this.searchTextBoxField);
        this.searchTextBoxField = null;
    };
    ClientQuickSearchPanel.prototype.createElements = function () {
        var quickSearchContainer = this.createQuickSearchContainer();
        var wrapper = this.createWrapper();
        quickSearchContainer.appendChild(wrapper);
        var quickSearchPanel = this.createQuickSearchPanel();
        wrapper.appendChild(quickSearchPanel);
        var toolbarElement = document.createElement('div');
        quickSearchPanel.appendChild(toolbarElement);
        new toolbar_1.default(toolbarElement, {
            items: this.createToolbarItems()
        });
    };
    ClientQuickSearchPanel.prototype.getSearchField = function () {
        if (!this.searchTextBoxField) {
            var rawElement = this.searchTextBox.element();
            var element = rawElement.querySelector ? rawElement : rawElement[0];
            this.searchTextBoxField = element.querySelector('input');
        }
        return this.searchTextBoxField;
    };
    ClientQuickSearchPanel.prototype.createQuickSearchContainer = function () {
        var richElement = this.owner.getMainElement();
        var quickSearchContainer = document.createElement('DIV');
        quickSearchContainer.id = this.owner.name + quick_search_panel_1.QuickSearchPanel.containerPostfix;
        dom_1.DomUtils.addClassName(quickSearchContainer, 'dxreQuickSearch');
        richElement.insertBefore(quickSearchContainer, richElement.childNodes[2]);
        return quickSearchContainer;
    };
    ClientQuickSearchPanel.prototype.createWrapper = function () {
        var wrapper = document.createElement('DIV');
        wrapper.id = this.owner.name + quick_search_panel_1.QuickSearchPanel.wrapperPostfix;
        dom_1.DomUtils.addClassName(wrapper, 'dxreQuickSearchPanelWrapper');
        return wrapper;
    };
    ClientQuickSearchPanel.prototype.createQuickSearchPanel = function () {
        var quickSearchPanel = document.createElement('DIV');
        dom_1.DomUtils.addClassName(quickSearchPanel, 'dxreQuickSearchPanel');
        dom_1.DomUtils.addClassName(quickSearchPanel, dxt_theme_css_classes_1.DxtThemeCssClasses.BorderColor);
        dom_1.DomUtils.addClassName(quickSearchPanel, dxt_theme_css_classes_1.DxtThemeCssClasses.BackgroundColor);
        return quickSearchPanel;
    };
    ClientQuickSearchPanel.prototype.getButtonId = function (postfix) {
        return this.owner.name + postfix;
    };
    ClientQuickSearchPanel.prototype.createToolbarItems = function () {
        return [
            this.crateToolbarTextBoxTemplate(),
            this.createToolbarButtonTemplate(quick_search_panel_1.QuickSearchPanel.prevBtnPostfix, 'dxre-icon-Prev'),
            this.createToolbarButtonTemplate(quick_search_panel_1.QuickSearchPanel.nextBtnPostfix, 'dxre-icon-Next'),
            this.createToolbarButtonTemplate(quick_search_panel_1.QuickSearchPanel.collapseBtnPostfix, 'dxre-icon-ViewMergedData'),
            this.createToolbarButtonTemplate(quick_search_panel_1.QuickSearchPanel.closeBtnPostfix, 'dxre-icon-ClearHeaderAndFooter')
        ];
    };
    ClientQuickSearchPanel.prototype.crateToolbarTextBoxTemplate = function () {
        var _this = this;
        return {
            widget: 'dxTextBox',
            options: {
                onInitialized: function (e) { return _this.searchTextBox = e.component; },
                stylingMode: 'filled',
                width: '250px'
            }
        };
    };
    ClientQuickSearchPanel.prototype.createToolbarButtonTemplate = function (postfix, icon) {
        return {
            widget: 'dxButton',
            options: {
                stylingMode: "text",
                icon: dxt_utils_1.DxtUtils.correctIconName(icon),
                elementAttr: { id: this.getButtonId(postfix) }
            }
        };
    };
    return ClientQuickSearchPanel;
}(quick_search_panel_1.QuickSearchPanel));
exports.ClientQuickSearchPanel = ClientQuickSearchPanel;
