// devexpress-richedit/lib/docx/import/destination/numbering/numberings-destination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var map_creator_1 = require("../../../../base-utils/map-creator");
var character_properties_1 = require("../../../../core/model/character/character-properties");
var list_level_properties_1 = require("../../../../core/model/numbering-lists/list-level-properties");
var numbering_helper_1 = require("../../../../core/model/numbering-lists/numbering-helper");
var numbering_list_1 = require("../../../../core/model/numbering-lists/numbering-list");
var control_1 = require("../../../../core/model/options/control");
var paragraph_properties_1 = require("../../../../core/model/paragraph/paragraph-properties");
var paragraph_style_1 = require("../../../../core/model/paragraph/paragraph-style");
var rich_utils_1 = require("../../../../core/model/rich-utils");
var list_1 = require("@devexpress/utils/lib/utils/list");
var string_1 = require("@devexpress/utils/lib/utils/string");
var translation_tables_1 = require("../../../translation-table/translation-tables");
var constants_1 = require("../../../utils/constants");
var open_xml_abstract_numbering_info_1 = require("../../list/open-xml-abstract-numbering-info");
var open_xml_list_level_override_1 = require("../../list/open-xml-list-level-override");
var open_xml_numbering_list_info_1 = require("../../list/open-xml-numbering-list-info");
var run_properties_base_destination_1 = require("../character-properties/run-properties-base-destination");
var destination_1 = require("../destination");
var paragraph_properties_base_destination_1 = require("../paragraph-properties/paragraph-properties-base-destination");
var NumberingsDestination = (function (_super) {
    tslib_1.__extends(NumberingsDestination, _super);
    function NumberingsDestination(data) {
        return _super.call(this, data) || this;
    }
    Object.defineProperty(NumberingsDestination.prototype, "elementHandlerTable", {
        get: function () {
            return NumberingsDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    NumberingsDestination.onAbstractNumbering = function (data, _reader) {
        return new AbstractNumberingListDestination(data);
    };
    NumberingsDestination.onNumbering = function (data, _reader) {
        return new NumberingListDestination(data);
    };
    NumberingsDestination.handlerTable = new map_creator_1.MapCreator()
        .add('abstractNum', NumberingsDestination.onAbstractNumbering)
        .add('num', NumberingsDestination.onNumbering)
        .get();
    return NumberingsDestination;
}(destination_1.ElementDestination));
exports.NumberingsDestination = NumberingsDestination;
var AbstractNumberingListDestination = (function (_super) {
    tslib_1.__extends(AbstractNumberingListDestination, _super);
    function AbstractNumberingListDestination(data) {
        var _this = _super.call(this, data) || this;
        _this.list = new numbering_list_1.AbstractNumberingList(data.documentModel);
        _this.info = new open_xml_abstract_numbering_info_1.OpenXmlAbstractNumberingInfo();
        return _this;
    }
    Object.defineProperty(AbstractNumberingListDestination.prototype, "elementHandlerTable", {
        get: function () {
            return AbstractNumberingListDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    AbstractNumberingListDestination.getThis = function (data) {
        return data.destinationStack.getThis();
    };
    AbstractNumberingListDestination.onLevel = function (data, _reader) {
        return new NumberingLevelDestination(data, AbstractNumberingListDestination.getThis(data).list);
    };
    AbstractNumberingListDestination.onMultilevelType = function (data, _reader) {
        return new NumberingListMultiLevelTypeDestination(data, AbstractNumberingListDestination.getThis(data).list);
    };
    AbstractNumberingListDestination.onName = function (data, _reader) {
        return new NumberingListNameDestination(data, AbstractNumberingListDestination.getThis(data).list);
    };
    AbstractNumberingListDestination.onUniqueId = function (data, _reader) {
        return new AbstractNumberingListUniqueIdDestination(data, AbstractNumberingListDestination.getThis(data).list);
    };
    AbstractNumberingListDestination.onNumberingStyleLink = function (data, _reader) {
        return new NumberingListNumStyleLinkDestination(data, AbstractNumberingListDestination.getThis(data).info);
    };
    AbstractNumberingListDestination.onStyleLink = function (_data, _reader) {
        return null;
    };
    AbstractNumberingListDestination.onTemplate = function (_data, _reader) {
        return null;
    };
    AbstractNumberingListDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.id = reader.getAttributeNS('abstractNumId', this.data.constants.namespaces[constants_1.DocxNsType.WordProcessing].namespace);
                return [2];
            });
        });
    };
    AbstractNumberingListDestination.prototype.processElementClose = function (_reader) {
        if (string_1.StringUtils.isNullOrEmpty(this.id))
            return;
        var innerId = this.list.innerId;
        var index;
        var abstractNumberingList = list_1.ListUtils.elementBy(this.data.documentModel.abstractNumberingLists, function (list) { return list.innerId == innerId; });
        if (innerId != -1 && abstractNumberingList)
            index = this.data.documentModel.abstractNumberingLists.indexOf(abstractNumberingList);
        else {
            index = this.data.documentModel.abstractNumberingLists.length;
            this.data.documentModel.abstractNumberingLists.push(this.list);
        }
        this.info.abstractNumberingListId = this.id;
        this.info.abstractNumberingIndex = index;
        this.data.stylesImporter.addAbstractListInfo(this.info);
    };
    AbstractNumberingListDestination.handlerTable = new map_creator_1.MapCreator()
        .add('lvl', AbstractNumberingListDestination.onLevel)
        .add('multiLevelType', AbstractNumberingListDestination.onMultilevelType)
        .add('name', AbstractNumberingListDestination.onName)
        .add('nsid', AbstractNumberingListDestination.onUniqueId)
        .add('numStyleLink', AbstractNumberingListDestination.onNumberingStyleLink)
        .add('styleLink', AbstractNumberingListDestination.onStyleLink)
        .add('tmpl', AbstractNumberingListDestination.onTemplate)
        .get();
    return AbstractNumberingListDestination;
}(destination_1.ElementDestination));
exports.AbstractNumberingListDestination = AbstractNumberingListDestination;
var NumberingListDestination = (function (_super) {
    tslib_1.__extends(NumberingListDestination, _super);
    function NumberingListDestination(data) {
        var _this = _super.call(this, data) || this;
        _this.listInfo = new open_xml_numbering_list_info_1.OpenXmlNumberingListInfo();
        return _this;
    }
    Object.defineProperty(NumberingListDestination.prototype, "elementHandlerTable", {
        get: function () {
            return NumberingListDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    NumberingListDestination.getThis = function (data) {
        return data.destinationStack.getThis();
    };
    NumberingListDestination.onAbstractNumberingId = function (data, _reader) {
        return new AbstractNumberingListReferenceDestination(data, NumberingListDestination.getThis(data).listInfo);
    };
    NumberingListDestination.onLevelOverride = function (data, _reader) {
        return new ListLevelOverrideDestination(data, NumberingListDestination.getThis(data).listInfo);
    };
    NumberingListDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.listInfo.id = this.data.readerHelper.getWpSTIntegerValue(reader, 'numId', Number.MIN_VALUE);
                return [2];
            });
        });
    };
    NumberingListDestination.prototype.processElementClose = function (_reader) {
        if (string_1.StringUtils.isNullOrEmpty(this.listInfo.abstractNumberingListId) || this.listInfo.id == Number.MIN_VALUE)
            return;
        this.data.stylesImporter.addListInfo(this.listInfo);
    };
    NumberingListDestination.handlerTable = new map_creator_1.MapCreator()
        .add('abstractNumId', NumberingListDestination.onAbstractNumberingId)
        .add('lvlOverride', NumberingListDestination.onLevelOverride)
        .get();
    return NumberingListDestination;
}(destination_1.ElementDestination));
exports.NumberingListDestination = NumberingListDestination;
var AbstractNumberingLeafElementDestination = (function (_super) {
    tslib_1.__extends(AbstractNumberingLeafElementDestination, _super);
    function AbstractNumberingLeafElementDestination(data, list) {
        var _this = _super.call(this, data) || this;
        _this.list = list;
        return _this;
    }
    return AbstractNumberingLeafElementDestination;
}(destination_1.LeafElementDestination));
exports.AbstractNumberingLeafElementDestination = AbstractNumberingLeafElementDestination;
var NumberingListNameDestination = (function (_super) {
    tslib_1.__extends(NumberingListNameDestination, _super);
    function NumberingListNameDestination(data, list) {
        return _super.call(this, data, list) || this;
    }
    NumberingListNameDestination.prototype.processElementOpen = function (_reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2];
            });
        });
    };
    return NumberingListNameDestination;
}(AbstractNumberingLeafElementDestination));
exports.NumberingListNameDestination = NumberingListNameDestination;
var AbstractNumberingListUniqueIdDestination = (function (_super) {
    tslib_1.__extends(AbstractNumberingListUniqueIdDestination, _super);
    function AbstractNumberingListUniqueIdDestination(data, list) {
        return _super.call(this, data, list) || this;
    }
    AbstractNumberingListUniqueIdDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var value;
            return tslib_1.__generator(this, function (_a) {
                value = this.data.readerHelper.getWpSTIntegerValue(reader, 'val', Number.MIN_VALUE, 16);
                if (value != Number.MIN_VALUE) {
                    this.list.innerId = Math.abs(value);
                }
                return [2];
            });
        });
    };
    return AbstractNumberingListUniqueIdDestination;
}(AbstractNumberingLeafElementDestination));
exports.AbstractNumberingListUniqueIdDestination = AbstractNumberingListUniqueIdDestination;
var NumberingListMultiLevelTypeDestination = (function (_super) {
    tslib_1.__extends(NumberingListMultiLevelTypeDestination, _super);
    function NumberingListMultiLevelTypeDestination(data, list) {
        return _super.call(this, data, list) || this;
    }
    NumberingListMultiLevelTypeDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var listType;
            return tslib_1.__generator(this, function (_a) {
                listType = this.data.readerHelper.getWpEnumValue(reader, 'val', translation_tables_1.TranslationTables.numberingListTypeTable.importMap, numbering_list_1.NumberingType.Simple);
                if (listType != numbering_list_1.NumberingType.MultiLevel)
                    this.setHybridListType();
                return [2];
            });
        });
    };
    NumberingListMultiLevelTypeDestination.prototype.setHybridListType = function () {
        var _this = this;
        var count = this.list.levels.length;
        for (var i = 0; i < count; i++) {
            var listLevel = this.list.levels[i];
            listLevel.changeListLevelProperties(function (properties) {
                if (properties.templateCode === 0)
                    properties.templateCode = numbering_helper_1.NumberingHelper.generateNewTemplateCode(_this.data.documentModel);
            });
        }
    };
    return NumberingListMultiLevelTypeDestination;
}(AbstractNumberingLeafElementDestination));
exports.NumberingListMultiLevelTypeDestination = NumberingListMultiLevelTypeDestination;
var NumberingListNumStyleLinkDestination = (function (_super) {
    tslib_1.__extends(NumberingListNumStyleLinkDestination, _super);
    function NumberingListNumStyleLinkDestination(data, info) {
        var _this = _super.call(this, data) || this;
        _this.info = info;
        return _this;
    }
    NumberingListNumStyleLinkDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var styleId;
            return tslib_1.__generator(this, function (_a) {
                styleId = reader.getAttributeNS('val', this.data.constants.wordProcessingNamespaceConst);
                if (styleId)
                    this.info.numberingStyleReferenceIndex = styleId;
                return [2];
            });
        });
    };
    return NumberingListNumStyleLinkDestination;
}(destination_1.LeafElementDestination));
exports.NumberingListNumStyleLinkDestination = NumberingListNumStyleLinkDestination;
var AbstractNumberingListReferenceDestination = (function (_super) {
    tslib_1.__extends(AbstractNumberingListReferenceDestination, _super);
    function AbstractNumberingListReferenceDestination(data, listInfo) {
        var _this = _super.call(this, data) || this;
        _this.listInfo = listInfo;
        return _this;
    }
    AbstractNumberingListReferenceDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.listInfo.abstractNumberingListId = reader.getAttributeNS('val', this.data.constants.namespaces[constants_1.DocxNsType.WordProcessing].namespace);
                return [2];
            });
        });
    };
    return AbstractNumberingListReferenceDestination;
}(destination_1.LeafElementDestination));
exports.AbstractNumberingListReferenceDestination = AbstractNumberingListReferenceDestination;
var ListLevelOverrideDestination = (function (_super) {
    tslib_1.__extends(ListLevelOverrideDestination, _super);
    function ListLevelOverrideDestination(data, listInfo) {
        var _this = _super.call(this, data) || this;
        _this.listInfo = listInfo;
        _this.levelOverride = new open_xml_list_level_override_1.OpenXmlListLevelOverride(_this.data.documentModel);
        return _this;
    }
    Object.defineProperty(ListLevelOverrideDestination.prototype, "elementHandlerTable", {
        get: function () {
            return ListLevelOverrideDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    ListLevelOverrideDestination.getThis = function (data) {
        return data.destinationStack.getThis();
    };
    ListLevelOverrideDestination.onLevelOverride = function (data, _reader) {
        var thisObject = ListLevelOverrideDestination.getThis(data);
        thisObject.overrideRead = true;
        return new ListLevelOverrideLevelDestination(data, thisObject.levelOverride);
    };
    ListLevelOverrideDestination.onLevelStartOverride = function (data, _reader) {
        var thisObject = ListLevelOverrideDestination.getThis(data);
        thisObject.overrideRead = true;
        return new ListLevelOverrideStartDestination(data, thisObject.levelOverride);
    };
    ListLevelOverrideDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.levelOverride.levelIndex = this.data.readerHelper.getWpSTIntegerValue(reader, 'ilvl', Number.MIN_VALUE);
                return [2];
            });
        });
    };
    ListLevelOverrideDestination.prototype.processElementClose = function (_reader) {
        if (this.levelOverride.levelIndex < 0 || this.levelOverride.levelIndex >= 9 || !this.overrideRead)
            return;
        this.listInfo.levelOverrides.push(this.levelOverride);
    };
    ListLevelOverrideDestination.handlerTable = new map_creator_1.MapCreator()
        .add('lvl', ListLevelOverrideDestination.onLevelOverride)
        .add('startOverride', ListLevelOverrideDestination.onLevelStartOverride)
        .get();
    return ListLevelOverrideDestination;
}(destination_1.LeafElementDestination));
exports.ListLevelOverrideDestination = ListLevelOverrideDestination;
var NumberingLevelBaseDestination = (function (_super) {
    tslib_1.__extends(NumberingLevelBaseDestination, _super);
    function NumberingLevelBaseDestination(data) {
        return _super.call(this, data) || this;
    }
    Object.defineProperty(NumberingLevelBaseDestination.prototype, "elementHandlerTable", {
        get: function () {
            return NumberingLevelBaseDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NumberingLevelBaseDestination.prototype, "levelIndex", {
        get: function () { return this._levelIndex; },
        set: function (val) { this._levelIndex = val; },
        enumerable: true,
        configurable: true
    });
    ;
    ;
    NumberingLevelBaseDestination.getThis = function (data) {
        return data.destinationStack.getThis();
    };
    NumberingLevelBaseDestination.createElementHandlerTable = function () {
        var result = new map_creator_1.MapCreator()
            .add('isLgl', NumberingLevelBaseDestination.onLegalNumbering)
            .add('lvlJc', NumberingLevelBaseDestination.onTextAlignment)
            .add('lvlRestart', NumberingLevelBaseDestination.onRestart)
            .add('lvlText', NumberingLevelBaseDestination.onText)
            .add('start', NumberingLevelBaseDestination.onStart)
            .add('suff', NumberingLevelBaseDestination.onSuffix)
            .add('numFmt', NumberingLevelBaseDestination.onNumberingFormat)
            .add('pPr', NumberingLevelBaseDestination.onParagraphProperties)
            .add('rPr', NumberingLevelBaseDestination.onRunProperties)
            .add('legacy', NumberingLevelBaseDestination.onLegacy);
        return result;
    };
    NumberingLevelBaseDestination.onTextAlignment = function (data, _reader) {
        return new ListLevelTextAlignmentDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onRestart = function (data, _reader) {
        return new ListLevelRestartDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onText = function (data, _reader) {
        return new ListLevelFormatStringDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onStart = function (data, _reader) {
        return new ListLevelStartDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onSuffix = function (data, _reader) {
        return new ListLevelSuffixDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onNumberingFormat = function (data, _reader) {
        return new ListLevelNumberingFormatDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onParagraphProperties = function (data, _reader) {
        return new ListLevelParagraphPropertiesDestination(data, NumberingLevelBaseDestination.getThis(data).level);
    };
    NumberingLevelBaseDestination.onRunProperties = function (data, _reader) {
        return new ListLevelRunPropertiesDestination(data, NumberingLevelBaseDestination.getThis(data).level);
    };
    NumberingLevelBaseDestination.onLegalNumbering = function (data, _reader) {
        return new ListLevelLegalNumberingDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.onLegacy = function (data, _reader) {
        return new ListLevelLegacyDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelBaseDestination.handlerTable = NumberingLevelBaseDestination.createElementHandlerTable()
        .get();
    return NumberingLevelBaseDestination;
}(destination_1.ElementDestination));
exports.NumberingLevelBaseDestination = NumberingLevelBaseDestination;
var NumberingLevelDestination = (function (_super) {
    tslib_1.__extends(NumberingLevelDestination, _super);
    function NumberingLevelDestination(data, list) {
        var _this = _super.call(this, data) || this;
        _this.list = list;
        return _this;
    }
    Object.defineProperty(NumberingLevelDestination.prototype, "elementHandlerTable", {
        get: function () {
            return NumberingLevelDestination.handlerTable;
        },
        enumerable: true,
        configurable: true
    });
    NumberingLevelDestination.onParagraphStyleReference = function (data, _reader) {
        return new ListLevelParagraphStyleReferenceDestination(data, NumberingLevelBaseDestination.getThis(data));
    };
    NumberingLevelDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.levelIndex = this.data.readerHelper.getWpSTIntegerValue(reader, 'ilvl', 0);
                this.level = this.list.createLevel(this.levelIndex);
                this.levelProperties = this.level.getListLevelProperties().clone();
                this.levelProperties.templateCode = this.data.readerHelper.getWpSTIntegerValue(reader, 'tplc', 0, 16);
                return [2];
            });
        });
    };
    NumberingLevelDestination.prototype.processElementClose = function (_reader) {
        this.level.setListLevelProperties(this.levelProperties);
        if (this.levelIndex < 0 || this.levelIndex >= 9)
            return;
        this.list.levels[this.levelIndex] = this.level;
    };
    NumberingLevelDestination.handlerTable = NumberingLevelBaseDestination.createElementHandlerTable()
        .add('pStyle', NumberingLevelDestination.onParagraphStyleReference)
        .get();
    return NumberingLevelDestination;
}(NumberingLevelBaseDestination));
exports.NumberingLevelDestination = NumberingLevelDestination;
var ListLevelOverrideLevelDestination = (function (_super) {
    tslib_1.__extends(ListLevelOverrideLevelDestination, _super);
    function ListLevelOverrideLevelDestination(data, levelOverride) {
        var _this = _super.call(this, data) || this;
        _this.levelOverride = levelOverride;
        _this.level = levelOverride.getOverrideListLevel();
        _this.levelProperties = _this.level.getListLevelProperties().clone();
        return _this;
    }
    Object.defineProperty(ListLevelOverrideLevelDestination.prototype, "levelIndex", {
        get: function () {
            return this.levelOverride.levelIndex;
        },
        enumerable: true,
        configurable: true
    });
    ListLevelOverrideLevelDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                _super.prototype.processElementOpen.call(this, reader);
                this.level.changeListLevelProperties(function (prop) { return prop.templateCode = _this.data.readerHelper.getWpSTIntegerValue(reader, 'tplc', 0, 16); });
                return [2];
            });
        });
    };
    ListLevelOverrideLevelDestination.prototype.processElementClose = function (_reader) {
        this.level.setListLevelProperties(this.levelProperties);
    };
    return ListLevelOverrideLevelDestination;
}(NumberingLevelBaseDestination));
exports.ListLevelOverrideLevelDestination = ListLevelOverrideLevelDestination;
var ListLevelOverrideStartDestination = (function (_super) {
    tslib_1.__extends(ListLevelOverrideStartDestination, _super);
    function ListLevelOverrideStartDestination(data, levelOverride) {
        var _this = _super.call(this, data) || this;
        _this.levelOverride = levelOverride;
        return _this;
    }
    ListLevelOverrideStartDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.levelOverride.newStart = this.data.readerHelper.getWpSTIntegerValue(reader, 'val', 0);
                return [2];
            });
        });
    };
    return ListLevelOverrideStartDestination;
}(destination_1.LeafElementDestination));
exports.ListLevelOverrideStartDestination = ListLevelOverrideStartDestination;
var ListLevelElementDestination = (function (_super) {
    tslib_1.__extends(ListLevelElementDestination, _super);
    function ListLevelElementDestination(data, levelDestination) {
        var _this = _super.call(this, data) || this;
        _this.levelDestination = levelDestination;
        return _this;
    }
    Object.defineProperty(ListLevelElementDestination.prototype, "properties", {
        get: function () {
            return this.levelDestination.levelProperties;
        },
        enumerable: true,
        configurable: true
    });
    return ListLevelElementDestination;
}(destination_1.LeafElementDestination));
exports.ListLevelElementDestination = ListLevelElementDestination;
var ListLevelTextAlignmentDestination = (function (_super) {
    tslib_1.__extends(ListLevelTextAlignmentDestination, _super);
    function ListLevelTextAlignmentDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelTextAlignmentDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.properties.alignment = this.data.readerHelper.getWpEnumValue(reader, 'val', translation_tables_1.TranslationTables.listNumberAlignmentTable.importMap, list_level_properties_1.ListNumberAlignment.Left);
                return [2];
            });
        });
    };
    return ListLevelTextAlignmentDestination;
}(ListLevelElementDestination));
exports.ListLevelTextAlignmentDestination = ListLevelTextAlignmentDestination;
var ListLevelStartDestination = (function (_super) {
    tslib_1.__extends(ListLevelStartDestination, _super);
    function ListLevelStartDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelStartDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.properties.start = Math.max(0, this.data.readerHelper.getWpSTIntegerValue(reader, 'val'));
                return [2];
            });
        });
    };
    return ListLevelStartDestination;
}(ListLevelElementDestination));
exports.ListLevelStartDestination = ListLevelStartDestination;
var ListLevelLegacyDestination = (function (_super) {
    tslib_1.__extends(ListLevelLegacyDestination, _super);
    function ListLevelLegacyDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelLegacyDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.properties.legacy = true;
                this.properties.legacySpace = this.data.readerHelper.getWpSTIntegerValue(reader, 'legacySpace', 0);
                this.properties.legacyIndent = this.data.readerHelper.getWpSTIntegerValue(reader, 'legacyIndent', 0);
                return [2];
            });
        });
    };
    return ListLevelLegacyDestination;
}(ListLevelElementDestination));
exports.ListLevelLegacyDestination = ListLevelLegacyDestination;
var ListLevelFormatStringDestination = (function (_super) {
    tslib_1.__extends(ListLevelFormatStringDestination, _super);
    function ListLevelFormatStringDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelFormatStringDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var formatString;
            return tslib_1.__generator(this, function (_a) {
                formatString = reader.getAttributeNS('val', this.data.constants.namespaces[constants_1.DocxNsType.WordProcessing].namespace);
                if (formatString == null)
                    formatString = '';
                if (!control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingBulleted))
                    formatString = this.replaceStringFormatInBulletNumbering(formatString);
                if (!control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingSimple) &&
                    control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingBulleted)) {
                    formatString = this.replaceStringFormatInSimpleNumbering();
                    this.levelDestination.level.getCharacterProperties().fontInfo.name = 'Symbol';
                }
                this.properties.displayFormatString = this.convertFormatString(formatString);
                return [2];
            });
        });
    };
    ListLevelFormatStringDestination.prototype.replaceStringFormatInBulletNumbering = function (value) {
        var result = this.convertFormatString(value);
        if (result == value)
            return '%1.';
        return value;
    };
    ListLevelFormatStringDestination.prototype.replaceStringFormatInSimpleNumbering = function () {
        return '·';
    };
    ListLevelFormatStringDestination.prototype.convertFormatString = function (value) {
        var regex = new RegExp(/%\d/);
        var result = value.replace('{', '{{').replace('}', '}}');
        var match = result.match(regex);
        if (match) {
            do {
                var matchString = match[0];
                var numberIndex = matchString.search(/\d/);
                var numb = parseInt(matchString.substr(numberIndex, 1), 10);
                result = result.replace(regex, '{' + (numb - 1) + '}');
                match = result.match(regex);
            } while (match);
        }
        return result;
    };
    return ListLevelFormatStringDestination;
}(ListLevelElementDestination));
exports.ListLevelFormatStringDestination = ListLevelFormatStringDestination;
var ListLevelRestartDestination = (function (_super) {
    tslib_1.__extends(ListLevelRestartDestination, _super);
    function ListLevelRestartDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelRestartDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var index;
            return tslib_1.__generator(this, function (_a) {
                index = Math.min(9, Math.max(0, this.data.readerHelper.getWpSTIntegerValue(reader, 'val')));
                this.properties.relativeRestartLevel = this.levelDestination.levelIndex - index;
                return [2];
            });
        });
    };
    return ListLevelRestartDestination;
}(ListLevelElementDestination));
exports.ListLevelRestartDestination = ListLevelRestartDestination;
var ListLevelSuffixDestination = (function (_super) {
    tslib_1.__extends(ListLevelSuffixDestination, _super);
    function ListLevelSuffixDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelSuffixDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.properties.separator = String.fromCharCode(this.data.readerHelper.getWpEnumValue(reader, 'val', translation_tables_1.TranslationTables.listNumberSeparatorTable.importMap, rich_utils_1.RichUtils.specialCharacters.TabMark.charCodeAt(0)));
                return [2];
            });
        });
    };
    return ListLevelSuffixDestination;
}(ListLevelElementDestination));
exports.ListLevelSuffixDestination = ListLevelSuffixDestination;
var ListLevelNumberingFormatDestination = (function (_super) {
    tslib_1.__extends(ListLevelNumberingFormatDestination, _super);
    function ListLevelNumberingFormatDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelNumberingFormatDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var format;
            return tslib_1.__generator(this, function (_a) {
                format = this.data.readerHelper.getWpEnumValue(reader, 'val', translation_tables_1.TranslationTables.pageNumberingFormatTable.importMap, list_level_properties_1.NumberingFormat.Decimal);
                if (format == list_level_properties_1.NumberingFormat.Bullet && !control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingBulleted) &&
                    control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingSimple))
                    format = list_level_properties_1.NumberingFormat.Decimal;
                else if (format == list_level_properties_1.NumberingFormat.Decimal && !control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingSimple) &&
                    control_1.ControlOptions.isEnabled(this.data.documentModel.options.numberingBulleted))
                    format = list_level_properties_1.NumberingFormat.Bullet;
                this.properties.format = format;
                return [2];
            });
        });
    };
    return ListLevelNumberingFormatDestination;
}(ListLevelElementDestination));
exports.ListLevelNumberingFormatDestination = ListLevelNumberingFormatDestination;
var ListLevelLegalNumberingDestination = (function (_super) {
    tslib_1.__extends(ListLevelLegalNumberingDestination, _super);
    function ListLevelLegalNumberingDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelLegalNumberingDestination.prototype.processElementOpen = function (reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.properties.convertPreviousLevelNumberingToDecimal = this.data.readerHelper.getWpSTOnOffValue(reader, 'val');
                return [2];
            });
        });
    };
    return ListLevelLegalNumberingDestination;
}(ListLevelElementDestination));
exports.ListLevelLegalNumberingDestination = ListLevelLegalNumberingDestination;
var ListLevelParagraphStyleReferenceDestination = (function (_super) {
    tslib_1.__extends(ListLevelParagraphStyleReferenceDestination, _super);
    function ListLevelParagraphStyleReferenceDestination(data, levelDestination) {
        return _super.call(this, data, levelDestination) || this;
    }
    ListLevelParagraphStyleReferenceDestination.prototype.processElementOpen = function (_reader) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2];
            });
        });
    };
    return ListLevelParagraphStyleReferenceDestination;
}(ListLevelElementDestination));
exports.ListLevelParagraphStyleReferenceDestination = ListLevelParagraphStyleReferenceDestination;
var ListLevelRunPropertiesDestination = (function (_super) {
    tslib_1.__extends(ListLevelRunPropertiesDestination, _super);
    function ListLevelRunPropertiesDestination(data, level) {
        var _this = _super.call(this, data, null) || this;
        _this.level = level;
        _this.characterProperties = character_properties_1.MaskedCharacterProperties.createDefault(_this.data.documentModel);
        return _this;
    }
    ListLevelRunPropertiesDestination.prototype.processElementClose = function (_reader) {
        this.level.setCharacterProperties(this.characterProperties);
    };
    return ListLevelRunPropertiesDestination;
}(run_properties_base_destination_1.RunPropertiesBaseDestination));
exports.ListLevelRunPropertiesDestination = ListLevelRunPropertiesDestination;
var ListLevelParagraphPropertiesDestination = (function (_super) {
    tslib_1.__extends(ListLevelParagraphPropertiesDestination, _super);
    function ListLevelParagraphPropertiesDestination(data, level) {
        var _this = _super.call(this, data, null, new paragraph_style_1.TabProperties()) || this;
        _this.paragraphProperties = paragraph_properties_1.MaskedParagraphProperties.createDefault(_this.data.documentModel);
        _this.level = level;
        return _this;
    }
    Object.defineProperty(ListLevelParagraphPropertiesDestination.prototype, "numberingId", {
        get: function () {
            return -1;
        },
        set: function (_value) { },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListLevelParagraphPropertiesDestination.prototype, "listLevelIndex", {
        get: function () {
            return -1;
        },
        set: function (_value) { },
        enumerable: true,
        configurable: true
    });
    ListLevelParagraphPropertiesDestination.prototype.processElementClose = function (_reader) {
        this.level.setParagraphProperties(this.paragraphProperties);
    };
    return ListLevelParagraphPropertiesDestination;
}(paragraph_properties_base_destination_1.ParagraphPropertiesBaseDestination));
exports.ListLevelParagraphPropertiesDestination = ListLevelParagraphPropertiesDestination;
