// devexpress-richedit/lib/core/model/runs/inline-text-box-run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var properties_bundle_1 = require("../../rich-utils/properties-bundle");
var non_visual_drawing_object_info_1 = require("../manipulators/picture-manipulator/non-visual-drawing-object-info");
var run_base_1 = require("./run-base");
var run_type_1 = require("./run-type");
var InlineTextBoxRun = (function (_super) {
    tslib_1.__extends(InlineTextBoxRun, _super);
    function InlineTextBoxRun(startOffset, paragraph, charPropsBundle, shape, subDocId, size, textBoxProperties, containerProperties) {
        if (containerProperties === void 0) { containerProperties = new non_visual_drawing_object_info_1.NonVisualDrawingObjectInfo(); }
        var _this = _super.call(this, startOffset, paragraph, charPropsBundle, shape, subDocId, textBoxProperties, containerProperties) || this;
        _this.size = size;
        return _this;
    }
    InlineTextBoxRun.prototype.getType = function () {
        return run_type_1.RunType.InlineTextBoxRun;
    };
    InlineTextBoxRun.prototype.clone = function () {
        return new InlineTextBoxRun(this.startOffset, this.paragraph, new properties_bundle_1.MaskedCharacterPropertiesBundle(this.maskedCharacterProperties, this.characterStyle), this.shape.clone(), this.subDocId, this.size.clone(), this.textBoxProperties.clone());
    };
    InlineTextBoxRun.prototype.cloneToNewSubDocument = function (subDocument) {
        return new InlineTextBoxRun(this.startOffset, subDocument.getParagraphByPosition(this.paragraph.startLogPosition.value), new properties_bundle_1.MaskedCharacterPropertiesBundle(subDocument.documentModel.cache.maskedCharacterPropertiesCache.getItem(this.maskedCharacterProperties), subDocument.documentModel.stylesManager.getCharacterStyleByName(this.characterStyle.styleName)), this.shape.clone(), this.subDocId, this.size.clone(), this.textBoxProperties.clone());
    };
    InlineTextBoxRun.prototype.copyFrom = function (obj) {
        _super.prototype.copyFrom.call(this, obj);
        this.size = obj.size.clone();
    };
    return InlineTextBoxRun;
}(run_base_1.TextBoxRun));
exports.InlineTextBoxRun = InlineTextBoxRun;
