// devexpress-richedit/lib/core/model/runs/text-run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var properties_bundle_1 = require("../../rich-utils/properties-bundle");
var run_base_1 = require("./run-base");
var run_type_1 = require("./run-type");
var TextRun = (function (_super) {
    tslib_1.__extends(TextRun, _super);
    function TextRun(startOffset, length, paragraph, charPropsBundle) {
        var _this = _super.call(this, startOffset, paragraph, charPropsBundle) || this;
        _this.length = length;
        return _this;
    }
    TextRun.prototype.getLength = function () {
        return this.length;
    };
    TextRun.prototype.setLength = function (newLength) {
        this.length = newLength;
    };
    TextRun.prototype.getType = function () {
        return run_type_1.RunType.TextRun;
    };
    TextRun.prototype.incLength = function (additionalLength) {
        this.length += additionalLength;
    };
    TextRun.prototype.clone = function () {
        return new TextRun(this.startOffset, this.length, this.paragraph, new properties_bundle_1.MaskedCharacterPropertiesBundle(this.maskedCharacterProperties, this.characterStyle));
    };
    TextRun.prototype.cloneToNewSubDocument = function (subDocument) {
        return new TextRun(this.startOffset, this.length, subDocument.getParagraphByPosition(this.paragraph.startLogPosition.value), new properties_bundle_1.MaskedCharacterPropertiesBundle(subDocument.documentModel.cache.maskedCharacterPropertiesCache.getItem(this.maskedCharacterProperties), subDocument.documentModel.stylesManager.getCharacterStyleByName(this.characterStyle.styleName)));
    };
    TextRun.prototype.copyFrom = function (obj) {
        _super.prototype.copyFrom.call(this, obj);
        this.length = obj.length;
    };
    return TextRun;
}(run_base_1.RunBase));
exports.TextRun = TextRun;
