// devexpress-richedit/lib/client/public/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var docvar_args_1 = require("../../document-processor/docvar-args");
exports.CalculateDocumentVariableEventArgs = docvar_args_1.CalculateDocumentVariableEventArgs;
exports.EventArgs = docvar_args_1.EventArgs;
var Event = (function () {
    function Event() {
        this._handlerInfoList = [];
    }
    Event.prototype.addHandler = function (handler, executionContext) {
        this.removeHandler(handler, executionContext);
        var handlerInfo = new HandlerInfo(handler, executionContext);
        this._handlerInfoList.push(handlerInfo);
    };
    Event.prototype.removeHandler = function (handler, executionContext) {
        for (var i = this._handlerInfoList.length - 1; i >= 0; i--) {
            var handlerInfo = this._handlerInfoList[i];
            if (handlerInfo.handler == handler && (!handlerInfo.executionContext && !executionContext || handlerInfo.executionContext == executionContext))
                this._handlerInfoList.splice(i, 1);
        }
    };
    Event.prototype.clearHandlers = function () {
        this._handlerInfoList = [];
    };
    Event.prototype._fireEvent = function (obj, args) {
        for (var i = 0, handlerInfo = void 0; handlerInfo = this._handlerInfoList[i]; i++)
            handlerInfo.handler.call(handlerInfo.executionContext, obj, args);
    };
    Event.prototype.isEmpty = function () {
        return this._handlerInfoList.length == 0;
    };
    return Event;
}());
exports.Event = Event;
var HandlerInfo = (function () {
    function HandlerInfo(handler, executionContext) {
        this.handler = handler;
        this.executionContext = executionContext;
    }
    return HandlerInfo;
}());
var RichEditEvent = (function (_super) {
    tslib_1.__extends(RichEditEvent, _super);
    function RichEditEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RichEditEvent;
}(Event));
exports.RichEditEvent = RichEditEvent;
var SelectionChangedEvent = (function (_super) {
    tslib_1.__extends(SelectionChangedEvent, _super);
    function SelectionChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SelectionChangedEvent;
}(RichEditEvent));
exports.SelectionChangedEvent = SelectionChangedEvent;
var DocumentLoadedEvent = (function (_super) {
    tslib_1.__extends(DocumentLoadedEvent, _super);
    function DocumentLoadedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DocumentLoadedEvent;
}(RichEditEvent));
exports.DocumentLoadedEvent = DocumentLoadedEvent;
var DocumentChangedEvent = (function (_super) {
    tslib_1.__extends(DocumentChangedEvent, _super);
    function DocumentChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DocumentChangedEvent;
}(RichEditEvent));
exports.DocumentChangedEvent = DocumentChangedEvent;
var DocumentFormattedEvent = (function (_super) {
    tslib_1.__extends(DocumentFormattedEvent, _super);
    function DocumentFormattedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DocumentFormattedEvent;
}(RichEditEvent));
exports.DocumentFormattedEvent = DocumentFormattedEvent;
var ActiveSubDocumentChangedEvent = (function (_super) {
    tslib_1.__extends(ActiveSubDocumentChangedEvent, _super);
    function ActiveSubDocumentChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ActiveSubDocumentChangedEvent;
}(RichEditEvent));
exports.ActiveSubDocumentChangedEvent = ActiveSubDocumentChangedEvent;
var GotFocusEvent = (function (_super) {
    tslib_1.__extends(GotFocusEvent, _super);
    function GotFocusEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return GotFocusEvent;
}(RichEditEvent));
exports.GotFocusEvent = GotFocusEvent;
var LostFocusEvent = (function (_super) {
    tslib_1.__extends(LostFocusEvent, _super);
    function LostFocusEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return LostFocusEvent;
}(RichEditEvent));
exports.LostFocusEvent = LostFocusEvent;
var HyperlinkClickEvent = (function (_super) {
    tslib_1.__extends(HyperlinkClickEvent, _super);
    function HyperlinkClickEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return HyperlinkClickEvent;
}(RichEditEvent));
exports.HyperlinkClickEvent = HyperlinkClickEvent;
var PointerDownEvent = (function (_super) {
    tslib_1.__extends(PointerDownEvent, _super);
    function PointerDownEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PointerDownEvent;
}(RichEditEvent));
exports.PointerDownEvent = PointerDownEvent;
var PointerUpEvent = (function (_super) {
    tslib_1.__extends(PointerUpEvent, _super);
    function PointerUpEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PointerUpEvent;
}(RichEditEvent));
exports.PointerUpEvent = PointerUpEvent;
var KeyDownEvent = (function (_super) {
    tslib_1.__extends(KeyDownEvent, _super);
    function KeyDownEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return KeyDownEvent;
}(RichEditEvent));
exports.KeyDownEvent = KeyDownEvent;
var KeyUpEvent = (function (_super) {
    tslib_1.__extends(KeyUpEvent, _super);
    function KeyUpEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return KeyUpEvent;
}(RichEditEvent));
exports.KeyUpEvent = KeyUpEvent;
var CalculateDocumentVariableEvent = (function (_super) {
    tslib_1.__extends(CalculateDocumentVariableEvent, _super);
    function CalculateDocumentVariableEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CalculateDocumentVariableEvent;
}(RichEditEvent));
exports.CalculateDocumentVariableEvent = CalculateDocumentVariableEvent;
var ContentInsertedEvent = (function (_super) {
    tslib_1.__extends(ContentInsertedEvent, _super);
    function ContentInsertedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ContentInsertedEvent;
}(RichEditEvent));
exports.ContentInsertedEvent = ContentInsertedEvent;
var ContentRemovedEvent = (function (_super) {
    tslib_1.__extends(ContentRemovedEvent, _super);
    function ContentRemovedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ContentRemovedEvent;
}(RichEditEvent));
exports.ContentRemovedEvent = ContentRemovedEvent;
var CharacterPropertiesChangedEvent = (function (_super) {
    tslib_1.__extends(CharacterPropertiesChangedEvent, _super);
    function CharacterPropertiesChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CharacterPropertiesChangedEvent;
}(RichEditEvent));
exports.CharacterPropertiesChangedEvent = CharacterPropertiesChangedEvent;
var ParagraphPropertiesChangedEvent = (function (_super) {
    tslib_1.__extends(ParagraphPropertiesChangedEvent, _super);
    function ParagraphPropertiesChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ParagraphPropertiesChangedEvent;
}(RichEditEvent));
exports.ParagraphPropertiesChangedEvent = ParagraphPropertiesChangedEvent;
var AutoCorrectEvent = (function (_super) {
    tslib_1.__extends(AutoCorrectEvent, _super);
    function AutoCorrectEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AutoCorrectEvent;
}(RichEditEvent));
exports.AutoCorrectEvent = AutoCorrectEvent;
var SavingEvent = (function (_super) {
    tslib_1.__extends(SavingEvent, _super);
    function SavingEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SavingEvent;
}(RichEditEvent));
exports.SavingEvent = SavingEvent;
var SavedEvent = (function (_super) {
    tslib_1.__extends(SavedEvent, _super);
    function SavedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SavedEvent;
}(RichEditEvent));
exports.SavedEvent = SavedEvent;
var CustomCommandExecutedEvent = (function (_super) {
    tslib_1.__extends(CustomCommandExecutedEvent, _super);
    function CustomCommandExecutedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CustomCommandExecutedEvent;
}(RichEditEvent));
exports.CustomCommandExecutedEvent = CustomCommandExecutedEvent;
var PdfExportingEvent = (function (_super) {
    tslib_1.__extends(PdfExportingEvent, _super);
    function PdfExportingEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PdfExportingEvent;
}(RichEditEvent));
exports.PdfExportingEvent = PdfExportingEvent;
var PdfExportedEvent = (function (_super) {
    tslib_1.__extends(PdfExportedEvent, _super);
    function PdfExportedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PdfExportedEvent;
}(RichEditEvent));
exports.PdfExportedEvent = PdfExportedEvent;
var CommandStateChangedEvent = (function (_super) {
    tslib_1.__extends(CommandStateChangedEvent, _super);
    function CommandStateChangedEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CommandStateChangedEvent;
}(RichEditEvent));
exports.CommandStateChangedEvent = CommandStateChangedEvent;
var ContextMenuShowingEvent = (function (_super) {
    tslib_1.__extends(ContextMenuShowingEvent, _super);
    function ContextMenuShowingEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ContextMenuShowingEvent;
}(RichEditEvent));
exports.ContextMenuShowingEvent = ContextMenuShowingEvent;
var CalculateDocumentVariableAsyncEvent = (function (_super) {
    tslib_1.__extends(CalculateDocumentVariableAsyncEvent, _super);
    function CalculateDocumentVariableAsyncEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CalculateDocumentVariableAsyncEvent;
}(RichEditEvent));
exports.CalculateDocumentVariableAsyncEvent = CalculateDocumentVariableAsyncEvent;
var HyperlinkClickEventArgs = (function (_super) {
    tslib_1.__extends(HyperlinkClickEventArgs, _super);
    function HyperlinkClickEventArgs(htmlEvent, handled, targetUri, hyperlinkType, hyperlink) {
        var _this = _super.call(this) || this;
        _this.handled = handled;
        _this.htmlEvent = htmlEvent;
        _this.hyperlinkType = hyperlinkType;
        _this.targetUri = targetUri;
        _this.hyperlink = hyperlink;
        return _this;
    }
    return HyperlinkClickEventArgs;
}(docvar_args_1.EventArgs));
exports.HyperlinkClickEventArgs = HyperlinkClickEventArgs;
var PointerEventArgs = (function (_super) {
    tslib_1.__extends(PointerEventArgs, _super);
    function PointerEventArgs(htmlEvent) {
        var _this = _super.call(this) || this;
        _this.handled = false;
        _this.htmlEvent = htmlEvent;
        return _this;
    }
    return PointerEventArgs;
}(docvar_args_1.EventArgs));
exports.PointerEventArgs = PointerEventArgs;
var KeyboardEventArgs = (function (_super) {
    tslib_1.__extends(KeyboardEventArgs, _super);
    function KeyboardEventArgs(htmlEvent) {
        var _this = _super.call(this) || this;
        _this.handled = false;
        _this.htmlEvent = htmlEvent;
        return _this;
    }
    return KeyboardEventArgs;
}(docvar_args_1.EventArgs));
exports.KeyboardEventArgs = KeyboardEventArgs;
var ContentChangedEventArgs = (function (_super) {
    tslib_1.__extends(ContentChangedEventArgs, _super);
    function ContentChangedEventArgs(subDocumentId, interval) {
        var _this = _super.call(this) || this;
        _this.subDocumentId = subDocumentId;
        _this.interval = interval;
        return _this;
    }
    return ContentChangedEventArgs;
}(docvar_args_1.EventArgs));
exports.ContentChangedEventArgs = ContentChangedEventArgs;
var ContentRemovedEventArgs = (function (_super) {
    tslib_1.__extends(ContentRemovedEventArgs, _super);
    function ContentRemovedEventArgs(subDocumentId, interval, removedText) {
        var _this = _super.call(this, subDocumentId, interval) || this;
        _this.removedText = removedText;
        return _this;
    }
    return ContentRemovedEventArgs;
}(ContentChangedEventArgs));
exports.ContentRemovedEventArgs = ContentRemovedEventArgs;
var ParagraphPropertiesChangedEventArgs = (function (_super) {
    tslib_1.__extends(ParagraphPropertiesChangedEventArgs, _super);
    function ParagraphPropertiesChangedEventArgs(subDocumentId, paragraphIndex) {
        var _this = _super.call(this) || this;
        _this.subDocumentId = subDocumentId;
        _this.paragraphIndex = paragraphIndex;
        return _this;
    }
    return ParagraphPropertiesChangedEventArgs;
}(docvar_args_1.EventArgs));
exports.ParagraphPropertiesChangedEventArgs = ParagraphPropertiesChangedEventArgs;
var AutoCorrectEventArgs = (function (_super) {
    tslib_1.__extends(AutoCorrectEventArgs, _super);
    function AutoCorrectEventArgs(text, interval) {
        var _this = _super.call(this) || this;
        _this.handled = false;
        _this.text = text;
        _this.interval = interval;
        return _this;
    }
    return AutoCorrectEventArgs;
}(docvar_args_1.EventArgs));
exports.AutoCorrectEventArgs = AutoCorrectEventArgs;
var SavingEventArgs = (function (_super) {
    tslib_1.__extends(SavingEventArgs, _super);
    function SavingEventArgs(base64, fileName, format, reason) {
        var _this = _super.call(this) || this;
        _this.handled = false;
        _this.base64 = base64;
        _this.fileName = fileName;
        _this.format = format;
        _this.reason = reason;
        return _this;
    }
    return SavingEventArgs;
}(docvar_args_1.EventArgs));
exports.SavingEventArgs = SavingEventArgs;
var SavedEventArgs = (function (_super) {
    tslib_1.__extends(SavedEventArgs, _super);
    function SavedEventArgs(success, reason) {
        var _this = _super.call(this) || this;
        _this.success = success;
        _this.reason = reason;
        return _this;
    }
    return SavedEventArgs;
}(docvar_args_1.EventArgs));
exports.SavedEventArgs = SavedEventArgs;
var CustomCommandExecutedEventArgs = (function (_super) {
    tslib_1.__extends(CustomCommandExecutedEventArgs, _super);
    function CustomCommandExecutedEventArgs(commandName, parameter) {
        var _this = _super.call(this) || this;
        _this.commandName = commandName;
        _this.parameter = parameter;
        return _this;
    }
    return CustomCommandExecutedEventArgs;
}(docvar_args_1.EventArgs));
exports.CustomCommandExecutedEventArgs = CustomCommandExecutedEventArgs;
var DocumentFormattedEventArgs = (function (_super) {
    tslib_1.__extends(DocumentFormattedEventArgs, _super);
    function DocumentFormattedEventArgs(pageCount) {
        var _this = _super.call(this) || this;
        _this.pageCount = pageCount;
        return _this;
    }
    return DocumentFormattedEventArgs;
}(docvar_args_1.EventArgs));
exports.DocumentFormattedEventArgs = DocumentFormattedEventArgs;
var PdfExportingEventArgs = (function (_super) {
    tslib_1.__extends(PdfExportingEventArgs, _super);
    function PdfExportingEventArgs(base64, blob, blobStream, handled) {
        var _this = _super.call(this) || this;
        _this.base64 = base64;
        _this.blob = blob;
        _this.blobStream = blobStream;
        _this.handled = handled;
        return _this;
    }
    return PdfExportingEventArgs;
}(docvar_args_1.EventArgs));
exports.PdfExportingEventArgs = PdfExportingEventArgs;
var PdfExportedEventArgs = (function (_super) {
    tslib_1.__extends(PdfExportedEventArgs, _super);
    function PdfExportedEventArgs(success) {
        var _this = _super.call(this) || this;
        _this.success = success;
        return _this;
    }
    return PdfExportedEventArgs;
}(docvar_args_1.EventArgs));
exports.PdfExportedEventArgs = PdfExportedEventArgs;
var CommandStateChangedEventArgs = (function (_super) {
    tslib_1.__extends(CommandStateChangedEventArgs, _super);
    function CommandStateChangedEventArgs(commands) {
        var _this = _super.call(this) || this;
        _this.commands = commands;
        return _this;
    }
    return CommandStateChangedEventArgs;
}(docvar_args_1.EventArgs));
exports.CommandStateChangedEventArgs = CommandStateChangedEventArgs;
var ContextMenuShowingEventArgs = (function (_super) {
    tslib_1.__extends(ContextMenuShowingEventArgs, _super);
    function ContextMenuShowingEventArgs(contextMenu) {
        var _this = _super.call(this) || this;
        _this.contextMenu = contextMenu;
        return _this;
    }
    return ContextMenuShowingEventArgs;
}(docvar_args_1.EventArgs));
exports.ContextMenuShowingEventArgs = ContextMenuShowingEventArgs;
var DocumentLinkType;
(function (DocumentLinkType) {
    DocumentLinkType[DocumentLinkType["Hyperlink"] = 0] = "Hyperlink";
    DocumentLinkType[DocumentLinkType["Bookmark"] = 1] = "Bookmark";
    DocumentLinkType[DocumentLinkType["EmailAddress"] = 2] = "EmailAddress";
    DocumentLinkType[DocumentLinkType["Document"] = 3] = "Document";
})(DocumentLinkType = exports.DocumentLinkType || (exports.DocumentLinkType = {}));
