// devexpress-richedit/lib/client/public/rich-edit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var document_1 = require("../../base-api/document");
var history_1 = require("../../base-api/history");
var key_code_1 = require("../../base-api/key-code");
exports.KeyCode = key_code_1.KeyCode;
var layout_api_1 = require("../../base-api/layout-api");
var loading_panel_1 = require("../../base-api/loading-panel");
exports.LoadingPanel = loading_panel_1.LoadingPanelApi;
var selection_api_1 = require("../../base-api/selection-api");
var sub_document_1 = require("../../base-api/sub-document");
exports.SubDocumentBase = sub_document_1.SubDocumentBaseApi;
var utils_1 = require("../../base-utils/utils");
var client_command_1 = require("../../base/commands/client-command");
var assign_shortcut_command_1 = require("../../base/commands/shortcuts/assign-shortcut-command");
var i_rich_edit_core_1 = require("../../base/interfaces/i-rich-edit-core");
var document_format_1 = require("../../core/document-format");
var base64_1 = require("@devexpress/utils/lib/utils/base64");
var common_1 = require("@devexpress/utils/lib/utils/common");
var file_1 = require("@devexpress/utils/lib/utils/file");
var key_1 = require("@devexpress/utils/lib/utils/key");
var mime_type_1 = require("@devexpress/utils/lib/utils/mime-type");
var docvar_args_1 = require("../../document-processor/docvar-args");
var processor_1 = require("../../document-processor/public/processor");
exports.DocumentProcessorBase = processor_1.DocumentProcessorBaseApi;
var field_1 = require("../../model-api/field");
exports.FieldName = field_1.FieldNameApi;
var model_font_1 = require("../../model-api/fonts/model-font");
exports.Font = model_font_1.ModelFontApi;
var enum_1 = require("../../model-api/formats/enum");
exports.DocumentFormat = enum_1.DocumentFormatApi;
var exporter_1 = require("../../model-api/formats/exporter");
var range_permission_1 = require("../../model-api/range-permission");
exports.RangePermission = range_permission_1.RangePermissionApi;
var unit_converter_1 = require("../../model-api/unit-converter");
var pdf_1 = require("../../pdf/api/pdf");
var client_rich_edit_1 = require("../client-rich-edit");
var download_document_command_1 = require("../commands/download-document-command");
var export_document_command_1 = require("../commands/export-document-command");
var mail_merge_command_1 = require("../commands/mail-merge-command");
var open_document_command_1 = require("../commands/open-document-command");
var authentication_options_1 = require("./api/authentication-options");
exports.AuthenticationOptions = authentication_options_1.AuthenticationOptionsApi;
var mail_merge_options_1 = require("./api/mail-merge-options");
exports.MailMergeOptions = mail_merge_options_1.MailMergeOptionsApi;
var range_permission_options_1 = require("./api/range-permission-options");
exports.RangePermissionOptions = range_permission_options_1.RangePermissionOptionsApi;
var simple_view_1 = require("./api/simple-view");
exports.Paddings = simple_view_1.Paddings;
var client_events_1 = require("./client-events");
exports.Events = client_events_1.Events;
var commands_1 = require("./commands/commands");
var enum_2 = require("./commands/enum");
exports.ContextMenuCommandId = enum_2.ContextMenuCommandId;
exports.FileTabCommandId = enum_2.FileTabCommandId;
exports.FloatingObjectsFormatTabCommandId = enum_2.FloatingObjectsFormatTabCommandId;
exports.HeaderAndFooterTabCommandId = enum_2.HeaderAndFooterTabCommandId;
exports.HomeTabCommandId = enum_2.HomeTabCommandId;
exports.InsertTabCommandId = enum_2.InsertTabCommandId;
exports.MailMergeTabCommandId = enum_2.MailMergeTabCommandId;
exports.PageLayoutTabCommandId = enum_2.PageLayoutTabCommandId;
exports.ReferencesTabCommandId = enum_2.ReferencesTabCommandId;
exports.TableDesignTabCommandId = enum_2.TableDesignTabCommandId;
exports.TableLayoutTabCommandId = enum_2.TableLayoutTabCommandId;
exports.ViewTabCommandId = enum_2.ViewTabCommandId;
var menu_1 = require("./context-menu/menu");
var document_processor_1 = require("./document-processor");
exports.DocumentProcessor = document_processor_1.DocumentProcessorApi;
var settings_1 = require("./nusa/settings");
var utils_2 = require("./utils");
var RichEditPublic = (function () {
    function RichEditPublic(htmlElement, options) {
        var _this = this;
        this._native = new client_rich_edit_1.ClientRichEdit(htmlElement, options, this);
        this.document = new document_1.RichEditDocumentBaseApi(this._native.core);
        this.selection = new selection_api_1.RichEditSelectionApi(this._native.core);
        this.layout = new layout_api_1.RichEditLayoutApi(this._native.core);
        this.history = new history_1.HistoryApi(this._native.core);
        this.unitConverter = new unit_converter_1.UnitConverterApi();
        this.mailMergeOptions = new mail_merge_options_1.MailMergeOptionsApi(this._native);
        this.authenticationOptions = new authentication_options_1.AuthenticationOptionsApi(this._native);
        this.simpleViewSettings = new simple_view_1.SimpleViewSettings(this._native);
        this.contextMenu = new menu_1.ContextMenuRuntime(this._native);
        this.rangePermissionOptions = new range_permission_options_1.RangePermissionOptionsApi(this._native);
        this.nusaSettings = new settings_1.NusaSettings(this._native);
        this._native.allowDocumentLoadedEventCall = false;
        var initialDoc = this._native.document;
        var afterDocumentOpened = function (openDocCallback) {
            if (_this.isDisposed)
                return;
            if (_this._native.onInit) {
                var initFunc = common_1.isString(_this._native.onInit) ? utils_1.convertToFunction(_this._native.onInit) : _this._native.onInit;
                initFunc(_this, new docvar_args_1.EventArgs());
            }
            openDocCallback();
            _this._native.allowDocumentLoadedEventCall = true;
            if (!_this._native.core.activeDocumentImporter)
                _this._native.raiseDocumentLoaded();
            _this._native.activeContextTabManager.forbidChangeActiveTabIndex = false;
        };
        if (initialDoc && common_1.isDefined(initialDoc.content)) {
            var onDocumentLoaded_1 = common_1.isDefined(initialDoc.onLoaded) ?
                (common_1.isString(initialDoc.onLoaded) ? utils_1.convertToFunction(initialDoc.onLoaded) : initialDoc.onLoaded) :
                function () { };
            this.openDocument(initialDoc.content, initialDoc.name, initialDoc.format, function (success) {
                afterDocumentOpened(function () { return onDocumentLoaded_1(success); });
            });
        }
        else {
            this.newDocument();
            afterDocumentOpened(function () { });
        }
    }
    Object.defineProperty(RichEditPublic.prototype, "events", {
        get: function () { return this._native.events; },
        enumerable: true,
        configurable: true
    });
    RichEditPublic.prototype.exportToBase64 = function (callback, documentFormat) {
        var format = common_1.isDefined(documentFormat) ? documentFormat : this._native.core.getExportDocumentFormat();
        exporter_1.exportModelToBase64(this._native.core.modelManager.modelManipulator, format, callback);
    };
    RichEditPublic.prototype.exportToBlob = function (callback, documentFormat) {
        var format = common_1.isDefined(documentFormat) ? documentFormat : this._native.core.getExportDocumentFormat();
        exporter_1.exportModelToBlob(this._native.core.modelManager.modelManipulator, format, callback);
    };
    RichEditPublic.prototype.exportToArrayBuffer = function (callback, documentFormat) {
        this.exportToBlob(function (blob) {
            var reader = new FileReader();
            reader.onloadend = function () { return callback(reader.result); };
            reader.readAsArrayBuffer(blob);
        }, documentFormat);
    };
    RichEditPublic.prototype.exportToFile = function (callback, documentFormat) {
        var format = common_1.isDefined(documentFormat) ? documentFormat : this._native.core.getExportDocumentFormat();
        var mimeType = '';
        switch (format) {
            case document_format_1.DocumentFormat.Rtf:
                mimeType = mime_type_1.RtfMimeType;
                break;
            case document_format_1.DocumentFormat.OpenXml:
                mimeType = mime_type_1.OpenXmlMimeType;
                break;
            case document_format_1.DocumentFormat.PlainText:
                mimeType = mime_type_1.PlainTextMimeType;
                break;
        }
        var fileName = open_document_command_1.OpenDocumentCommand.getFileNameWithoutExtension(this.documentName, format);
        var extension = utils_2.Utils.documentFormatToExtension(format);
        exporter_1.exportModelToBlob(this._native.core.modelManager.modelManipulator, format, function (blob) {
            return callback(file_1.FileUtils.createFile([blob], fileName + extension, { type: mimeType }));
        });
    };
    RichEditPublic.prototype.exportToPdf = function (documentName, options) {
        var _this = this;
        var core = this._native.core;
        pdf_1.pdfExport(core, function (blob, stream) {
            base64_1.Base64Utils.fromBlobAsArrayBuffer(blob, function (base64) {
                var handled = _this._native.raisePdfExporting(base64, blob, stream, false);
                var exportUrl = core.modelManager.richOptions.pdf.exportUrl;
                if (!handled && exportUrl)
                    _this._native.sendExportPdfRequest(base64, _this._native.core.documentInfo.getFileNameForDownload(documentName), exportUrl);
            });
        }, options);
    };
    RichEditPublic.prototype.downloadPdf = function (documentName, options) {
        pdf_1.downloadPdf(this._native.core, this._native.core.documentInfo.getFileNameForDownload(documentName), options);
    };
    RichEditPublic.prototype.adjust = function () {
        this._native.adjustControl();
    };
    Object.defineProperty(RichEditPublic.prototype, "isDocumentImported", {
        get: function () {
            return !this._native.core.activeDocumentImporter;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "fullScreen", {
        get: function () {
            return this._native.isInFullScreenMode;
        },
        set: function (value) {
            var command = this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.FullScreen);
            command.execute(true, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "viewType", {
        get: function () {
            return this._native.core.innerClientProperties.viewsSettings.viewType;
        },
        set: function (type) {
            var command = this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.ChangeViewType);
            command.execute(true, type);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "readOnly", {
        get: function () {
            return this._native.core.readOnly == i_rich_edit_core_1.ReadOnlyMode.Persistent;
        },
        set: function (value) {
            if (this.readOnly != value) {
                this._native.core.readOnly = value ? i_rich_edit_core_1.ReadOnlyMode.Persistent : i_rich_edit_core_1.ReadOnlyMode.None;
                this._native.core.barHolder.updateItemsState();
                this._native.core.horizontalRulerControl.update();
                this._native.core.beginUpdate();
                var currState = this._native.core.innerClientProperties.showHiddenSymbols;
                var cmd = this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.ToggleShowWhitespace);
                cmd.execute(this._native.core.commandManager.isPublicApiCall, !currState);
                cmd.execute(this._native.core.commandManager.isPublicApiCall, currState);
                this._native.core.endUpdate();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "isDisposed", {
        get: function () {
            return !this._native.core;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "showHorizontalRuler", {
        get: function () {
            return this._native.core.horizontalRulerControl.getVisible();
        },
        set: function (value) {
            var command = this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.ToggleShowHorizontalRuler);
            command.execute(true, value);
        },
        enumerable: true,
        configurable: true
    });
    RichEditPublic.prototype.focus = function () {
        this._native.core.focusManager.captureFocus();
    };
    Object.defineProperty(RichEditPublic.prototype, "fileName", {
        get: function () {
            return this._native.core.documentInfo.fileName;
        },
        set: function (name) {
            this._native.core.documentInfo.fileName = name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "documentSaveFormat", {
        get: function () {
            return this._native.core.saveDocumentFormat;
        },
        set: function (format) {
            this._native.core.saveDocumentFormat = format;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "documentName", {
        get: function () {
            return this._native.core.documentInfo.fileName;
        },
        set: function (name) {
            this._native.core.documentInfo.fileName = name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "documentFormat", {
        get: function () {
            return this._native.core.documentInfo.documentFormat;
        },
        set: function (format) {
            this._native.core.documentInfo.documentFormat = format;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "documentExtension", {
        get: function () {
            return utils_2.Utils.documentFormatToExtension(this._native.core.documentInfo.documentFormat);
        },
        set: function (filePath) {
            var docFormat = utils_2.Utils.getDocumentFormat(filePath);
            if (docFormat === null)
                console.warn("Error: unknown extension. richEdit.documentExtension = \"" + filePath + "\"");
            else
                this._native.core.documentInfo.documentFormat = docFormat;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RichEditPublic.prototype, "hasUnsavedChanges", {
        get: function () { return !this._native.documentSaved; },
        set: function (value) { this._native.documentSaved = !value; },
        enumerable: true,
        configurable: true
    });
    RichEditPublic.prototype.newDocument = function () {
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.CreateNewDocumentLocally)
            .execute(true);
    };
    RichEditPublic.prototype.openDocument = function (fileContent, documentName, documentFormat, callback) {
        var fileInfo = common_1.isDefined(fileContent) ?
            new open_document_command_1.FileInfo(callback ? (function (success, _reason) { return callback(success); }) : null, fileContent, documentName, documentFormat) :
            null;
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.OpenDocumentLocally)
            .execute(true, fileInfo);
    };
    RichEditPublic.prototype.saveDocument = function (documentFormat, reason, documentName) {
        var oldReadonly = this._native.core.readOnly;
        this._native.core.readOnly = i_rich_edit_core_1.ReadOnlyMode.None;
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.ExportDocument)
            .execute(true, new export_document_command_1.ExportDocumentCommandOptions(documentFormat, reason, documentName));
        this._native.core.readOnly = oldReadonly;
    };
    RichEditPublic.prototype.mailMerge = function (callback, mergeMode, documentFormat, exportFrom, exportRecordsCount) {
        if (mergeMode === void 0) { mergeMode = MergeMode.NewParagraph; }
        if (documentFormat === void 0) { documentFormat = enum_1.DocumentFormatApi.OpenXml; }
        if (!this._native.rawDataSource) {
            console.warn('No data source');
            return;
        }
        var params = new mail_merge_command_1.MailMergeCommandParameters(callback, mergeMode, documentFormat, exportFrom, exportRecordsCount);
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.MailMergeOnClient)
            .execute(true, params);
    };
    RichEditPublic.prototype.downloadDocument = function (documentFormat, documentName) {
        if (documentFormat === void 0) { documentFormat = enum_1.DocumentFormatApi.OpenXml; }
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.DownloadDocumentLocally)
            .execute(true, new download_document_command_1.DownloadDocumentParameters(documentFormat, documentName));
    };
    RichEditPublic.prototype.printDocument = function (mode) {
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.PrintDocumentOnClient).execute(true, mode);
    };
    RichEditPublic.prototype.assignShortcut = function (shortcut, callback) {
        this._native.core.commandManager.getCommand(client_command_1.RichEditClientCommand.AssignShortcut)
            .execute(true, new assign_shortcut_command_1.AssignShortcutCommandOptions(this._native.core, key_1.KeyUtils.getShortcutCode(shortcut.keyCode, shortcut.ctrl, shortcut.shift, shortcut.alt, shortcut.meta), callback));
    };
    Object.defineProperty(RichEditPublic.prototype, "printMode", {
        get: function () { return this._native.core.modelManager.richOptions.printing.mode; },
        set: function (val) { this._native.core.modelManager.richOptions.printing.mode = val; },
        enumerable: true,
        configurable: true
    });
    RichEditPublic.prototype.updateRibbon = function (callback) {
        var oldRibbon = this._native.lastRibbonSettings;
        var oldRibbonBar = this._native.barHolder.ribbon;
        if (oldRibbonBar)
            oldRibbon.activeTabIndex = oldRibbonBar.getActiveTabIndex();
        callback(oldRibbon);
        this._native.core.beginUpdate();
        this._native.initRibbon(oldRibbon, this._native.core.modelManager.richOptions.fonts);
        this.adjust();
        this._native.core.endUpdate();
        this._native.core.barHolder.updateItemsState();
    };
    RichEditPublic.prototype.beginUpdate = function () {
        this._native.core.beginUpdate();
    };
    RichEditPublic.prototype.endUpdate = function () {
        this._native.core.endUpdate();
    };
    Object.defineProperty(RichEditPublic.prototype, "loadingPanel", {
        get: function () {
            return new loading_panel_1.LoadingPanelApi(this._native.core);
        },
        enumerable: true,
        configurable: true
    });
    RichEditPublic.prototype.createDocumentProcessor = function (options) {
        if (options === void 0) { options = {}; }
        var modelManager = this._native.core.modelManager;
        var procOptions = modelManager.richOptions.clone();
        procOptions.spellChecker.isEnabled = false;
        var proc = new document_processor_1.DocumentProcessorApi(procOptions);
        if (options.cloneCurrentModel)
            proc._importInnerDocument(modelManager.model.clone());
        else
            proc._initByEmptyModel();
        return proc;
    };
    RichEditPublic.prototype.dispose = function () {
        this._native.dispose();
        this.nusaSettings.unregister();
    };
    RichEditPublic.prototype.executeCommand = function (commandId, parameter) {
        return commands_1.executeApiCommand(this._native.core.commandManager, commandId, parameter);
    };
    RichEditPublic.prototype.getCommandState = function (commandId) {
        var _a;
        return (_a = commands_1.getApiCommandState(this._native.core.commandManager, commandId)) !== null && _a !== void 0 ? _a : new commands_1.CommandState(false, false);
    };
    RichEditPublic.prototype.setCommandEnabled = function (command, enabled) {
        var coreCommandId = command;
        this._native.core.commandManager.setCommandEnabled(coreCommandId, enabled);
        this._native.core.barHolder.forceUpdate();
    };
    return RichEditPublic;
}());
exports.RichEdit = RichEditPublic;
var ViewType;
(function (ViewType) {
    ViewType[ViewType["Simple"] = 0] = "Simple";
    ViewType[ViewType["PrintLayout"] = 1] = "PrintLayout";
})(ViewType = exports.ViewType || (exports.ViewType = {}));
var MergeMode;
(function (MergeMode) {
    MergeMode[MergeMode["NewParagraph"] = 0] = "NewParagraph";
    MergeMode[MergeMode["NewSection"] = 1] = "NewSection";
})(MergeMode = exports.MergeMode || (exports.MergeMode = {}));
var PrintModeApi;
(function (PrintModeApi) {
    PrintModeApi[PrintModeApi["Html"] = 1] = "Html";
    PrintModeApi[PrintModeApi["Pdf"] = 2] = "Pdf";
})(PrintModeApi || (PrintModeApi = {}));
exports.PrintMode = PrintModeApi;
var character_properties_1 = require("../../model-api/character-properties");
exports.CharacterProperties = character_properties_1.CharacterPropertiesApi;
exports.CharacterPropertiesScript = character_properties_1.CharacterPropertiesScriptApi;
var field_2 = require("../../model-api/field");
exports.HyperlinkInfo = field_2.HyperlinkInfoApi;
var interval_1 = require("../../model-api/interval");
exports.Interval = interval_1.IntervalApi;
var enums_1 = require("../../model-api/lists/enums");
exports.ListType = enums_1.ListTypeApi;
exports.ListLevelNumberAlignment = enums_1.ListLevelNumberAlignmentApi;
exports.ListLevelFormat = enums_1.ListLevelFormatApi;
var list_level_settings_1 = require("../../model-api/lists/list-level-settings");
exports.ListLevelSettings = list_level_settings_1.ListLevelSettingsApi;
var lists_1 = require("../../model-api/lists/lists");
exports.List = lists_1.ListApi;
var paragraph_1 = require("../../model-api/paragraph");
exports.ParagraphAlignment = paragraph_1.ParagraphAlignmentApi;
exports.ParagraphLineSpacingType = paragraph_1.ParagraphLineSpacingTypeApi;
exports.ParagraphFirstLineIndent = paragraph_1.ParagraphFirstLineIndentApi;
exports.ParagraphProperties = paragraph_1.ParagraphPropertiesApi;
exports.Paragraph = paragraph_1.ParagraphApi;
var size_1 = require("../../model-api/size");
exports.Margins = size_1.MarginsApi;
exports.Size = size_1.SizeApi;
var sub_document_2 = require("../../model-api/sub-document");
exports.SubDocumentType = sub_document_2.SubDocumentTypeApi;
exports.HeaderFooterType = sub_document_2.HeaderFooterTypeApi;
exports.SectionBreakType = sub_document_2.SectionBreakTypeApi;
exports.SubDocument = sub_document_2.SubDocumentApi;
var bookmark_1 = require("../../model-api/bookmark");
exports.Bookmark = bookmark_1.BookmarkApi;
var bookmark_2 = require("../../base-api/bookmark");
exports.BookmarkBase = bookmark_2.BookmarkBaseApi;
var section_1 = require("../../model-api/section");
exports.PaperSize = section_1.PaperSizeApi;
var table_1 = require("../../model-api/table/table");
exports.Table = table_1.TableApi;
var section_2 = require("../../model-api/section");
exports.Section = section_2.SectionApi;
var sub_documents_collection_1 = require("../../base-api/collections/sub-documents-collection");
exports.SubDocumentCollectionBase = sub_documents_collection_1.SubDocumentCollectionBaseApi;
var key_code_2 = require("../../base-api/key-code");
exports.ShortcutOptions = key_code_2.ShortcutOptions;
var document_2 = require("../../model-api/document");
exports.RichEditDocumentBase = document_2.RichEditDocumentApi;
var bookmark_collection_1 = require("../../base-api/collections/bookmark-collection");
exports.BookmarkCollectionBase = bookmark_collection_1.BookmarkCollectionBaseApi;
var field_3 = require("../../model-api/field");
exports.Field = field_3.FieldApi;
var field_collection_1 = require("../../model-api/collections/field-collection");
exports.UpdateFieldsOptions = field_collection_1.UpdateFieldsOptionsApi;
var field_4 = require("../../model-api/field");
exports.Hyperlink = field_4.HyperlinkApi;
var table_cell_1 = require("../../model-api/table/table-cell");
exports.TableCell = table_cell_1.TableCellApi;
var table_row_1 = require("../../model-api/table/table-row");
exports.TableRow = table_row_1.TableRowApi;
var image_enums_1 = require("../../model-api/images/image-enums");
exports.WrapType = image_enums_1.WrapTypeApi;
exports.WrapSide = image_enums_1.WrapSideApi;
exports.FloatingObjectHorizontalAlignment = image_enums_1.FloatingObjectHorizontalAlignmentApi;
exports.FloatingObjectHorizontalAnchorElement = image_enums_1.FloatingObjectHorizontalAnchorElementApi;
exports.FloatingObjectHorizontalPositionType = image_enums_1.FloatingObjectHorizontalPositionTypeApi;
exports.FloatingObjectVerticalAnchorElement = image_enums_1.FloatingObjectVerticalAnchorElementApi;
exports.FloatingObjectVerticalAlignment = image_enums_1.FloatingObjectVerticalAlignmentApi;
exports.FloatingObjectVerticalPositionType = image_enums_1.FloatingObjectVerticalPositionTypeApi;
var image_interfaces_1 = require("../../model-api/images/image-interfaces");
exports.HorizontalAbsolutePosition = image_interfaces_1.HorizontalAbsolutePositionApi;
exports.HorizontalAlignedPosition = image_interfaces_1.HorizontalAlignedPositionApi;
exports.HorizontalRelativePosition = image_interfaces_1.HorizontalRelativePositionApi;
exports.VerticalAbsolutePosition = image_interfaces_1.VerticalAbsolutePositionApi;
exports.VerticalAlignedPosition = image_interfaces_1.VerticalAlignedPositionApi;
exports.VerticalRelativePosition = image_interfaces_1.VerticalRelativePositionApi;
var floating_image_1 = require("../../model-api/images/floating-image");
exports.FloatingImage = floating_image_1.FloatingImageApi;
var inline_image_1 = require("../../model-api/images/inline-image");
exports.InlineImage = inline_image_1.InlineImageApi;
var image_1 = require("../../model-api/images/image");
exports.Image = image_1.ImageApi;
var images_1 = require("../../model-api/images/images");
exports.Images = images_1.ImagesApi;
var image_iterator_1 = require("../../model-api/images/image-iterator");
exports.ImageIterator = image_iterator_1.ImageIteratorApi;
var item_1 = require("./context-menu/item");
exports.ContextMenuItem = item_1.ContextMenuItem;
var utils_3 = require("./utils");
exports.Utils = utils_3.Utils;
var characters_1 = require("./characters");
exports.Characters = characters_1.Characters;
tslib_1.__exportStar(require("./commands/enum"), exports);
