// devexpress-richedit/lib/rtf/import/model/section/general-section-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var VerticalAlignment;
(function (VerticalAlignment) {
    VerticalAlignment[VerticalAlignment["Top"] = 0] = "Top";
    VerticalAlignment[VerticalAlignment["Both"] = 1] = "Both";
    VerticalAlignment[VerticalAlignment["Center"] = 2] = "Center";
    VerticalAlignment[VerticalAlignment["Bottom"] = 3] = "Bottom";
})(VerticalAlignment = exports.VerticalAlignment || (exports.VerticalAlignment = {}));
var RtfGeneralSectionInfo = (function () {
    function RtfGeneralSectionInfo() {
    }
    RtfGeneralSectionInfo.prototype.copyFrom = function (obj) {
        this.onlyAllowEditingOfFormFields = obj.onlyAllowEditingOfFormFields;
        this.firstPagePaperSource = obj.firstPagePaperSource;
        this.otherPagePaperSource = obj.otherPagePaperSource;
        this.textDirection = obj.textDirection;
        this.verticalTextAlignment = obj.verticalTextAlignment;
        this.rightToLeft = obj.rightToLeft;
    };
    return RtfGeneralSectionInfo;
}());
exports.RtfGeneralSectionInfo = RtfGeneralSectionInfo;
