// devexpress-richedit/lib/base/commands/numbering/change-numbering-indent-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var list_level_paragraph_properties_history_items_1 = require("../../../core/model/history/items/list-level-paragraph-properties-history-items");
var numbering_list_history_items_1 = require("../../../core/model/history/items/numbering-list-history-items");
var paragraph_properties_1 = require("../../../core/model/paragraph/paragraph-properties");
var command_states_1 = require("../command-states");
var paragraph_indent_command_base_1 = require("../paragraph-properties/paragraph-indent-command-base");
var ChangeNumberingIndentCommandBase = (function (_super) {
    tslib_1.__extends(ChangeNumberingIndentCommandBase, _super);
    function ChangeNumberingIndentCommandBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChangeNumberingIndentCommandBase.prototype.getState = function (options) {
        if (options === void 0) { options = this.convertToCommandOptions(undefined); }
        return new command_states_1.SimpleCommandState(this.isEnabled(options));
    };
    ChangeNumberingIndentCommandBase.prototype.isEnabled = function (options) {
        return _super.prototype.isEnabled.call(this, options) && options.subDocument.getParagraphByPosition(options.intervalsInfo.position).isInList();
    };
    ChangeNumberingIndentCommandBase.prototype.executeCore = function (_state, options) {
        var paragraphIndices = options.subDocument.getParagraphIndicesByIntervals(this.selection.intervals);
        var firstParagraph = options.subDocument.paragraphs[paragraphIndices[0]];
        if (!firstParagraph.isInList())
            return false;
        this.history.beginTransaction();
        if (this.hasPreviousParagraphsInList(paragraphIndices[0], options.subDocument) || firstParagraph.getListLevelIndex() > 0)
            this.changeListLevelIndices(paragraphIndices, options.subDocument);
        else
            this.changeListLevelIndents(paragraphIndices, options.subDocument);
        this.history.endTransaction();
        return true;
    };
    ChangeNumberingIndentCommandBase.prototype.changeListLevelIndices = function (paragraphIndices, subDocument) {
        var paragraphIndicesLength = paragraphIndices.length;
        for (var i = 0; i < paragraphIndicesLength; i++) {
            var paragraphIndex = paragraphIndices[i];
            var paragraph = subDocument.paragraphs[paragraphIndex];
            var newListLevelIndex = this.getNewListLevelIndex(paragraph);
            if (newListLevelIndex !== paragraph.getListLevelIndex())
                this.history.addAndRedo(new numbering_list_history_items_1.AddParagraphToListHistoryItem(this.modelManipulator, subDocument, paragraphIndex, paragraph.numberingListIndex, newListLevelIndex));
        }
    };
    ChangeNumberingIndentCommandBase.prototype.changeListLevelIndents = function (paragraphIndices, subDocument) {
        var startParagraph = subDocument.paragraphs[paragraphIndices[0]];
        var tabs = this.getTabs(paragraphIndices, subDocument);
        var abstractNumberingListIndex = startParagraph.getAbstractNumberingListIndex();
        var abstractNumberingList = startParagraph.getAbstractNumberingList();
        var firstLevelProperties = abstractNumberingList.levels[0].getParagraphMergedProperties();
        var currentLeftIndent = this.getLeftIndentPosition(firstLevelProperties.leftIndent, firstLevelProperties.firstLineIndent, firstLevelProperties.firstLineIndentType);
        var nextListLevelIndent = this.getNextListLevelIndent(currentLeftIndent, tabs);
        this.assignNewIndent(abstractNumberingListIndex, nextListLevelIndent);
    };
    ChangeNumberingIndentCommandBase.prototype.assignNewIndent = function (abstractNumberingListIndex, nextListLevelIndent) {
        var abstractNumberingList = this.control.modelManager.model.abstractNumberingLists[abstractNumberingListIndex];
        var levels = abstractNumberingList.levels;
        var firstLevelProperties = abstractNumberingList.levels[0].getParagraphMergedProperties();
        var delta = this.calculateLeftIndentDelta(nextListLevelIndent, firstLevelProperties.leftIndent, firstLevelProperties.firstLineIndent, firstLevelProperties.firstLineIndentType);
        var levelCount = levels.length;
        for (var i = 0; i < levelCount; i++) {
            var level = levels[i];
            var levelProperties = level.getParagraphMergedProperties();
            var newLeftIndent = levelProperties.leftIndent + delta;
            if (newLeftIndent >= 0) {
                if (levelProperties.firstLineIndentType == paragraph_properties_1.ParagraphFirstLineIndent.Hanging) {
                    var firstLineLeftIndent = newLeftIndent - levelProperties.firstLineIndent;
                    if (firstLineLeftIndent < 0)
                        newLeftIndent -= firstLineLeftIndent;
                }
                if (i == 0 && levelProperties.leftIndent == newLeftIndent)
                    break;
                this.history.addAndRedo(new list_level_paragraph_properties_history_items_1.ListLevelParagraphLeftIndentHistoryItem(this.modelManipulator, true, abstractNumberingListIndex, i, newLeftIndent, true));
            }
            else if (i == 0)
                break;
        }
    };
    ChangeNumberingIndentCommandBase.prototype.calculateLeftIndentDelta = function (nextListLevelIndent, currentLeftIndent, firstLineIndent, firstLineIndentType) {
        return nextListLevelIndent - this.getLeftIndentPosition(currentLeftIndent, firstLineIndent, firstLineIndentType);
    };
    ChangeNumberingIndentCommandBase.prototype.hasPreviousParagraphsInList = function (paragraphIndex, subDocument) {
        var abstractNumberingListIndex = subDocument.paragraphs[paragraphIndex].getAbstractNumberingListIndex();
        for (var i = paragraphIndex - 1, prevParagraph; prevParagraph = subDocument.paragraphs[i]; i--) {
            if (prevParagraph.getAbstractNumberingListIndex() === abstractNumberingListIndex)
                return true;
        }
        return false;
    };
    ChangeNumberingIndentCommandBase.prototype.getLeftIndentPosition = function (currentLeftIndent, firstLineIndent, firstLineIndentType) {
        return firstLineIndentType === paragraph_properties_1.ParagraphFirstLineIndent.Hanging ? (currentLeftIndent - firstLineIndent) : currentLeftIndent;
    };
    return ChangeNumberingIndentCommandBase;
}(paragraph_indent_command_base_1.ParagraphIndentCommandBase));
exports.ChangeNumberingIndentCommandBase = ChangeNumberingIndentCommandBase;
var IncrementNumberingIndentCommand = (function (_super) {
    tslib_1.__extends(IncrementNumberingIndentCommand, _super);
    function IncrementNumberingIndentCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IncrementNumberingIndentCommand.prototype.getNextListLevelIndent = function (currentLeftIndent, tabs) {
        var nearestRightDefaultTab = this.getNearRightDefaultTab(currentLeftIndent);
        var nearestRightTab = this.getNearRightTab(currentLeftIndent, tabs);
        return (nearestRightDefaultTab < nearestRightTab || nearestRightTab == currentLeftIndent) ? nearestRightDefaultTab : nearestRightTab;
    };
    IncrementNumberingIndentCommand.prototype.getNewListLevelIndex = function (paragraph) {
        return Math.min(7, paragraph.getListLevelIndex() + 1);
    };
    return IncrementNumberingIndentCommand;
}(ChangeNumberingIndentCommandBase));
exports.IncrementNumberingIndentCommand = IncrementNumberingIndentCommand;
var DecrementNumberingIndentCommand = (function (_super) {
    tslib_1.__extends(DecrementNumberingIndentCommand, _super);
    function DecrementNumberingIndentCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DecrementNumberingIndentCommand.prototype.getNextListLevelIndent = function (currentLeftIndent, tabs) {
        var nearestLeftDefaultTab = this.getNearLeftDefaultTab(currentLeftIndent);
        var nearestLeftTab = this.getNearLeftTab(currentLeftIndent, tabs);
        return (nearestLeftDefaultTab > nearestLeftTab || nearestLeftTab == currentLeftIndent) ? nearestLeftDefaultTab : nearestLeftTab;
    };
    DecrementNumberingIndentCommand.prototype.getNewListLevelIndex = function (paragraph) {
        return Math.max(0, paragraph.getListLevelIndex() - 1);
    };
    return DecrementNumberingIndentCommand;
}(ChangeNumberingIndentCommandBase));
exports.DecrementNumberingIndentCommand = DecrementNumberingIndentCommand;
