// devexpress-richedit/lib/core/selection/selected-special-run-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var run_type_1 = require("../model/runs/run-type");
var SelectedSpecialRunInfo = (function () {
    function SelectedSpecialRunInfo() {
        this.noSpecRun();
    }
    SelectedSpecialRunInfo.prototype.noSpecRun = function () {
        this.setSpecRunInfo(null, -1, -1, false, -1);
    };
    SelectedSpecialRunInfo.prototype.setSpecRunInfo = function (parentSubDocument, picturePos, textBoxPos, isAnchoredObject, textBoxInnerSubDocumentId) {
        this.picturePos = picturePos;
        this.textBoxPos = textBoxPos;
        this.parentSubDocument = parentSubDocument;
        this.isAnchoredObject = isAnchoredObject;
        this.textBoxInnerSubDocumentId = textBoxInnerSubDocumentId;
    };
    SelectedSpecialRunInfo.prototype.init = function (intervalsInfo) {
        var activeSubDocument = intervalsInfo.subDocument;
        if (this.textBoxInnerSubDocumentId == activeSubDocument.id)
            return;
        if (!intervalsInfo.multiselection && intervalsInfo.interval.length == 1) {
            var pos = intervalsInfo.interval.start;
            var run = activeSubDocument.getRunByPosition(pos);
            switch (run.getType()) {
                case run_type_1.RunType.AnchoredTextBoxRun:
                    this.setSpecRunInfo(activeSubDocument, -1, pos, true, run.subDocId);
                    break;
                case run_type_1.RunType.AnchoredPictureRun:
                    this.setSpecRunInfo(activeSubDocument, pos, -1, true, -1);
                    break;
                case run_type_1.RunType.InlinePictureRun:
                    this.setSpecRunInfo(activeSubDocument, pos, -1, false, -1);
                    break;
                default: this.noSpecRun();
            }
        }
        else
            this.noSpecRun();
    };
    SelectedSpecialRunInfo.prototype.getTextBoxInnerSubDocumentId = function () {
        return this.textBoxInnerSubDocumentId;
    };
    Object.defineProperty(SelectedSpecialRunInfo.prototype, "isSelectedAnchorObject", {
        get: function () {
            return this.isAnchoredObject;
        },
        enumerable: true,
        configurable: true
    });
    SelectedSpecialRunInfo.prototype.getParentSubDocument = function () {
        return this.parentSubDocument;
    };
    SelectedSpecialRunInfo.prototype.getPosition = function () {
        return this.picturePos > -1 ? this.picturePos : this.textBoxPos;
    };
    SelectedSpecialRunInfo.prototype.getPicturePosition = function () {
        return this.picturePos;
    };
    SelectedSpecialRunInfo.prototype.getTextBoxPosition = function () {
        return this.textBoxPos;
    };
    SelectedSpecialRunInfo.prototype.isPictureSelected = function () {
        return this.picturePos > -1;
    };
    SelectedSpecialRunInfo.prototype.isTextBoxSelected = function () {
        return this.textBoxPos > -1;
    };
    SelectedSpecialRunInfo.prototype.isSelected = function () {
        return this.isPictureSelected() || this.isTextBoxSelected();
    };
    SelectedSpecialRunInfo.prototype.clone = function () {
        var obj = new SelectedSpecialRunInfo();
        obj.picturePos = this.picturePos;
        obj.textBoxPos = this.textBoxPos;
        obj.parentSubDocument = this.parentSubDocument;
        obj.isAnchoredObject = this.isAnchoredObject;
        obj.textBoxInnerSubDocumentId = this.textBoxInnerSubDocumentId;
        return obj;
    };
    return SelectedSpecialRunInfo;
}());
exports.SelectedSpecialRunInfo = SelectedSpecialRunInfo;
