// devexpress-richedit/lib/base/ui/quick-search-panel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var evt_1 = require("@devexpress/utils/lib/utils/evt");
var browser_1 = require("@devexpress/utils/lib/browser");
var dom_1 = require("@devexpress/utils/lib/utils/dom");
var key_1 = require("@devexpress/utils/lib/utils/key");
var client_command_1 = require("../commands/client-command");
var QuickSearchPanel = (function () {
    function QuickSearchPanel(owner, searchManager) {
        this.owner = owner;
        this.searchManager = searchManager;
        this.initialize();
    }
    Object.defineProperty(QuickSearchPanel.prototype, "core", {
        get: function () { return this.owner.core; },
        enumerable: true,
        configurable: true
    });
    QuickSearchPanel.prototype.dispose = function () {
        if (this.isVisible())
            this.onClose(true);
        dom_1.DomUtils.hideNode(this.resultLabel);
        this.owner = null;
        this.searchManager = null;
        clearTimeout(this.timerId);
        dom_1.DomUtils.hideNode(this.searchField);
    };
    QuickSearchPanel.prototype.initialize = function () {
        this.createElements();
        this.timerId = null;
        this.height = dom_1.DomUtils.getClearClientHeight(this.getQuickSearchPanelWrapper());
        this.attachEvents();
        this.initializeFields();
    };
    QuickSearchPanel.prototype.initializeFields = function () {
        if (this.searchManager.whatFind) {
            this.getSearchField().value = this.searchManager.whatFind;
            if (this.searchManager.foundIntervals.length)
                this.updateSearchInfo(-1);
            else
                this.findAll();
        }
    };
    QuickSearchPanel.prototype.createElements = function () { };
    QuickSearchPanel.prototype.attachEvents = function () {
        var searchField = this.getSearchField();
        var prevBtn = this.getPrevButton();
        var nextBtn = this.getNextButton();
        var collapseBtn = this.getCollapseButton();
        var closeBtn = this.getCloseButton();
        var that = this;
        closeBtn.onclick = function () { that.hide(true); };
        collapseBtn.onclick = function () { that.showFindReplaceDialog(); };
        prevBtn.onclick = function () {
            if (that.searchManager.foundIntervals.length)
                that.findPrev();
            else
                that.findAll();
        };
        nextBtn.onclick = function () {
            if (that.searchManager.foundIntervals.length)
                that.findNext();
            else
                that.findAll();
        };
        searchField.onkeydown = function (e) {
            var keyCode = key_1.KeyUtils.getEventKeyCode(e);
            if (keyCode === key_1.KeyCode.Esc) {
                that.hide(true);
            }
            else if (keyCode === key_1.KeyCode.Enter) {
                if (that.searchManager.foundIntervals.length)
                    if (e.shiftKey)
                        that.findPrev();
                    else
                        that.findNext();
                else
                    that.findAll();
                evt_1.EvtUtils.preventEvent(e);
            }
            else {
                var shortcutCode = key_1.KeyUtils.getShortcutCodeByEvent(e);
                if (shortcutCode === QuickSearchPanel.shortcuts.showDialog || shortcutCode === QuickSearchPanel.shortcuts.showPanel) {
                    evt_1.EvtUtils.preventEvent(e);
                    if (shortcutCode === QuickSearchPanel.shortcuts.showDialog)
                        that.showFindReplaceDialog();
                }
            }
        };
        searchField.onkeyup = function (e) {
            clearTimeout(that.timerId);
            var keyCode = key_1.KeyUtils.getEventKeyCode(e);
            if (keyCode === key_1.KeyCode.Enter)
                return;
            if (!that.getSearchField().value) {
                that.clearResult();
                that.searchManager.resetSearch();
                return;
            }
            that.timerId = setTimeout(function () {
                that.findAll();
            }, 700);
        };
    };
    QuickSearchPanel.prototype.findNext = function () {
        var index = this.searchManager.findNextIntervalIndex();
        if (index !== null) {
            this.searchManager.selectIntervalByIndex(index);
            this.searchManager.scrollToIntervalByIndex(index);
            this.updateSearchInfo(index);
        }
    };
    QuickSearchPanel.prototype.findPrev = function () {
        var index = this.searchManager.findPrevIntervalIndex();
        if (index !== null) {
            this.searchManager.selectIntervalByIndex(index);
            this.searchManager.scrollToIntervalByIndex(index);
            this.updateSearchInfo(index);
        }
    };
    QuickSearchPanel.prototype.findAll = function () {
        var whatFind = this.getSearchField().value;
        if (whatFind) {
            this.searchManager.findAll(whatFind, false);
            this.updateSearchInfo(-1);
            var index = this.searchManager.findNextIntervalIndex();
            if (index !== null)
                this.searchManager.scrollToIntervalByIndex(index);
        }
    };
    QuickSearchPanel.prototype.updateSearchInfo = function (index) {
        if (this.searchManager.foundIntervals.length) {
            var resultText = index >= 0 ?
                index + 1 + " " + this.core.stringResources.quickSearchPanel.of + " " + this.searchManager.foundIntervals.length :
                this.core.stringResources.quickSearchPanel.items + ": " + this.searchManager.foundIntervals.length;
            this.getResultLabel().innerHTML = resultText;
            return;
        }
        this.getResultLabel().innerHTML = this.core.stringResources.quickSearchPanel.noMatches;
    };
    QuickSearchPanel.prototype.clearResult = function () {
        this.getResultLabel().innerHTML = "";
    };
    QuickSearchPanel.prototype.getQuickSearchPanelWrapper = function () {
        return this.owner.getChildElement(QuickSearchPanel.wrapperPostfix);
    };
    QuickSearchPanel.prototype.getSearchField = function () {
        if (!this.searchField) {
            var inputElements = this.owner.getChildElement(QuickSearchPanel.wrapperPostfix).getElementsByTagName("INPUT");
            for (var i = 0, element = void 0; element = inputElements[i]; i++)
                if (element.type.toLowerCase() === "text") {
                    this.searchField = element;
                    break;
                }
        }
        return this.searchField;
    };
    QuickSearchPanel.prototype.getPrevButton = function () {
        return this.owner.getChildElement(QuickSearchPanel.prevBtnPostfix);
    };
    QuickSearchPanel.prototype.getNextButton = function () {
        return this.owner.getChildElement(QuickSearchPanel.nextBtnPostfix);
    };
    QuickSearchPanel.prototype.getCollapseButton = function () {
        return this.owner.getChildElement(QuickSearchPanel.collapseBtnPostfix);
    };
    QuickSearchPanel.prototype.getCloseButton = function () {
        return this.owner.getChildElement(QuickSearchPanel.closeBtnPostfix);
    };
    QuickSearchPanel.prototype.getResultLabel = function () {
        if (!this.resultLabel) {
            var inputElement = this.getSearchField();
            var tdContainer = inputElement.parentElement;
            var span = document.createElement("SPAN");
            tdContainer.appendChild(span);
            span.style.lineHeight = tdContainer.offsetHeight + "px";
            this.resultLabel = span;
        }
        return this.resultLabel;
    };
    QuickSearchPanel.prototype.NotifySearchReset = function () {
        this.clearResult();
    };
    QuickSearchPanel.prototype.showFindReplaceDialog = function () {
        this.core.commandManager.getCommand(client_command_1.RichEditClientCommand.Replace).execute(this.core.commandManager.isPublicApiCall);
    };
    QuickSearchPanel.prototype.show = function () {
        if (!this.isVisible() && !this.owner.hasActiveDialog()) {
            var panel = this.getQuickSearchPanelWrapper();
            panel.style.visibility = "visible";
            panel.style.height = this.height + "px";
            this.searchManager.onChanged.add(this);
            this.initializeFields();
        }
        dom_1.DomUtils.setFocus(this.getSearchField());
        this.getSearchField().select();
    };
    QuickSearchPanel.prototype.hide = function (resetSearch) {
        if (this.isVisible()) {
            var panel = this.getQuickSearchPanelWrapper();
            panel.style.height = "0px";
            panel.style.visibility = "hidden";
            this.onClose(resetSearch);
        }
    };
    QuickSearchPanel.prototype.onClose = function (resetSearch) {
        if (resetSearch)
            this.searchManager.resetSearch();
        this.searchManager.onChanged.remove(this);
        this.getSearchField().value = "";
        this.clearResult();
        this.core.focusManager.captureFocus();
    };
    QuickSearchPanel.prototype.isVisible = function () {
        var style = dom_1.DomUtils.getCurrentStyle(this.getQuickSearchPanelWrapper());
        return style.visibility !== "hidden";
    };
    QuickSearchPanel.shortcuts = {
        showPanel: key_1.KeyUtils.parseShortcutString(browser_1.Browser.MacOSPlatform ? "CMD+F" : "CTRL+F"),
        showDialog: key_1.KeyUtils.parseShortcutString(browser_1.Browser.MacOSPlatform ? "CMD+H" : "CTRL+H")
    };
    QuickSearchPanel.containerPostfix = "_QS";
    QuickSearchPanel.wrapperPostfix = "_QSPW";
    QuickSearchPanel.prevBtnPostfix = "_QSPPrevBtn";
    QuickSearchPanel.nextBtnPostfix = "_QSPNextBtn";
    QuickSearchPanel.collapseBtnPostfix = "_QSPCollapseBtn";
    QuickSearchPanel.closeBtnPostfix = "_QSPCloseBtn";
    return QuickSearchPanel;
}());
exports.QuickSearchPanel = QuickSearchPanel;
