// devexpress-richedit/lib/model-api/anchor-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AnchorInfoApi = (function () {
    function AnchorInfoApi() {
    }
    return AnchorInfoApi;
}());
exports.AnchorInfoApi = AnchorInfoApi;
var AnchorObjectTextWrapTypeApi;
(function (AnchorObjectTextWrapTypeApi) {
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["Inline"] = 0] = "Inline";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["TopAndBottom"] = 1] = "TopAndBottom";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["Tight"] = 2] = "Tight";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["Through"] = 3] = "Through";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["Square"] = 4] = "Square";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["BehindText"] = 5] = "BehindText";
    AnchorObjectTextWrapTypeApi[AnchorObjectTextWrapTypeApi["InFrontOfText"] = 6] = "InFrontOfText";
})(AnchorObjectTextWrapTypeApi = exports.AnchorObjectTextWrapTypeApi || (exports.AnchorObjectTextWrapTypeApi = {}));
var AnchorObjectTextWrapSideApi;
(function (AnchorObjectTextWrapSideApi) {
    AnchorObjectTextWrapSideApi[AnchorObjectTextWrapSideApi["Both"] = 0] = "Both";
    AnchorObjectTextWrapSideApi[AnchorObjectTextWrapSideApi["Left"] = 1] = "Left";
    AnchorObjectTextWrapSideApi[AnchorObjectTextWrapSideApi["Right"] = 2] = "Right";
    AnchorObjectTextWrapSideApi[AnchorObjectTextWrapSideApi["Largest"] = 3] = "Largest";
})(AnchorObjectTextWrapSideApi = exports.AnchorObjectTextWrapSideApi || (exports.AnchorObjectTextWrapSideApi = {}));
var AnchorObjectHorizontalPositionAlignmentApi;
(function (AnchorObjectHorizontalPositionAlignmentApi) {
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["None"] = 0] = "None";
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["Left"] = 1] = "Left";
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["Center"] = 2] = "Center";
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["Right"] = 3] = "Right";
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["Inside"] = 4] = "Inside";
    AnchorObjectHorizontalPositionAlignmentApi[AnchorObjectHorizontalPositionAlignmentApi["Outside"] = 5] = "Outside";
})(AnchorObjectHorizontalPositionAlignmentApi = exports.AnchorObjectHorizontalPositionAlignmentApi || (exports.AnchorObjectHorizontalPositionAlignmentApi = {}));
var AnchorObjectHorizontalPositionTypeApi;
(function (AnchorObjectHorizontalPositionTypeApi) {
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["Page"] = 0] = "Page";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["Character"] = 1] = "Character";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["Column"] = 2] = "Column";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["Margin"] = 3] = "Margin";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["LeftMargin"] = 4] = "LeftMargin";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["RightMargin"] = 5] = "RightMargin";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["InsideMargin"] = 6] = "InsideMargin";
    AnchorObjectHorizontalPositionTypeApi[AnchorObjectHorizontalPositionTypeApi["OutsideMargin"] = 7] = "OutsideMargin";
})(AnchorObjectHorizontalPositionTypeApi = exports.AnchorObjectHorizontalPositionTypeApi || (exports.AnchorObjectHorizontalPositionTypeApi = {}));
var AnchorObjectVerticalPositionTypeApi;
(function (AnchorObjectVerticalPositionTypeApi) {
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["Page"] = 0] = "Page";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["Line"] = 1] = "Line";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["Paragraph"] = 2] = "Paragraph";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["Margin"] = 3] = "Margin";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["TopMargin"] = 4] = "TopMargin";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["BottomMargin"] = 5] = "BottomMargin";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["InsideMargin"] = 6] = "InsideMargin";
    AnchorObjectVerticalPositionTypeApi[AnchorObjectVerticalPositionTypeApi["OutsideMargin"] = 7] = "OutsideMargin";
})(AnchorObjectVerticalPositionTypeApi = exports.AnchorObjectVerticalPositionTypeApi || (exports.AnchorObjectVerticalPositionTypeApi = {}));
var AnchorObjectVerticalPositionAlignmentApi;
(function (AnchorObjectVerticalPositionAlignmentApi) {
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["None"] = 0] = "None";
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["Top"] = 1] = "Top";
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["Center"] = 2] = "Center";
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["Bottom"] = 3] = "Bottom";
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["Inside"] = 4] = "Inside";
    AnchorObjectVerticalPositionAlignmentApi[AnchorObjectVerticalPositionAlignmentApi["Outside"] = 5] = "Outside";
})(AnchorObjectVerticalPositionAlignmentApi = exports.AnchorObjectVerticalPositionAlignmentApi || (exports.AnchorObjectVerticalPositionAlignmentApi = {}));
