// devexpress-dashboard/common/internal/_interfaces.js
/**
* DevExpress Dashboard (_interfaces.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _jquery_helpers_1 = require("../../data/_jquery-helpers");
exports.KeyCodes = {
    Esc: 27,
    Delete: 46,
    Z: 90,
    Y: 89,
    S: 83
};
var SingleItemSizeController = (function () {
    function SingleItemSizeController(_element, requestRepaint, itemMargins) {
        if (itemMargins === void 0) { itemMargins = 0; }
        this._element = _element;
        this.requestRepaint = requestRepaint;
        this.itemMargins = itemMargins;
        this.renderImmediately = true;
    }
    SingleItemSizeController.prototype.getWidth = function () {
        return _jquery_helpers_1.getWidth(this._element) - this.itemMargins;
    };
    SingleItemSizeController.prototype.getHeight = function () {
        return _jquery_helpers_1.getHeight(this._element) - this.itemMargins;
    };
    SingleItemSizeController.prototype.setConstraints = function (constraints) {
    };
    return SingleItemSizeController;
}());
exports.SingleItemSizeController = SingleItemSizeController;
var SingleTabItemSizeController = (function () {
    function SingleTabItemSizeController(requestRepaint, width, height) {
        this.requestRepaint = requestRepaint;
        this.width = width;
        this.height = height;
        this.renderImmediately = true;
    }
    SingleTabItemSizeController.prototype.getWidth = function () {
        return this.width();
    };
    SingleTabItemSizeController.prototype.getHeight = function () {
        return this.height();
    };
    SingleTabItemSizeController.prototype.setConstraints = function (constraints) {
    };
    return SingleTabItemSizeController;
}());
exports.SingleTabItemSizeController = SingleTabItemSizeController;
