// devexpress-richedit/lib/core/model/drawing/shape3d-properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var drawing_color_1 = require("./drawing-color");
var drawing_color_model_info_1 = require("./drawing-color-model-info");
var enums_1 = require("./enums");
var shape_bevel3d_properties_1 = require("./shape-bevel3d-properties");
var Shape3DProperties = (function () {
    function Shape3DProperties() {
        this.topBevel = new shape_bevel3d_properties_1.ShapeBevel3DProperties();
        this.bottomBevel = new shape_bevel3d_properties_1.ShapeBevel3DProperties();
        this.contourColor = new drawing_color_1.DrawingColor(drawing_color_model_info_1.DrawingColorModelInfo.empty);
        this.extrusionColor = new drawing_color_1.DrawingColor(drawing_color_model_info_1.DrawingColorModelInfo.empty);
        this.presetMaterial = enums_1.PresetMaterialType.WarmMatte;
    }
    Object.defineProperty(Shape3DProperties.prototype, "isDefault", {
        get: function () {
            return this.contourColor.isEmpty && this.extrusionColor.isEmpty && this.topBevel.isDefault && this.bottomBevel.isDefault &&
                this.presetMaterial == Shape3DProperties.defaultPresetMaterialType && this.extrusionHeight == Shape3DProperties.defaultExtrusionHeight &&
                this.contourWidth == Shape3DProperties.defaultContourWidth && this.shapeDepth == Shape3DProperties.defaultShapeDepth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Shape3DProperties.prototype, "type", {
        get: function () {
            return enums_1.DrawingText3DType.Shape3D;
        },
        enumerable: true,
        configurable: true
    });
    Shape3DProperties.prototype.clone = function () {
        var obj = new Shape3DProperties();
        obj.topBevel = this.topBevel;
        obj.bottomBevel = this.bottomBevel;
        obj.contourColor = this.contourColor.clone();
        obj.extrusionColor = this.extrusionColor.clone();
        obj.presetMaterial = this.presetMaterial;
        obj.extrusionHeight = this.extrusionHeight;
        obj.contourWidth = this.contourWidth;
        obj.shapeDepth = this.shapeDepth;
        return obj;
    };
    Shape3DProperties.defaultExtrusionHeight = 0;
    Shape3DProperties.defaultContourWidth = 0;
    Shape3DProperties.defaultShapeDepth = 0;
    Shape3DProperties.defaultPresetMaterialType = enums_1.PresetMaterialType.WarmMatte;
    return Shape3DProperties;
}());
exports.Shape3DProperties = Shape3DProperties;
