// devexpress-richedit/lib/base/scroll/model-states.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var layout_position_creator_1 = require("../../core/layout-engine/layout-position-creator");
var document_layout_details_level_1 = require("../../core/layout/document-layout-details-level");
var layout_position_1 = require("../../core/layout/layout-position");
var canvas_states_1 = require("./canvas-states");
var RelativePosition;
(function (RelativePosition) {
    RelativePosition[RelativePosition["Top"] = 0] = "Top";
    RelativePosition[RelativePosition["Bottom"] = 1] = "Bottom";
    RelativePosition[RelativePosition["Inside"] = 2] = "Inside";
})(RelativePosition = exports.RelativePosition || (exports.RelativePosition = {}));
var ScrollState = (function () {
    function ScrollState() {
    }
    ScrollState.prototype.byModelPosition = function (selection) {
        return this.byModelPositionCore(selection.activeSubDocument, selection.endOfLine, selection.pageIndex);
    };
    ScrollState.prototype.byModelPositionCore = function (subDocument, endOfLine, pageIndex) {
        return (new ModelStateByModelPosition(subDocument, endOfLine, pageIndex));
    };
    Object.defineProperty(ScrollState.prototype, "byScrollInfo", {
        get: function () {
            return (new ModelStateByPageInfo());
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScrollState.prototype, "nothing", {
        get: function () { return new ModelStateEmpty(); },
        enumerable: true,
        configurable: true
    });
    return ScrollState;
}());
exports.ScrollState = ScrollState;
var ModelStateByModelPosition = (function () {
    function ModelStateByModelPosition(subDocument, endOfLine, pageIndex) {
        this.subDocument = subDocument;
        this.endOfLine = endOfLine;
        this.pageIndex = pageIndex;
    }
    ModelStateByModelPosition.prototype.getModelPosition = function () {
        return this.modelPosition;
    };
    ModelStateByModelPosition.prototype.setModelPosition = function (modelPosition) {
        this.modelPosition = modelPosition;
        return this;
    };
    ModelStateByModelPosition.prototype.useCurrentPosition = function (selection) {
        return this.setModelPosition(selection.anchorPosition);
    };
    ModelStateByModelPosition.prototype.setVerticalOffset = function (getVerticalOffset) {
        this.getVerticalOffset = getVerticalOffset;
        return this;
    };
    ModelStateByModelPosition.prototype.useStdOffset = function () {
        return this.setVerticalOffset(function () { return 0; });
    };
    ModelStateByModelPosition.prototype.setRelativePosition = function (relativePosition) {
        this.relativePosition = relativePosition;
        return this;
    };
    ModelStateByModelPosition.prototype.useStdRelativePosition = function () {
        return this.setRelativePosition(RelativePosition.Inside);
    };
    ModelStateByModelPosition.prototype.getCanvasState = function (layout) {
        var modelPos = this.modelPosition;
        var cursorPos = this.subDocument.isMain()
            ? new layout_position_creator_1.LayoutPositionMainSubDocumentCreator(layout, this.subDocument, modelPos, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(this.endOfLine), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false))
            : new layout_position_creator_1.LayoutPositionOtherSubDocumentCreator(layout, this.subDocument, modelPos, this.pageIndex, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(this.endOfLine), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false));
        return cursorPos ? new canvas_states_1.CanvasState(cursorPos, this.relativePosition, this.getVerticalOffset) : null;
    };
    return ModelStateByModelPosition;
}());
exports.ModelStateByModelPosition = ModelStateByModelPosition;
var ModelStateByPageInfo = (function () {
    function ModelStateByPageInfo() {
    }
    ModelStateByPageInfo.prototype.setPageInfo = function (scrollTopInfo) {
        this.getVerticalOffset = function () { return scrollTopInfo.topPositionRelativePage; };
        this.pageIndex = scrollTopInfo.pageIndex;
        this.relativePosition = RelativePosition.Top;
        return this;
    };
    ModelStateByPageInfo.prototype.getCanvasState = function (layout) {
        return layout.isFullyFormatted || layout.isPageValid(this.pageIndex + 1) ?
            new canvas_states_1.CanvasState(new layout_position_1.LayoutPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Page).initByIndexes(this.pageIndex).applyObjectsAsMainSubDocument(layout, -1), this.relativePosition, this.getVerticalOffset) :
            null;
    };
    return ModelStateByPageInfo;
}());
exports.ModelStateByPageInfo = ModelStateByPageInfo;
var ModelStateEmpty = (function () {
    function ModelStateEmpty() {
    }
    ModelStateEmpty.prototype.getCanvasState = function (_layout) {
        return null;
    };
    return ModelStateEmpty;
}());
exports.ModelStateEmpty = ModelStateEmpty;
