// devexpress-richedit/lib/core/model/tables/properties-mergers/table-properties-merger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var flag_1 = require("@devexpress/utils/lib/class/flag");
var table_properties_1 = require("../properties/table-properties");
var table_base_structures_1 = require("../secondary-structures/table-base-structures");
var table_properties_merger_base_1 = require("./table-properties-merger-base");
var TablePropertiesMerger = (function (_super) {
    tslib_1.__extends(TablePropertiesMerger, _super);
    function TablePropertiesMerger() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMerger.prototype.getContainerFromConditionalStyle = function (condStyle) {
        return condStyle.tableProperties;
    };
    TablePropertiesMerger.prototype.canUseValue = function (props) {
        return !!(props.mask & this.getPropertyMask());
    };
    TablePropertiesMerger.prototype.getCondTableStyleFormattingListForThisContainer = function () {
        return TablePropertiesMerger.conditionalTableStyleFormattingPriority;
    };
    TablePropertiesMerger.prototype.getTableNotMergedProperty = function (container) {
        if (!this.canUseValue(container))
            return new table_properties_merger_base_1.TableMergerNotMergedPropertyResult(false, null);
        return new table_properties_merger_base_1.TableMergerNotMergedPropertyResult(true, this.getPropertyFromContainer(container));
    };
    TablePropertiesMerger.prototype.getNotMergedProperty = function () {
        return this.getTableNotMergedProperty(this.tablePropertiesException);
    };
    TablePropertiesMerger.conditionalTableStyleFormattingPriority = [
        table_base_structures_1.ConditionalTableStyleFormatting.WholeTable,
    ];
    return TablePropertiesMerger;
}(table_properties_merger_base_1.TablePropertiesMergerBase));
exports.TablePropertiesMerger = TablePropertiesMerger;
var TablePropertiesMergerIndent = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerIndent, _super);
    function TablePropertiesMergerIndent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerIndent.prototype.getPropertyFromContainer = function (container) {
        return container.indent;
    };
    TablePropertiesMergerIndent.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTableIndent;
    };
    return TablePropertiesMergerIndent;
}(TablePropertiesMerger));
exports.TablePropertiesMergerIndent = TablePropertiesMergerIndent;
var TablePropertiesMergerCellSpacing = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerCellSpacing, _super);
    function TablePropertiesMergerCellSpacing() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerCellSpacing.prototype.getPropertyFromContainer = function (container) {
        return container.cellSpacing;
    };
    TablePropertiesMergerCellSpacing.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseCellSpacing;
    };
    return TablePropertiesMergerCellSpacing;
}(TablePropertiesMerger));
exports.TablePropertiesMergerCellSpacing = TablePropertiesMergerCellSpacing;
var TablePropertiesMergerBorderBase = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderBase, _super);
    function TablePropertiesMergerBorderBase(isTableOuterBorder) {
        var _this = _super.call(this) || this;
        _this.isTableOuterBorder = isTableOuterBorder;
        return _this;
    }
    TablePropertiesMergerBorderBase.prototype.getPropertiesMaskCore = function (mask) {
        if (this.isTableOuterBorder)
            return mask;
        var flag = new flag_1.Flag(mask);
        if (flag.get(table_properties_1.TablePropertiesMask.UseLeftBorder) || flag.get(table_properties_1.TablePropertiesMask.UseRightBorder))
            flag.set(table_properties_1.TablePropertiesMask.UseInsideVerticalBorder, true);
        if (flag.get(table_properties_1.TablePropertiesMask.UseTopBorder) || flag.get(table_properties_1.TablePropertiesMask.UseBottomBorder))
            flag.set(table_properties_1.TablePropertiesMask.UseInsideHorizontalBorder, true);
        flag.set(table_properties_1.TablePropertiesMask.UseLeftBorder, false);
        flag.set(table_properties_1.TablePropertiesMask.UseRightBorder, false);
        flag.set(table_properties_1.TablePropertiesMask.UseTopBorder, false);
        flag.set(table_properties_1.TablePropertiesMask.UseBottomBorder, false);
        return flag.getValue();
    };
    return TablePropertiesMergerBorderBase;
}(TablePropertiesMerger));
exports.TablePropertiesMergerBorderBase = TablePropertiesMergerBorderBase;
var TablePropertiesMergerBorderLeft = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderLeft, _super);
    function TablePropertiesMergerBorderLeft(isTableOuterBorder) {
        if (isTableOuterBorder === void 0) { isTableOuterBorder = true; }
        return _super.call(this, isTableOuterBorder) || this;
    }
    TablePropertiesMergerBorderLeft.prototype.getPropertyFromContainer = function (container) {
        return container.borders.leftBorder;
    };
    TablePropertiesMergerBorderLeft.prototype.getPropertyMask = function () {
        return _super.prototype.getPropertiesMaskCore.call(this, table_properties_1.TablePropertiesMask.UseLeftBorder);
    };
    return TablePropertiesMergerBorderLeft;
}(TablePropertiesMergerBorderBase));
exports.TablePropertiesMergerBorderLeft = TablePropertiesMergerBorderLeft;
var TablePropertiesMergerBorderRight = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderRight, _super);
    function TablePropertiesMergerBorderRight(isTableOuterBorder) {
        if (isTableOuterBorder === void 0) { isTableOuterBorder = true; }
        return _super.call(this, isTableOuterBorder) || this;
    }
    TablePropertiesMergerBorderRight.prototype.getPropertyFromContainer = function (container) {
        return container.borders.rightBorder;
    };
    TablePropertiesMergerBorderRight.prototype.getPropertyMask = function () {
        return _super.prototype.getPropertiesMaskCore.call(this, table_properties_1.TablePropertiesMask.UseRightBorder);
    };
    return TablePropertiesMergerBorderRight;
}(TablePropertiesMergerBorderBase));
exports.TablePropertiesMergerBorderRight = TablePropertiesMergerBorderRight;
var TablePropertiesMergerBorderTop = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderTop, _super);
    function TablePropertiesMergerBorderTop(isTableOuterBorder) {
        if (isTableOuterBorder === void 0) { isTableOuterBorder = true; }
        return _super.call(this, isTableOuterBorder) || this;
    }
    TablePropertiesMergerBorderTop.prototype.getPropertyFromContainer = function (container) {
        return container.borders.topBorder;
    };
    TablePropertiesMergerBorderTop.prototype.getPropertyMask = function () {
        return _super.prototype.getPropertiesMaskCore.call(this, table_properties_1.TablePropertiesMask.UseTopBorder);
    };
    return TablePropertiesMergerBorderTop;
}(TablePropertiesMergerBorderBase));
exports.TablePropertiesMergerBorderTop = TablePropertiesMergerBorderTop;
var TablePropertiesMergerBorderBottom = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderBottom, _super);
    function TablePropertiesMergerBorderBottom(isTableOuterBorder) {
        if (isTableOuterBorder === void 0) { isTableOuterBorder = true; }
        return _super.call(this, isTableOuterBorder) || this;
    }
    TablePropertiesMergerBorderBottom.prototype.getPropertyFromContainer = function (container) {
        return container.borders.bottomBorder;
    };
    TablePropertiesMergerBorderBottom.prototype.getPropertyMask = function () {
        return _super.prototype.getPropertiesMaskCore.call(this, table_properties_1.TablePropertiesMask.UseBottomBorder);
    };
    return TablePropertiesMergerBorderBottom;
}(TablePropertiesMergerBorderBase));
exports.TablePropertiesMergerBorderBottom = TablePropertiesMergerBorderBottom;
var TablePropertiesMergerBorderVertical = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderVertical, _super);
    function TablePropertiesMergerBorderVertical() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerBorderVertical.prototype.getPropertyFromContainer = function (container) {
        return container.borders.insideVerticalBorder;
    };
    TablePropertiesMergerBorderVertical.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseInsideVerticalBorder;
    };
    return TablePropertiesMergerBorderVertical;
}(TablePropertiesMerger));
exports.TablePropertiesMergerBorderVertical = TablePropertiesMergerBorderVertical;
var TablePropertiesMergerBorderHorizontal = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerBorderHorizontal, _super);
    function TablePropertiesMergerBorderHorizontal() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerBorderHorizontal.prototype.getPropertyFromContainer = function (container) {
        return container.borders.insideHorizontalBorder;
    };
    TablePropertiesMergerBorderHorizontal.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseInsideHorizontalBorder;
    };
    return TablePropertiesMergerBorderHorizontal;
}(TablePropertiesMerger));
exports.TablePropertiesMergerBorderHorizontal = TablePropertiesMergerBorderHorizontal;
var TablePropertiesMergerMarginLeft = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerMarginLeft, _super);
    function TablePropertiesMergerMarginLeft() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerMarginLeft.prototype.getPropertyFromContainer = function (container) {
        return container.cellMargins.left;
    };
    TablePropertiesMergerMarginLeft.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseLeftMargin;
    };
    return TablePropertiesMergerMarginLeft;
}(TablePropertiesMerger));
exports.TablePropertiesMergerMarginLeft = TablePropertiesMergerMarginLeft;
var TablePropertiesMergerMarginRight = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerMarginRight, _super);
    function TablePropertiesMergerMarginRight() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerMarginRight.prototype.getPropertyFromContainer = function (container) {
        return container.cellMargins.right;
    };
    TablePropertiesMergerMarginRight.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseRightMargin;
    };
    return TablePropertiesMergerMarginRight;
}(TablePropertiesMerger));
exports.TablePropertiesMergerMarginRight = TablePropertiesMergerMarginRight;
var TablePropertiesMergerMarginTop = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerMarginTop, _super);
    function TablePropertiesMergerMarginTop() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerMarginTop.prototype.getPropertyFromContainer = function (container) {
        return container.cellMargins.top;
    };
    TablePropertiesMergerMarginTop.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTopMargin;
    };
    return TablePropertiesMergerMarginTop;
}(TablePropertiesMerger));
exports.TablePropertiesMergerMarginTop = TablePropertiesMergerMarginTop;
var TablePropertiesMergerMarginBottom = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerMarginBottom, _super);
    function TablePropertiesMergerMarginBottom() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerMarginBottom.prototype.getPropertyFromContainer = function (container) {
        return container.cellMargins.bottom;
    };
    TablePropertiesMergerMarginBottom.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseBottomMargin;
    };
    return TablePropertiesMergerMarginBottom;
}(TablePropertiesMerger));
exports.TablePropertiesMergerMarginBottom = TablePropertiesMergerMarginBottom;
var TablePropertiesMergerLayoutType = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerLayoutType, _super);
    function TablePropertiesMergerLayoutType() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerLayoutType.prototype.getPropertyFromContainer = function (container) {
        return container.layoutType;
    };
    TablePropertiesMergerLayoutType.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTableLayout;
    };
    return TablePropertiesMergerLayoutType;
}(TablePropertiesMerger));
exports.TablePropertiesMergerLayoutType = TablePropertiesMergerLayoutType;
var TablePropertiesMergerShadingInfo = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerShadingInfo, _super);
    function TablePropertiesMergerShadingInfo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerShadingInfo.prototype.getPropertyFromContainer = function (container) {
        return container.shadingInfo;
    };
    TablePropertiesMergerShadingInfo.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseShadingInfoIndex;
    };
    return TablePropertiesMergerShadingInfo;
}(TablePropertiesMerger));
exports.TablePropertiesMergerShadingInfo = TablePropertiesMergerShadingInfo;
var TablePropertiesMergerStyleColumnBandSize = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerStyleColumnBandSize, _super);
    function TablePropertiesMergerStyleColumnBandSize() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerStyleColumnBandSize.prototype.getPropertyFromContainer = function (container) {
        return container.tableStyleColumnBandSize;
    };
    TablePropertiesMergerStyleColumnBandSize.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTableStyleColBandSize;
    };
    return TablePropertiesMergerStyleColumnBandSize;
}(TablePropertiesMerger));
exports.TablePropertiesMergerStyleColumnBandSize = TablePropertiesMergerStyleColumnBandSize;
var TablePropertiesMergerStyleRowBandSize = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerStyleRowBandSize, _super);
    function TablePropertiesMergerStyleRowBandSize() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerStyleRowBandSize.prototype.getPropertyFromContainer = function (container) {
        return container.tableStyleRowBandSize;
    };
    TablePropertiesMergerStyleRowBandSize.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTableStyleRowBandSize;
    };
    return TablePropertiesMergerStyleRowBandSize;
}(TablePropertiesMerger));
exports.TablePropertiesMergerStyleRowBandSize = TablePropertiesMergerStyleRowBandSize;
var TablePropertiesMergerHorizontalAlignment = (function (_super) {
    tslib_1.__extends(TablePropertiesMergerHorizontalAlignment, _super);
    function TablePropertiesMergerHorizontalAlignment() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TablePropertiesMergerHorizontalAlignment.prototype.getPropertyFromContainer = function (container) {
        return container.tableRowAlignment;
    };
    TablePropertiesMergerHorizontalAlignment.prototype.getPropertyMask = function () {
        return table_properties_1.TablePropertiesMask.UseTableAlignment;
    };
    TablePropertiesMergerHorizontalAlignment.prototype.actionBeforeDefaultValue = function () {
        this.result = table_base_structures_1.TableRowAlignment.Left;
        return true;
    };
    return TablePropertiesMergerHorizontalAlignment;
}(TablePropertiesMerger));
exports.TablePropertiesMergerHorizontalAlignment = TablePropertiesMergerHorizontalAlignment;
