// devexpress-richedit/lib/base/mouse-handler/touch-handler/touch-handler-resize-box-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var resize_box_helper_1 = require("../resize-box-helper");
var touch_handler_state_base_1 = require("./touch-handler-state-base");
var TouchHandlerResizeBoxState = (function (_super) {
    tslib_1.__extends(TouchHandlerResizeBoxState, _super);
    function TouchHandlerResizeBoxState() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TouchHandlerResizeBoxState.prototype.start = function () {
        this.resizeBoxHelper = new resize_box_helper_1.ResizeBoxHelper(this.handler.control, this.handler.boxVisualizerManager.resizeBoxVisualizer);
    };
    TouchHandlerResizeBoxState.prototype.onTouchStart = function (evt) {
        this.resizeBoxHelper.start(evt);
    };
    TouchHandlerResizeBoxState.prototype.onTouchMove = function (evt) {
        this.resizeBoxHelper.move(evt);
        return false;
    };
    TouchHandlerResizeBoxState.prototype.onTouchEnd = function (evt) {
        this.resizeBoxHelper.end(evt);
        this.handler.switchToDefaultState();
    };
    return TouchHandlerResizeBoxState;
}(touch_handler_state_base_1.TouchHandlerStateBase));
exports.TouchHandlerResizeBoxState = TouchHandlerResizeBoxState;
