// devexpress-dashboard/designer/extensions/binding-panel.js
/**
* DevExpress Dashboard (binding-panel.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var control_options_1 = require("../../common/control-options");
var _options_manager_1 = require("../../common/internal/_options-manager");
var _jquery_helpers_1 = require("../../data/_jquery-helpers");
var data_dashboard_item_1 = require("../../model/items/data-dashboard-item");
var _base_item_surface_1 = require("../items/surfaces/_base-item-surface");
var _dashboard_item_menu_1 = require("../items/_dashboard-item-menu");
var name = 'item-binding-panel';
var nameAlias = 'itemBindingPanel';
var BindingPanelExtension = (function () {
    function BindingPanelExtension(dashboardControl, options) {
        if (options === void 0) { options = {}; }
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._subscriptions = [];
        this._optionsManager = new _options_manager_1.OptionsManager();
        this._optionsManager.initialize({
            alias: nameAlias,
            dashboardControl: dashboardControl,
            defaultOptions: {},
            eventsHolder: this,
            initOptions: options,
            optionChanged: function (args) { return null; }
        });
    }
    BindingPanelExtension.prototype._contextMenuSubscriber = function (dashboardItemMenu) {
        if (!!dashboardItemMenu) {
            if (this.dashboardControl._actualLayoutController().selectedDashboardItem() instanceof data_dashboard_item_1.DataDashboardItem) {
                this._updateDashboardItemMenu(dashboardItemMenu);
            }
        }
    };
    BindingPanelExtension.prototype._updateEmptyItemTemplate = function (dashboardLayout) {
        var _this = this;
        if (dashboardLayout && dashboardLayout.emptyItemTemplates) {
            dashboardLayout.emptyItemTemplates.push({
                name: 'dx-dashboard-binding-properties-empty-item',
                data: {
                    isDesignMode: this.dashboardControl.isDesignMode,
                    click: function () {
                        var contextMenuExtension = _this.dashboardControl.findExtension('item-menu');
                        if (contextMenuExtension) {
                            contextMenuExtension.menuItemClick(_this.name);
                        }
                    }
                }
            });
        }
    };
    BindingPanelExtension.prototype._updateDashboardItemMenu = function (menu) {
        var _this = this;
        if (menu instanceof _dashboard_item_menu_1.DashboardItemMenu && menu.selectedItemSurface instanceof _base_item_surface_1.DataDashboardItemSurface) {
            var surface = menu.selectedItemSurface;
            if (surface && surface.dataSections && surface.dataSections().length > 0) {
                surface.customizeDataItemContainerTabs = function (args) {
                    _this._optionsManager.raiseEvent('customizeDataItemContainerSections', args);
                };
                menu.contextMenuItems.push({
                    menuItemId: this.name,
                    icon: 'dx-dashboard-item-options',
                    hint: 'DashboardWebStringId.Bindings',
                    popoverClass: 'dx-dashboard-binding-panel',
                    panelWidth: _dashboard_item_menu_1.DashboardItemMenuSizes.BindingPanelPanelWidth,
                    templateName: 'dx-dashboard-binding-properties',
                    detailVisible: ko.observable(false),
                    customData: surface,
                    index: 100
                });
            }
        }
    };
    BindingPanelExtension.prototype.start = function () {
        var _this = this;
        this._updateEmptyItemTemplate(this.dashboardControl._actualLayoutController());
        this.dashboardControl._actualLayoutController.subscribe(function (layoutController) { return _this._updateEmptyItemTemplate(layoutController); });
        var contextMenuExtension = this.dashboardControl.findExtension('item-menu');
        if (contextMenuExtension) {
            this._subscriptions.push(contextMenuExtension._itemContextMenu.subscribe(this._contextMenuSubscriber, this));
            this._contextMenuSubscriber(contextMenuExtension._itemContextMenu());
        }
    };
    BindingPanelExtension.prototype.stop = function () {
        this._subscriptions.forEach(function (s) { return s.dispose(); });
        this._subscriptions = [];
    };
    return BindingPanelExtension;
}());
exports.BindingPanelExtension = BindingPanelExtension;
ko.bindingHandlers['parentHeightCalculator'] = {
    init: function (element, valueAccessor, allBindings, viewModel, bindingContext) {
    },
    update: function (element, valueAccessor, allBindings, viewModel, bindingContext) {
        viewModel.parentHeight = _jquery_helpers_1.getHeight(_jquery_helpers_1.closest(element, valueAccessor().selector));
    }
};
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new BindingPanelExtension(dashboardControl, options); };
control_options_1.extensionNameMap[name] = nameAlias;
