// devexpress-dashboard/designer/extensions/filter-panel-extension.js
/**
* DevExpress Dashboard (filter-panel-extension.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var control_options_1 = require("../../common/control-options");
var model_1 = require("../../model");
var _dashboard_item_menu_1 = require("../items/_dashboard-item-menu");
var _item_filter_properties_composer_1 = require("../items/_item-filter-properties-composer");
var name = 'item-filter-panel';
var nameAlias = 'itemFilterPanel';
var FilterPanelExtension = (function () {
    function FilterPanelExtension(dashboardControl) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._subscriptions = [];
    }
    FilterPanelExtension.prototype._contextMenuSubscriber = function (itemContextMenu) {
        if (!!itemContextMenu) {
            var item = this.dashboardControl._actualLayoutController().selectedDashboardItem();
            this._updateContextMenu(itemContextMenu, item);
        }
    };
    FilterPanelExtension.prototype._updateContextMenu = function (menu, item) {
        var _this = this;
        if (menu instanceof _dashboard_item_menu_1.DashboardItemMenu && item instanceof model_1.DataDashboardItem) {
            menu.contextMenuItems.push({
                menuItemId: this.name,
                icon: 'dx-dashboard-item-filter',
                title: 'DashboardWebStringId.Filters',
                panelWidth: _dashboard_item_menu_1.DashboardItemMenuSizes.OptionsPanelWidth,
                templateName: 'dx-dashboard-options',
                detailVisible: ko.observable(false),
                showMenu: function () {
                    var composer = new _item_filter_properties_composer_1.ItemFilterPropertiesComposer(_this.dashboardControl._dataSourceBrowser);
                    var tabs = composer.composeTabs(item);
                    menu.propertiesController.mainModel({
                        data: {
                            model: item,
                            propertiesTabs: ko.observableArray(tabs),
                        }
                    });
                },
                customData: {
                    propertiesController: menu.propertiesController,
                },
                index: 150
            });
        }
    };
    FilterPanelExtension.prototype.start = function () {
        var contextMenuExtension = this.dashboardControl.findExtension('item-menu');
        if (contextMenuExtension) {
            this._subscriptions.push(contextMenuExtension._itemContextMenu.subscribe(this._contextMenuSubscriber, this));
            this._contextMenuSubscriber(contextMenuExtension._itemContextMenu());
        }
    };
    FilterPanelExtension.prototype.stop = function () {
        this._subscriptions.forEach(function (s) { return s.dispose(); });
        this._subscriptions = [];
    };
    return FilterPanelExtension;
}());
exports.FilterPanelExtension = FilterPanelExtension;
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new FilterPanelExtension(dashboardControl); };
control_options_1.extensionNameMap[name] = nameAlias;
