// devexpress-dashboard/designer/extensions/open-dashboard.js
/**
* DevExpress Dashboard (open-dashboard.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var array_store_1 = require("devextreme/data/array_store");
var control_options_1 = require("../../common/control-options");
var _jquery_helpers_1 = require("../../data/_jquery-helpers");
var toolbox_items_1 = require("../toolbox-extension/toolbox-items");
var name = 'open-dashboard';
var nameAlias = 'openDashboard';
var OpenDashboardExtension = (function () {
    function OpenDashboardExtension(dashboardControl) {
        var _this = this;
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._menuItem = new toolbox_items_1.DashboardMenuItem(this.name, 'DashboardWebStringId.DashboardMenuOpen', 108, 79);
        this._menuItem.template = 'dx-dashboard-form-open';
        this._menuItem.data = {
            dashboardsListStore: function (loadOptions) {
                var def = _jquery_helpers_1.createJQueryDeferred();
                _this.dashboardControl.requestDashboardList().done(function (items) {
                    new array_store_1.default(items).load({ filter: loadOptions.searchValue ? [loadOptions.searchExpr, loadOptions.searchOperation, loadOptions.searchValue] : null }).then(function (data) {
                        def.resolve(data);
                    });
                });
                return def;
            },
            openDashboard: function (e) {
                var toolboxExtension = _this.dashboardControl.findExtension('toolbox');
                if (toolboxExtension) {
                    toolboxExtension.menuVisible(false);
                }
                var openDashboardCallback = function () { return _this.loadDashboard(e.itemData.id); };
                var saveExtension = _this.dashboardControl.findExtension('save-dashboard');
                if (saveExtension) {
                    saveExtension.ensureDashboardSaved(openDashboardCallback);
                }
                else {
                    openDashboardCallback();
                }
            }
        };
    }
    OpenDashboardExtension.prototype.start = function () {
        var toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.push(this._menuItem);
        }
    };
    OpenDashboardExtension.prototype.stop = function () {
        var toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.remove(this._menuItem);
        }
    };
    OpenDashboardExtension.prototype.loadDashboard = function (dashboardId) {
        return this.dashboardControl.loadDashboard(dashboardId);
    };
    return OpenDashboardExtension;
}());
exports.OpenDashboardExtension = OpenDashboardExtension;
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new OpenDashboardExtension(dashboardControl); };
control_options_1.extensionNameMap[name] = nameAlias;
