// devexpress-richedit/lib/core/model/paragraph/paragraph-property-descriptors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var errors_1 = require("@devexpress/utils/lib/errors");
var comparers_1 = require("@devexpress/utils/lib/utils/comparers");
var border_info_1 = require("../borders/border-info");
var paragraph_properties_history_items_1 = require("../history/items/paragraph-properties-history-items");
var json_paragraph_enums_1 = require("../json/enums/json-paragraph-enums");
var shading_info_1 = require("../shadings/shading-info");
var paragraph_properties_1 = require("./paragraph-properties");
var ParagraphPropertiesFirstLineIndentDescriptor = (function () {
    function ParagraphPropertiesFirstLineIndentDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesFirstLineIndentDescriptor.prototype.setProp = function (props, newValue) {
        props.firstLineIndent = newValue;
    };
    ParagraphPropertiesFirstLineIndentDescriptor.prototype.getProp = function (props) {
        return props.firstLineIndent;
    };
    ParagraphPropertiesFirstLineIndentDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseFirstLineIndent;
    };
    ParagraphPropertiesFirstLineIndentDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphFirstLineIndentHistoryItem;
    };
    ParagraphPropertiesFirstLineIndentDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.FirstLineIndent;
    };
    return ParagraphPropertiesFirstLineIndentDescriptor;
}());
exports.ParagraphPropertiesFirstLineIndentDescriptor = ParagraphPropertiesFirstLineIndentDescriptor;
var ParagraphPropertiesWidowOrphanControlDescriptor = (function () {
    function ParagraphPropertiesWidowOrphanControlDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = true;
    }
    ParagraphPropertiesWidowOrphanControlDescriptor.prototype.setProp = function (props, newValue) {
        props.widowOrphanControl = newValue;
    };
    ParagraphPropertiesWidowOrphanControlDescriptor.prototype.getProp = function (props) {
        return props.widowOrphanControl;
    };
    ParagraphPropertiesWidowOrphanControlDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseWidowOrphanControl;
    };
    ParagraphPropertiesWidowOrphanControlDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphWidowOrphanControlHistoryItem;
    };
    ParagraphPropertiesWidowOrphanControlDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.WidowOrphanControl;
    };
    return ParagraphPropertiesWidowOrphanControlDescriptor;
}());
exports.ParagraphPropertiesWidowOrphanControlDescriptor = ParagraphPropertiesWidowOrphanControlDescriptor;
var ParagraphPropertiesFirstLineIndentTypeDescriptor = (function () {
    function ParagraphPropertiesFirstLineIndentTypeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = paragraph_properties_1.ParagraphFirstLineIndent.None;
    }
    ParagraphPropertiesFirstLineIndentTypeDescriptor.prototype.setProp = function (props, newValue) {
        props.firstLineIndentType = newValue;
    };
    ParagraphPropertiesFirstLineIndentTypeDescriptor.prototype.getProp = function (props) {
        return props.firstLineIndentType;
    };
    ParagraphPropertiesFirstLineIndentTypeDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseFirstLineIndent;
    };
    ParagraphPropertiesFirstLineIndentTypeDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphFirstLineIndentTypeHistoryItem;
    };
    ParagraphPropertiesFirstLineIndentTypeDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.FirstLineIndentType;
    };
    return ParagraphPropertiesFirstLineIndentTypeDescriptor;
}());
exports.ParagraphPropertiesFirstLineIndentTypeDescriptor = ParagraphPropertiesFirstLineIndentTypeDescriptor;
var ParagraphPropertiesAfterAutoSpacingDescriptor = (function () {
    function ParagraphPropertiesAfterAutoSpacingDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesAfterAutoSpacingDescriptor.prototype.setProp = function (props, newValue) {
        props.afterAutoSpacing = newValue;
    };
    ParagraphPropertiesAfterAutoSpacingDescriptor.prototype.getProp = function (props) {
        return props.afterAutoSpacing;
    };
    ParagraphPropertiesAfterAutoSpacingDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseAfterAutoSpacing;
    };
    ParagraphPropertiesAfterAutoSpacingDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphAfterAutoSpacingHistoryItem;
    };
    ParagraphPropertiesAfterAutoSpacingDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.AfterAutoSpacing;
    };
    return ParagraphPropertiesAfterAutoSpacingDescriptor;
}());
exports.ParagraphPropertiesAfterAutoSpacingDescriptor = ParagraphPropertiesAfterAutoSpacingDescriptor;
var ParagraphPropertiesOutlineLevelDescriptor = (function () {
    function ParagraphPropertiesOutlineLevelDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesOutlineLevelDescriptor.prototype.setProp = function (props, newValue) {
        props.outlineLevel = newValue;
    };
    ParagraphPropertiesOutlineLevelDescriptor.prototype.getProp = function (props) {
        return props.outlineLevel;
    };
    ParagraphPropertiesOutlineLevelDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseOutlineLevel;
    };
    ParagraphPropertiesOutlineLevelDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphOutlineLevelHistoryItem;
    };
    ParagraphPropertiesOutlineLevelDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.OutlineLevel;
    };
    return ParagraphPropertiesOutlineLevelDescriptor;
}());
exports.ParagraphPropertiesOutlineLevelDescriptor = ParagraphPropertiesOutlineLevelDescriptor;
var ParagraphPropertiesBeforeAutoSpacingDescriptor = (function () {
    function ParagraphPropertiesBeforeAutoSpacingDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesBeforeAutoSpacingDescriptor.prototype.setProp = function (props, newValue) {
        props.beforeAutoSpacing = newValue;
    };
    ParagraphPropertiesBeforeAutoSpacingDescriptor.prototype.getProp = function (props) {
        return props.beforeAutoSpacing;
    };
    ParagraphPropertiesBeforeAutoSpacingDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseBeforeAutoSpacing;
    };
    ParagraphPropertiesBeforeAutoSpacingDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphBeforeAutoSpacingHistoryItem;
    };
    ParagraphPropertiesBeforeAutoSpacingDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.BeforeAutoSpacing;
    };
    return ParagraphPropertiesBeforeAutoSpacingDescriptor;
}());
exports.ParagraphPropertiesBeforeAutoSpacingDescriptor = ParagraphPropertiesBeforeAutoSpacingDescriptor;
var ParagraphPropertiesPageBreakBeforeDescriptor = (function () {
    function ParagraphPropertiesPageBreakBeforeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesPageBreakBeforeDescriptor.prototype.setProp = function (props, newValue) {
        props.pageBreakBefore = newValue;
    };
    ParagraphPropertiesPageBreakBeforeDescriptor.prototype.getProp = function (props) {
        return props.pageBreakBefore;
    };
    ParagraphPropertiesPageBreakBeforeDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UsePageBreakBefore;
    };
    ParagraphPropertiesPageBreakBeforeDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphPageBreakBeforeHistoryItem;
    };
    ParagraphPropertiesPageBreakBeforeDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.PageBreakBefore;
    };
    return ParagraphPropertiesPageBreakBeforeDescriptor;
}());
exports.ParagraphPropertiesPageBreakBeforeDescriptor = ParagraphPropertiesPageBreakBeforeDescriptor;
var ParagraphPropertiesRightIndentDescriptor = (function () {
    function ParagraphPropertiesRightIndentDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesRightIndentDescriptor.prototype.setProp = function (props, newValue) {
        props.rightIndent = newValue;
    };
    ParagraphPropertiesRightIndentDescriptor.prototype.getProp = function (props) {
        return props.rightIndent;
    };
    ParagraphPropertiesRightIndentDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseRightIndent;
    };
    ParagraphPropertiesRightIndentDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphRightIndentHistoryItem;
    };
    ParagraphPropertiesRightIndentDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.RightIndent;
    };
    return ParagraphPropertiesRightIndentDescriptor;
}());
exports.ParagraphPropertiesRightIndentDescriptor = ParagraphPropertiesRightIndentDescriptor;
var ParagraphPropertiesSuppressHyphenationDescriptor = (function () {
    function ParagraphPropertiesSuppressHyphenationDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesSuppressHyphenationDescriptor.prototype.setProp = function (props, newValue) {
        props.suppressHyphenation = newValue;
    };
    ParagraphPropertiesSuppressHyphenationDescriptor.prototype.getProp = function (props) {
        return props.suppressHyphenation;
    };
    ParagraphPropertiesSuppressHyphenationDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseSuppressHyphenation;
    };
    ParagraphPropertiesSuppressHyphenationDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphSuppressHyphenationHistoryItem;
    };
    ParagraphPropertiesSuppressHyphenationDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.SuppressHyphenation;
    };
    return ParagraphPropertiesSuppressHyphenationDescriptor;
}());
exports.ParagraphPropertiesSuppressHyphenationDescriptor = ParagraphPropertiesSuppressHyphenationDescriptor;
var ParagraphPropertiesLineSpacingDescriptor = (function () {
    function ParagraphPropertiesLineSpacingDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesLineSpacingDescriptor.prototype.setProp = function (props, newValue) {
        props.lineSpacing = newValue;
    };
    ParagraphPropertiesLineSpacingDescriptor.prototype.getProp = function (props) {
        return props.lineSpacing;
    };
    ParagraphPropertiesLineSpacingDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseLineSpacing;
    };
    ParagraphPropertiesLineSpacingDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphLineSpacingHistoryItem;
    };
    ParagraphPropertiesLineSpacingDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.LineSpacing;
    };
    return ParagraphPropertiesLineSpacingDescriptor;
}());
exports.ParagraphPropertiesLineSpacingDescriptor = ParagraphPropertiesLineSpacingDescriptor;
var ParagraphPropertiesSuppressLineNumbersDescriptor = (function () {
    function ParagraphPropertiesSuppressLineNumbersDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesSuppressLineNumbersDescriptor.prototype.setProp = function (props, newValue) {
        props.suppressLineNumbers = newValue;
    };
    ParagraphPropertiesSuppressLineNumbersDescriptor.prototype.getProp = function (props) {
        return props.suppressLineNumbers;
    };
    ParagraphPropertiesSuppressLineNumbersDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseSuppressLineNumbers;
    };
    ParagraphPropertiesSuppressLineNumbersDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphSuppressLineNumbersHistoryItem;
    };
    ParagraphPropertiesSuppressLineNumbersDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.SuppressLineNumbers;
    };
    return ParagraphPropertiesSuppressLineNumbersDescriptor;
}());
exports.ParagraphPropertiesSuppressLineNumbersDescriptor = ParagraphPropertiesSuppressLineNumbersDescriptor;
var ParagraphPropertiesKeepLinesTogetherDescriptor = (function () {
    function ParagraphPropertiesKeepLinesTogetherDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesKeepLinesTogetherDescriptor.prototype.setProp = function (props, newValue) {
        props.keepLinesTogether = newValue;
    };
    ParagraphPropertiesKeepLinesTogetherDescriptor.prototype.getProp = function (props) {
        return props.keepLinesTogether;
    };
    ParagraphPropertiesKeepLinesTogetherDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseKeepLinesTogether;
    };
    ParagraphPropertiesKeepLinesTogetherDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphKeepLinesTogetherHistoryItem;
    };
    ParagraphPropertiesKeepLinesTogetherDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.KeepLinesTogether;
    };
    return ParagraphPropertiesKeepLinesTogetherDescriptor;
}());
exports.ParagraphPropertiesKeepLinesTogetherDescriptor = ParagraphPropertiesKeepLinesTogetherDescriptor;
var ParagraphPropertiesKeepWithNextDescriptor = (function () {
    function ParagraphPropertiesKeepWithNextDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesKeepWithNextDescriptor.prototype.setProp = function (props, newValue) {
        props.keepWithNext = newValue;
    };
    ParagraphPropertiesKeepWithNextDescriptor.prototype.getProp = function (props) {
        return props.keepWithNext;
    };
    ParagraphPropertiesKeepWithNextDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseKeepWithNext;
    };
    ParagraphPropertiesKeepWithNextDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphKeepWithNextHistoryItem;
    };
    ParagraphPropertiesKeepWithNextDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.KeepWithNext;
    };
    return ParagraphPropertiesKeepWithNextDescriptor;
}());
exports.ParagraphPropertiesKeepWithNextDescriptor = ParagraphPropertiesKeepWithNextDescriptor;
var ParagraphPropertiesShadingInfoIndexDescriptor = (function () {
    function ParagraphPropertiesShadingInfoIndexDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = shading_info_1.ShadingInfo.noColor;
    }
    ParagraphPropertiesShadingInfoIndexDescriptor.prototype.setProp = function (props, newValue) {
        props.shadingInfo = newValue;
    };
    ParagraphPropertiesShadingInfoIndexDescriptor.prototype.getProp = function (props) {
        return props.shadingInfo;
    };
    ParagraphPropertiesShadingInfoIndexDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseShadingInfoIndex;
    };
    ParagraphPropertiesShadingInfoIndexDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphShadingInfoIndexHistoryItem;
    };
    ParagraphPropertiesShadingInfoIndexDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.ShadingInfo;
    };
    return ParagraphPropertiesShadingInfoIndexDescriptor;
}());
exports.ParagraphPropertiesShadingInfoIndexDescriptor = ParagraphPropertiesShadingInfoIndexDescriptor;
var ParagraphPropertiesLeftIndentDescriptor = (function () {
    function ParagraphPropertiesLeftIndentDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesLeftIndentDescriptor.prototype.setProp = function (props, newValue) {
        props.leftIndent = newValue;
    };
    ParagraphPropertiesLeftIndentDescriptor.prototype.getProp = function (props) {
        return props.leftIndent;
    };
    ParagraphPropertiesLeftIndentDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseLeftIndent;
    };
    ParagraphPropertiesLeftIndentDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphLeftIndentHistoryItem;
    };
    ParagraphPropertiesLeftIndentDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.LeftIndent;
    };
    return ParagraphPropertiesLeftIndentDescriptor;
}());
exports.ParagraphPropertiesLeftIndentDescriptor = ParagraphPropertiesLeftIndentDescriptor;
var ParagraphPropertiesLineSpacingTypeDescriptor = (function () {
    function ParagraphPropertiesLineSpacingTypeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = paragraph_properties_1.ParagraphLineSpacingType.Single;
    }
    ParagraphPropertiesLineSpacingTypeDescriptor.prototype.setProp = function (props, newValue) {
        props.lineSpacingType = newValue;
    };
    ParagraphPropertiesLineSpacingTypeDescriptor.prototype.getProp = function (props) {
        return props.lineSpacingType;
    };
    ParagraphPropertiesLineSpacingTypeDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseLineSpacing;
    };
    ParagraphPropertiesLineSpacingTypeDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphLineSpacingTypeHistoryItem;
    };
    ParagraphPropertiesLineSpacingTypeDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.LineSpacingType;
    };
    return ParagraphPropertiesLineSpacingTypeDescriptor;
}());
exports.ParagraphPropertiesLineSpacingTypeDescriptor = ParagraphPropertiesLineSpacingTypeDescriptor;
var ParagraphPropertiesAlignmentDescriptor = (function () {
    function ParagraphPropertiesAlignmentDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = paragraph_properties_1.ParagraphAlignment.Left;
    }
    ParagraphPropertiesAlignmentDescriptor.prototype.setProp = function (props, newValue) {
        props.alignment = newValue;
    };
    ParagraphPropertiesAlignmentDescriptor.prototype.getProp = function (props) {
        return props.alignment;
    };
    ParagraphPropertiesAlignmentDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseAlignment;
    };
    ParagraphPropertiesAlignmentDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphAlignmentHistoryItem;
    };
    ParagraphPropertiesAlignmentDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.Alignment;
    };
    return ParagraphPropertiesAlignmentDescriptor;
}());
exports.ParagraphPropertiesAlignmentDescriptor = ParagraphPropertiesAlignmentDescriptor;
var ParagraphPropertiesContextualSpacingDescriptor = (function () {
    function ParagraphPropertiesContextualSpacingDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesContextualSpacingDescriptor.prototype.setProp = function (props, newValue) {
        props.contextualSpacing = newValue;
    };
    ParagraphPropertiesContextualSpacingDescriptor.prototype.getProp = function (props) {
        return props.contextualSpacing;
    };
    ParagraphPropertiesContextualSpacingDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseContextualSpacing;
    };
    ParagraphPropertiesContextualSpacingDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphContextualSpacingHistoryItem;
    };
    ParagraphPropertiesContextualSpacingDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.ContextualSpacing;
    };
    return ParagraphPropertiesContextualSpacingDescriptor;
}());
exports.ParagraphPropertiesContextualSpacingDescriptor = ParagraphPropertiesContextualSpacingDescriptor;
var ParagraphPropertiesSpacingBeforeDescriptor = (function () {
    function ParagraphPropertiesSpacingBeforeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesSpacingBeforeDescriptor.prototype.setProp = function (props, newValue) {
        props.spacingBefore = newValue;
    };
    ParagraphPropertiesSpacingBeforeDescriptor.prototype.getProp = function (props) {
        return props.spacingBefore;
    };
    ParagraphPropertiesSpacingBeforeDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseSpacingBefore;
    };
    ParagraphPropertiesSpacingBeforeDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphSpacingBeforeHistoryItem;
    };
    ParagraphPropertiesSpacingBeforeDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.SpacingBefore;
    };
    return ParagraphPropertiesSpacingBeforeDescriptor;
}());
exports.ParagraphPropertiesSpacingBeforeDescriptor = ParagraphPropertiesSpacingBeforeDescriptor;
var ParagraphPropertiesSpacingAfterDescriptor = (function () {
    function ParagraphPropertiesSpacingAfterDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesSpacingAfterDescriptor.prototype.setProp = function (props, newValue) {
        props.spacingAfter = newValue;
    };
    ParagraphPropertiesSpacingAfterDescriptor.prototype.getProp = function (props) {
        return props.spacingAfter;
    };
    ParagraphPropertiesSpacingAfterDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseSpacingAfter;
    };
    ParagraphPropertiesSpacingAfterDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphSpacingAfterHistoryItem;
    };
    ParagraphPropertiesSpacingAfterDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.SpacingAfter;
    };
    return ParagraphPropertiesSpacingAfterDescriptor;
}());
exports.ParagraphPropertiesSpacingAfterDescriptor = ParagraphPropertiesSpacingAfterDescriptor;
var ParagraphPropertiesRightToLeftDescriptor = (function () {
    function ParagraphPropertiesRightToLeftDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    ParagraphPropertiesRightToLeftDescriptor.prototype.setProp = function (props, newValue) {
        props.rightToLeft = newValue;
    };
    ParagraphPropertiesRightToLeftDescriptor.prototype.getProp = function (props) {
        return props.rightToLeft;
    };
    ParagraphPropertiesRightToLeftDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseRightToLeft;
    };
    ParagraphPropertiesRightToLeftDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphRightToLeftHistoryItem;
    };
    ParagraphPropertiesRightToLeftDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.RightToLeft;
    };
    return ParagraphPropertiesRightToLeftDescriptor;
}());
exports.ParagraphPropertiesRightToLeftDescriptor = ParagraphPropertiesRightToLeftDescriptor;
var ParagraphPropertiesLeftBorderDescriptor = (function () {
    function ParagraphPropertiesLeftBorderDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new border_info_1.BorderInfo();
    }
    ParagraphPropertiesLeftBorderDescriptor.prototype.setProp = function (props, newValue) {
        props.leftBorder = newValue;
    };
    ParagraphPropertiesLeftBorderDescriptor.prototype.getProp = function (props) {
        return props.leftBorder;
    };
    ParagraphPropertiesLeftBorderDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseLeftBorder;
    };
    ParagraphPropertiesLeftBorderDescriptor.prototype.getHistoryItemConstructor = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    ParagraphPropertiesLeftBorderDescriptor.prototype.getJSONProperty = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    return ParagraphPropertiesLeftBorderDescriptor;
}());
exports.ParagraphPropertiesLeftBorderDescriptor = ParagraphPropertiesLeftBorderDescriptor;
var ParagraphPropertiesRightBorderDescriptor = (function () {
    function ParagraphPropertiesRightBorderDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new border_info_1.BorderInfo();
    }
    ParagraphPropertiesRightBorderDescriptor.prototype.setProp = function (props, newValue) {
        props.rightBorder = newValue;
    };
    ParagraphPropertiesRightBorderDescriptor.prototype.getProp = function (props) {
        return props.rightBorder;
    };
    ParagraphPropertiesRightBorderDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseRightBorder;
    };
    ParagraphPropertiesRightBorderDescriptor.prototype.getHistoryItemConstructor = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    ParagraphPropertiesRightBorderDescriptor.prototype.getJSONProperty = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    return ParagraphPropertiesRightBorderDescriptor;
}());
exports.ParagraphPropertiesRightBorderDescriptor = ParagraphPropertiesRightBorderDescriptor;
var ParagraphPropertiesTopBorderDescriptor = (function () {
    function ParagraphPropertiesTopBorderDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new border_info_1.BorderInfo();
    }
    ParagraphPropertiesTopBorderDescriptor.prototype.setProp = function (props, newValue) {
        props.topBorder = newValue;
    };
    ParagraphPropertiesTopBorderDescriptor.prototype.getProp = function (props) {
        return props.topBorder;
    };
    ParagraphPropertiesTopBorderDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseTopBorder;
    };
    ParagraphPropertiesTopBorderDescriptor.prototype.getHistoryItemConstructor = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    ParagraphPropertiesTopBorderDescriptor.prototype.getJSONProperty = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    return ParagraphPropertiesTopBorderDescriptor;
}());
exports.ParagraphPropertiesTopBorderDescriptor = ParagraphPropertiesTopBorderDescriptor;
var ParagraphPropertiesBottomBorderDescriptor = (function () {
    function ParagraphPropertiesBottomBorderDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new border_info_1.BorderInfo();
    }
    ParagraphPropertiesBottomBorderDescriptor.prototype.setProp = function (props, newValue) {
        props.bottomBorder = newValue;
    };
    ParagraphPropertiesBottomBorderDescriptor.prototype.getProp = function (props) {
        return props.bottomBorder;
    };
    ParagraphPropertiesBottomBorderDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseBottomBorder;
    };
    ParagraphPropertiesBottomBorderDescriptor.prototype.getHistoryItemConstructor = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    ParagraphPropertiesBottomBorderDescriptor.prototype.getJSONProperty = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    return ParagraphPropertiesBottomBorderDescriptor;
}());
exports.ParagraphPropertiesBottomBorderDescriptor = ParagraphPropertiesBottomBorderDescriptor;
var ParagraphPropertiesBetweenBorderDescriptor = (function () {
    function ParagraphPropertiesBetweenBorderDescriptor() {
        this.binaryEquals = comparers_1.Equals.object;
        this.defaultValue = new border_info_1.BorderInfo();
    }
    ParagraphPropertiesBetweenBorderDescriptor.prototype.setProp = function (props, newValue) {
        props.betweenBorder = newValue;
    };
    ParagraphPropertiesBetweenBorderDescriptor.prototype.getProp = function (props) {
        return props.betweenBorder;
    };
    ParagraphPropertiesBetweenBorderDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseBetweenBorder;
    };
    ParagraphPropertiesBetweenBorderDescriptor.prototype.getHistoryItemConstructor = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    ParagraphPropertiesBetweenBorderDescriptor.prototype.getJSONProperty = function () {
        throw Error(errors_1.Errors.NotImplemented);
    };
    return ParagraphPropertiesBetweenBorderDescriptor;
}());
exports.ParagraphPropertiesBetweenBorderDescriptor = ParagraphPropertiesBetweenBorderDescriptor;
var ParagraphPropertiesDivIdDescriptor = (function () {
    function ParagraphPropertiesDivIdDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 0;
    }
    ParagraphPropertiesDivIdDescriptor.prototype.setProp = function (props, newValue) {
        props.divId = newValue;
    };
    ParagraphPropertiesDivIdDescriptor.prototype.getProp = function (props) {
        return props.divId;
    };
    ParagraphPropertiesDivIdDescriptor.prototype.maskValue = function () {
        return paragraph_properties_1.ParagraphPropertiesMask.UseDivId;
    };
    ParagraphPropertiesDivIdDescriptor.prototype.getHistoryItemConstructor = function () {
        return paragraph_properties_history_items_1.ParagraphDivIdHistoryItem;
    };
    ParagraphPropertiesDivIdDescriptor.prototype.getJSONProperty = function () {
        return json_paragraph_enums_1.JSONParagraphFormattingProperty.DivId;
    };
    return ParagraphPropertiesDivIdDescriptor;
}());
exports.ParagraphPropertiesDivIdDescriptor = ParagraphPropertiesDivIdDescriptor;
