// devexpress-richedit/lib/rtf/export/exporters/table/rtf-table-row-properties-exporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var table_base_structures_1 = require("../../../../core/model/tables/secondary-structures/table-base-structures");
var table_units_1 = require("../../../../core/model/tables/secondary-structures/table-units");
var rtf_export_sr_1 = require("../../../translation-table/rtf-export-sr");
var rtf_shading_info_export_helper_1 = require("../../helpers/rtf-shading-info-export-helper");
var rtf_properties_exporter_1 = require("../rtf-properties-exporter");
var RtfTableRowPropertiesExporter = (function (_super) {
    tslib_1.__extends(RtfTableRowPropertiesExporter, _super);
    function RtfTableRowPropertiesExporter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RtfTableRowPropertiesExporter.prototype.writeLastRowMark = function () {
        this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableLastRow);
    };
    RtfTableRowPropertiesExporter.prototype.writeHalfSpaceBetweenCells = function (val) {
        if (val > 0)
            this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableHalfSpaceBetweenCells, val);
    };
    RtfTableRowPropertiesExporter.prototype.writeRowAlignment = function (value) {
        var defaultGeneralSettings = this.documentModel.defaultTableRowProperties;
        if (value == defaultGeneralSettings.tableRowAlignment)
            return;
        switch (value) {
            case table_base_structures_1.TableRowAlignment.Center:
                this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableRowCenterAlignment);
                break;
            case table_base_structures_1.TableRowAlignment.Left:
                this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableRowLeftAlignment);
                break;
            case table_base_structures_1.TableRowAlignment.Right:
                this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableRowRightAlignment);
                break;
            default:
                break;
        }
    };
    RtfTableRowPropertiesExporter.prototype.writeRowHeader = function (header) {
        var defaultGeneralSettings = this.documentModel.defaultTableRowProperties;
        if (header != defaultGeneralSettings.header)
            this.writeBoolCommand(rtf_export_sr_1.RtfExportSR.TableRowHeader, header);
    };
    RtfTableRowPropertiesExporter.prototype.writeRowCantSplit = function (cantSplit) {
        var defaultGeneralSettings = this.documentModel.defaultTableRowProperties;
        if (cantSplit != defaultGeneralSettings.cantSplit)
            this.writeBoolCommand(rtf_export_sr_1.RtfExportSR.TableRowCantSplit, cantSplit);
    };
    RtfTableRowPropertiesExporter.prototype.writeRowHeight = function (height) {
        var defaultHeight = table_units_1.TableHeightUnit.createDefault();
        if (height.type == defaultHeight.type)
            return;
        switch (height.type) {
            case table_units_1.TableHeightUnitType.Auto:
                this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableRowHeight, 0);
                break;
            case table_units_1.TableHeightUnitType.Minimum:
                this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableRowHeight, height.value);
                break;
            case table_units_1.TableHeightUnitType.Exact:
                this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableRowHeight, -height.value);
                break;
        }
    };
    RtfTableRowPropertiesExporter.prototype.writeWidthBefore = function (widthBefore) {
        this.writeWidthUnit(widthBefore, rtf_export_sr_1.RtfExportSR.TableRowWidthBeforeType, rtf_export_sr_1.RtfExportSR.TableRowWidthBefore);
    };
    RtfTableRowPropertiesExporter.prototype.writeWidthAfter = function (widthAfter) {
        this.writeWidthUnit(widthAfter, rtf_export_sr_1.RtfExportSR.TableRowWidthAfterType, rtf_export_sr_1.RtfExportSR.TableRowWidthAfter);
    };
    RtfTableRowPropertiesExporter.prototype.writeRowCellSpacing = function (cellSpacing) {
        this.writeWidthUnitInTwips(cellSpacing, rtf_export_sr_1.RtfExportSR.TableCellSpacingLeftType, rtf_export_sr_1.RtfExportSR.TableCellSpacingLeft);
        this.writeWidthUnitInTwips(cellSpacing, rtf_export_sr_1.RtfExportSR.TableCellSpacingBottomType, rtf_export_sr_1.RtfExportSR.TableCellSpacingBottom);
        this.writeWidthUnitInTwips(cellSpacing, rtf_export_sr_1.RtfExportSR.TableCellSpacingRightType, rtf_export_sr_1.RtfExportSR.TableCellSpacingRight);
        this.writeWidthUnitInTwips(cellSpacing, rtf_export_sr_1.RtfExportSR.TableCellSpacingTopType, rtf_export_sr_1.RtfExportSR.TableCellSpacingTop);
    };
    RtfTableRowPropertiesExporter.prototype.writeRowShading = function (shadingInfo) {
        rtf_shading_info_export_helper_1.RtfShadingInfoExportHelper.exportShadingForeColorIndex(this.rtfBuilder, this.rtfExportHelper, this.documentModel.colorProvider, shadingInfo, rtf_export_sr_1.RtfExportSR.TableRowForegroundColor);
        rtf_shading_info_export_helper_1.RtfShadingInfoExportHelper.exportShadingBackColorIndex(this.rtfBuilder, this.rtfExportHelper, this.documentModel.colorProvider, shadingInfo, rtf_export_sr_1.RtfExportSR.TableRowBackgroundColor);
        rtf_shading_info_export_helper_1.RtfShadingInfoExportHelper.exportShadingPattern(this.rtfBuilder, this.documentModel.colorProvider, shadingInfo, rtf_export_sr_1.RtfExportSR.TableRowShading);
        var shadingPatternKeyword = rtf_export_sr_1.RtfExportSR.RunShadingPatternTable[shadingInfo.shadingPattern];
        if (shadingPatternKeyword)
            this.rtfBuilder.writeCommand(shadingPatternKeyword);
    };
    return RtfTableRowPropertiesExporter;
}(rtf_properties_exporter_1.RtfPropertiesExporter));
exports.RtfTableRowPropertiesExporter = RtfTableRowPropertiesExporter;
