// devexpress-richedit/lib/rtf/export/exporters/table/rtf-table-properties-exporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var table_properties_1 = require("../../../../core/model/tables/properties/table-properties");
var table_base_structures_1 = require("../../../../core/model/tables/secondary-structures/table-base-structures");
var rtf_export_sr_1 = require("../../../translation-table/rtf-export-sr");
var rtf_properties_exporter_1 = require("../rtf-properties-exporter");
var RtfTablePropertiesExporter = (function (_super) {
    tslib_1.__extends(RtfTablePropertiesExporter, _super);
    function RtfTablePropertiesExporter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RtfTablePropertiesExporter.prototype.writeRTLRow = function () {
        this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.RTLRow);
    };
    RtfTablePropertiesExporter.prototype.writeRowLeft = function (left) {
        this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableRowLeft, left);
    };
    RtfTablePropertiesExporter.prototype.writeTableBorders = function (borders) {
        var defaultBorder = this.documentModel.defaultTableCellProperties.borders.bottomBorder;
        var topBorder = borders.topBorder;
        var leftBorder = borders.leftBorder;
        var bottomBorder = borders.bottomBorder;
        var rightBorder = borders.rightBorder;
        var innerHorizontalBorder = borders.insideHorizontalBorder;
        var innerVerticalBorder = borders.insideVerticalBorder;
        if (!defaultBorder.equals(topBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableTopBorder);
            this.writeBorderProperties(topBorder);
        }
        if (!defaultBorder.equals(leftBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableLeftBorder);
            this.writeBorderProperties(leftBorder);
        }
        if (!defaultBorder.equals(bottomBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableBottomBorder);
            this.writeBorderProperties(bottomBorder);
        }
        if (!defaultBorder.equals(rightBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableRightBorder);
            this.writeBorderProperties(rightBorder);
        }
        if (!defaultBorder.equals(innerHorizontalBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableHorizontalBorder);
            this.writeBorderProperties(innerHorizontalBorder);
        }
        if (!defaultBorder.equals(innerVerticalBorder)) {
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableVerticalBorder);
            this.writeBorderProperties(innerVerticalBorder);
        }
    };
    RtfTablePropertiesExporter.prototype.writeTableWidth = function (preferredWidth) {
        this.writeWidthUnit(preferredWidth, rtf_export_sr_1.RtfExportSR.TablePreferredWidthType, rtf_export_sr_1.RtfExportSR.TablePreferredWidth);
    };
    RtfTablePropertiesExporter.prototype.writeTableLayout = function (value) {
        if (value != table_base_structures_1.TableLayoutType.Fixed) {
            this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableLayout, value);
        }
    };
    RtfTablePropertiesExporter.prototype.writeTableCellMargins = function (cellMargins) {
        var leftMargin = cellMargins.left;
        var rightMargin = cellMargins.right;
        var bottomMargin = cellMargins.bottom;
        var topMargin = cellMargins.top;
        if (this.shouldExportCellMargin(leftMargin))
            this.writeWidthUnitInTwips(leftMargin, rtf_export_sr_1.RtfExportSR.TableCellMarginsLeftType, rtf_export_sr_1.RtfExportSR.TableCellMarginsLeft);
        if (this.shouldExportCellMargin(bottomMargin))
            this.writeWidthUnitInTwips(bottomMargin, rtf_export_sr_1.RtfExportSR.TableCellMarginsBottomType, rtf_export_sr_1.RtfExportSR.TableCellMarginsBottom);
        if (this.shouldExportCellMargin(rightMargin))
            this.writeWidthUnitInTwips(rightMargin, rtf_export_sr_1.RtfExportSR.TableCellMarginsRightType, rtf_export_sr_1.RtfExportSR.TableCellMarginsRight);
        if (this.shouldExportCellMargin(topMargin))
            this.writeWidthUnitInTwips(topMargin, rtf_export_sr_1.RtfExportSR.TableCellMarginsTopType, rtf_export_sr_1.RtfExportSR.TableCellMarginsTop);
    };
    RtfTablePropertiesExporter.prototype.writeTableLook = function (value) {
        this.documentModel.defaultTableProperties;
        if ((value & table_base_structures_1.TableLookTypes.ApplyFirstColumn) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableApplyFirstColumn);
        if ((value & table_base_structures_1.TableLookTypes.ApplyFirstRow) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableApplyFirstRow);
        if ((value & table_base_structures_1.TableLookTypes.ApplyLastColumn) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableApplyLastColumn);
        if ((value & table_base_structures_1.TableLookTypes.ApplyLastRow) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableApplyLastRow);
        if ((value & table_base_structures_1.TableLookTypes.DoNotApplyColumnBanding) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableDoNotApplyColumnBanding);
        if ((value & table_base_structures_1.TableLookTypes.DoNotApplyRowBanding) != 0)
            this.rtfBuilder.writeCommand(rtf_export_sr_1.RtfExportSR.TableDoNotApplyRowBanding);
    };
    RtfTablePropertiesExporter.prototype.writeTableIndent = function (tableIndent) {
        this.writeWidthUnit(tableIndent, rtf_export_sr_1.RtfExportSR.TableIndentType, rtf_export_sr_1.RtfExportSR.TableIndent, true);
    };
    RtfTablePropertiesExporter.prototype.writeBandSizes = function (info) {
        var exportRowBand = info.getUseValue(table_properties_1.TablePropertiesMask.UseTableStyleRowBandSize);
        var exportColBand = info.getUseValue(table_properties_1.TablePropertiesMask.UseTableStyleColBandSize);
        if (info.tableStyleRowBandSize != 0 && (exportRowBand || info.tableStyleRowBandSize > 1))
            this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableStyleRowBandSize, info.tableStyleRowBandSize);
        if (info.tableStyleColumnBandSize != 0 && (exportColBand || info.tableStyleColumnBandSize > 1))
            this.rtfBuilder.writeIntegerCommand(rtf_export_sr_1.RtfExportSR.TableStyleColumnBandSize, info.tableStyleColumnBandSize);
    };
    return RtfTablePropertiesExporter;
}(rtf_properties_exporter_1.RtfPropertiesExporter));
exports.RtfTablePropertiesExporter = RtfTablePropertiesExporter;
