// devexpress-richedit/lib/base/commands/selection/go-to-next-page-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var layout_position_creator_1 = require("../../../core/layout-engine/layout-position-creator");
var document_layout_details_level_1 = require("../../../core/layout/document-layout-details-level");
var layout_point_1 = require("../../../core/layout/layout-point");
var selection_command_base_1 = require("./selection-command-base");
var GoToNextPageCommandBase = (function (_super) {
    tslib_1.__extends(GoToNextPageCommandBase, _super);
    function GoToNextPageCommandBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GoToNextPageCommandBase.prototype.getPosition = function () {
        var selection = this.selection;
        var initPosition = selection.forwardDirection ? selection.lastSelectedInterval.end : selection.lastSelectedInterval.start;
        var layoutPosition = new layout_position_creator_1.LayoutPositionMainSubDocumentCreator(this.control.layout, this.selection.activeSubDocument, initPosition, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
            .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(selection.endOfLine), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false));
        if (!layoutPosition)
            return -1;
        var charOffset = initPosition - layoutPosition.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box);
        var x = layoutPosition.pageArea.x + layoutPosition.column.x + layoutPosition.row.x + layoutPosition.box.x + layoutPosition.box.getCharOffsetXInPixels(this.control.measurer, charOffset);
        var y = layoutPosition.pageArea.y + layoutPosition.column.y + layoutPosition.row.y + layoutPosition.box.y;
        var siblingPageIndex = layoutPosition.pageIndex + 1;
        var siblingPage = this.control.layoutFormatterManager.forceFormatPage(siblingPageIndex);
        if (siblingPage) {
            var point = new layout_point_1.LayoutPoint(siblingPageIndex, x, y);
            var htr = this.control.hitTestManager.calculate(point, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character, this.selection.activeSubDocument);
            if (!this.extendSelection())
                htr.correctAsVisibleBox();
            return htr.getPosition();
        }
        else
            return this.selection.activeSubDocument.getDocumentEndPosition() - 1;
    };
    GoToNextPageCommandBase.prototype.isEnabled = function () {
        return _super.prototype.isEnabled.call(this) && this.selection.activeSubDocument.isMain();
    };
    return GoToNextPageCommandBase;
}(selection_command_base_1.SelectionCommandBase));
exports.GoToNextPageCommandBase = GoToNextPageCommandBase;
var GoToNextPageCommand = (function (_super) {
    tslib_1.__extends(GoToNextPageCommand, _super);
    function GoToNextPageCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GoToNextPageCommand.prototype.executeCore = function (_state, _parameter) {
        var selection = this.selection;
        var position = this.getPosition();
        if (position < 0)
            return false;
        selection.deprecatedSetSelection(position, position, selection.endOfLine, selection.keepX, true);
        return true;
    };
    GoToNextPageCommand.prototype.extendSelection = function () {
        return false;
    };
    return GoToNextPageCommand;
}(GoToNextPageCommandBase));
exports.GoToNextPageCommand = GoToNextPageCommand;
var ExtendGoToNextPageCommand = (function (_super) {
    tslib_1.__extends(ExtendGoToNextPageCommand, _super);
    function ExtendGoToNextPageCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExtendGoToNextPageCommand.prototype.executeCore = function (_state, _parameter) {
        var position = this.getPosition();
        if (position < 0)
            return false;
        this.selection.changeState(function (newState) { return newState.extendLastInterval(position); });
        return true;
    };
    ExtendGoToNextPageCommand.prototype.extendSelection = function () {
        return true;
    };
    return ExtendGoToNextPageCommand;
}(GoToNextPageCommandBase));
exports.ExtendGoToNextPageCommand = ExtendGoToNextPageCommand;
