// devexpress-richedit/bin/webpack.config.js
/**
* DevExpress WebRichEdit (webpack.config.js)
* Version: 21.2.3
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs
*/
const path = require('path');
const TerserPlugin = require('terser-webpack-plugin');
const { handleExternals, jsZipExternals, devextremeExternals } = require('./webpack-externals');
const source = path.resolve(__dirname, 'index-custom.js');
const scriptFileName = 'dx.richedit';
const libraryName = ['DevExpress']; // this creates window.DevExpress namespace for bundle

module.exports = {
    // This bundle decreases the size of DevExtreme scripts by 2.76 Mb
    // Uncomment a variable to exclude the corresponding module from the bundle
    // Register the excluded module on your page
    // <script src="/node_modules/jszip/dist/jszip.min.js"></script>
    // <script src="/node_modules/devextreme/dist/js/dx.all.js"> </script>
    externals: [ handleExternals([/*jsZipExternals, devextremeExternals*/]) ],
    entry: {
        [scriptFileName]: source,
        [scriptFileName + '.min']: source,
    },
    output: {
        path: path.resolve(__dirname, '..', 'dist', 'custom'),
        filename: '[name].js',
        library: libraryName
    },
    optimization: {
        minimize: true,
        minimizer: [new TerserPlugin({test: /\.min\.js$/})],
    },
};
