// devexpress-dashboard/viewer-parts/viewer/item-widget-event-args.js
/**
* DevExpress Dashboard (item-widget-event-args.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("devextreme/ui/widget/ui.widget");
var _common_1 = require("../../data/_common");
var ChartContext = (function () {
    function ChartContext(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    ChartContext.prototype.getDashboardItemSeries = function (seriesOptions) {
        return this._dashboardItem
            .panes()
            .reduce(function (acc, pane) { return acc.concat(pane.series()); }, [])
            .filter(function (series) { return seriesOptions[_common_1.DashboardDataIdField] && series._getDataId() === seriesOptions[_common_1.DashboardDataIdField]; })[0];
    };
    return ChartContext;
}());
exports.ChartContext = ChartContext;
var GridContext = (function () {
    function GridContext(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    GridContext.prototype.getDashboardItemColumn = function (columnOptions) {
        return this._dashboardItem
            .columns()
            .filter(function (column) { return columnOptions[_common_1.DashboardDataIdField] && column._getDataId() === columnOptions[_common_1.DashboardDataIdField]; })[0];
    };
    return GridContext;
}());
exports.GridContext = GridContext;
var GaugeContext = (function () {
    function GaugeContext(_dashboardItem) {
        this._dashboardItem = _dashboardItem;
    }
    GaugeContext.prototype.getDashboardItemGauge = function (gaugeOptions) {
        return this._dashboardItem
            .gauges()
            .filter(function (gauge) { return gaugeOptions[_common_1.DashboardDataIdField] && gauge._getDataId() === gaugeOptions[_common_1.DashboardDataIdField]; })[0];
    };
    return GaugeContext;
}());
exports.GaugeContext = GaugeContext;
