// devexpress-richedit/lib/core/model/history/states/history-item-state-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var errors_1 = require("@devexpress/utils/lib/errors");
var size_1 = require("@devexpress/utils/lib/geometry/size");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var border_info_1 = require("../../borders/border-info");
var sizes_1 = require("../../floating-objects/sizes");
var font_info_1 = require("../../fonts/font-info");
var json_masked_character_properties_converter_1 = require("../../json/importers/json-masked-character-properties-converter");
var json_masked_paragraph_properties_converter_1 = require("../../json/importers/json-masked-paragraph-properties-converter");
var json_shape_converter_1 = require("../../json/importers/json-shape-converter");
var json_tab_converter_1 = require("../../json/importers/json-tab-converter");
var json_columns_section_properties_converter_1 = require("../../json/importers/section/json-columns-section-properties-converter");
var json_anchor_info_converter_1 = require("../../json/importers/sub-document/json-anchor-info-converter");
var json_text_box_properties_converter_1 = require("../../json/importers/sub-document/json-text-box-properties-converter");
var json_anchor_text_box_size_converter_1 = require("../../json/importers/sub-document/sizes/json-anchor-text-box-size-converter");
var json_inline_picture_size_converter_1 = require("../../json/importers/sub-document/sizes/json-inline-picture-size-converter");
var table_units_1 = require("../../tables/secondary-structures/table-units");
var HistoryItemIntervalStateObject = (function () {
    function HistoryItemIntervalStateObject(interval, value) {
        this.interval = interval.clone();
        this.value = value;
    }
    HistoryItemIntervalStateObject.prototype.merge = function (object) {
        this.interval.length += object.interval.length;
    };
    HistoryItemIntervalStateObject.prototype.canMerge = function (object) {
        return this.interval.end === object.interval.start && this.isEqualValue(object);
    };
    HistoryItemIntervalStateObject.prototype.isEqualValue = function (object) {
        return this.value === object.value;
    };
    HistoryItemIntervalStateObject.prototype.toJSON = function (withPostData) {
        return [this.interval.start, this.interval.length, this.getPropertyValueForJSON(this.value, withPostData)];
    };
    HistoryItemIntervalStateObject.prototype.getPropertyValueForJSON = function (value, _withPostData) {
        if (value instanceof font_info_1.FontInfo)
            return value.name;
        return value;
    };
    return HistoryItemIntervalStateObject;
}());
exports.HistoryItemIntervalStateObject = HistoryItemIntervalStateObject;
var HistoryItemTabStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTabStateObject, _super);
    function HistoryItemTabStateObject(interval, tabInfo) {
        return _super.call(this, interval, tabInfo) || this;
    }
    HistoryItemTabStateObject.prototype.isEqualValue = function (object) {
        return this.value.position == object.value.position;
    };
    HistoryItemTabStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_tab_converter_1.JSONTabConverter.convertToJSON(value);
    };
    return HistoryItemTabStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemTabStateObject = HistoryItemTabStateObject;
var HistoryItemTextBufferStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTextBufferStateObject, _super);
    function HistoryItemTextBufferStateObject(startPosition, text) {
        return _super.call(this, new fixed_1.FixedInterval(startPosition, text.length), text) || this;
    }
    HistoryItemTextBufferStateObject.prototype.canMerge = function (_stateValue) {
        return false;
    };
    return HistoryItemTextBufferStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemTextBufferStateObject = HistoryItemTextBufferStateObject;
var HistoryItemIntervalStyleStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalStyleStateObject, _super);
    function HistoryItemIntervalStyleStateObject(interval, style) {
        return _super.call(this, interval, style) || this;
    }
    HistoryItemIntervalStyleStateObject.prototype.getPropertyValueForJSON = function (value) {
        return value.styleName;
    };
    return HistoryItemIntervalStyleStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalStyleStateObject = HistoryItemIntervalStyleStateObject;
var HistoryItemIntervalParagraphPropertiesStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalParagraphPropertiesStateObject, _super);
    function HistoryItemIntervalParagraphPropertiesStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalParagraphPropertiesStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_masked_paragraph_properties_converter_1.JSONMaskedParagraphPropertiesConverter.convertToJSON(value);
    };
    return HistoryItemIntervalParagraphPropertiesStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalParagraphPropertiesStateObject = HistoryItemIntervalParagraphPropertiesStateObject;
var HistoryItemIntervalCharacterPropertiesStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalCharacterPropertiesStateObject, _super);
    function HistoryItemIntervalCharacterPropertiesStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalCharacterPropertiesStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_masked_character_properties_converter_1.JSONMaskedCharacterPropertiesConverter.convertToJSON(value);
    };
    return HistoryItemIntervalCharacterPropertiesStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalCharacterPropertiesStateObject = HistoryItemIntervalCharacterPropertiesStateObject;
var HistoryItemIntervalUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalUseStateObject, _super);
    function HistoryItemIntervalUseStateObject(interval, value, use) {
        var _this = _super.call(this, interval, value) || this;
        _this.use = use;
        return _this;
    }
    HistoryItemIntervalUseStateObject.prototype.canMerge = function (stateValue) {
        return _super.prototype.canMerge.call(this, stateValue) && this.use === stateValue.use;
    };
    HistoryItemIntervalUseStateObject.prototype.toJSON = function () {
        return _super.prototype.toJSON.call(this).concat([this.use ? 1 : 0]);
    };
    return HistoryItemIntervalUseStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalUseStateObject = HistoryItemIntervalUseStateObject;
var HistoryItemSectionStateObject = (function () {
    function HistoryItemSectionStateObject(sectionIndex, value) {
        this.value = value;
        this.sectionIndex = sectionIndex;
    }
    HistoryItemSectionStateObject.prototype.toJSON = function () {
        return [this.sectionIndex, this.getPropertyValueForJSON(this.value)];
    };
    HistoryItemSectionStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof Array)
            return json_columns_section_properties_converter_1.JSONColumnsSectionPropertiesConverter.convertToJSON(value);
        return value;
    };
    HistoryItemSectionStateObject.prototype.canMerge = function (_obj) { return false; };
    HistoryItemSectionStateObject.prototype.merge = function (_obj) { };
    return HistoryItemSectionStateObject;
}());
exports.HistoryItemSectionStateObject = HistoryItemSectionStateObject;
var HistoryItemListLevelStateObject = (function () {
    function HistoryItemListLevelStateObject(isAbstractNumberingList, numberingListIndex, listLevelIndex, value) {
        this.isAbstractNumberingList = isAbstractNumberingList;
        this.numberingListIndex = numberingListIndex;
        this.listLevelIndex = listLevelIndex;
        this.value = value;
    }
    HistoryItemListLevelStateObject.prototype.toJSON = function () {
        return [this.isAbstractNumberingList ? 1 : 0, this.numberingListIndex, this.listLevelIndex, this.getPropertyValueForJSON(this.value)];
    };
    HistoryItemListLevelStateObject.prototype.canMerge = function (_obj) {
        return false;
    };
    HistoryItemListLevelStateObject.prototype.merge = function (_obj) {
        throw new Error(errors_1.Errors.NotImplemented);
    };
    HistoryItemListLevelStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof font_info_1.FontInfo)
            return value.name;
        return value;
    };
    return HistoryItemListLevelStateObject;
}());
exports.HistoryItemListLevelStateObject = HistoryItemListLevelStateObject;
var HistoryItemListLevelUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemListLevelUseStateObject, _super);
    function HistoryItemListLevelUseStateObject(isAbstractNumberingList, numberingListIndex, listLevelIndex, value, use) {
        var _this = _super.call(this, isAbstractNumberingList, numberingListIndex, listLevelIndex, value) || this;
        _this.use = use;
        return _this;
    }
    HistoryItemListLevelUseStateObject.prototype.toJSON = function () {
        return _super.prototype.toJSON.call(this).concat([this.use ? 1 : 0]);
    };
    return HistoryItemListLevelUseStateObject;
}(HistoryItemListLevelStateObject));
exports.HistoryItemListLevelUseStateObject = HistoryItemListLevelUseStateObject;
var HistoryItemTableStateObject = (function () {
    function HistoryItemTableStateObject(tableStartPosition, tableNestedLevel, tableIndex, value) {
        this.tableIndex = tableIndex;
        this.value = value;
        this.tableStartPosition = tableStartPosition;
        this.tableNestedLevel = tableNestedLevel;
    }
    HistoryItemTableStateObject.prototype.toJSON = function () {
        return [this.tableStartPosition, this.tableNestedLevel, this.getPropertyValueForJSON(this.value)];
    };
    HistoryItemTableStateObject.prototype.canMerge = function (_obj) {
        return false;
    };
    HistoryItemTableStateObject.prototype.merge = function (_obj) {
        throw new Error(errors_1.Errors.NotImplemented);
    };
    HistoryItemTableStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof table_units_1.TableWidthUnit)
            return [value.type, value.value];
        if (value instanceof table_units_1.TableHeightUnit)
            return [value.type, value.value];
        if (value instanceof border_info_1.BorderInfo)
            return [value.color.toJSON(), value.frame, value.offset, value.shadow, value.style, value.width];
        return value;
    };
    return HistoryItemTableStateObject;
}());
exports.HistoryItemTableStateObject = HistoryItemTableStateObject;
var HistoryItemTableUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTableUseStateObject, _super);
    function HistoryItemTableUseStateObject(tableStartPosition, tableNestedLevel, tableIndex, value, use) {
        var _this = _super.call(this, tableStartPosition, tableNestedLevel, tableIndex, value) || this;
        _this.use = use;
        return _this;
    }
    HistoryItemTableUseStateObject.prototype.toJSON = function () {
        return _super.prototype.toJSON.call(this).concat([this.use ? 1 : 0]);
    };
    return HistoryItemTableUseStateObject;
}(HistoryItemTableStateObject));
exports.HistoryItemTableUseStateObject = HistoryItemTableUseStateObject;
var HistoryItemTableComplexUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTableComplexUseStateObject, _super);
    function HistoryItemTableComplexUseStateObject(tableStartPosition, tableNestedLevel, tableIndex, value, uses) {
        var _this = _super.call(this, tableStartPosition, tableNestedLevel, tableIndex, value) || this;
        _this.uses = uses;
        return _this;
    }
    HistoryItemTableComplexUseStateObject.prototype.toJSON = function () {
        var uses = [];
        var usesLength = this.uses.length;
        for (var i = 0; i < usesLength; i++)
            uses.push(this.uses[i] ? 1 : 0);
        return _super.prototype.toJSON.call(this).concat([uses]);
    };
    HistoryItemTableComplexUseStateObject.prototype.getPropertyValueForJSON = function (value) {
        var result = [];
        var length = value.length;
        for (var i = 0; i < length; i++)
            result.push(_super.prototype.getPropertyValueForJSON.call(this, value[i]));
        return result;
    };
    return HistoryItemTableComplexUseStateObject;
}(HistoryItemTableStateObject));
exports.HistoryItemTableComplexUseStateObject = HistoryItemTableComplexUseStateObject;
var HistoryItemTableCellStateObject = (function () {
    function HistoryItemTableCellStateObject(tableStartPosition, tableNestedLevel, tableIndex, rowIndex, cellIndex, value) {
        this.tableIndex = tableIndex;
        this.rowIndex = rowIndex;
        this.cellIndex = cellIndex;
        this.value = value;
        this.tableStartPosition = tableStartPosition;
        this.tableNestedLevel = tableNestedLevel;
    }
    HistoryItemTableCellStateObject.prototype.toJSON = function () {
        return [this.tableStartPosition, this.tableNestedLevel, this.rowIndex, this.cellIndex, this.getPropertyValueForJSON(this.value)];
    };
    HistoryItemTableCellStateObject.prototype.canMerge = function (_obj) {
        return false;
    };
    HistoryItemTableCellStateObject.prototype.merge = function (_obj) {
        throw new Error(errors_1.Errors.NotImplemented);
    };
    HistoryItemTableCellStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof table_units_1.TableWidthUnit)
            return [value.type, value.value];
        if (value instanceof table_units_1.TableHeightUnit)
            return [value.type, value.value];
        if (value instanceof border_info_1.BorderInfo)
            return [value.color.toJSON(), value.frame, value.offset, value.shadow, value.style, value.width];
        return value;
    };
    return HistoryItemTableCellStateObject;
}());
exports.HistoryItemTableCellStateObject = HistoryItemTableCellStateObject;
var HistoryItemTableCellUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTableCellUseStateObject, _super);
    function HistoryItemTableCellUseStateObject(tableStartPosition, tableNestedLevel, tableIndex, rowIndex, cellIndex, value, use) {
        var _this = _super.call(this, tableStartPosition, tableNestedLevel, tableIndex, rowIndex, cellIndex, value) || this;
        _this.use = use;
        return _this;
    }
    HistoryItemTableCellUseStateObject.prototype.toJSON = function () {
        return _super.prototype.toJSON.call(this).concat([this.use ? 1 : 0]);
    };
    return HistoryItemTableCellUseStateObject;
}(HistoryItemTableCellStateObject));
exports.HistoryItemTableCellUseStateObject = HistoryItemTableCellUseStateObject;
var HistoryItemTableCellComplexUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTableCellComplexUseStateObject, _super);
    function HistoryItemTableCellComplexUseStateObject(tableStartPosition, tableNestedLevel, tableIndex, rowIndex, cellIndex, value, uses) {
        var _this = _super.call(this, tableStartPosition, tableNestedLevel, tableIndex, rowIndex, cellIndex, value) || this;
        _this.uses = uses;
        return _this;
    }
    HistoryItemTableCellComplexUseStateObject.prototype.toJSON = function () {
        var uses = [];
        var usesLength = this.uses.length;
        for (var i = 0; i < usesLength; i++) {
            uses.push(this.uses[i] ? 1 : 0);
        }
        return _super.prototype.toJSON.call(this).concat([uses]);
    };
    HistoryItemTableCellComplexUseStateObject.prototype.getPropertyValueForJSON = function (value) {
        var result = [];
        var length = value.length;
        for (var i = 0; i < length; i++)
            result.push(_super.prototype.getPropertyValueForJSON.call(this, value[i]));
        return result;
    };
    return HistoryItemTableCellComplexUseStateObject;
}(HistoryItemTableCellStateObject));
exports.HistoryItemTableCellComplexUseStateObject = HistoryItemTableCellComplexUseStateObject;
var HistoryItemTableRowStateObject = (function () {
    function HistoryItemTableRowStateObject(tableStartPosition, tableNestedLevel, tableIndex, rowIndex, value) {
        this.tableIndex = tableIndex;
        this.rowIndex = rowIndex;
        this.value = value;
        this.tableStartPosition = tableStartPosition;
        this.tableNestedLevel = tableNestedLevel;
    }
    HistoryItemTableRowStateObject.prototype.toJSON = function () {
        return [this.tableStartPosition, this.tableNestedLevel, this.rowIndex, this.getPropertyValueForJSON(this.value)];
    };
    HistoryItemTableRowStateObject.prototype.canMerge = function (_obj) {
        return false;
    };
    HistoryItemTableRowStateObject.prototype.merge = function (_obj) {
        throw new Error(errors_1.Errors.NotImplemented);
    };
    HistoryItemTableRowStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof table_units_1.TableWidthUnit)
            return [value.type, value.value];
        if (value instanceof table_units_1.TableHeightUnit)
            return [value.type, value.value];
        return value;
    };
    return HistoryItemTableRowStateObject;
}());
exports.HistoryItemTableRowStateObject = HistoryItemTableRowStateObject;
var HistoryItemTableRowUseStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemTableRowUseStateObject, _super);
    function HistoryItemTableRowUseStateObject(tableStartPosition, tableNestedLevel, tableIndex, rowIndex, value, use) {
        var _this = _super.call(this, tableStartPosition, tableNestedLevel, tableIndex, rowIndex, value) || this;
        _this.use = use;
        return _this;
    }
    HistoryItemTableRowUseStateObject.prototype.toJSON = function () {
        return _super.prototype.toJSON.call(this).concat([this.use ? 1 : 0]);
    };
    return HistoryItemTableRowUseStateObject;
}(HistoryItemTableRowStateObject));
exports.HistoryItemTableRowUseStateObject = HistoryItemTableRowUseStateObject;
var HistoryItemIntervalAnchorInfoStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalAnchorInfoStateObject, _super);
    function HistoryItemIntervalAnchorInfoStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalAnchorInfoStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_anchor_info_converter_1.JSONAnchorInfoConverter.convertToJSON(value);
    };
    return HistoryItemIntervalAnchorInfoStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalAnchorInfoStateObject = HistoryItemIntervalAnchorInfoStateObject;
var HistoryItemIntervalAnchorSizeStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalAnchorSizeStateObject, _super);
    function HistoryItemIntervalAnchorSizeStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalAnchorSizeStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof sizes_1.PictureSize)
            return json_inline_picture_size_converter_1.JSONInlinePictureSizeConverterConverter.convertToJSON(value);
        if (value instanceof sizes_1.AnchorTextBoxSize)
            return json_anchor_text_box_size_converter_1.JSONAnchorTextBoxSizeConverterConverter.convertToJSON(value);
        return value;
    };
    return HistoryItemIntervalAnchorSizeStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalAnchorSizeStateObject = HistoryItemIntervalAnchorSizeStateObject;
var HistoryItemIntervalShapeStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalShapeStateObject, _super);
    function HistoryItemIntervalShapeStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalShapeStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_shape_converter_1.JSONShapeConverter.convertToJSON(value);
    };
    return HistoryItemIntervalShapeStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalShapeStateObject = HistoryItemIntervalShapeStateObject;
var HistoryItemIntervalAnchoredTextBoxPropertiesStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemIntervalAnchoredTextBoxPropertiesStateObject, _super);
    function HistoryItemIntervalAnchoredTextBoxPropertiesStateObject(interval, properties) {
        return _super.call(this, interval, properties) || this;
    }
    HistoryItemIntervalAnchoredTextBoxPropertiesStateObject.prototype.getPropertyValueForJSON = function (value) {
        return json_text_box_properties_converter_1.JSONTextBoxPropertiesConverter.convertToJSON(value);
    };
    return HistoryItemIntervalAnchoredTextBoxPropertiesStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemIntervalAnchoredTextBoxPropertiesStateObject = HistoryItemIntervalAnchoredTextBoxPropertiesStateObject;
var HistoryItemInlineObjectPropertiesStateObject = (function (_super) {
    tslib_1.__extends(HistoryItemInlineObjectPropertiesStateObject, _super);
    function HistoryItemInlineObjectPropertiesStateObject() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HistoryItemInlineObjectPropertiesStateObject.prototype.getPropertyValueForJSON = function (value) {
        if (value instanceof size_1.Size)
            return [value.width, value.height];
        else
            return value;
    };
    return HistoryItemInlineObjectPropertiesStateObject;
}(HistoryItemIntervalStateObject));
exports.HistoryItemInlineObjectPropertiesStateObject = HistoryItemInlineObjectPropertiesStateObject;
