// devexpress-richedit/lib/core/model/numbering-lists/numbering-list-counters-calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NumberingListCountersCalculator = (function () {
    function NumberingListCountersCalculator(list) {
        this.counters = [];
        this.usedListIndecies = {};
        this.currentParagraphIndex = -1;
        this.list = list;
        for (var i = 0, listLevel; listLevel = this.list.levels[i]; i++) {
            this.counters.push(listLevel.getListLevelProperties().start - 1);
        }
    }
    NumberingListCountersCalculator.prototype.calculateCounters = function (paragraph) {
        this.currentParagraphIndex = -1;
        return this.calculateNextCounters(paragraph);
    };
    NumberingListCountersCalculator.prototype.calculateNextCounters = function (paragraph) {
        var paragraphs = paragraph.subDocument.paragraphs;
        if (paragraphs[this.currentParagraphIndex] !== paragraph) {
            var abstractNumberingList = paragraph.getAbstractNumberingList();
            for (var i = this.currentParagraphIndex + 1, currentParagraph; currentParagraph = paragraph.subDocument.paragraphs[i]; i++) {
                this.currentParagraphIndex++;
                if (this.shouldAdvanceListLevelCounters(currentParagraph, abstractNumberingList))
                    this.advanceListLevelCounters(currentParagraph, currentParagraph.getListLevelIndex());
                if (currentParagraph === paragraph)
                    break;
            }
        }
        return this.getActualRangeCounters(paragraph.getListLevelIndex());
    };
    NumberingListCountersCalculator.prototype.getLastCounters = function (paragraph) {
        return this.getActualRangeCounters(paragraph.getListLevelIndex());
    };
    NumberingListCountersCalculator.prototype.shouldAdvanceListLevelCounters = function (paragraph, abstractNumberingList) {
        return paragraph.getAbstractNumberingList() === abstractNumberingList;
    };
    NumberingListCountersCalculator.prototype.advanceListLevelCounters = function (paragraph, listLevelIndex) {
        var numberingListIndex = paragraph.getNumberingListIndex();
        var numberingList = paragraph.subDocument.documentModel.numberingLists[numberingListIndex];
        var level = numberingList.levels[listLevelIndex];
        if (level.overrideStart && !this.usedListIndecies[numberingListIndex]) {
            this.usedListIndecies[numberingListIndex] = true;
            this.counters[listLevelIndex] = level.getNewStart();
        }
        else
            this.counters[listLevelIndex]++;
        this.advanceSkippedLevelCounters(listLevelIndex);
        this.restartNextLevelCounters(listLevelIndex);
    };
    NumberingListCountersCalculator.prototype.advanceSkippedLevelCounters = function (listLevelIndex) {
        for (var i = 0; i < listLevelIndex; i++) {
            var listLevel = this.list.levels[i];
            if (this.counters[i] == listLevel.getListLevelProperties().start - 1)
                this.counters[i]++;
        }
    };
    NumberingListCountersCalculator.prototype.restartNextLevelCounters = function (listLevelIndex) {
        var restartedLevels = [];
        restartedLevels[listLevelIndex] = true;
        var countersLength = this.counters.length;
        for (var i = listLevelIndex + 1; i < countersLength; i++) {
            var listLevelProperties = this.list.levels[i].getListLevelProperties();
            if (!listLevelProperties.suppressRestart) {
                var restartLevel = i - listLevelProperties.relativeRestartLevel - 1;
                if (restartLevel >= 0 && restartLevel < countersLength && restartedLevels[restartLevel]) {
                    this.counters[i] = listLevelProperties.start - 1;
                    restartedLevels[i] = true;
                }
            }
        }
    };
    NumberingListCountersCalculator.prototype.getActualRangeCounters = function (listLevelIndex) {
        var rangeCounters = [];
        for (var i = 0; i <= listLevelIndex; i++)
            rangeCounters[i] = this.counters[i];
        return rangeCounters;
    };
    return NumberingListCountersCalculator;
}());
exports.NumberingListCountersCalculator = NumberingListCountersCalculator;
