// devexpress-richedit/lib/core/model/numbering-lists/numbering-list-index-calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NumberingListIndexCalculator = (function () {
    function NumberingListIndexCalculator(subDocument, targetType, startIndex) {
        this.subDocument = subDocument;
        this.targetType = targetType;
        this.startIndex = startIndex;
    }
    NumberingListIndexCalculator.prototype.getTargetListInfo = function (paragraphIndices) {
        if (this.startIndex === 1)
            return null;
        var targetParagraph;
        var startParagraphIndex = paragraphIndices[0];
        for (var i = startParagraphIndex - 1; i >= 0; i--) {
            if (this.hasSameParagraphType(i)) {
                var targetParagraph = this.subDocument.paragraphs[i];
                if (this.startIndex > 1) {
                    var listCounters = this.subDocument.documentModel.getRangeListCounters(targetParagraph);
                    if (listCounters.length !== 1 || listCounters[0] !== this.startIndex - 1)
                        return null;
                }
                break;
            }
        }
        if (!targetParagraph && this.hasSameParagraphType(startParagraphIndex + 1))
            targetParagraph = this.subDocument.paragraphs[startParagraphIndex + 1];
        if (targetParagraph) {
            return {
                listIndex: targetParagraph.getNumberingListIndex(),
                listlevelIndex: Math.max(0, targetParagraph.getListLevelIndex())
            };
        }
        return null;
    };
    NumberingListIndexCalculator.prototype.hasSameParagraphType = function (paragraphIndex) {
        var paragraph = this.subDocument.paragraphs[paragraphIndex];
        return paragraph && paragraph.isInList() && paragraph.getNumberingList().getListType() === this.targetType;
    };
    return NumberingListIndexCalculator;
}());
exports.NumberingListIndexCalculator = NumberingListIndexCalculator;
