// devexpress-richedit/lib/core/model/numbering-lists/numbering-list-id-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var number_1 = require("@devexpress/utils/lib/utils/map/number");
var ListIdProviderBase = (function () {
    function ListIdProviderBase(documentModel) {
        this.documentModel = documentModel;
    }
    return ListIdProviderBase;
}());
exports.ListIdProviderBase = ListIdProviderBase;
var NumberingListIdProvider = (function (_super) {
    tslib_1.__extends(NumberingListIdProvider, _super);
    function NumberingListIdProvider() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.lastId = 0;
        return _this;
    }
    NumberingListIdProvider.prototype.getNextId = function () {
        this.lastId++;
        return this.lastId;
    };
    NumberingListIdProvider.prototype.clone = function (model) {
        var result = new NumberingListIdProvider(model);
        result.lastId = this.lastId;
        return result;
    };
    return NumberingListIdProvider;
}(ListIdProviderBase));
exports.NumberingListIdProvider = NumberingListIdProvider;
var AbstractNumberingListIdProvider = (function (_super) {
    tslib_1.__extends(AbstractNumberingListIdProvider, _super);
    function AbstractNumberingListIdProvider() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.lastId = 0;
        return _this;
    }
    AbstractNumberingListIdProvider.prototype.getMap = function () {
        if (!this.map) {
            this.map = {};
            for (var i = 0, list; list = this.documentModel.abstractNumberingLists[i]; i++)
                this.map[list.innerId] = true;
        }
        return this.map;
    };
    AbstractNumberingListIdProvider.prototype.getNextId = function () {
        var map = this.getMap();
        do {
            this.lastId++;
        } while (map[this.lastId]);
        map[this.lastId] = true;
        return this.lastId;
    };
    AbstractNumberingListIdProvider.prototype.clone = function (model) {
        var result = new AbstractNumberingListIdProvider(model);
        result.map = number_1.NumberMapUtils.shallowCopy(this.map);
        return result;
    };
    return AbstractNumberingListIdProvider;
}(ListIdProviderBase));
exports.AbstractNumberingListIdProvider = AbstractNumberingListIdProvider;
