// devexpress-richedit/lib/document-processor/docvar-args.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var EventArgs = (function () {
    function EventArgs() {
    }
    return EventArgs;
}());
exports.EventArgs = EventArgs;
var CalculateDocumentVariableEventArgs = (function (_super) {
    tslib_1.__extends(CalculateDocumentVariableEventArgs, _super);
    function CalculateDocumentVariableEventArgs(fieldInterval, variableName, args) {
        var _this = _super.call(this) || this;
        _this.fieldInterval = fieldInterval;
        _this.variableName = variableName;
        _this.args = args;
        _this.value = null;
        _this.keepLastParagraph = false;
        return _this;
    }
    return CalculateDocumentVariableEventArgs;
}(EventArgs));
exports.CalculateDocumentVariableEventArgs = CalculateDocumentVariableEventArgs;
var DocumentVariableData = (function () {
    function DocumentVariableData(callback, fieldInterval, variableName, args) {
        this.callback = callback;
        this.fieldInterval = fieldInterval;
        this.variableName = variableName;
        this.args = args;
    }
    return DocumentVariableData;
}());
exports.DocumentVariableData = DocumentVariableData;
var CalculateDocumentVariableAsyncEventArgs = (function (_super) {
    tslib_1.__extends(CalculateDocumentVariableAsyncEventArgs, _super);
    function CalculateDocumentVariableAsyncEventArgs(data) {
        var _this = _super.call(this) || this;
        _this.data = data;
        return _this;
    }
    return CalculateDocumentVariableAsyncEventArgs;
}(EventArgs));
exports.CalculateDocumentVariableAsyncEventArgs = CalculateDocumentVariableAsyncEventArgs;
