// devexpress-richedit/lib/base/selection/input-position-model-changes-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var enums_1 = require("../../core/model/changes/enums");
var batch_updatable_1 = require("@devexpress/utils/lib/class/batch-updatable");
var flag_1 = require("@devexpress/utils/lib/class/flag");
var InputPositionEventType;
(function (InputPositionEventType) {
    InputPositionEventType[InputPositionEventType["None"] = 0] = "None";
    InputPositionEventType[InputPositionEventType["FullReset"] = 1] = "FullReset";
    InputPositionEventType[InputPositionEventType["ResetSectionProperties"] = 2] = "ResetSectionProperties";
    InputPositionEventType[InputPositionEventType["CheckSelectionChange"] = 4] = "CheckSelectionChange";
})(InputPositionEventType || (InputPositionEventType = {}));
var InputPositionModelChangesListener = (function (_super) {
    tslib_1.__extends(InputPositionModelChangesListener, _super);
    function InputPositionModelChangesListener(inputPosition, selection) {
        var _this = _super.call(this) || this;
        _this.updateEnabled = true;
        _this.inputPosition = inputPosition;
        _this.selection = selection;
        return _this;
    }
    InputPositionModelChangesListener.prototype.onUpdateUnlocked = function (occurredEvents) {
        this.applyEvent(occurredEvents);
    };
    InputPositionModelChangesListener.prototype.NotifySelectionChanged = function () {
        this.applyEvent(this.getSelectionChangeEvent());
    };
    InputPositionModelChangesListener.prototype.modelChanged = function (change) {
        var occurredEvents = this.getModelChangeEvent(change);
        if (this.isUpdateLocked())
            this.registerOccurredEvent(occurredEvents);
        else
            this.applyEvent(occurredEvents);
    };
    InputPositionModelChangesListener.prototype.applyEvent = function (occurredEvents) {
        if (!this.updateEnabled)
            return;
        var flag = new flag_1.Flag(occurredEvents);
        if (flag.get(InputPositionEventType.CheckSelectionChange))
            flag.add(this.getSelectionChangeEvent());
        if (flag.get(InputPositionEventType.FullReset))
            this.inputPosition.reset();
        else if (flag.get(InputPositionEventType.ResetSectionProperties))
            this.inputPosition.resetSectionMergedProperties();
    };
    InputPositionModelChangesListener.prototype.getSelectionChangeEvent = function () {
        return this.selection.shouldResetInputPosition() ?
            InputPositionEventType.FullReset :
            InputPositionEventType.None;
        ;
    };
    InputPositionModelChangesListener.prototype.getModelChangeEvent = function (change) {
        switch (change.type) {
            case enums_1.ModelChangeType.DifferentOddAndEvenPages:
            case enums_1.ModelChangeType.PageColor:
                return InputPositionEventType.ResetSectionProperties;
            case enums_1.ModelChangeType.DefaultTabWidth:
            case enums_1.ModelChangeType.HeaderFooterCreated:
            case enums_1.ModelChangeType.HeaderFooterIndexChanged:
            case enums_1.ModelChangeType.LoadFontInfo:
            case enums_1.ModelChangeType.SectionFormattingChanged:
            case enums_1.ModelChangeType.CreateStyleLink:
            case enums_1.ModelChangeType.DeleteStyleLink:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.AbstractNumberingListAdded:
            case enums_1.ModelChangeType.AbstractNumberingListDeleted:
            case enums_1.ModelChangeType.NumberingListAdded:
            case enums_1.ModelChangeType.NumberingListDeleted:
            case enums_1.ModelChangeType.ListLevelPropertyChanged:
            case enums_1.ModelChangeType.ListLevelParagraphPropertyChanged:
            case enums_1.ModelChangeType.ListLevelCharacterPropertyChanged:
            case enums_1.ModelChangeType.IOverrideListLevelChanged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.BookmarkCreated:
            case enums_1.ModelChangeType.BookmarkDeleted:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.TabInserted:
            case enums_1.ModelChangeType.TabDeleted:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.LoadPicturesInfo:
            case enums_1.ModelChangeType.InlinePicturesUpdated:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.InlinePictureInserted:
            case enums_1.ModelChangeType.AnchoredPictureInserted:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.AnchoredPictureSizeChanged:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.InlineObjectRunPropertyChanged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.FieldInserted:
            case enums_1.ModelChangeType.FieldDeleted:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.HyperlinkInfoChanged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.FieldsShowCodeChanged:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.SimpleRunInserted:
            case enums_1.ModelChangeType.TextBufferChanged:
            case enums_1.ModelChangeType.ParagraphInserted:
            case enums_1.ModelChangeType.SectionInserted:
            case enums_1.ModelChangeType.AnchoredTextBoxInserted:
            case enums_1.ModelChangeType.AnchorObjectRemoved:
            case enums_1.ModelChangeType.IntervalRemoved:
            case enums_1.ModelChangeType.ParagraphMerged:
            case enums_1.ModelChangeType.SectionMerged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.CharacterFormattingChanged:
            case enums_1.ModelChangeType.CharacterPropertiesChanged:
            case enums_1.ModelChangeType.ParagraphFormattingChanged:
            case enums_1.ModelChangeType.ParagraphPropertiesChanged:
            case enums_1.ModelChangeType.ParagraphAndCharacterMergedPropertiesReset:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.AnchoredTextBoxSizeChanged:
                return InputPositionEventType.None;
            case enums_1.ModelChangeType.AnchoredTextBoxPropertiesChanged:
            case enums_1.ModelChangeType.AnchorInfoPropertyChanged:
            case enums_1.ModelChangeType.ShapeChanged:
            case enums_1.ModelChangeType.ShapePropertyChanged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.CharacterStyleApplied:
            case enums_1.ModelChangeType.ParagraphStyleApplied:
            case enums_1.ModelChangeType.TableStyleChanged:
            case enums_1.ModelChangeType.ParagraphNumberingListChanged:
                return InputPositionEventType.FullReset;
            case enums_1.ModelChangeType.TableCreated:
            case enums_1.ModelChangeType.TableRemoved:
            case enums_1.ModelChangeType.TableStartPositionShifted:
            case enums_1.ModelChangeType.TableCellPropertyChanged:
            case enums_1.ModelChangeType.TablePropertyChanged:
            case enums_1.ModelChangeType.TableRowPropertyChanged:
            case enums_1.ModelChangeType.TableCellSplittedHorizontally:
            case enums_1.ModelChangeType.TableCellMergedHorizontally:
            case enums_1.ModelChangeType.TableRowInserted:
            case enums_1.ModelChangeType.TableRowRemoved:
            case enums_1.ModelChangeType.TableCellRemoved:
            case enums_1.ModelChangeType.TableCellInserted:
                return InputPositionEventType.FullReset;
            default:
                return InputPositionEventType.None;
        }
    };
    return InputPositionModelChangesListener;
}(batch_updatable_1.BatchUpdatableObject));
exports.InputPositionModelChangesListener = InputPositionModelChangesListener;
