// devexpress-richedit/lib/base/rich-utils/debug/chrome-dev-tools-custom-formatters/short-string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pair_1 = require("@devexpress/utils/lib/class/pair");
var base_formatter_1 = require("./base-formatter");
var elements_1 = require("./elements");
var ShortString = (function () {
    function ShortString(str, shortLen) {
        if (shortLen === void 0) { shortLen = ShortString.SHORT_LEN; }
        this.str = str;
        this.setLen(shortLen);
    }
    Object.defineProperty(ShortString.prototype, "fullString", {
        get: function () { return new elements_1.SpanElement().setText("\"" + this.str + "\"").setColor(base_formatter_1.BaseFormatter.stringColor); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShortString.prototype, "shortString", {
        get: function () {
            if (!this.canShowAsShort())
                return this.fullString;
            return new elements_1.SpanElement()
                .setNode(new elements_1.SpanElement().setText("\"" + this.str.substr(0, this.halfOfShortLen) + "").setColor(base_formatter_1.BaseFormatter.stringColor))
                .setNode(new elements_1.SpanElement().setText("~").setColor(base_formatter_1.BaseFormatter.blackColor))
                .setNode(new elements_1.SpanElement().setText(this.str.substring(this.str.length - this.halfOfShortLen) + "\"").setColor(base_formatter_1.BaseFormatter.stringColor));
        },
        enumerable: true,
        configurable: true
    });
    ShortString.prototype.setLen = function (shortLen) {
        this.halfOfShortLen = Math.floor(shortLen / 2);
        return this;
    };
    ShortString.prototype.canShowAsShort = function () {
        return this.str.length > this.halfOfShortLen * 2;
    };
    ShortString.SHORT_LEN = 14;
    return ShortString;
}());
exports.ShortString = ShortString;
var CCF_ShortString = (function (_super) {
    tslib_1.__extends(CCF_ShortString, _super);
    function CCF_ShortString() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CCF_ShortString.prototype.isHandleObject = function (obj) {
        return obj instanceof ShortString;
    };
    CCF_ShortString.prototype.getShortDescription = function (_config) {
        return this.stdShow(new base_formatter_1.StdProps([new pair_1.Pair("", this.curr.shortString)]).showAsLine().setBoundChars("", ""));
    };
    CCF_ShortString.prototype.availableFullDescription = function (_config) {
        return this.curr.canShowAsShort();
    };
    CCF_ShortString.prototype.getFullDescription = function (_config) {
        return this.stdShow(new base_formatter_1.StdProps([new pair_1.Pair("", this.curr.fullString)]).showAsColumn());
    };
    CCF_ShortString._foo = base_formatter_1.BaseFormatter.addToFormattersList(new CCF_ShortString());
    return CCF_ShortString;
}(base_formatter_1.BaseFormatter));
exports.CCF_ShortString = CCF_ShortString;
