// devexpress-richedit/lib/rtf/import/destination/fields/code-field-destination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var destination_type_1 = require("../utils/destination-type");
var field_sub_destination_1 = require("./field-sub-destination");
var CodeFieldDestination = (function (_super) {
    tslib_1.__extends(CodeFieldDestination, _super);
    function CodeFieldDestination() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(CodeFieldDestination.prototype, "destinationType", {
        get: function () { return destination_type_1.DestinationType.CodeFieldDestination; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeFieldDestination.prototype, "currentInfo", {
        get: function () { return this.importer.importers.field.fields.last; },
        enumerable: true,
        configurable: true
    });
    CodeFieldDestination.prototype.createInstance = function () {
        return new CodeFieldDestination(this.importer);
    };
    CodeFieldDestination.prototype.processTextCore = function (text) {
        if (this.shouldSkipField())
            return;
        if (this.detectShapeField(text)) {
            this.currentInfo.isShapeField = true;
        }
        _super.prototype.processTextCore.call(this, text);
    };
    CodeFieldDestination.prototype.processCharCore = function (ch) {
        if (this.shouldSkipField())
            return;
        _super.prototype.processCharCore.call(this, ch);
    };
    CodeFieldDestination.prototype.startNewField = function () {
        if (this.shouldSkipField())
            return;
        _super.prototype.startNewField.call(this);
    };
    CodeFieldDestination.prototype.detectShapeField = function (text) {
        return text.toUpperCase().indexOf("SHAPE") >= 0;
    };
    CodeFieldDestination.prototype.shouldSkipField = function () {
        return this.currentInfo.isShapeField;
    };
    return CodeFieldDestination;
}(field_sub_destination_1.FieldSubDestination));
exports.CodeFieldDestination = CodeFieldDestination;
