// devexpress-dashboard/designer/conditional-formatting/custom-style-settings/_popover-color-picker.js
/**
* DevExpress Dashboard (_popover-color-picker.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _default_1 = require("../../../data/localization/_default");
var PopoverColorPicker = (function () {
    function PopoverColorPicker(setColor) {
        var _this = this;
        this.setColor = setColor;
        this.color = ko.observable('');
        this.visible = ko.observable(false);
        this.buttonItems = ko.observableArray([
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: _default_1.getLocalizationById('DashboardWebStringId.ButtonConfirm'), onClick: function () {
                        _this.setColor(_this.color());
                        _this.visible(false);
                    } } },
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: _default_1.getLocalizationById('DashboardStringId.ButtonCancel'), onClick: function () { return _this.visible(false); } } }
        ]);
    }
    PopoverColorPicker.prototype.show = function (color) {
        this.color(color);
        this.visible(true);
    };
    return PopoverColorPicker;
}());
exports.PopoverColorPicker = PopoverColorPicker;
