// devexpress-richedit/lib/rtf/import/keyword-tables/sub-document/sub-document-keywords.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var map_creator_1 = require("../../../../base-utils/map-creator");
var form_field_properties_1 = require("../../../../core/model/fields/form-field-properties");
var rtf_drawing_keywords_1 = require("../../../translation-table/rtf-drawing-keywords");
var text_after_destination_1 = require("../../destination/base/text-after-destination");
var text_before_destination_1 = require("../../destination/base/text-before-destination");
var bookmark_end_destination_1 = require("../../destination/bookmark/bookmark-end-destination");
var bookmark_start_destination_1 = require("../../destination/bookmark/bookmark-start-destination");
var document_variable_destination_1 = require("../../destination/document-variable-destination");
var form_field_destination_1 = require("../../destination/fields/form-field-destination");
var mail_merge_destination_1 = require("../../destination/fields/mail-merge-destination");
var destination_old_paragraph_numbering_1 = require("../../destination/numbering-list/destination-old-paragraph-numbering");
var destination_old_section_numbering_level_1 = require("../../destination/numbering-list/destination-old-section-numbering-level");
var picture_destination_1 = require("../../destination/picture/picture-destination");
var range_permission_end_destination_1 = require("../../destination/range-permission/range-permission-end-destination");
var range_permission_start_destination_1 = require("../../destination/range-permission/range-permission-start-destination");
var destination_sub_document_1 = require("../../destination/sub-document/destination-sub-document");
var character_properties_1 = require("../character-properties");
var paragraph_properties_1 = require("../paragraph-properties");
var table_keywords_1 = require("./table-keywords");
exports.addCommonSymbolsAndObjectsKeywords = new map_creator_1.MapCreator()
    .add("pict", function (data) { return data.destination = new picture_destination_1.PictureDestination(data); })
    .add("shp", destination_sub_document_1.DestinationSubDocument.onShapeKeyword)
    .add("shppict", destination_sub_document_1.DestinationSubDocument.onShapePictureKeyword)
    .add("nonshppict", destination_sub_document_1.DestinationSubDocument.onNonShapePictureKeyword)
    .add(rtf_drawing_keywords_1.RtfDrawingKeywords.ShapeGroup, destination_sub_document_1.DestinationSubDocument.onShapeGroupKeyword)
    .add("line", destination_sub_document_1.DestinationSubDocument.onLineBreakKeyword)
    .add("u", destination_sub_document_1.DestinationSubDocument.onUnicodeKeyword)
    .add("tab", destination_sub_document_1.DestinationSubDocument.onTabKeyword)
    .add("emdash", destination_sub_document_1.DestinationSubDocument.onEmDashKeyword)
    .add("endash", destination_sub_document_1.DestinationSubDocument.onEnDashKeyword)
    .add("lquote", destination_sub_document_1.DestinationSubDocument.onLeftSingleQuoteKeyword)
    .add("rquote", destination_sub_document_1.DestinationSubDocument.onRightSingleQuoteKeyword)
    .add("ldblquote", destination_sub_document_1.DestinationSubDocument.onLeftDoubleQuoteKeyword)
    .add("rdblquote", destination_sub_document_1.DestinationSubDocument.onRightDoubleQuoteKeyword)
    .add("bullet", destination_sub_document_1.DestinationSubDocument.onBulletKeyword)
    .add("emspace", destination_sub_document_1.DestinationSubDocument.onEmSpaceKeyword)
    .add("enspace", destination_sub_document_1.DestinationSubDocument.onEnSpaceKeyword)
    .add("qmspace", destination_sub_document_1.DestinationSubDocument.onQmSpaceKeyword)
    .add("field", destination_sub_document_1.DestinationSubDocument.onFieldStartKeyword)
    .add("formfield", function (data) {
    if (data.importers.field.fields.count > 0) {
        data.importers.field.fields.peek().formFieldProperties = new form_field_properties_1.FormFieldProperties();
        data.destination = new form_field_destination_1.FormFieldDestination(data);
    }
})
    .add("tc", destination_sub_document_1.DestinationSubDocument.onTableOfContentsEntryKeyword)
    .add("tcl", destination_sub_document_1.DestinationSubDocument.onTableOfContentsEntryLevelNumberKeyword)
    .add("tcf", destination_sub_document_1.DestinationSubDocument.onTableOfContentsEntryTypeTableKeyword)
    .add("mailmerge", function (data) { return data.destination = new mail_merge_destination_1.MailMergeDestination(data); })
    .add("bkmkstart", function (data) { return data.destination = new bookmark_start_destination_1.BookmarkStartDestination(data); })
    .add("bkmkend", function (data) { return data.destination = new bookmark_end_destination_1.BookmarkEndDestination(data); })
    .add("protstart", function (data) { return data.destination = new range_permission_start_destination_1.RangePermissionStartDestination(data); })
    .add("protend", function (data) { return data.destination = new range_permission_end_destination_1.RangePermissionEndDestination(data); })
    .add("docvar", function (data) { return data.destination = new document_variable_destination_1.DocumentVariableDestination(data); })
    .add("dxcustomrundata", destination_sub_document_1.DestinationSubDocument.onDxCustomRunDataKeyword)
    .add("zwj", destination_sub_document_1.DestinationSubDocument.onZeroWidthJoiner)
    .add("zwnj", destination_sub_document_1.DestinationSubDocument.onZeroWidthNonJoiner)
    .add("zwbo", destination_sub_document_1.DestinationSubDocument.onZeroWidthBreakOpportunity)
    .add("zwnbo", destination_sub_document_1.DestinationSubDocument.onZeroWidthNonBreakOpportunity)
    .get();
exports.addCommonNumberingListsKeywords = new map_creator_1.MapCreator()
    .add("ls", destination_sub_document_1.DestinationSubDocument.onListOverride)
    .add("ilvl", destination_sub_document_1.DestinationSubDocument.onListLevel)
    .add("listtext", destination_sub_document_1.DestinationSubDocument.onListText)
    .add("pntext", destination_sub_document_1.DestinationSubDocument.onListText)
    .add("pnseclvl", function (data, parameterValue) { return data.destination = new destination_old_section_numbering_level_1.DestinationOldSectionNumberingLevel(data, parameterValue); })
    .add("pn", function (data) { return data.destination = new destination_old_paragraph_numbering_1.DestinationOldParagraphNumbering(data); })
    .add("pntxta", function (data) { return data.destination = new text_after_destination_1.TextAfterDestination(data); })
    .add("pntxtb", function (data) { return data.destination = new text_before_destination_1.TextBeforeDestination(data); })
    .get();
exports.addCommonTabKeywords = new map_creator_1.MapCreator()
    .add("tqr", destination_sub_document_1.DestinationSubDocument.onTabRightKeyword)
    .add("tqc", destination_sub_document_1.DestinationSubDocument.onTabCenterKeyword)
    .add("tqdec", destination_sub_document_1.DestinationSubDocument.onTabDecimalKeyword)
    .add("tldot", destination_sub_document_1.DestinationSubDocument.onTabLeaderDotsKeyword)
    .add("tlmdot", destination_sub_document_1.DestinationSubDocument.onTabLeaderMiddleDotsKeyword)
    .add("tlhyph", destination_sub_document_1.DestinationSubDocument.onTabLeaderHyphensKeyword)
    .add("tlul", destination_sub_document_1.DestinationSubDocument.onTabLeaderUnderlineKeyword)
    .add("tlth", destination_sub_document_1.DestinationSubDocument.onTabLeaderThickLineKeyword)
    .add("tleq", destination_sub_document_1.DestinationSubDocument.onTabLeaderEqualSignKeyword)
    .add("tx", destination_sub_document_1.DestinationSubDocument.onTabPositionKeyword)
    .add("tb", destination_sub_document_1.DestinationSubDocument.onBarTabKeyword)
    .get();
exports.addCommonCharacterKeywords = new map_creator_1.MapCreator()
    .add("plain", destination_sub_document_1.DestinationSubDocument.onPlainKeyword)
    .add("cs", destination_sub_document_1.DestinationSubDocument.onCharacterStyleIndex)
    .append(character_properties_1.characterPropertiesKeywords)
    .get();
exports.addCommonParagraphKeywords = new map_creator_1.MapCreator()
    .add("par", destination_sub_document_1.DestinationSubDocument.onParKeyword)
    .add("pard", destination_sub_document_1.DestinationSubDocument.onResetParagraphPropertiesKeyword)
    .add("s", destination_sub_document_1.DestinationSubDocument.onParagraphStyleIndex)
    .add("yts", destination_sub_document_1.DestinationSubDocument.onTableStyleIndexForRowOrCell)
    .add("ltrpar", destination_sub_document_1.DestinationSubDocument.onLeftToRightParagraphKeyword)
    .add("rtlpar", destination_sub_document_1.DestinationSubDocument.onRightToLeftParagraphKeyword)
    .append(paragraph_properties_1.addParagraphPropertiesKeywords)
    .get();
exports.subDocumentKeyword = new map_creator_1.MapCreator()
    .append(exports.addCommonCharacterKeywords)
    .append(exports.addCommonParagraphKeywords)
    .append(exports.addCommonSymbolsAndObjectsKeywords)
    .append(exports.addCommonTabKeywords)
    .append(exports.addCommonNumberingListsKeywords)
    .append(table_keywords_1.appendTableKeywords)
    .get();
