// devexpress-dashboard/designer/expression-editor/_filter-expression-editor.js
/**
* DevExpress Dashboard (_filter-expression-editor.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var ko = require("knockout");
var FilterExpressionEditorViewModel = (function () {
    function FilterExpressionEditorViewModel(params) {
        var _this = this;
        this.params = params;
        this.editorTemplate = 'dx-filtereditor-popup';
        this.criteriaString = ko.observable();
        this.editor = new analytics_widgets_1.FilterEditor(params.options, params.fieldListProvider, false, params.displayNameProvider);
        ko.computed(function () {
            var displayExpressionConverter = new analytics_internal_1.DisplayExpressionConverter(params.displayNameProvider);
            displayExpressionConverter.toDisplayExpression(params.options().path(), params.options().value()).done(function (result) {
                _this.criteriaString(result);
            }).fail(function () {
                _this.criteriaString(params.options().value());
            });
        });
    }
    FilterExpressionEditorViewModel.prototype.edit = function () {
        this.editor.popupVisible(true);
    };
    FilterExpressionEditorViewModel.prototype.remove = function () {
        this.params.options().value(null);
    };
    return FilterExpressionEditorViewModel;
}());
exports.FilterExpressionEditorViewModel = FilterExpressionEditorViewModel;
