// devexpress-richedit/lib/base/commands/command-states.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var SimpleCommandState = (function () {
    function SimpleCommandState(enabled, value) {
        this.visible = true;
        this.denyUpdateValue = false;
        this.enabled = enabled;
        this.value = value;
    }
    return SimpleCommandState;
}());
exports.SimpleCommandState = SimpleCommandState;
var IntervalCommandState = (function (_super) {
    tslib_1.__extends(IntervalCommandState, _super);
    function IntervalCommandState(enabled, interval, value) {
        var _this = _super.call(this, enabled, value) || this;
        _this.interval = interval;
        return _this;
    }
    return IntervalCommandState;
}(SimpleCommandState));
exports.IntervalCommandState = IntervalCommandState;
var IntervalCommandStateEx = (function (_super) {
    tslib_1.__extends(IntervalCommandStateEx, _super);
    function IntervalCommandStateEx(enabled, intervals, value) {
        var _this = _super.call(this, enabled, value) || this;
        _this.intervals = intervals;
        return _this;
    }
    return IntervalCommandStateEx;
}(SimpleCommandState));
exports.IntervalCommandStateEx = IntervalCommandStateEx;
