// devexpress-richedit/lib/core/model/section/section-property-descriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var comparers_1 = require("@devexpress/utils/lib/utils/comparers");
var section_properties_history_items_1 = require("../history/items/section-properties-history-items");
var json_section_enums_1 = require("../json/enums/json-section-enums");
var enums_1 = require("./enums");
var paper_kind_1 = require("./paper-kind");
var section_column_properties_1 = require("./section-column-properties");
var SectionPropertiesMarginLeftDescriptor = (function () {
    function SectionPropertiesMarginLeftDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 1440;
    }
    SectionPropertiesMarginLeftDescriptor.prototype.setProp = function (props, newValue) {
        props.marginLeft = newValue;
    };
    SectionPropertiesMarginLeftDescriptor.prototype.getProp = function (props) {
        return props.marginLeft;
    };
    SectionPropertiesMarginLeftDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionMarginLeftHistoryItem;
    };
    SectionPropertiesMarginLeftDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.MarginLeft;
    };
    return SectionPropertiesMarginLeftDescriptor;
}());
exports.SectionPropertiesMarginLeftDescriptor = SectionPropertiesMarginLeftDescriptor;
var SectionPropertiesMarginTopDescriptor = (function () {
    function SectionPropertiesMarginTopDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 1440;
    }
    SectionPropertiesMarginTopDescriptor.prototype.setProp = function (props, newValue) {
        props.marginTop = newValue;
    };
    SectionPropertiesMarginTopDescriptor.prototype.getProp = function (props) {
        return props.marginTop;
    };
    SectionPropertiesMarginTopDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionMarginTopHistoryItem;
    };
    SectionPropertiesMarginTopDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.MarginTop;
    };
    return SectionPropertiesMarginTopDescriptor;
}());
exports.SectionPropertiesMarginTopDescriptor = SectionPropertiesMarginTopDescriptor;
var SectionPropertiesMarginRightDescriptor = (function () {
    function SectionPropertiesMarginRightDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 1440;
    }
    SectionPropertiesMarginRightDescriptor.prototype.setProp = function (props, newValue) {
        props.marginRight = newValue;
    };
    SectionPropertiesMarginRightDescriptor.prototype.getProp = function (props) {
        return props.marginRight;
    };
    SectionPropertiesMarginRightDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionMarginRightHistoryItem;
    };
    SectionPropertiesMarginRightDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.MarginRight;
    };
    return SectionPropertiesMarginRightDescriptor;
}());
exports.SectionPropertiesMarginRightDescriptor = SectionPropertiesMarginRightDescriptor;
var SectionPropertiesMarginBottomDescriptor = (function () {
    function SectionPropertiesMarginBottomDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 1440;
    }
    SectionPropertiesMarginBottomDescriptor.prototype.setProp = function (props, newValue) {
        props.marginBottom = newValue;
    };
    SectionPropertiesMarginBottomDescriptor.prototype.getProp = function (props) {
        return props.marginBottom;
    };
    SectionPropertiesMarginBottomDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionMarginBottomHistoryItem;
    };
    SectionPropertiesMarginBottomDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.MarginBottom;
    };
    return SectionPropertiesMarginBottomDescriptor;
}());
exports.SectionPropertiesMarginBottomDescriptor = SectionPropertiesMarginBottomDescriptor;
var SectionPropertiesFooterOffsetDescriptor = (function () {
    function SectionPropertiesFooterOffsetDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 720;
    }
    SectionPropertiesFooterOffsetDescriptor.prototype.setProp = function (props, newValue) {
        props.footerOffset = newValue;
    };
    SectionPropertiesFooterOffsetDescriptor.prototype.getProp = function (props) {
        return props.footerOffset;
    };
    SectionPropertiesFooterOffsetDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionFooterOffsetHistoryItem;
    };
    SectionPropertiesFooterOffsetDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.FooterOffset;
    };
    return SectionPropertiesFooterOffsetDescriptor;
}());
exports.SectionPropertiesFooterOffsetDescriptor = SectionPropertiesFooterOffsetDescriptor;
var SectionPropertiesHeaderOffsetDescriptor = (function () {
    function SectionPropertiesHeaderOffsetDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 720;
    }
    SectionPropertiesHeaderOffsetDescriptor.prototype.setProp = function (props, newValue) {
        props.headerOffset = newValue;
    };
    SectionPropertiesHeaderOffsetDescriptor.prototype.getProp = function (props) {
        return props.headerOffset;
    };
    SectionPropertiesHeaderOffsetDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionHeaderOffsetHistoryItem;
    };
    SectionPropertiesHeaderOffsetDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.HeaderOffset;
    };
    return SectionPropertiesHeaderOffsetDescriptor;
}());
exports.SectionPropertiesHeaderOffsetDescriptor = SectionPropertiesHeaderOffsetDescriptor;
var SectionPropertiesColumnCountDescriptor = (function () {
    function SectionPropertiesColumnCountDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 1;
    }
    SectionPropertiesColumnCountDescriptor.prototype.setProp = function (props, newValue) {
        props.columnCount = newValue;
    };
    SectionPropertiesColumnCountDescriptor.prototype.getProp = function (props) {
        return props.columnCount;
    };
    SectionPropertiesColumnCountDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionColumnCountHistoryItem;
    };
    SectionPropertiesColumnCountDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.ColumnCount;
    };
    return SectionPropertiesColumnCountDescriptor;
}());
exports.SectionPropertiesColumnCountDescriptor = SectionPropertiesColumnCountDescriptor;
var SectionPropertiesSpaceDescriptor = (function () {
    function SectionPropertiesSpaceDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 720;
    }
    SectionPropertiesSpaceDescriptor.prototype.setProp = function (props, newValue) {
        props.space = newValue;
    };
    SectionPropertiesSpaceDescriptor.prototype.getProp = function (props) {
        return props.space;
    };
    SectionPropertiesSpaceDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionSpaceHistoryItem;
    };
    SectionPropertiesSpaceDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.Space;
    };
    return SectionPropertiesSpaceDescriptor;
}());
exports.SectionPropertiesSpaceDescriptor = SectionPropertiesSpaceDescriptor;
var SectionPropertiesEqualWidthColumnsDescriptor = (function () {
    function SectionPropertiesEqualWidthColumnsDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = true;
    }
    SectionPropertiesEqualWidthColumnsDescriptor.prototype.setProp = function (props, newValue) {
        props.equalWidthColumns = newValue;
    };
    SectionPropertiesEqualWidthColumnsDescriptor.prototype.getProp = function (props) {
        return props.equalWidthColumns;
    };
    SectionPropertiesEqualWidthColumnsDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionEqualWidthColumnsHistoryItem;
    };
    SectionPropertiesEqualWidthColumnsDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.EqualWidthColumns;
    };
    return SectionPropertiesEqualWidthColumnsDescriptor;
}());
exports.SectionPropertiesEqualWidthColumnsDescriptor = SectionPropertiesEqualWidthColumnsDescriptor;
var SectionPropertiesColumnsInfoDescriptor = (function () {
    function SectionPropertiesColumnsInfoDescriptor() {
        this.binaryEquals = section_column_properties_1.SectionColumnProperties.equalsColumnsInfoBinary;
        this.defaultValue = [];
    }
    SectionPropertiesColumnsInfoDescriptor.prototype.setProp = function (props, newValue) {
        props.columnsInfo = newValue;
    };
    SectionPropertiesColumnsInfoDescriptor.prototype.getProp = function (props) {
        return props.columnsInfo;
    };
    SectionPropertiesColumnsInfoDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionColumnsInfoHistoryItem;
    };
    SectionPropertiesColumnsInfoDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.ColumnsInfo;
    };
    return SectionPropertiesColumnsInfoDescriptor;
}());
exports.SectionPropertiesColumnsInfoDescriptor = SectionPropertiesColumnsInfoDescriptor;
var SectionPropertiesPageWidthDescriptor = (function () {
    function SectionPropertiesPageWidthDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 11906;
    }
    SectionPropertiesPageWidthDescriptor.prototype.setProp = function (props, newValue) {
        props.pageWidth = newValue;
    };
    SectionPropertiesPageWidthDescriptor.prototype.getProp = function (props) {
        return props.pageWidth;
    };
    SectionPropertiesPageWidthDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionPageWidthHistoryItem;
    };
    SectionPropertiesPageWidthDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.PageWidth;
    };
    return SectionPropertiesPageWidthDescriptor;
}());
exports.SectionPropertiesPageWidthDescriptor = SectionPropertiesPageWidthDescriptor;
var SectionPropertiesPageHeightDescriptor = (function () {
    function SectionPropertiesPageHeightDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = 16838;
    }
    SectionPropertiesPageHeightDescriptor.prototype.setProp = function (props, newValue) {
        props.pageHeight = newValue;
    };
    SectionPropertiesPageHeightDescriptor.prototype.getProp = function (props) {
        return props.pageHeight;
    };
    SectionPropertiesPageHeightDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionPageHeightHistoryItem;
    };
    SectionPropertiesPageHeightDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.PageHeight;
    };
    return SectionPropertiesPageHeightDescriptor;
}());
exports.SectionPropertiesPageHeightDescriptor = SectionPropertiesPageHeightDescriptor;
var SectionPropertiesStartTypeDescriptor = (function () {
    function SectionPropertiesStartTypeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = enums_1.SectionStartType.NextPage;
    }
    SectionPropertiesStartTypeDescriptor.prototype.setProp = function (props, newValue) {
        props.startType = newValue;
    };
    SectionPropertiesStartTypeDescriptor.prototype.getProp = function (props) {
        return props.startType;
    };
    SectionPropertiesStartTypeDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionStartTypeHistoryItem;
    };
    SectionPropertiesStartTypeDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.StartType;
    };
    return SectionPropertiesStartTypeDescriptor;
}());
exports.SectionPropertiesStartTypeDescriptor = SectionPropertiesStartTypeDescriptor;
var SectionPropertiesLandscapeDescriptor = (function () {
    function SectionPropertiesLandscapeDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    SectionPropertiesLandscapeDescriptor.prototype.setProp = function (props, newValue) {
        props.landscape = newValue;
    };
    SectionPropertiesLandscapeDescriptor.prototype.getProp = function (props) {
        return props.landscape;
    };
    SectionPropertiesLandscapeDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionLandscapeHistoryItem;
    };
    SectionPropertiesLandscapeDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.Landscape;
    };
    return SectionPropertiesLandscapeDescriptor;
}());
exports.SectionPropertiesLandscapeDescriptor = SectionPropertiesLandscapeDescriptor;
var SectionPropertiesDifferentFirstPageDescriptor = (function () {
    function SectionPropertiesDifferentFirstPageDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = false;
    }
    SectionPropertiesDifferentFirstPageDescriptor.prototype.setProp = function (props, newValue) {
        props.differentFirstPage = newValue;
    };
    SectionPropertiesDifferentFirstPageDescriptor.prototype.getProp = function (props) {
        return props.differentFirstPage;
    };
    SectionPropertiesDifferentFirstPageDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionDifferentFirstPageHistoryItem;
    };
    SectionPropertiesDifferentFirstPageDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.DifferentFirstPage;
    };
    return SectionPropertiesDifferentFirstPageDescriptor;
}());
exports.SectionPropertiesDifferentFirstPageDescriptor = SectionPropertiesDifferentFirstPageDescriptor;
var SectionPropertiesPaperKindDescriptor = (function () {
    function SectionPropertiesPaperKindDescriptor() {
        this.binaryEquals = comparers_1.Equals.simpleType;
        this.defaultValue = paper_kind_1.PaperKind.Letter;
    }
    SectionPropertiesPaperKindDescriptor.prototype.setProp = function (props, newValue) {
        props.paperKind = newValue;
    };
    SectionPropertiesPaperKindDescriptor.prototype.getProp = function (props) {
        return props.paperKind;
    };
    SectionPropertiesPaperKindDescriptor.prototype.getHistoryItemConstructor = function () {
        return section_properties_history_items_1.SectionPaperKindHistoryItem;
    };
    SectionPropertiesPaperKindDescriptor.prototype.getJSONProperty = function () {
        return json_section_enums_1.JSONSectionProperty.PaperKind;
    };
    return SectionPropertiesPaperKindDescriptor;
}());
exports.SectionPropertiesPaperKindDescriptor = SectionPropertiesPaperKindDescriptor;
var SectionPropertyDescriptor = (function () {
    function SectionPropertyDescriptor() {
    }
    SectionPropertyDescriptor.marginLeft = new SectionPropertiesMarginLeftDescriptor();
    SectionPropertyDescriptor.marginTop = new SectionPropertiesMarginTopDescriptor();
    SectionPropertyDescriptor.marginRight = new SectionPropertiesMarginRightDescriptor();
    SectionPropertyDescriptor.marginBottom = new SectionPropertiesMarginBottomDescriptor();
    SectionPropertyDescriptor.footerOffset = new SectionPropertiesFooterOffsetDescriptor();
    SectionPropertyDescriptor.headerOffset = new SectionPropertiesHeaderOffsetDescriptor();
    SectionPropertyDescriptor.columnCount = new SectionPropertiesColumnCountDescriptor();
    SectionPropertyDescriptor.space = new SectionPropertiesSpaceDescriptor();
    SectionPropertyDescriptor.equalWidthColumns = new SectionPropertiesEqualWidthColumnsDescriptor();
    SectionPropertyDescriptor.columnsInfo = new SectionPropertiesColumnsInfoDescriptor();
    SectionPropertyDescriptor.pageWidth = new SectionPropertiesPageWidthDescriptor();
    SectionPropertyDescriptor.pageHeight = new SectionPropertiesPageHeightDescriptor();
    SectionPropertyDescriptor.startType = new SectionPropertiesStartTypeDescriptor();
    SectionPropertyDescriptor.landscape = new SectionPropertiesLandscapeDescriptor();
    SectionPropertyDescriptor.differentFirstPage = new SectionPropertiesDifferentFirstPageDescriptor();
    SectionPropertyDescriptor.paperKind = new SectionPropertiesPaperKindDescriptor();
    SectionPropertyDescriptor.ALL_FIELDS = [
        SectionPropertyDescriptor.marginLeft,
        SectionPropertyDescriptor.marginTop,
        SectionPropertyDescriptor.marginRight,
        SectionPropertyDescriptor.marginBottom,
        SectionPropertyDescriptor.footerOffset,
        SectionPropertyDescriptor.headerOffset,
        SectionPropertyDescriptor.columnCount,
        SectionPropertyDescriptor.space,
        SectionPropertyDescriptor.equalWidthColumns,
        SectionPropertyDescriptor.columnsInfo,
        SectionPropertyDescriptor.pageWidth,
        SectionPropertyDescriptor.pageHeight,
        SectionPropertyDescriptor.startType,
        SectionPropertyDescriptor.landscape,
        SectionPropertyDescriptor.differentFirstPage,
        SectionPropertyDescriptor.paperKind
    ];
    return SectionPropertyDescriptor;
}());
exports.SectionPropertyDescriptor = SectionPropertyDescriptor;
