// devexpress-dashboard/designer/conditional-formatting/_style-settings-adapters.js
/**
* DevExpress Dashboard (_style-settings-adapters.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _default_1 = require("../../data/localization/_default");
var color_1 = require("../../model/color");
var appearance_settings_1 = require("../../model/format-rules/style-settings/appearance-settings");
var bar_style_settings_1 = require("../../model/format-rules/style-settings/bar-style-settings");
var color_style_settings_1 = require("../../model/format-rules/style-settings/color-style-settings");
var icon_settings_1 = require("../../model/format-rules/style-settings/icon-settings");
var _appearance_settings_1 = require("../../model/format-rules/style-settings/metadata/_appearance-settings");
var _icon_settings_1 = require("../../model/format-rules/style-settings/metadata/_icon-settings");
var _style_settings_base_1 = require("../../model/format-rules/style-settings/metadata/_style-settings-base");
var _appearance_settings_provider_1 = require("../../viewer-parts/conditional-formatting/_appearance-settings-provider");
var _style_settings_provider_1 = require("../../viewer-parts/conditional-formatting/_style-settings-provider");
var _cssHelper_1 = require("../../viewer-parts/viewer/_cssHelper");
var _style_settings_palette_1 = require("./_style-settings-palette");
var colorStyleSettingsAdapter = {
    getCssStyles: function (item) {
        var predefinedColor = item.predefinedColor();
        if (predefinedColor === 'Custom' || predefinedColor === _style_settings_base_1.emptyStyleType) {
            return 'background-color: ' + (item.color() && item.color().css);
        }
        else {
            return _cssHelper_1.convertCssPropertyWrappersToString(_appearance_settings_provider_1.appearanceSettingsProvider.toCssProperties(predefinedColor));
        }
    },
    getCssClasses: function (item, isEmptyAllowed, isRange) {
        var isEmpty = isEmptyAllowed && item.predefinedColor() === _style_settings_base_1.emptyStyleType;
        var isLabel = colorStyleSettingsAdapter.isEmptyCustomStyle(item) || (colorStyleSettingsAdapter.getLabelText(item, isRange, false) !== '');
        return [
            isEmpty ? 'dx-dashboard-appearance-item-empty' : '',
            isLabel ? 'dx-dashboard-appearance-item-label' : '',
            isRange ? 'dx-dashboard-range-appearance-item' : 'dx-dashboard-appearance-item',
        ];
    },
    getLocalizedCaption: function (item) { return _default_1.getLocalizationById(_appearance_settings_1.appearanceType.values[item.predefinedColor()]); },
    getLabelText: function (item, isRange, isGradient) {
        var labelText = _style_settings_palette_1.Palette.getLabelText(item.predefinedColor(), isGradient ? '(Automatic)' : 'No Style');
        return (!isRange || isRange && item.predefinedColor() !== _style_settings_base_1.emptyStyleType) ? labelText : '';
    },
    getPredefinedStyle: function (item) { return item.predefinedColor(); },
    setPredefinedStyle: function (item, style) { return item.predefinedColor(style); },
    hasCustomStyle: function (item) { return item.predefinedColor() === 'Custom'; },
    isEmptyCustomStyle: function (item) { return colorStyleSettingsAdapter.hasCustomStyle(item) && !item.color(); },
    getCustomColor: function (item) { return item.color() && item.color().css; },
    getCustomAppearance: function (item) { return null; },
    setCustomColor: function (item, style) { return item.color(style && color_1.Color.fromRgbaString(style)); },
    setCustomAppearance: function () { },
};
var iconStyleSettingsAdapter = {
    getCssStyles: function () { return ''; },
    getCssClasses: function (item, isEmptyAllowed, isRange) {
        if (isEmptyAllowed && item.iconType() === _style_settings_base_1.emptyStyleType) {
            return [
                'dx-dashboard-appearance-item',
                'dx-dashboard-appearance-item-label',
                'dx-dashboard-appearance-item-empty'
            ];
        }
        else {
            return [
                isRange ? 'dx-dashboard-rangeicon' : 'dx-dashboard-icon-item',
                _style_settings_provider_1.styleSettingsProvider.toIconCssClass(item.iconType())
            ];
        }
    },
    getLocalizedCaption: function (item) { return _default_1.getLocalizationById(_icon_settings_1.iconType.values[item.iconType()]); },
    getLabelText: function (item, isRange, isGradient) {
        var labelText = _style_settings_palette_1.Palette.getLabelText(item.iconType(), isGradient ? '(Automatic)' : 'No Style');
        return (!isRange || isRange && item.iconType() !== _style_settings_base_1.emptyStyleType) ? labelText : '';
    },
    getPredefinedStyle: function (item) { return item.iconType(); },
    setPredefinedStyle: function (item, style) { return item.iconType(style); },
    hasCustomStyle: function () { return false; },
    isEmptyCustomStyle: function () { return false; },
    getCustomColor: function () { return null; },
    getCustomAppearance: function () { return null; },
    setCustomColor: function () { },
    setCustomAppearance: function () { },
};
var appearanceStyleSettingsAdapter = {
    getCssStyles: function (item) {
        if (item.appearanceType() !== _style_settings_base_1.emptyStyleType && !appearanceStyleSettingsAdapter.hasCustomStyle(item)) {
            return _cssHelper_1.convertCssPropertyWrappersToString(_appearance_settings_provider_1.appearanceSettingsProvider.toCssProperties(item.appearanceType()));
        }
        else {
            var backColor = item.backColor();
            return backColor ? 'background-color: ' + backColor.css + ';' : '';
        }
    },
    getCssClasses: function (item, isEmptyAllowed, isRange, isGradient) {
        if (isGradient === void 0) { isGradient = false; }
        var _isEmptyAppearanceItem = function () { return isEmptyAllowed && item.appearanceType() === _style_settings_base_1.emptyStyleType; };
        var _isTransparent = function () { return item.appearanceType() === 'GradientTransparent'; };
        var _isLabel = function () { return (item.appearanceType() === 'GradientTransparent')
            || (appearanceStyleSettingsAdapter.getLabelText(item, isRange, isGradient) !== '')
            || (appearanceStyleSettingsAdapter.isEmptyCustomStyle(item)); };
        return [
            isRange ? 'dx-dashboard-range-appearance-item' : 'dx-dashboard-appearance-item',
            _isEmptyAppearanceItem() ? 'dx-dashboard-appearance-item-empty' : '',
            _isLabel() ? 'dx-dashboard-appearance-item-label' : '',
            _isTransparent() && !isRange ? 'dx-dashboard-appearance-item-transparent' : ''
        ];
    },
    getLocalizedCaption: function (item) { return _default_1.getLocalizationById(_appearance_settings_1.appearanceType.values[item.appearanceType()]); },
    getLabelText: function (item, isRange, isGradient) {
        var labelText = _style_settings_palette_1.Palette.getLabelText(item.appearanceType(), isGradient ? '(Automatic)' : 'No Style');
        return (!isRange || isRange && item.appearanceType() !== _style_settings_base_1.emptyStyleType) ? labelText : '';
    },
    getPredefinedStyle: function (item) { return item.appearanceType(); },
    setPredefinedStyle: function (item, style) { return item.appearanceType(style); },
    hasCustomStyle: function (item) { return item.appearanceType() === 'Custom'; },
    isEmptyCustomStyle: function (item) {
        return appearanceStyleSettingsAdapter.hasCustomStyle(item)
            && !item.foreColor() && !item.backColor()
            && item.fontStyle() === _appearance_settings_1.fontStyle.defaultVal
            && item.fontFamily() === _appearance_settings_1.fontFamily.defaultVal;
    },
    getCustomColor: function (item) { return item.backColor() && item.backColor().css; },
    getCustomAppearance: function (item) {
        return !appearanceStyleSettingsAdapter.hasCustomStyle(item) || appearanceStyleSettingsAdapter.isEmptyCustomStyle(item) ? null : ({
            fontFamily: item.fontFamily(),
            fontStyle: item.fontStyle(),
            backColor: item.backColor() && item.backColor().css,
            foreColor: item.foreColor() && item.foreColor().css
        });
    },
    setCustomColor: function (item, style) { return item.backColor(style && color_1.Color.fromRgbaString(style)); },
    setCustomAppearance: function (item, style) {
        item.fontFamily(style && style.fontFamily);
        item.fontStyle(style ? style.fontStyle : null);
        item.backColor(style && style.backColor && color_1.Color.fromRgbaString(style.backColor));
        item.foreColor(style && style.foreColor && color_1.Color.fromRgbaString(style.foreColor));
    },
};
function styleSettingsAdapter(item, itemType) {
    var adapter;
    switch (itemType || item.constructor) {
        case appearance_settings_1.AppearanceSettings:
            itemType = appearance_settings_1.AppearanceSettings;
            adapter = appearanceStyleSettingsAdapter;
            break;
        case icon_settings_1.IconSettings:
            itemType = icon_settings_1.IconSettings;
            adapter = iconStyleSettingsAdapter;
            break;
        case bar_style_settings_1.BarStyleSettings:
            itemType = bar_style_settings_1.BarStyleSettings;
            adapter = colorStyleSettingsAdapter;
            break;
        case color_style_settings_1.ColorStyleSettings:
            itemType = color_style_settings_1.ColorStyleSettings;
            adapter = colorStyleSettingsAdapter;
            break;
        default:
            throw new Error('Unknown StyleSettings type');
    }
    return {
        itemFactory: function () { return new itemType(); },
        getCssStyles: function () { return adapter.getCssStyles(item); },
        getCssClasses: function (isEmptyAllowed, isRange, isGradient) { return adapter.getCssClasses(item, isEmptyAllowed, isRange, isGradient); },
        getLocalizedCaption: function () { return adapter.getLocalizedCaption(item); },
        getLabelText: function (isRange, isGradient) { return adapter.getLabelText(item, isRange, isGradient); },
        getPredefinedStyle: function () { return adapter.getPredefinedStyle(item); },
        setPredefinedStyle: function (style) { return adapter.setPredefinedStyle(item, style); },
        hasCustomStyle: function () { return adapter.hasCustomStyle(item); },
        isEmptyCustomStyle: function () { return adapter.isEmptyCustomStyle(item); },
        getCustomColor: function () { return adapter.getCustomColor(item); },
        getCustomAppearance: function () { return adapter.getCustomAppearance(item); },
        setCustomColor: function (style) { return adapter.setCustomColor(item, style); },
        setCustomAppearance: function (style) { return adapter.setCustomAppearance(item, style); },
    };
}
exports.styleSettingsAdapter = styleSettingsAdapter;
