// devexpress-dashboard/designer/calculations/_expression-editor.js
/**
* DevExpress Dashboard (_expression-editor.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _expression_editor_1 = require("../expression-editor/_expression-editor");
var _expression_editor_functions_1 = require("../expression-editor/_expression-editor-functions");
var CalculationExpressionEditor = (function () {
    function CalculationExpressionEditor(params) {
        this.params = params;
        this.editorTemplate = 'dx-dashboard-calculation-expression-editor';
        this.editor = new ExpressionEditorWrapper(params);
        this.criteriaString = ko.computed(function () {
            var criteria = params.options().value();
            return criteria;
        });
    }
    CalculationExpressionEditor.prototype.edit = function () {
        this.editor.show();
    };
    CalculationExpressionEditor.prototype.remove = function () {
        this.params.options().value(null);
    };
    return CalculationExpressionEditor;
}());
exports.CalculationExpressionEditor = CalculationExpressionEditor;
var ExpressionEditorWrapper = (function () {
    function ExpressionEditorWrapper(params) {
        this.params = params;
        this.viewModel = ko.observable();
    }
    ExpressionEditorWrapper.prototype.show = function () {
        if (this.viewModel()) {
            this.viewModel().dispose();
        }
        var expressionEditorOptions = this.params.options();
        expressionEditorOptions.functions = _expression_editor_functions_1.getExpressionEditorFunctions();
        var editor = new _expression_editor_1.DashboardExpressionEditor(expressionEditorOptions, this.params.fieldListProvider, this.params.displayNameProvider);
        editor.textAreaValue(expressionEditorOptions.value());
        editor.popupVisible(true);
        this.viewModel(editor);
    };
    return ExpressionEditorWrapper;
}());
exports.ExpressionEditorWrapper = ExpressionEditorWrapper;
