// devexpress-richedit/lib/client-ribbon/client-ribbon/toolbar-items/toolbar-select-box-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var toolbar_interactive_item_1 = require("./toolbar-interactive-item");
var select_box_toolbar_item_template_creator_1 = require("./toolbar-item-template-creators/select-box-toolbar-item-template-creator");
var ToolbarSelectBoxItem = (function (_super) {
    tslib_1.__extends(ToolbarSelectBoxItem, _super);
    function ToolbarSelectBoxItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolbarSelectBoxItem.prototype.getBuildTemplateStrategy = function () {
        return new select_box_toolbar_item_template_creator_1.SelectBoxToolbarItemTemplateCreator({
            itemOptions: this.options,
            onInitialized: this.getOnInitializedHandler(),
            onValueChanged: this.getOnValueChangedHandler()
        });
    };
    ToolbarSelectBoxItem.prototype.setValue = function (value) {
        this.widget.option('value', value);
    };
    ToolbarSelectBoxItem.prototype.getOnValueChangedHandler = function () {
        var _this = this;
        return this.onCommandExecuted ? function (e) {
            if (!e.event)
                return;
            _this.onCommandExecuted({
                item: _this,
                parameter: e.value
            });
        } : undefined;
    };
    return ToolbarSelectBoxItem;
}(toolbar_interactive_item_1.ToolbarInteractiveItem));
exports.ToolbarSelectBoxItem = ToolbarSelectBoxItem;
