// devexpress-richedit/lib/client-ribbon/client-ribbon/toolbar-items/toolbar-button-group-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var toolbar_interactive_item_1 = require("./toolbar-interactive-item");
var button_group_toolbar_item_template_creator_1 = require("./toolbar-item-template-creators/button-group-toolbar-item-template-creator");
var ToolbarButtonGroupItem = (function (_super) {
    tslib_1.__extends(ToolbarButtonGroupItem, _super);
    function ToolbarButtonGroupItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolbarButtonGroupItem.prototype.getBuildTemplateStrategy = function () {
        return new button_group_toolbar_item_template_creator_1.ButtonGroupToolbarItemTemplateCreator({
            itemOptions: this.options,
            onInitialized: this.getOnInitializedHandler(),
            onClick: this.getOnClickHandler()
        });
    };
    ToolbarButtonGroupItem.prototype.setValue = function (_value) {
    };
    ToolbarButtonGroupItem.prototype.getOnClickHandler = function () {
        var _this = this;
        return this.onCommandExecuted ? function () {
            _this.onCommandExecuted({
                item: _this,
                parameter: null
            });
        } : undefined;
    };
    ToolbarButtonGroupItem.ToggleStateClassName = 'dx-r-toggle';
    ToolbarButtonGroupItem.ToggleButtonDataProperty = 'dx-ri-value';
    return ToolbarButtonGroupItem;
}(toolbar_interactive_item_1.ToolbarInteractiveItem));
exports.ToolbarButtonGroupItem = ToolbarButtonGroupItem;
