// devexpress-richedit/lib/client-ribbon/client-ribbon/toolbar-items/toolbar-number-box-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var toolbar_interactive_item_1 = require("./toolbar-interactive-item");
var number_box_toolbar_item_template_creator_1 = require("./toolbar-item-template-creators/number-box-toolbar-item-template-creator");
var ToolbarNumberBoxItem = (function (_super) {
    tslib_1.__extends(ToolbarNumberBoxItem, _super);
    function ToolbarNumberBoxItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolbarNumberBoxItem.prototype.getBuildTemplateStrategy = function () {
        return new number_box_toolbar_item_template_creator_1.NumberBoxToolbarItemTemplateCreator({
            itemOptions: this.options,
            onInitialized: this.getOnInitializedHandler(),
            onValueChanged: this.getOnValueChangedHandler()
        });
    };
    ToolbarNumberBoxItem.prototype.setValue = function (value) {
        this.widget.option('value', value);
    };
    ToolbarNumberBoxItem.prototype.getOnValueChangedHandler = function () {
        var _this = this;
        return this.onCommandExecuted ? function (e) {
            if (!e.event)
                return;
            _this.onCommandExecuted({
                item: _this,
                parameter: e.value
            });
        } : undefined;
    };
    return ToolbarNumberBoxItem;
}(toolbar_interactive_item_1.ToolbarInteractiveItem));
exports.ToolbarNumberBoxItem = ToolbarNumberBoxItem;
