// devexpress-richedit/lib/client/dialogs/delete-table-cells-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var localization_1 = require("devextreme/localization");
var dialog_insert_table_cells_command_1 = require("../../base/commands/dialogs/dialog-insert-table-cells-command");
var dialog_base_1 = require("./dialog-base");
var DeleteTableCellsDialog = (function (_super) {
    tslib_1.__extends(DeleteTableCellsDialog, _super);
    function DeleteTableCellsDialog() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DeleteTableCellsDialog.prototype.getMaxWidth = function () {
        return 300;
    };
    DeleteTableCellsDialog.prototype.getTitle = function () {
        return localization_1.formatMessage("ASPxRichEditStringId.DeleteTableCellsTitle");
    };
    DeleteTableCellsDialog.prototype.getFormOptions = function () {
        return {
            labelLocation: 'top',
            colCount: 1,
            items: [{
                    dataField: 'tableCellOperation',
                    editorType: 'dxRadioGroup',
                    label: { visible: false },
                    editorOptions: {
                        items: [
                            { text: localization_1.formatMessage("ASPxRichEditStringId.DeleteCells_ShiftCellsLeft"), value: dialog_insert_table_cells_command_1.TableCellOperation.ShiftToTheHorizontally },
                            { text: localization_1.formatMessage("ASPxRichEditStringId.DeleteCells_ShiftCellsUp"), value: dialog_insert_table_cells_command_1.TableCellOperation.ShiftToTheVertically },
                            { text: localization_1.formatMessage("ASPxRichEditStringId.DeleteCells_DeleteEntireRow"), value: dialog_insert_table_cells_command_1.TableCellOperation.RowOperation },
                            { text: localization_1.formatMessage("ASPxRichEditStringId.DeleteCells_DeleteEntireColumn"), value: dialog_insert_table_cells_command_1.TableCellOperation.ColumnOperation },
                        ],
                        valueExpr: "value",
                        value: this.parameters.tableCellOperation
                    }
                }]
        };
    };
    DeleteTableCellsDialog.prototype.updateParameters = function (parameters, data) {
        parameters.tableCellOperation = data.tableCellOperation;
    };
    return DeleteTableCellsDialog;
}(dialog_base_1.DialogBase));
exports.DeleteTableCellsDialog = DeleteTableCellsDialog;
