// devexpress-richedit/lib/client/dialogs/hyperlink-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var localization_1 = require("devextreme/localization");
var dialog_base_1 = require("./dialog-base");
var HyperlinkDialog = (function (_super) {
    tslib_1.__extends(HyperlinkDialog, _super);
    function HyperlinkDialog() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mailtoPrefix = 'mailto:';
        _this.subjectPrefix = '?subject=';
        return _this;
    }
    HyperlinkDialog.prototype.getTitle = function () {
        return localization_1.formatMessage("ASPxRichEditStringId.HyperlinkTitle");
    };
    HyperlinkDialog.prototype.getMaxWidth = function () {
        return 500;
    };
    HyperlinkDialog.prototype.getFormOptions = function () {
        var _this = this;
        return {
            labelLocation: 'top',
            colCount: 1,
            items: [
                {
                    itemType: "tabbed",
                    tabPanelOptions: {
                        deferRendering: false,
                        selectedIndex: this.parameters.anchor ? 1 : this.isLinkMailTo(this.parameters.url) ? 2 : 0,
                        onInitialized: function (e) { _this.tabPanel = e.component; }
                    },
                    tabs: [
                        {
                            title: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_WebPage"),
                            items: [
                                {
                                    dataField: 'url',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Url"), location: 'left' },
                                    editorOptions: {
                                        value: this.isLinkMailTo(this.parameters.url) ? '' : this.parameters.url ? this.parameters.url : 'https://'
                                    }
                                },
                                {
                                    dataField: 'text',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Text"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.text,
                                        disabled: !this.parameters.canChangeDisplayText
                                    }
                                },
                                {
                                    dataField: 'tooltip',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_ToolTip"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.tooltip
                                    }
                                }
                            ]
                        },
                        {
                            title: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_PlaceInThisDocument"),
                            items: [
                                {
                                    dataField: 'anchor',
                                    editorType: 'dxSelectBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Bookmark"), location: 'left' },
                                    editorOptions: {
                                        items: this.parameters.bookmarkNames,
                                        value: this.parameters.anchor
                                    }
                                },
                                {
                                    dataField: 'text',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Text"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.text,
                                        disabled: !this.parameters.canChangeDisplayText
                                    }
                                },
                                {
                                    dataField: 'tooltip',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_ToolTip"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.tooltip
                                    }
                                }
                            ]
                        },
                        {
                            title: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_EmailAddress"),
                            items: [
                                {
                                    dataField: 'email',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_EmailTo"), location: 'left' },
                                    editorOptions: {
                                        value: this.getEmail(this.parameters.url)
                                    }
                                },
                                {
                                    dataField: 'subject',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Subject"), location: 'left' },
                                    editorOptions: {
                                        value: this.getSubject(this.parameters.url)
                                    }
                                },
                                {
                                    dataField: 'text',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_Text"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.text,
                                        disabled: !this.parameters.canChangeDisplayText
                                    }
                                },
                                {
                                    dataField: 'tooltip',
                                    editorType: 'dxTextBox',
                                    label: { text: localization_1.formatMessage("ASPxRichEditStringId.Hyperlink_ToolTip"), location: 'left' },
                                    editorOptions: {
                                        value: this.parameters.tooltip
                                    }
                                }
                            ]
                        }
                    ]
                }
            ]
        };
    };
    HyperlinkDialog.prototype.isLinkMailTo = function (url) {
        return url.toLowerCase().indexOf(this.mailtoPrefix) > -1;
    };
    HyperlinkDialog.prototype.getSubject = function (url) {
        var subjectIndex = url.toLowerCase().indexOf(this.subjectPrefix);
        return subjectIndex == -1 ? '' : decodeURI(url.substring(subjectIndex + this.subjectPrefix.length));
    };
    HyperlinkDialog.prototype.getEmail = function (url) {
        if (!this.isLinkMailTo(url))
            return '';
        var mailtoIndex = url.toLowerCase().indexOf(this.mailtoPrefix), subjectIndex = url.toLowerCase().indexOf(this.subjectPrefix), endIndex = subjectIndex > -1 ? subjectIndex : url.length;
        return url.substring(mailtoIndex + this.mailtoPrefix.length, endIndex);
    };
    HyperlinkDialog.prototype.updateParameters = function (parameters, data) {
        parameters.text = data.text;
        parameters.tooltip = data.tooltip;
        var tabPanelSelectedIndex = this.tabPanel.option('selectedIndex');
        if (tabPanelSelectedIndex == 0)
            parameters.url = data.url;
        else if (tabPanelSelectedIndex == 1)
            parameters.anchor = data.anchor;
        else {
            var subject = data.subject ? this.subjectPrefix + data.subject : '';
            parameters.url = this.mailtoPrefix + data.email + subject;
        }
    };
    return HyperlinkDialog;
}(dialog_base_1.DialogBase));
exports.HyperlinkDialog = HyperlinkDialog;
