// devexpress-richedit/lib/client/dialogs/page-setup-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var localization_1 = require("devextreme/localization");
var dialog_page_setup_command_1 = require("../../base/commands/dialogs/dialog-page-setup-command");
var enums_1 = require("../../core/model/section/enums");
var paper_kind_1 = require("../../core/model/section/paper-kind");
var size_1 = require("@devexpress/utils/lib/geometry/size");
var dialog_base_1 = require("./dialog-base");
var PageSetupDialog = (function (_super) {
    tslib_1.__extends(PageSetupDialog, _super);
    function PageSetupDialog() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.lockOrientation = false;
        _this.lockKind = false;
        _this.lockSize = false;
        return _this;
    }
    PageSetupDialog.prototype.getTitle = function () {
        return localization_1.formatMessage('ASPxRichEditStringId.PageSetupTitle');
    };
    PageSetupDialog.prototype.getMaxWidth = function () {
        return 500;
    };
    PageSetupDialog.prototype.getFormOptions = function () {
        var _this = this;
        return {
            labelLocation: 'top',
            colCount: 1,
            items: [
                {
                    itemType: 'tabbed',
                    tabPanelOptions: {
                        deferRendering: false,
                        selectedIndex: this.parameters.initialTab
                    },
                    tabs: [
                        {
                            title: localization_1.formatMessage('ASPxRichEditStringId.Margins'),
                            items: [
                                {
                                    itemType: 'group',
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Margins'),
                                    colCount: 2,
                                    items: [
                                        {
                                            dataField: 'marginTop',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Top'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.marginTop,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'marginLeft',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Left'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.marginLeft,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'marginBottom',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Bottom'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.marginBottom,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'marginRight',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Right'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.marginRight,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        }
                                    ]
                                },
                                {
                                    itemType: 'group',
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Orientation'),
                                    items: [
                                        {
                                            dataField: 'landscape',
                                            editorType: 'dxRadioGroup',
                                            label: { visible: false },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Portrait'), value: false },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Landscape'), value: true }
                                                ],
                                                valueExpr: 'value',
                                                value: this.parameters.landscape,
                                                onValueChanged: function () { _this.onOrientationChanged(); },
                                                onInitialized: function (e) { _this.orientationRadioGroup = e.component; }
                                            }
                                        }
                                    ]
                                }
                            ]
                        },
                        {
                            title: localization_1.formatMessage('ASPxRichEditStringId.Paper'),
                            items: [
                                {
                                    itemType: 'group',
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.PaperSize'),
                                    colCount: 2,
                                    items: [
                                        {
                                            editorType: 'dxSelectBox',
                                            label: { visible: false },
                                            colSpan: 2,
                                            editorOptions: {
                                                items: this.getPaperKinds(),
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                onValueChanged: function (e) { _this.onPaperKindChanged(e.component.option('value')); },
                                                onInitialized: function (e) { _this.pageKindSelectBox = e.component; }
                                            }
                                        },
                                        {
                                            dataField: 'pageWidth',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Width'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.pageWidth,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##',
                                                onValueChanged: function () { _this.onPaperSizeChanged(); },
                                                onInitialized: function (e) { _this.pageWidthNumberBox = e.component; }
                                            }
                                        },
                                        {
                                            dataField: 'pageHeight',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Height'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.pageHeight,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##',
                                                onValueChanged: function () { _this.onPaperSizeChanged(); },
                                                onInitialized: function (e) { _this.pageHeightNumberBox = e.component; }
                                            }
                                        }
                                    ]
                                }
                            ]
                        },
                        {
                            title: localization_1.formatMessage('ASPxRichEditStringId.Layout'),
                            items: [
                                {
                                    itemType: 'group',
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Section'),
                                    items: [
                                        {
                                            dataField: 'startType',
                                            editorType: 'dxSelectBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.SectionStart'), location: 'left' },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Continuous'), value: enums_1.SectionStartType.Continuous },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.NewPage'), value: enums_1.SectionStartType.NextPage },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.EvenPage'), value: enums_1.SectionStartType.EvenPage },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.OddPage'), value: enums_1.SectionStartType.OddPage }
                                                ],
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                value: this.parameters.startType
                                            }
                                        }
                                    ]
                                },
                                {
                                    itemType: 'group',
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.HeadersAndFooters'),
                                    items: [
                                        {
                                            dataField: 'headerDifferentOddAndEven',
                                            editorType: 'dxCheckBox',
                                            label: { visible: false },
                                            editorOptions: {
                                                value: !!this.parameters.headerDifferentOddAndEven,
                                                text: localization_1.formatMessage('ASPxRichEditStringId.DifferentOddAndEven')
                                            }
                                        },
                                        {
                                            dataField: 'headerDifferentFirstPage',
                                            editorType: 'dxCheckBox',
                                            label: { visible: false },
                                            editorOptions: {
                                                value: !!this.parameters.headerDifferentFirstPage,
                                                text: localization_1.formatMessage('ASPxRichEditStringId.DifferentFirstPage')
                                            }
                                        },
                                        {
                                            dataField: 'headerOffset',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.FromEdge') + ' ' + localization_1.formatMessage('ASPxRichEditStringId.Header'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.headerOffset,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'footerOffset',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.FromEdge') + ' ' + localization_1.formatMessage('ASPxRichEditStringId.Footer'), location: 'left' },
                                            editorOptions: {
                                                value: this.parameters.footerOffset,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                },
                {
                    dataField: 'applyTo',
                    editorType: 'dxSelectBox',
                    label: { text: localization_1.formatMessage('ASPxRichEditStringId.ApplyTo'), location: 'left' },
                    editorOptions: {
                        items: [
                            { text: localization_1.formatMessage('ASPxRichEditStringId.WholeDocument'), value: dialog_page_setup_command_1.SectionPropertiesApplyType.WholeDocument },
                            { text: localization_1.formatMessage('ASPxRichEditStringId.CurrentSection'), value: dialog_page_setup_command_1.SectionPropertiesApplyType.CurrentSection },
                            { text: localization_1.formatMessage('ASPxRichEditStringId.SelectedSections'), value: dialog_page_setup_command_1.SectionPropertiesApplyType.SelectedSections },
                            { text: localization_1.formatMessage('ASPxRichEditStringId.ThisPointForward'), value: dialog_page_setup_command_1.SectionPropertiesApplyType.ThisPointForward }
                        ],
                        valueExpr: 'value',
                        displayExpr: 'text',
                        value: this.parameters.applyTo
                    }
                }
            ]
        };
    };
    PageSetupDialog.prototype.afterShowing = function () {
        if (this.parameters.pageWidth && this.parameters.pageHeight)
            this.onPaperSizeChanged();
    };
    PageSetupDialog.prototype.getPaperKinds = function () {
        var enumTexts = Object.keys(paper_kind_1.PaperKind).map(function (key) { return paper_kind_1.PaperKind[key]; }).filter(function (v) { return typeof v === 'string'; });
        var enumValues = Object.keys(paper_kind_1.PaperKind).map(function (key) { return paper_kind_1.PaperKind[key]; }).filter(function (v) { return typeof v === 'number'; });
        return enumValues.map(function (value, index) { return { text: enumTexts[index], value: value }; });
    };
    PageSetupDialog.prototype.onPaperKindChanged = function (kind) {
        this.lockKind = true;
        var size = paper_kind_1.PaperSizeConverter.calculatePaperSize(kind);
        if (!this.lockSize) {
            this.pageWidthNumberBox.option('value', this.richedit.uiUnitConverter.twipsToUI(size.width));
            this.pageHeightNumberBox.option('value', this.richedit.uiUnitConverter.twipsToUI(size.height));
        }
        this.updateOrientation(size.width > size.height);
        this.lockKind = false;
    };
    PageSetupDialog.prototype.onPaperSizeChanged = function () {
        this.lockSize = true;
        var width = this.pageWidthNumberBox.option('value');
        var height = this.pageHeightNumberBox.option('value');
        this.updatePaperKind(width, height);
        this.updateOrientation(width > height);
        this.lockSize = false;
    };
    PageSetupDialog.prototype.onOrientationChanged = function () {
        this.lockOrientation = true;
        var landscape = this.orientationRadioGroup.option('value');
        var width = this.pageWidthNumberBox.option('value');
        var height = this.pageHeightNumberBox.option('value');
        if (landscape) {
            this.pageWidthNumberBox.option('value', Math.max(width, height));
            this.pageHeightNumberBox.option('value', Math.min(width, height));
            this.updatePaperKind(Math.max(width, height), Math.min(width, height));
        }
        else {
            this.pageWidthNumberBox.option('value', Math.min(width, height));
            this.pageHeightNumberBox.option('value', Math.max(width, height));
            this.updatePaperKind(Math.min(width, height), Math.max(width, height));
        }
        this.lockOrientation = false;
    };
    PageSetupDialog.prototype.updatePaperKind = function (width, height) {
        if (this.lockKind)
            return;
        var size = new size_1.Size(this.richedit.uiUnitConverter.UIToTwips(width), this.richedit.uiUnitConverter.UIToTwips(height));
        var paperKind = paper_kind_1.PaperSizeConverter.calculatePaperKind(size, 0, this.richedit.uiUnitConverter.UIToTwips(0.01));
        if (paperKind == paper_kind_1.PaperKind.Custom)
            paperKind = paper_kind_1.PaperSizeConverter.calculatePaperKind(new size_1.Size(size.height, size.width), 0, this.richedit.uiUnitConverter.UIToTwips(0.01));
        this.pageKindSelectBox.option('value', paperKind >= 0 ? paperKind : null);
    };
    PageSetupDialog.prototype.updateOrientation = function (landscape) {
        if (!this.lockOrientation)
            this.orientationRadioGroup.option('value', landscape);
    };
    PageSetupDialog.prototype.updateParameters = function (parameters, data) {
        parameters.marginTop = data.marginTop;
        parameters.marginLeft = data.marginLeft;
        parameters.marginBottom = data.marginBottom;
        parameters.marginRight = data.marginRight;
        parameters.landscape = data.landscape;
        parameters.pageWidth = data.pageWidth;
        parameters.pageHeight = data.pageHeight;
        parameters.startType = data.startType;
        parameters.headerDifferentOddAndEven = data.headerDifferentOddAndEven;
        parameters.headerDifferentFirstPage = data.headerDifferentFirstPage;
        parameters.headerOffset = data.headerOffset;
        parameters.footerOffset = data.footerOffset;
        parameters.applyTo = data.applyTo;
    };
    return PageSetupDialog;
}(dialog_base_1.DialogBase));
exports.PageSetupDialog = PageSetupDialog;
