// devexpress-richedit/lib/client/dialogs/paragraph-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var localization_1 = require("devextreme/localization");
var number_box_1 = require("devextreme/ui/number_box");
var client_command_1 = require("../../base/commands/client-command");
var paragraph_properties_1 = require("../../core/model/paragraph/paragraph-properties");
var dialog_base_1 = require("./dialog-base");
var ParagraphDialog = (function (_super) {
    tslib_1.__extends(ParagraphDialog, _super);
    function ParagraphDialog() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ParagraphDialog.prototype.getTitle = function () {
        return localization_1.formatMessage('ASPxRichEditStringId.ParagraphTitle');
    };
    ParagraphDialog.prototype.getMaxWidth = function () {
        return 650;
    };
    ParagraphDialog.prototype.getFormOptions = function () {
        var _this = this;
        return {
            items: [{
                    itemType: "tabbed",
                    tabPanelOptions: {
                        deferRendering: false
                    },
                    tabs: [{
                            title: localization_1.formatMessage('ASPxRichEditStringId.ParagraphTab0'),
                            items: [{
                                    itemType: "group",
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.General'),
                                    colCount: 2,
                                    items: [{
                                            dataField: 'alignment',
                                            editorType: 'dxSelectBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Alignment'), location: 'top' },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Left'), value: paragraph_properties_1.ParagraphAlignment.Left },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Right'), value: paragraph_properties_1.ParagraphAlignment.Right },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Centered'), value: paragraph_properties_1.ParagraphAlignment.Center },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Justified'), value: paragraph_properties_1.ParagraphAlignment.Justify }
                                                ],
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                value: this.parameters.alignment
                                            }
                                        },
                                        {
                                            dataField: 'outlineLevel',
                                            editorType: 'dxSelectBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.OutlineLevel'), location: 'top' },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.BodyText'), value: 0 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 1', value: 1 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 2', value: 2 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 3', value: 3 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 4', value: 4 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 5', value: 5 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 6', value: 6 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 7', value: 7 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 8', value: 8 },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Level') + ' 9', value: 9 }
                                                ],
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                value: this.parameters.outlineLevel
                                            }
                                        }]
                                },
                                {
                                    itemType: "group",
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Indentation'),
                                    colCount: 4,
                                    items: [{
                                            dataField: 'leftIndent',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Left'), location: 'top' },
                                            editorOptions: {
                                                value: this.parameters.leftIndent,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'rightIndent',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Right'), location: 'top' },
                                            editorOptions: {
                                                value: this.parameters.rightIndent,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'firstLineIndentType',
                                            editorType: 'dxSelectBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Special'), location: 'top' },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.none'), value: paragraph_properties_1.ParagraphFirstLineIndent.None },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.FirstLine'), value: paragraph_properties_1.ParagraphFirstLineIndent.Indented },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Hanging'), value: paragraph_properties_1.ParagraphFirstLineIndent.Hanging }
                                                ],
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                value: this.parameters.firstLineIndentType
                                            }
                                        },
                                        {
                                            dataField: 'firstLineIndent',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.By'), location: 'top' },
                                            editorOptions: {
                                                value: this.parameters.firstLineIndent,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        }]
                                },
                                {
                                    itemType: "group",
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Spacing'),
                                    colCount: 4,
                                    items: [{
                                            dataField: 'spacingBefore',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.Before'), location: 'top' },
                                            editorOptions: {
                                                value: this.parameters.spacingBefore,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'spacingAfter',
                                            editorType: 'dxNumberBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.After'), location: 'top' },
                                            editorOptions: {
                                                value: this.parameters.spacingAfter,
                                                showSpinButtons: true,
                                                step: 0.1,
                                                format: '#0.##'
                                            }
                                        },
                                        {
                                            dataField: 'lineSpacingType',
                                            editorType: 'dxSelectBox',
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.LineSpacing'), location: 'top' },
                                            editorOptions: {
                                                items: [
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Single'), value: paragraph_properties_1.ParagraphLineSpacingType.Single },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.spacing_1_5_lines'), value: paragraph_properties_1.ParagraphLineSpacingType.Sesquialteral },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Double'), value: paragraph_properties_1.ParagraphLineSpacingType.Double },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Multiple'), value: paragraph_properties_1.ParagraphLineSpacingType.Multiple },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.Exactly'), value: paragraph_properties_1.ParagraphLineSpacingType.Exactly },
                                                    { text: localization_1.formatMessage('ASPxRichEditStringId.AtLeast'), value: paragraph_properties_1.ParagraphLineSpacingType.AtLeast }
                                                ],
                                                valueExpr: 'value',
                                                displayExpr: 'text',
                                                value: this.parameters.lineSpacingType,
                                                onValueChanged: function (e) { _this.updateAtNumberBox(e.component.option('value')); }
                                            }
                                        },
                                        {
                                            label: { text: localization_1.formatMessage('ASPxRichEditStringId.At'), location: 'top' },
                                            template: function () {
                                                var element = document.createElement('div');
                                                _this.atNumberBox = new number_box_1.default(element, {
                                                    showSpinButtons: true,
                                                    format: '#0.##'
                                                });
                                                return element;
                                            }
                                        }]
                                },
                                {
                                    dataField: 'contextualSpacing',
                                    editorType: 'dxCheckBox',
                                    label: { visible: false },
                                    editorOptions: {
                                        value: this.parameters.contextualSpacing,
                                        text: localization_1.formatMessage('ASPxRichEditStringId.NoSpace')
                                    }
                                }]
                        }, {
                            title: localization_1.formatMessage('ASPxRichEditStringId.ParagraphTab1'),
                            items: [{
                                    itemType: "group",
                                    caption: localization_1.formatMessage('ASPxRichEditStringId.Pagination'),
                                    items: [
                                        {
                                            dataField: 'pageBreakBefore',
                                            editorType: 'dxCheckBox',
                                            label: { visible: false },
                                            editorOptions: {
                                                value: this.parameters.pageBreakBefore,
                                                text: localization_1.formatMessage('ASPxRichEditStringId.PBB')
                                            }
                                        }
                                    ]
                                }]
                        }]
                }]
        };
    };
    ParagraphDialog.prototype.afterShowing = function () {
        var data = this.form.option("formData");
        var type = data.lineSpacingType;
        this.updateAtNumberBox(type);
    };
    ParagraphDialog.prototype.updateAtNumberBox = function (type) {
        switch (type) {
            case paragraph_properties_1.ParagraphLineSpacingType.Single:
            case paragraph_properties_1.ParagraphLineSpacingType.Sesquialteral:
            case paragraph_properties_1.ParagraphLineSpacingType.Double:
                this.atNumberBox.option('value', undefined);
                this.atNumberBox.option('disabled', true);
                break;
            case paragraph_properties_1.ParagraphLineSpacingType.Multiple:
                this.atNumberBox.option('step', 0.5);
                this.atNumberBox.option('disabled', false);
                this.atNumberBox.option('value', this.parameters.lineSpacingMultiple);
                this.atNumberBox.option('min', 1);
                break;
            default:
                this.atNumberBox.option('step', 0.1);
                this.atNumberBox.option('disabled', false);
                this.atNumberBox.option('value', this.parameters.lineSpacing);
                this.atNumberBox.option('min', undefined);
                break;
        }
    };
    ParagraphDialog.prototype.getToolbarItems = function () {
        return [this.getTabsToolbarItem(), this.getOkToolbarItem(), this.getCancelToolbarItem()];
    };
    ParagraphDialog.prototype.getTabsToolbarItem = function () {
        var _this = this;
        return {
            widget: 'dxButton',
            location: 'before',
            toolbar: 'bottom',
            options: {
                text: localization_1.formatMessage('ASPxRichEditStringId.TabsTitle') + '...',
                onClick: function () {
                    _this.popupDialog.hide();
                    var commandManager = _this.richedit.commandManager;
                    commandManager.getCommand(client_command_1.RichEditClientCommand.ShowTabsForm).execute(commandManager.isPublicApiCall);
                }
            }
        };
    };
    ParagraphDialog.prototype.updateParameters = function (parameters, data) {
        parameters.alignment = data.alignment;
        parameters.outlineLevel = data.outlineLevel;
        parameters.leftIndent = data.leftIndent;
        parameters.rightIndent = data.rightIndent;
        parameters.firstLineIndentType = data.firstLineIndentType;
        parameters.firstLineIndent = data.firstLineIndent;
        parameters.spacingBefore = data.spacingBefore;
        parameters.spacingAfter = data.spacingAfter;
        var lineSpacingType = data.lineSpacingType;
        parameters.lineSpacingType = lineSpacingType;
        if (lineSpacingType == paragraph_properties_1.ParagraphLineSpacingType.Multiple)
            parameters.lineSpacingMultiple = data.lineSpacingMultiple;
        if (lineSpacingType == paragraph_properties_1.ParagraphLineSpacingType.Exactly || lineSpacingType == paragraph_properties_1.ParagraphLineSpacingType.AtLeast)
            parameters.lineSpacing = data.lineSpacing;
        parameters.contextualSpacing = data.contextualSpacing;
        parameters.keepLinesTogether = data.keepLinesTogether;
        parameters.pageBreakBefore = data.pageBreakBefore;
    };
    return ParagraphDialog;
}(dialog_base_1.DialogBase));
exports.ParagraphDialog = ParagraphDialog;
