// devexpress-richedit/lib/base/commands/floating-objects/context-item-floating-objects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var command_base_1 = require("../command-base");
var command_states_1 = require("../command-states");
var ContextItemFloatingObjects = (function (_super) {
    tslib_1.__extends(ContextItemFloatingObjects, _super);
    function ContextItemFloatingObjects() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContextItemFloatingObjects.prototype.getState = function () {
        var specialRunInfo = this.selection.specialRunInfo;
        var state = new command_states_1.SimpleCommandState(!this.selection.activeSubDocument.isTextBox(), false);
        state.visible = specialRunInfo.isSelected() && state.enabled &&
            this.getFloatingObjectParentSubDocument().isEditable([new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)]);
        return state;
    };
    ContextItemFloatingObjects.prototype.canModify = function () {
        return true;
    };
    ContextItemFloatingObjects.prototype.executeCore = function (_state, _parameter) {
        return false;
    };
    return ContextItemFloatingObjects;
}(command_base_1.CommandBase));
exports.ContextItemFloatingObjects = ContextItemFloatingObjects;
var ChangeFloatingObjectTextWrapTypeMenu = (function (_super) {
    tslib_1.__extends(ChangeFloatingObjectTextWrapTypeMenu, _super);
    function ChangeFloatingObjectTextWrapTypeMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ChangeFloatingObjectTextWrapTypeMenu;
}(ContextItemFloatingObjects));
exports.ChangeFloatingObjectTextWrapTypeMenu = ChangeFloatingObjectTextWrapTypeMenu;
var FloatingObjectBringForwardMenu = (function (_super) {
    tslib_1.__extends(FloatingObjectBringForwardMenu, _super);
    function FloatingObjectBringForwardMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FloatingObjectBringForwardMenu;
}(ContextItemFloatingObjects));
exports.FloatingObjectBringForwardMenu = FloatingObjectBringForwardMenu;
var FloatingObjectSendBackwardMenu = (function (_super) {
    tslib_1.__extends(FloatingObjectSendBackwardMenu, _super);
    function FloatingObjectSendBackwardMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FloatingObjectSendBackwardMenu;
}(ContextItemFloatingObjects));
exports.FloatingObjectSendBackwardMenu = FloatingObjectSendBackwardMenu;
