// devexpress-richedit/lib/base/commands/floating-objects/set-floating-object-alignment-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var enums_1 = require("../../../core/model/floating-objects/enums");
var anchor_info_property_history_item_1 = require("../../../core/model/history/items/floating-objects/anchor-info-property-history-item");
var run_type_1 = require("../../../core/model/runs/run-type");
var sub_document_1 = require("../../../core/model/sub-document");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var client_command_1 = require("../client-command");
var command_base_1 = require("../command-base");
var command_states_1 = require("../command-states");
var SetFloatingObjectAlignmentCommandBase = (function (_super) {
    tslib_1.__extends(SetFloatingObjectAlignmentCommandBase, _super);
    function SetFloatingObjectAlignmentCommandBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalPositionType = enums_1.AnchorObjectHorizontalPositionType.Margin;
        _this.anchorObjectVerticalPositionType = enums_1.AnchorObjectVerticalPositionType.Margin;
        return _this;
    }
    SetFloatingObjectAlignmentCommandBase.prototype.getRelatedCommands = function () {
        var _a;
        return _a = {},
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectTopLeftAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectTopCenterAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectTopRightAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectMiddleLeftAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectMiddleCenterAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectMiddleRightAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectBottomLeftAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectBottomCenterAlignment] = true,
            _a[client_command_1.RichEditClientCommand.SetFloatingObjectBottomRightAlignment] = true,
            _a;
    };
    SetFloatingObjectAlignmentCommandBase.prototype.getState = function () {
        return new command_states_1.SimpleCommandState(this.isEnabled(), this.getStateValue());
    };
    SetFloatingObjectAlignmentCommandBase.prototype.getStateValue = function () {
        var value = false;
        if (this.isEnabled()) {
            var specialRunInfo = this.selection.specialRunInfo;
            var anchoredRun = specialRunInfo.getParentSubDocument()
                .getRunByPosition(specialRunInfo.getPosition());
            if (anchoredRun.anchorInfo.horizontalPositionAlignment === this.anchorObjectHorizontalAlignment
                && anchoredRun.anchorInfo.verticalPositionAlignment === this.anchorObjectVerticalAlignment
                && anchoredRun.anchorInfo.horizontalPositionType === this.anchorObjectHorizontalPositionType
                && anchoredRun.anchorInfo.verticalPositionType === this.anchorObjectVerticalPositionType)
                value = true;
        }
        return value;
    };
    SetFloatingObjectAlignmentCommandBase.prototype.canModify = function () {
        return true;
    };
    SetFloatingObjectAlignmentCommandBase.prototype.isEnabled = function () {
        var specialRunInfo = this.selection.specialRunInfo;
        return _super.prototype.isEnabled.call(this) && specialRunInfo.isSelected() && specialRunInfo.isSelectedAnchorObject &&
            this.getFloatingObjectParentSubDocument().isEditable([new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)]);
    };
    SetFloatingObjectAlignmentCommandBase.prototype.executeCore = function (_state, _parameter) {
        var modelManipulator = this.modelManipulator;
        var specialRunInfo = this.selection.specialRunInfo;
        var subDocument = specialRunInfo.getParentSubDocument();
        var run = subDocument.getRunByPosition(specialRunInfo.getPosition());
        var anchoredRun = run.getType() == run_type_1.RunType.AnchoredPictureRun ? run : run;
        if (anchoredRun.anchorInfo.horizontalPositionAlignment !== this.anchorObjectHorizontalAlignment || anchoredRun.anchorInfo.verticalPositionAlignment !== this.anchorObjectVerticalAlignment) {
            this.history.beginTransaction();
            this.history.addAndRedo(new anchor_info_property_history_item_1.AnchorInfoPropertyHistoryItem(modelManipulator, new sub_document_1.SubDocumentInterval(subDocument, new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)), this.anchorObjectHorizontalAlignment, modelManipulator.floatingObject.anchorInfo.horizontalPositionAlignment));
            this.history.addAndRedo(new anchor_info_property_history_item_1.AnchorInfoPropertyHistoryItem(modelManipulator, new sub_document_1.SubDocumentInterval(subDocument, new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)), this.anchorObjectVerticalAlignment, modelManipulator.floatingObject.anchorInfo.verticalPositionAlignment));
            this.history.addAndRedo(new anchor_info_property_history_item_1.AnchorInfoPropertyHistoryItem(modelManipulator, new sub_document_1.SubDocumentInterval(subDocument, new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)), this.anchorObjectHorizontalPositionType, modelManipulator.floatingObject.anchorInfo.horizontalPositionType));
            this.history.addAndRedo(new anchor_info_property_history_item_1.AnchorInfoPropertyHistoryItem(modelManipulator, new sub_document_1.SubDocumentInterval(subDocument, new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)), this.anchorObjectVerticalPositionType, modelManipulator.floatingObject.anchorInfo.verticalPositionType));
            this.history.endTransaction();
        }
        return true;
    };
    return SetFloatingObjectAlignmentCommandBase;
}(command_base_1.CommandBase));
exports.SetFloatingObjectAlignmentCommandBase = SetFloatingObjectAlignmentCommandBase;
var SetFloatingObjectTopLeftAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectTopLeftAlignmentCommand, _super);
    function SetFloatingObjectTopLeftAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Left;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Top;
        return _this;
    }
    return SetFloatingObjectTopLeftAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectTopLeftAlignmentCommand = SetFloatingObjectTopLeftAlignmentCommand;
var SetFloatingObjectTopCenterAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectTopCenterAlignmentCommand, _super);
    function SetFloatingObjectTopCenterAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Center;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Top;
        return _this;
    }
    return SetFloatingObjectTopCenterAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectTopCenterAlignmentCommand = SetFloatingObjectTopCenterAlignmentCommand;
var SetFloatingObjectTopRightAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectTopRightAlignmentCommand, _super);
    function SetFloatingObjectTopRightAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Right;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Top;
        return _this;
    }
    return SetFloatingObjectTopRightAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectTopRightAlignmentCommand = SetFloatingObjectTopRightAlignmentCommand;
var SetFloatingObjectMiddleLeftAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectMiddleLeftAlignmentCommand, _super);
    function SetFloatingObjectMiddleLeftAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Left;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Center;
        return _this;
    }
    return SetFloatingObjectMiddleLeftAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectMiddleLeftAlignmentCommand = SetFloatingObjectMiddleLeftAlignmentCommand;
var SetFloatingObjectMiddleCenterAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectMiddleCenterAlignmentCommand, _super);
    function SetFloatingObjectMiddleCenterAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Center;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Center;
        return _this;
    }
    return SetFloatingObjectMiddleCenterAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectMiddleCenterAlignmentCommand = SetFloatingObjectMiddleCenterAlignmentCommand;
var SetFloatingObjectMiddleRightAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectMiddleRightAlignmentCommand, _super);
    function SetFloatingObjectMiddleRightAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Right;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Center;
        return _this;
    }
    return SetFloatingObjectMiddleRightAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectMiddleRightAlignmentCommand = SetFloatingObjectMiddleRightAlignmentCommand;
var SetFloatingObjectBottomLeftAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectBottomLeftAlignmentCommand, _super);
    function SetFloatingObjectBottomLeftAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Left;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Bottom;
        return _this;
    }
    return SetFloatingObjectBottomLeftAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectBottomLeftAlignmentCommand = SetFloatingObjectBottomLeftAlignmentCommand;
var SetFloatingObjectBottomCenterAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectBottomCenterAlignmentCommand, _super);
    function SetFloatingObjectBottomCenterAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Center;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Bottom;
        return _this;
    }
    return SetFloatingObjectBottomCenterAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectBottomCenterAlignmentCommand = SetFloatingObjectBottomCenterAlignmentCommand;
var SetFloatingObjectBottomRightAlignmentCommand = (function (_super) {
    tslib_1.__extends(SetFloatingObjectBottomRightAlignmentCommand, _super);
    function SetFloatingObjectBottomRightAlignmentCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.anchorObjectHorizontalAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.Right;
        _this.anchorObjectVerticalAlignment = enums_1.AnchorObjectVerticalPositionAlignment.Bottom;
        return _this;
    }
    return SetFloatingObjectBottomRightAlignmentCommand;
}(SetFloatingObjectAlignmentCommandBase));
exports.SetFloatingObjectBottomRightAlignmentCommand = SetFloatingObjectBottomRightAlignmentCommand;
