// devexpress-richedit/lib/base/commands/paragraph-properties/add-spacing-after-paragraph-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var paragraph_properties_1 = require("../../../core/model/paragraph/paragraph-properties");
var unit_converter_1 = require("@devexpress/utils/lib/class/unit-converter");
var client_command_1 = require("../client-command");
var change_paragraph_properties_command_base_1 = require("./change-paragraph-properties-command-base");
var LineSpacingParagraphCommandBase = (function (_super) {
    tslib_1.__extends(LineSpacingParagraphCommandBase, _super);
    function LineSpacingParagraphCommandBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LineSpacingParagraphCommandBase.prototype.getRelatedCommands = function () {
        var _a;
        return _a = {},
            _a[client_command_1.RichEditClientCommand.AddSpacingAfterParagraph] = true,
            _a[client_command_1.RichEditClientCommand.AddSpacingBeforeParagraph] = true,
            _a[client_command_1.RichEditClientCommand.SetSingleParagraphSpacing] = true,
            _a[client_command_1.RichEditClientCommand.SetSesquialteralParagraphSpacing] = true,
            _a[client_command_1.RichEditClientCommand.SetDoubleParagraphSpacing] = true,
            _a[client_command_1.RichEditClientCommand.RemoveSpacingAfterParagraph] = true,
            _a[client_command_1.RichEditClientCommand.RemoveSpacingBeforeParagraph] = true,
            _a;
    };
    return LineSpacingParagraphCommandBase;
}(change_paragraph_properties_command_base_1.ChangeParagraphPropertiesCommandBase));
exports.LineSpacingParagraphCommandBase = LineSpacingParagraphCommandBase;
var AddSpacingAfterParagraphCommand = (function (_super) {
    tslib_1.__extends(AddSpacingAfterParagraphCommand, _super);
    function AddSpacingAfterParagraphCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AddSpacingAfterParagraphCommand.prototype.getDescriptor = function () {
        return paragraph_properties_1.ParagraphPropertyDescriptor.spacingAfter;
    };
    AddSpacingAfterParagraphCommand.prototype.isEnabled = function () {
        return _super.prototype.isEnabled.call(this) && this.getCurrentValue() === 0;
    };
    AddSpacingAfterParagraphCommand.prototype.DEPRECATEDConvertOptionsParameter = function (parameter) {
        return parameter == null ? unit_converter_1.UnitConverter.pointsToTwips(12) : parameter;
    };
    return AddSpacingAfterParagraphCommand;
}(LineSpacingParagraphCommandBase));
exports.AddSpacingAfterParagraphCommand = AddSpacingAfterParagraphCommand;
