// devexpress-dashboard/model/custom-properties/_custom-properties-utils.js
/**
* DevExpress Dashboard (_custom-properties-utils.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _base_metadata_1 = require("../metadata/_base-metadata");
var custom_properties_1 = require("./custom-properties");
var custom_properties_metadata_1 = require("./custom-properties-metadata");
function _convertCustomPropertiesMetadata(customPropertiesMeta) {
    return (customPropertiesMeta || [])
        .map(function (d) {
        var serializationInfo = {
            propertyName: d.propertyName,
            modelName: d.propertyName,
            category: _base_metadata_1.PropertyCategory.ViewModel,
            defaultVal: d.defaultValue,
        };
        if (d.valueType === 'boolean') {
            serializationInfo.from = _base_metadata_1.parseBool;
        }
        if (d.valueType === 'number') {
            serializationInfo.from = _base_metadata_1.floatFromModel;
        }
        return serializationInfo;
    });
}
function _getCustomPropertiesSerializationInfoCore(customPropertiesMeta) {
    return {
        propertyName: 'customProperties',
        modelName: 'CustomProperties',
        category: _base_metadata_1.PropertyCategory.ViewModel,
        type: custom_properties_1.CustomProperties,
        info: _convertCustomPropertiesMetadata(customPropertiesMeta)
    };
}
exports._getCustomPropertiesSerializationInfoCore = _getCustomPropertiesSerializationInfoCore;
function getCustomPropertiesSerializationInfo(owner) {
    var filteredCustomProperties = custom_properties_metadata_1._customPropertiesMeta
        .filter(function (metadata) { return owner instanceof metadata.ownerType; });
    return _getCustomPropertiesSerializationInfoCore(filteredCustomProperties);
}
exports.getCustomPropertiesSerializationInfo = getCustomPropertiesSerializationInfo;
