// devexpress-dashboard/designer/data-source-wizard/wizards/_edit-query-wizard.js
/**
* DevExpress Dashboard (_edit-query-wizard.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var analytics_wizard_1 = require("@devexpress/analytics-core/analytics-wizard");
var $ = require("jquery");
var _default_1 = require("../../../data/localization/_default");
var data_source_wizard_model_1 = require("../models/data-source-wizard-model");
var _data_source_wizard_model_1 = require("../models/_data-source-wizard-model");
var choose_data_source_type_page_1 = require("../pages/choose-data-source-type-page");
var choose_olap_connection_string_page_1 = require("../pages/choose-olap-connection-string-page");
var _choose_json_schema_page_1 = require("../pages/_choose-json-schema-page");
var _create_query_page_1 = require("../pages/_create-query-page");
var _parameters_item_provider_1 = require("../pages/_parameters-item-provider");
var _data_source_wizard_page_iterator_1 = require("./_data-source-wizard-page-iterator");
var DashboardQueryWizardIterator = (function (_super) {
    __extends(DashboardQueryWizardIterator, _super);
    function DashboardQueryWizardIterator(_createNew, factory, stateManager) {
        var _this = _super.call(this, factory, stateManager) || this;
        _this._createNew = _createNew;
        return _this;
    }
    DashboardQueryWizardIterator.prototype.getInitialPage = function () {
        if (this._createNew) {
            return _super.prototype.getInitialPage.call(this);
        }
        return analytics_wizard_1.SqlDataSourceWizardPageId.ConfigureQueryPage;
    };
    return DashboardQueryWizardIterator;
}(_data_source_wizard_page_iterator_1.DataSourceWizardPageIteratorBase));
exports.DashboardQueryWizardIterator = DashboardQueryWizardIterator;
var DashboardQueryWizard = (function (_super) {
    __extends(DashboardQueryWizard, _super);
    function DashboardQueryWizard() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.title = _default_1.getLocalizationById('DashboardWebStringId.DataSources.DashboardDataSourceWizard');
        _this._container = function (element) { return $(element).closest('.dx-dashboard-widget-container'); };
        _this._extendCssClass = 'dxrd-sqldatasource-wizard';
        return _this;
    }
    return DashboardQueryWizard;
}(analytics_wizard_1.PopupWizard));
exports.DashboardQueryWizard = DashboardQueryWizard;
var DashboardDataSourceWizardState = (function () {
    function DashboardDataSourceWizardState() {
        this.sqlDataSourceWizard = {};
        this.jsonDataSourceWizard = {};
        this.olapDataSourceWizard = {};
        this.objectDataSourceWizard = {};
    }
    Object.defineProperty(DashboardDataSourceWizardState.prototype, "dashboardDataSourceType", {
        get: function () {
            return this.dataSourceType && data_source_wizard_model_1.ToDashboardDataSourceType(this.dataSourceType) || undefined;
        },
        set: function (value) {
            this.dataSourceType = data_source_wizard_model_1.ToDataSourceTypeNumber(value);
        },
        enumerable: true,
        configurable: true
    });
    return DashboardDataSourceWizardState;
}());
exports.DashboardDataSourceWizardState = DashboardDataSourceWizardState;
function createDashboardQueryWizard(requestWrapper, parameters, disableCustomSql, allowCreateNewJsonConnection, wizardSettings, dashboardConnectionStrings) {
    var factory = new analytics_wizard_1.PageFactory();
    var wizardOptions = new _data_source_wizard_model_1.DashboardDataSourceWizardOptions();
    wizardOptions.callbacks = _create_query_page_1.createQueryPageCallback(requestWrapper, parameters);
    wizardOptions.disableCustomSql = disableCustomSql;
    wizardOptions.requestWrapper = requestWrapper;
    wizardOptions.wizardSettings = wizardSettings;
    wizardOptions.allowCreateNewJsonConnection = allowCreateNewJsonConnection || false;
    wizardOptions.connectionStrings = dashboardConnectionStrings;
    choose_data_source_type_page_1._registerChooseDataSourceTypePage(factory, wizardOptions);
    analytics_wizard_1._registerChooseSqlConnectionPage(factory, dashboardConnectionStrings.sql);
    choose_olap_connection_string_page_1._registerOlapConnectionStringsPage(factory, dashboardConnectionStrings.olap);
    analytics_wizard_1._registerChooseJsonConnectionPage(factory, wizardOptions);
    analytics_wizard_1._registerConfigureQueryPage(factory, wizardOptions);
    analytics_wizard_1._registerConfigureParametersPage(factory, requestWrapper);
    analytics_wizard_1._registerChooseJsonSourcePage(factory, requestWrapper, function () { return new _parameters_item_provider_1.ParametersItemProvider(parameters()); });
    _choose_json_schema_page_1._registerChooseJsonSchemaPage(factory, requestWrapper);
    analytics_wizard_1._registerChooseObjectDataSourceTypesPage(factory, wizardOptions);
    analytics_wizard_1._registerChooseObjectDataSourceDataMembersPage(factory, wizardOptions);
    analytics_wizard_1._registerConfigureObjectDataSourceParametersPage(factory, function () { return new _parameters_item_provider_1.ParametersItemProvider(parameters()); });
    return new DashboardQueryWizard(factory);
}
exports.createDashboardQueryWizard = createDashboardQueryWizard;
