// devexpress-richedit/lib/base/commands/character-properties/change-font-back-color-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var character_property_descriptor_1 = require("../../../core/model/character/character-property-descriptor");
var color_1 = require("../../../core/model/color/color");
var color_model_info_1 = require("../../../core/model/color/color-model-info");
var shading_info_1 = require("../../../core/model/shadings/shading-info");
var change_character_properties_command_base_1 = require("./change-character-properties-command-base");
var ChangeFontHighlightColorCommand = (function (_super) {
    tslib_1.__extends(ChangeFontHighlightColorCommand, _super);
    function ChangeFontHighlightColorCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChangeFontHighlightColorCommand.prototype.getDescriptor = function () {
        return character_property_descriptor_1.CharacterPropertyDescriptor.highlightColor;
    };
    ChangeFontHighlightColorCommand.prototype.getValueForState = function (val) {
        return val ? val.toRgb(this.control.modelManager.model.colorProvider) : undefined;
    };
    ChangeFontHighlightColorCommand.prototype.isLockUpdateValue = function () {
        return true;
    };
    ChangeFontHighlightColorCommand.prototype.DEPRECATEDConvertOptionsParameter = function (parameter) {
        return color_model_info_1.ColorModelInfo.makeByColor(color_1.ColorHelper.anyToColor(parameter, color_1.ColorHelper.NO_COLOR));
    };
    return ChangeFontHighlightColorCommand;
}(change_character_properties_command_base_1.ChangeCharacterPropertiesCommandBase));
exports.ChangeFontHighlightColorCommand = ChangeFontHighlightColorCommand;
var ChangeFontShadingInfoCommand = (function (_super) {
    tslib_1.__extends(ChangeFontShadingInfoCommand, _super);
    function ChangeFontShadingInfoCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChangeFontShadingInfoCommand.prototype.getDescriptor = function () {
        return character_property_descriptor_1.CharacterPropertyDescriptor.shadingInfo;
    };
    ChangeFontShadingInfoCommand.prototype.getValueForState = function (val) {
        return val ? val.getActualColor(this.colorProvider) : undefined;
    };
    ChangeFontShadingInfoCommand.prototype.isLockUpdateValue = function () {
        return true;
    };
    ChangeFontShadingInfoCommand.prototype.DEPRECATEDConvertOptionsParameter = function (parameter) {
        return shading_info_1.ShadingInfo.createByColor(color_model_info_1.ColorModelInfo.makeByColor(color_1.ColorHelper.anyToColor(parameter, color_1.ColorHelper.NO_COLOR)));
    };
    return ChangeFontShadingInfoCommand;
}(change_character_properties_command_base_1.ChangeCharacterPropertiesCommandBase));
exports.ChangeFontShadingInfoCommand = ChangeFontShadingInfoCommand;
