// devexpress-richedit/lib/client/public/ribbon/creator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var errors_1 = require("@devexpress/utils/lib/errors");
var list_1 = require("@devexpress/utils/lib/utils/list");
var ribbon_items_data_1 = require("../../bars/rich-edit-ribbon/ribbon-items-data");
var base_1 = require("./items/base");
var button_1 = require("./items/button");
var color_box_1 = require("./items/color-box");
var menu_1 = require("./items/menu");
var number_box_1 = require("./items/number-box");
var select_box_1 = require("./items/select-box");
var sub_menu_1 = require("./items/sub-menu");
var ribbon_1 = require("./ribbon");
var tab_1 = require("./tab");
function createRibbon() {
    var ribbon = new ribbon_1.Ribbon();
    addTabs(ribbon);
    addContextTabCategories(ribbon);
    return ribbon;
}
exports.createRibbon = createRibbon;
function addTabs(ribbon, tabs) {
    if (tabs === void 0) { tabs = ribbon_items_data_1.RibbonItemsData.getDefaultItems(); }
    tabs.forEach(function (tab) { return ribbon.insertTab(createTab(tab)); });
}
exports.addTabs = addTabs;
function addContextTabCategories(ribbon, contextTabCategories) {
    if (contextTabCategories === void 0) { contextTabCategories = ribbon_items_data_1.RibbonItemsData.getDefaultContextItemsCategories(); }
    contextTabCategories.forEach(function (context) {
        return context.items.forEach(function (tab) { return ribbon.insertTab(createTab(tab, context.name)); });
    });
}
exports.addContextTabCategories = addContextTabCategories;
function createTab(rawTab, contextName) {
    var items = createItems(rawTab.items);
    return contextName === undefined ?
        new tab_1.RibbonTab(rawTab.title, rawTab.id, items, rawTab.localizationId) :
        new tab_1.RibbonContextTab(rawTab.title, rawTab.id, contextName, items, rawTab.localizationId);
}
function createItems(rawItems) {
    return list_1.ListUtils.map(rawItems, function (rawItem) {
        switch (rawItem.type) {
            case 'Button':
                return new button_1.RibbonButtonItem(rawItem.name, rawItem.text, {
                    showText: rawItem.alwaysShowText,
                    toggleMode: rawItem.isToggleMode,
                    icon: rawItem.icon,
                    beginGroup: rawItem.beginGroup,
                    localizationId: rawItem.localizationId,
                    selected: rawItem.selected
                });
            case 'ButtonGroup':
                throw new Error(errors_1.Errors.NotImplemented);
            case 'ColorBox':
                return new color_box_1.RibbonColorBoxItem(rawItem.name, rawItem.text, rawItem.value, {
                    beginGroup: rawItem.beginGroup,
                    localizationId: rawItem.localizationId,
                    textOptions: rawItem.textOptions
                });
            case 'Menu':
                return new menu_1.RibbonMenuItem(rawItem.name, rawItem.text, createSubItems(rawItem.items), {
                    beginGroup: rawItem.beginGroup,
                    icon: rawItem.icon,
                    showText: rawItem.alwaysShowText,
                    localizationId: rawItem.localizationId
                });
            case 'NumberBox':
                return new number_box_1.RibbonNumberBoxItem(rawItem.name, rawItem.text, {
                    beginGroup: rawItem.beginGroup,
                    min: rawItem.min,
                    max: rawItem.max,
                    step: rawItem.step,
                    width: rawItem.width,
                    localizationId: rawItem.localizationId,
                    format: rawItem.format,
                    value: rawItem.value
                });
            case 'SelectBox':
                return new select_box_1.RibbonSelectBoxItem(rawItem.name, rawItem.dataSource, {
                    beginGroup: rawItem.beginGroup,
                    width: rawItem.width,
                    displayExpr: rawItem.displayExpr,
                    valueExpr: rawItem.valueExpr,
                    value: rawItem.value,
                    showClearButton: rawItem.showClearButton,
                    placeholder: rawItem.placeholder,
                    acceptCustomValue: rawItem.acceptCustomValue,
                    onCustomItemCreating: rawItem.onCustomItemCreating,
                    textOptions: rawItem.textOptions,
                    _localizeDataSourceItems: rawItem._localizeDataSourceItems
                });
            default: throw new Error(errors_1.Errors.InternalException);
        }
    });
    function createSubItems(items) {
        if (!items)
            return [];
        return list_1.ListUtils.map(items, function (item) {
            return new sub_menu_1.RibbonSubMenuItem(item.name, item.text, createSubItems(item.items), {
                icon: item.icon,
                beginGroup: item.beginGroup,
                localizationId: item.localizationId,
            });
        });
    }
}
function createInnerTab(tab) {
    return {
        id: tab.id,
        title: tab.title,
        visible: true,
        items: createInnerItems(tab.items),
        localizationId: tab.localizationId,
    };
}
exports.createInnerTab = createInnerTab;
function createInnerItems(items) {
    return list_1.ListUtils.map(items, function (item) {
        switch (item.type) {
            case base_1.RibbonItemType.Button:
                return {
                    type: 'Button',
                    name: item.id,
                    beginGroup: item.beginGroup,
                    text: item.text,
                    icon: item.icon,
                    alwaysShowText: item.showText,
                    isToggleMode: item.toggleMode,
                    selected: item.selected,
                    localizationId: item.localizationId,
                };
            case base_1.RibbonItemType.ColorBox:
                return {
                    type: 'ColorBox',
                    text: item.text,
                    value: item.value,
                    textOptions: item.textOptions,
                    name: item.id,
                    beginGroup: item.beginGroup,
                    localizationId: item.localizationId,
                };
            case base_1.RibbonItemType.Menu:
                return {
                    type: 'Menu',
                    items: getSubMenuItems(item.items),
                    name: item.id,
                    beginGroup: item.beginGroup,
                    text: item.text,
                    icon: item.icon,
                    alwaysShowText: item.showText,
                    localizationId: item.localizationId,
                };
            case base_1.RibbonItemType.NumberBox:
                return {
                    type: 'NumberBox',
                    min: item.min,
                    max: item.max,
                    step: item.step,
                    text: item.text,
                    width: item.width,
                    name: item.id,
                    beginGroup: item.beginGroup,
                    localizationId: item.localizationId,
                    format: item.format,
                    value: item.value
                };
            case base_1.RibbonItemType.SelectBox:
                return {
                    type: 'SelectBox',
                    dataSource: item.dataSource,
                    width: item.width,
                    displayExpr: item.displayExpr,
                    valueExpr: item.valueExpr,
                    name: item.id,
                    beginGroup: item.beginGroup,
                    value: item.value,
                    textOptions: item.textOptions,
                    showClearButton: item.showClearButton,
                    placeholder: item.placeholder,
                    acceptCustomValue: item.acceptCustomValue,
                    onCustomItemCreating: item.onCustomItemCreating,
                    _localizeDataSourceItems: item._localizeDataSourceItems
                };
            default:
                throw new Error(item.type == base_1.RibbonItemType.SubMenu ?
                    "Sub menu item must be placed under menu item" :
                    errors_1.Errors.NotImplemented);
        }
    });
    function getSubMenuItems(items) {
        if (!items)
            return undefined;
        return list_1.ListUtils.map(items, function (item) {
            return {
                text: item.text,
                name: item.id,
                icon: item.icon,
                items: getSubMenuItems(item.items),
                beginGroup: item.beginGroup,
                localizationId: item.localizationId,
            };
        });
    }
}
exports.createInnerItems = createInnerItems;
