// devexpress-richedit/lib/model-api/collections/drawing-object-collection-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var parameter_checker_1 = require("../../api-utils/api-utils/parameter-checker");
var model_iterator_1 = require("../../core/model/model-iterator");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var list_1 = require("@devexpress/utils/lib/utils/list");
var model_parameter_checker_1 = require("../api-utils/model-parameter-checker");
var collection_1 = require("./collection");
var RunWithPosition = (function () {
    function RunWithPosition(run, position) {
        this.run = run;
        this.position = position;
    }
    return RunWithPosition;
}());
exports.RunWithPosition = RunWithPosition;
var DrawingObjectCollectionBase = (function (_super) {
    tslib_1.__extends(DrawingObjectCollectionBase, _super);
    function DrawingObjectCollectionBase(processor, subDocument) {
        var _this = _super.call(this, processor) || this;
        _this._subDocument = subDocument;
        return _this;
    }
    DrawingObjectCollectionBase.prototype.foreach = function (callback) {
        if (callback === void 0) { callback = function () { }; }
        var iterator = new model_iterator_1.ModelIterator(this._subDocument, true);
        iterator.setPosition(0);
        do {
            if (list_1.ListUtils.unsafeAnyOf(this._runTypes, function (type) { return type === iterator.run.getType(); })) {
                var pos = this._subDocument.positionManager.registerPosition(iterator.getAbsolutePosition());
                callback(this._getItem(new RunWithPosition(iterator.run, iterator.getAbsolutePosition())));
                iterator.setPosition(pos.value);
                this._subDocument.positionManager.unregisterPosition(pos);
            }
        } while (iterator.moveToNextChar());
    };
    DrawingObjectCollectionBase.prototype.find = function (position) {
        var interval = parameter_checker_1.ApiParametersChecker.check(position, 1, false, [
            parameter_checker_1.ApiParametersChecker.numberDescriptor("position", function (n) { return new fixed_1.FixedInterval(n, 0); }, 0, this._subDocument.getDocumentEndPosition() - 1),
            model_parameter_checker_1.ModelParametersChecker.intervalDescriptor("interval", function (interval) { return new fixed_1.FixedInterval(interval.start, interval.length); }, 0, this._subDocument.getDocumentEndPosition())
        ]);
        var result = [];
        var iterator = new model_iterator_1.ModelIterator(this._subDocument, true);
        iterator.setPosition(interval.start);
        do {
            if (list_1.ListUtils.unsafeAnyOf(this._runTypes, function (type) { return type === iterator.run.getType(); }))
                result.push(this._getItem(new RunWithPosition(iterator.run, iterator.getAbsolutePosition())));
        } while (iterator.moveToNextRun() && iterator.getAbsolutePosition() < interval.end);
        return result;
    };
    DrawingObjectCollectionBase.prototype._getCoreItems = function () {
        var result = [];
        var iterator = new model_iterator_1.ModelIterator(this._subDocument, true);
        iterator.setPosition(0);
        do {
            if (list_1.ListUtils.unsafeAnyOf(this._runTypes, function (type) { return type === iterator.run.getType(); }))
                result.push({ run: iterator.run, position: iterator.getAbsolutePosition() });
        } while (iterator.moveToNextChar());
        return result;
    };
    return DrawingObjectCollectionBase;
}(collection_1.Collection));
exports.DrawingObjectCollectionBase = DrawingObjectCollectionBase;
