// devexpress-richedit/lib/base/ui/ruler/controls/indent/left.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var min_max_1 = require("@devexpress/utils/lib/class/min-max");
var document_renderer_1 = require("../../../../canvas/renderes/common/document-renderer");
var client_command_1 = require("../../../../commands/client-command");
var ruler_paragraph_indents_command_1 = require("../../../../commands/ruler/ruler-paragraph-indents-command");
var settings_1 = require("../../settings");
var base_1 = require("../base");
var base_2 = require("./base");
var LEFT_INDENT_DRAG_HANDLE_BODY = settings_1.RICH_EDIT_CLASS_NAME_PREFIX + "leftIndentDragHandleBody";
var RulerLeftIndentDragHandle = (function (_super) {
    tslib_1.__extends(RulerLeftIndentDragHandle, _super);
    function RulerLeftIndentDragHandle(modelData, controls) {
        var _this = _super.call(this, modelData, controls) || this;
        _this.isActionLeftIndent = false;
        _this.topElement = document_renderer_1.DocumentRenderer.renderContainer(_this.modelData.settings.styles.leftIndentImage.spriteCssClass);
        _this.bodyElement = document_renderer_1.DocumentRenderer.renderContainer(LEFT_INDENT_DRAG_HANDLE_BODY);
        _this.rootElement.appendChild(_this.topElement);
        _this.rootElement.appendChild(_this.bodyElement);
        _this.controls.ruler.rootElement.appendChild(_this.rootElement);
        _this.adjustByTop();
        var mainElementWidth = _this.topElement.offsetWidth;
        _this.bodyElement.style.width = mainElementWidth + "px";
        var style = _this.rootElement.style;
        style.height = _this.topElement.offsetHeight + _this.bodyElement.offsetHeight + "px";
        style.width = mainElementWidth + "px";
        style.marginTop = _this.controls.divisions.height / 2 + "px";
        _this.bodyElement.title = _this.modelData.settings.titles.leftIndent;
        _this.topElement.title = _this.modelData.settings.titles.hangingIndent;
        _this.leftCorrection = Math.round(_this.rootElement.offsetWidth / 2);
        _this._heightOfProtrudingPart = _this.rootElement.offsetHeight - _this.controls.divisions.height / 2;
        return _this;
    }
    Object.defineProperty(RulerLeftIndentDragHandle.prototype, "heightOfProtrudingPart", {
        get: function () { return this.modelData.settings.showLeftIndent ? this._heightOfProtrudingPart : 0; },
        enumerable: true,
        configurable: true
    });
    RulerLeftIndentDragHandle.prototype.getRootClassName = function () { return this.modelData.settings.styles.leftIndent.className; };
    RulerLeftIndentDragHandle.prototype.adjustByTop = function () {
        var mainElementHeight = this.rootElement.offsetHeight;
        var divisionsControlHeight = this.controls.divisions.height;
        this.rootElement.style.marginTop = (mainElementHeight - divisionsControlHeight) / 2 + "px";
    };
    RulerLeftIndentDragHandle.prototype.getModelState = function () {
        var state = this.modelData.commandManager.getCommand(client_command_1.RichEditClientCommand.RulerParagraphLeftIndents).getState();
        return new base_1.RulerModelState(state.value.hanging, state.enabled);
    };
    RulerLeftIndentDragHandle.prototype.updateView = function () {
        var newViewState = this.controls.paragraphLeftPosition + this.currModelState.modelValue;
        if (newViewState != this.viewState) {
            this.viewState = newViewState;
            if (this.modelData.settings.showLeftIndent)
                this.rootElement.style.left = this.viewState + settings_1.RULLER_NUMBER_CORRECTION - this.leftCorrection + "px";
            else
                this.setVisible(false);
        }
    };
    RulerLeftIndentDragHandle.prototype.canHandle = function (source) {
        if (this.bodyElement == source) {
            this.isActionLeftIndent = true;
            return true;
        }
        else if (this.topElement == source) {
            this.isActionLeftIndent = false;
            return true;
        }
        return false;
    };
    RulerLeftIndentDragHandle.prototype.onMouseUp = function () {
        var commandValue = new ruler_paragraph_indents_command_1.RulerParagraphLeftIndentsCommandValue(null, this.currModelState.modelValue, this.controls.firstLineIndent.currModelState.modelValue);
        commandValue.setIntervalsInfo(this.modelData.selection.intervalsInfo);
        this.modelData.commandManager.getCommand(client_command_1.RichEditClientCommand.RulerParagraphLeftIndents)
            .execute(this.modelData.commandManager.isPublicApiCall, commandValue);
        this.finishHandle();
    };
    RulerLeftIndentDragHandle.prototype.onEscPress = function () {
        _super.prototype.onEscPress.call(this);
        this.controls.firstLineIndent.currModelState = this.controls.firstLineIndent.prevModelState.clone();
    };
    RulerLeftIndentDragHandle.prototype.calculateNewModelState = function (distance) {
        var parLeftPos = this.controls.paragraphLeftPosition;
        if (this.isActionLeftIndent) {
            var minPrevValue = Math.min(this.prevModelState.modelValue, this.controls.firstLineIndent.prevModelState.modelValue);
            var maxPrevValue = Math.max(this.prevModelState.modelValue, this.controls.firstLineIndent.prevModelState.modelValue);
            var diff = maxPrevValue - minPrevValue;
            var newModelValue = this.controls.chooseClosestAnchorPosition(parLeftPos + minPrevValue + distance, [parLeftPos + minPrevValue], new min_max_1.MinMaxNumber(0, Math.max(0, this.controls.rightIndent.viewStateRelativeLeft - diff - base_2.RulerMinDistanceBetweenIndents))) -
                parLeftPos;
            if (this.prevModelState.modelValue < this.controls.firstLineIndent.prevModelState.modelValue) {
                this.currModelState.modelValue = newModelValue;
                this.controls.firstLineIndent.currModelState.modelValue = newModelValue + diff;
            }
            else {
                this.currModelState.modelValue = newModelValue + diff;
                this.controls.firstLineIndent.currModelState.modelValue = newModelValue;
            }
        }
        else {
            this.currModelState.modelValue = this.controls.chooseClosestAnchorPosition(parLeftPos + this.prevModelState.modelValue + distance, [parLeftPos + this.prevModelState.modelValue], new min_max_1.MinMaxNumber(0, Math.max(0, this.controls.rightIndent.viewStateRelativeLeft - base_2.RulerMinDistanceBetweenIndents))) - parLeftPos;
        }
    };
    return RulerLeftIndentDragHandle;
}(base_2.RulerBaseIndentControl));
exports.RulerLeftIndentDragHandle = RulerLeftIndentDragHandle;
