// devexpress-richedit/lib/base/canvas/renderes/selection-renderers/range-permission-renderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var layout_page_1 = require("../../../../core/layout/main-structures/layout-page");
var log_1 = require("../../../../core/rich-utils/debug/logger/base-logger/log");
var log_source_1 = require("../../../../core/rich-utils/debug/logger/base-logger/log-source");
var document_renderer_1 = require("../common/document-renderer");
var selection_renderer_base_1 = require("./selection-renderer-base");
var RangePermissionRenderer = (function (_super) {
    tslib_1.__extends(RangePermissionRenderer, _super);
    function RangePermissionRenderer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.layoutPageFlag = layout_page_1.LayoutPageFlags.IsRangePermissionsRendered;
        return _this;
    }
    RangePermissionRenderer.prototype.renderAllPageSelection = function (layout, pageIndex, force) {
        _super.prototype.renderAllPageSelection.call(this, layout, pageIndex, force);
        log_1.Log.print(log_source_1.LogSource.SelectionRenderer, "renderAllPageRangePermissions", "pageIndex: " + pageIndex);
    };
    RangePermissionRenderer.prototype.renderItem = function (item, layoutPage) {
        var selectionElement = _super.prototype.renderItem.call(this, item, layoutPage);
        selectionElement.style.backgroundColor = item.color;
        return selectionElement;
    };
    RangePermissionRenderer.prototype.closeDocument = function () {
        this.getLayoutSelectionInfo().reset();
    };
    RangePermissionRenderer.prototype.getLayoutSelectionInfo = function () {
        return this.layoutSelection.rangePermissionInfo;
    };
    RangePermissionRenderer.prototype.getContainer = function (pageElement) {
        return document_renderer_1.DocumentRenderer.getRangePermissionsContainerCore(pageElement);
    };
    return RangePermissionRenderer;
}(selection_renderer_base_1.SelectionRendererBase));
exports.RangePermissionRenderer = RangePermissionRenderer;
