// devexpress-richedit/lib/base/canvas/renderes/selection-renderers/selection-renderer-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var z_index_helper_1 = require("../../../../core/canvas/renderes/z-index-helper");
var layout_page_1 = require("../../../../core/layout/main-structures/layout-page");
var layout_selection_items_1 = require("../../../../core/layout/selection/layout-selection-items");
var dom_1 = require("@devexpress/utils/lib/utils/dom");
var document_renderer_1 = require("../common/document-renderer");
var SelectionRendererBase = (function () {
    function SelectionRendererBase(cache, layoutSelection) {
        this.cache = cache;
        this.layoutSelection = layoutSelection;
    }
    SelectionRendererBase.prototype.closeDocument = function () { };
    SelectionRendererBase.prototype.applySelectionChanges = function (layout) {
        var _this = this;
        var _loop_1 = function (pageChange) {
            var layoutPage = layout.pages[pageChange.index];
            if (!layoutPage || !layoutPage.flags.get(layout_page_1.LayoutPageFlags.ContentRendered))
                return "continue";
            layoutPage.flags.set(this_1.layoutPageFlag, true);
            for (var _i = 0, _a = pageChange.areaChanges; _i < _a.length; _i++) {
                var areaChange = _a[_i];
                document_renderer_1.DocumentRenderer.applyChange(this_1.cache[pageChange.index].page, areaChange, this_1.getContainer, function () { return _this.renderItem(areaChange.selection, layoutPage); });
            }
        };
        var this_1 = this;
        for (var _i = 0, _a = this.getLayoutSelectionInfo().changes; _i < _a.length; _i++) {
            var pageChange = _a[_i];
            _loop_1(pageChange);
        }
        this.getLayoutSelectionInfo().changesApplied();
    };
    SelectionRendererBase.prototype.renderAllPageSelection = function (layout, pageIndex, force) {
        var pageCache = this.cache[pageIndex];
        if (!pageCache)
            return;
        var container = this.getContainer(pageCache.page);
        var pageInfo = this.getLayoutSelectionInfo().pageInfos[pageIndex];
        var layoutPage = layout.pages[pageIndex];
        if (!container || (!force && container.childNodes.length != 0) || !pageInfo || !layoutPage)
            return;
        layoutPage.flags.set(this.layoutPageFlag, true);
        if (force)
            dom_1.DomUtils.clearInnerHtml(container);
        for (var _i = 0, _a = pageInfo.oldItems; _i < _a.length; _i++) {
            var item = _a[_i];
            container.appendChild(this.renderItem(item, layoutPage));
        }
    };
    SelectionRendererBase.prototype.renderItem = function (item, layoutPage) {
        var ancObjHolder = layoutPage.anchoredObjectHolder;
        var level;
        switch (item.floatingObjectId) {
            case layout_selection_items_1.LayoutSelectionItem.mainPageAreaSelection:
                level = layoutPage.renderLevelCalculator.renderMainPageAreaLevel;
                break;
            case layout_selection_items_1.LayoutSelectionItem.headerFooterPageAreaSelection:
                level = layoutPage.renderLevelCalculator.renderHeaderFooterPageAreasLevel;
                break;
            default: level = layoutPage.renderLevelCalculator.getRenderLevel(ancObjHolder.getObjById(item.floatingObjectId).rendererLevel);
        }
        var selectionElement = document_renderer_1.DocumentRenderer.renderContainer(item.baseClassName + " " + z_index_helper_1.ZIndexHelper.getClassName(level, item.zIndexClassType));
        dom_1.DomUtils.setStyleSizeAndPosition(selectionElement.style, item);
        return selectionElement;
    };
    return SelectionRendererBase;
}());
exports.SelectionRendererBase = SelectionRendererBase;
