// devexpress-richedit/lib/base/canvas/renderes/selection-renderers/touch-selection-circle-elements-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_1 = require("@devexpress/utils/lib/browser");
var point_1 = require("@devexpress/utils/lib/geometry/point");
var dom_1 = require("@devexpress/utils/lib/utils/dom");
var list_1 = require("@devexpress/utils/lib/utils/list");
var document_renderer_1 = require("../common/document-renderer");
var TouchSelectionCircleElementsManager = (function () {
    function TouchSelectionCircleElementsManager(cache, layoutSelection) {
        this.firstElement = document_renderer_1.DocumentRenderer.renderContainer(TouchSelectionCircleElementsManager.BAR_CLASS_NAME);
        this.secondElement = document_renderer_1.DocumentRenderer.renderContainer(TouchSelectionCircleElementsManager.BAR_CLASS_NAME);
        this.layoutSelection = layoutSelection;
        this.cache = cache;
    }
    TouchSelectionCircleElementsManager.prototype.getSecondElement = function () {
        return this.secondElement;
    };
    TouchSelectionCircleElementsManager.prototype.update = function () {
        if (!browser_1.Browser.TouchUI || this.layoutSelection.firstPageIndexWithSelection < 0 || !this.cache.length)
            return;
        if (!this.isCreated()) {
            var serviceContainer = document_renderer_1.DocumentRenderer.getServiceContainerCore(this.cache[this.layoutSelection.firstPageIndexWithSelection].page);
            serviceContainer.appendChild(this.firstElement);
            serviceContainer.appendChild(this.secondElement);
            this.radius = this.firstElement.offsetWidth / 2;
        }
        var firstSelectionItem = this.layoutSelection.selectionInfo.pageInfos[this.layoutSelection.firstPageIndexWithSelection].oldItems[0];
        dom_1.DomUtils.setStylePosition(this.firstElement.style, new point_1.Point(firstSelectionItem.x - this.radius, firstSelectionItem.bottom));
        var lastSelectionItem = list_1.ListUtils.last(this.layoutSelection.selectionInfo.pageInfos[this.layoutSelection.lastPageIndexWithSelection].oldItems);
        dom_1.DomUtils.setStylePosition(this.secondElement.style, new point_1.Point(lastSelectionItem.right - this.radius, lastSelectionItem.bottom));
    };
    TouchSelectionCircleElementsManager.prototype.setVisibilityTouchBars = function (visible) {
        this.setFirstTouchBarVisibility(visible);
        this.setSecondTouchBarVisibility(visible);
    };
    TouchSelectionCircleElementsManager.prototype.setFirstTouchBarVisibility = function (visible) {
        this.setVisible(this.firstElement, visible);
    };
    TouchSelectionCircleElementsManager.prototype.setSecondTouchBarVisibility = function (visible) {
        this.setVisible(this.secondElement, visible);
    };
    TouchSelectionCircleElementsManager.prototype.setVisible = function (element, visible) {
        if (this.isElementVisible(element) != visible)
            element.style.display = visible ? "" : "none";
    };
    TouchSelectionCircleElementsManager.prototype.isElementVisible = function (element) {
        return element.style.display != "none";
    };
    TouchSelectionCircleElementsManager.prototype.isCreated = function () {
        var serviceContainer = document_renderer_1.DocumentRenderer.getServiceContainerCore(this.cache[this.layoutSelection.firstPageIndexWithSelection].page);
        return !!dom_1.DomUtils.getChildNodesByClassName(serviceContainer, TouchSelectionCircleElementsManager.BAR_CLASS_NAME).length;
    };
    TouchSelectionCircleElementsManager.BAR_CLASS_NAME = "dxreSelBar";
    return TouchSelectionCircleElementsManager;
}());
exports.TouchSelectionCircleElementsManager = TouchSelectionCircleElementsManager;
