// devexpress-richedit/lib/core/model/find-replace-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var layout_box_iterator_main_sub_document_1 = require("../layout-engine/layout-box-iterator/layout-box-iterator-main-sub-document");
var layout_box_iterator_other_sub_document_1 = require("../layout-engine/layout-box-iterator/layout-box-iterator-other-sub-document");
var layout_position_creator_1 = require("../layout-engine/layout-position-creator");
var document_layout_details_level_1 = require("../layout/document-layout-details-level");
var layout_box_1 = require("../layout/main-structures/layout-boxes/layout-box");
var properties_bundle_1 = require("../rich-utils/properties-bundle");
var insert_text_history_item_1 = require("./history/items/insert-text-history-item");
var insert_text_manipulator_params_1 = require("./manipulators/text-manipulator/insert-text-manipulator-params");
var rich_utils_1 = require("./rich-utils");
var run_type_1 = require("./runs/run-type");
var sub_document_1 = require("./sub-document");
var FindReplaceState;
(function (FindReplaceState) {
    FindReplaceState[FindReplaceState["Start"] = 0] = "Start";
    FindReplaceState[FindReplaceState["Found"] = 1] = "Found";
    FindReplaceState[FindReplaceState["DocumentBegin"] = 2] = "DocumentBegin";
    FindReplaceState[FindReplaceState["DocumentEnd"] = 3] = "DocumentEnd";
    FindReplaceState[FindReplaceState["SearchEnd"] = 4] = "SearchEnd";
})(FindReplaceState = exports.FindReplaceState || (exports.FindReplaceState = {}));
var SearchDirection;
(function (SearchDirection) {
    SearchDirection[SearchDirection["Up"] = 1] = "Up";
    SearchDirection[SearchDirection["Down"] = 2] = "Down";
    SearchDirection[SearchDirection["All"] = 3] = "All";
})(SearchDirection = exports.SearchDirection || (exports.SearchDirection = {}));
var FindReplaceHelper = (function () {
    function FindReplaceHelper(modelManager, layoutFormatterManager, pageIndex, subDocument, layout, storeSelection) {
        this.modelManager = modelManager;
        this.layoutFormatterManager = layoutFormatterManager;
        this.pageIndex = pageIndex;
        this.subDocument = subDocument;
        this.layout = layout;
        this.storeSelection = storeSelection;
    }
    FindReplaceHelper.prototype.setSearchParams = function (whatFind, replaceWith, searchDirection, matchCase, regularExpression, findWithPosition, wholeWordsOnly) {
        if (findWithPosition < 0 || findWithPosition > this.subDocument.getDocumentEndPosition())
            throw "In FindReplaceHelper setSearchParams findWithPosition < 0 || findWithPosition > this.subDocument.getDocumentEndPosition()";
        if (wholeWordsOnly && regularExpression)
            throw "In FindReplaceHelper setSearchParams can't set wholeWordsOnly && regularExpression";
        if (wholeWordsOnly && FindReplaceHelper.isCanSetWholeWordsOnlyForThisExpression(whatFind) >= 0)
            throw "In FindReplaceHelper setSearchParams whatFind expression consider not Alphanumeric char[" + FindReplaceHelper.isCanSetWholeWordsOnlyForThisExpression(whatFind) + "]";
        var recalculateSuppFunc = (this.template != (matchCase ? whatFind : whatFind.toUpperCase())) ||
            (this.regularExpression != regularExpression) || (this.matchCase != matchCase) ||
            (this.searchDirection == SearchDirection.Up && searchDirection != SearchDirection.Up) ||
            (this.searchDirection != SearchDirection.Up && searchDirection == SearchDirection.Up);
        this.replaceWith = replaceWith;
        this.searchDirection = searchDirection;
        this.regularExpression = regularExpression;
        this.matchCase = matchCase;
        this.wholeWordsOnly = wholeWordsOnly;
        this.template = this.matchCase ? whatFind : whatFind.toUpperCase();
        this.lastFound = null;
        this.state = FindReplaceState.Start;
        this.beginOrStartDocumentReach = false;
        this.findWithPosition = findWithPosition;
        this.currentPos = findWithPosition;
        this.templateLength = whatFind.length;
        if (!this.wholeWordsOnly && recalculateSuppFunc || !this.supportFunction)
            this.crateSupportFunction();
    };
    FindReplaceHelper.isCanSetWholeWordsOnlyForThisExpression = function (expression) {
        for (var i = 0; i < expression.length; i++)
            if (!expression[i].match(rich_utils_1.RichUtils.isAlphanumeric))
                return i;
        return -1;
    };
    FindReplaceHelper.prototype.findNext = function () {
        if (this.state == FindReplaceState.SearchEnd)
            this.setSearchParams(this.template, this.replaceWith, this.searchDirection, this.matchCase, this.regularExpression, this.findWithPosition, this.wholeWordsOnly);
        var newState;
        switch (this.searchDirection) {
            case SearchDirection.Down:
                newState = this.findNextDown();
                break;
            case SearchDirection.All:
                newState = this.findNextAll();
                break;
            case SearchDirection.Up:
                newState = this.findNextUp();
                break;
        }
        if (newState == FindReplaceState.DocumentEnd || newState == FindReplaceState.DocumentBegin)
            this.beginOrStartDocumentReach = true;
        return newState;
    };
    FindReplaceHelper.prototype.findNextDown = function () {
        switch (this.state) {
            case FindReplaceState.Start:
            case FindReplaceState.Found:
                var isBelowStartPos = this.currentPos >= this.findWithPosition;
                if (this.findNextDownInternal(this.currentPos, isBelowStartPos ? this.subDocument.getDocumentEndPosition() : this.findWithPosition))
                    this.state = FindReplaceState.Found;
                else {
                    if (!isBelowStartPos || this.findWithPosition == 0)
                        this.state = FindReplaceState.SearchEnd;
                    else
                        this.state = FindReplaceState.DocumentEnd;
                }
                break;
            case FindReplaceState.DocumentEnd:
                if (this.findWithPosition == 0)
                    this.state = FindReplaceState.SearchEnd;
                else {
                    if (this.findNextDownInternal(0, this.findWithPosition))
                        this.state = FindReplaceState.Found;
                    else
                        this.state = FindReplaceState.SearchEnd;
                }
                break;
        }
        return this.state;
    };
    FindReplaceHelper.prototype.findNextUp = function () {
        switch (this.state) {
            case FindReplaceState.Start:
            case FindReplaceState.Found:
                var isAboveStartPos = this.currentPos <= this.findWithPosition;
                if (this.beginOrStartDocumentReach && isAboveStartPos) {
                    this.lastFound = null;
                    this.state = FindReplaceState.SearchEnd;
                    return this.state;
                }
                if (this.findNextUpInternal(isAboveStartPos ? 0 : this.findWithPosition, this.currentPos))
                    this.state = FindReplaceState.Found;
                else {
                    if (!isAboveStartPos || this.findWithPosition == this.subDocument.getDocumentEndPosition() - 1)
                        this.state = FindReplaceState.SearchEnd;
                    else
                        this.state = FindReplaceState.DocumentBegin;
                }
                break;
            case FindReplaceState.DocumentBegin:
                var docEnd = this.subDocument.getDocumentEndPosition();
                if (this.findWithPosition == docEnd - 1)
                    this.state = FindReplaceState.SearchEnd;
                else {
                    if (this.findNextUpInternal(this.findWithPosition, docEnd - 1))
                        this.state = FindReplaceState.Found;
                    else
                        this.state = FindReplaceState.SearchEnd;
                }
                break;
        }
        return this.state;
    };
    FindReplaceHelper.prototype.findNextAll = function () {
        while (this.findNextDown() == FindReplaceState.DocumentEnd)
            ;
        return this.state;
    };
    FindReplaceHelper.prototype.findNextDownInternal = function (lowerPosition, greaterPosition) {
        var charIterator = new ForwardCharacterIterator(this.modelManager, this.layoutFormatterManager, this.pageIndex, this.subDocument, lowerPosition, greaterPosition);
        if (this.wholeWordsOnly)
            return this.findNextDownWholeWordsOnly(charIterator, lowerPosition, greaterPosition);
        var offset = 0;
        var posWhenStartEquivalents = -1;
        while (charIterator.nextChar()) {
            if (!this.matchCase)
                charIterator.char = charIterator.char.toUpperCase();
            while (offset > 0 && this.template[offset] != charIterator.char) {
                offset = this.supportFunction[offset - 1];
                posWhenStartEquivalents++;
            }
            if (this.template[offset] == charIterator.char) {
                if (offset == 0)
                    posWhenStartEquivalents = charIterator.getCurrLogPosition();
                offset++;
            }
            if (offset == this.templateLength) {
                this.lastFound = fixed_1.FixedInterval.fromPositions(posWhenStartEquivalents, charIterator.getCurrLogPosition() + 1);
                this.currentPos = this.lastFound.end;
                return true;
            }
        }
        this.currentPos = greaterPosition;
        this.lastFound = null;
        return false;
    };
    FindReplaceHelper.prototype.findNextDownWholeWordsOnly = function (charIterator, lowerPosition, greaterPosition) {
        var offset = 0;
        if (this.layout.pages[0].getPosition() < lowerPosition &&
            this.isCharCanBeInWord((this.subDocument.isMain() ? new layout_position_creator_1.LayoutPositionMainSubDocumentCreator(this.layout, this.subDocument, lowerPosition - 1, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                : new layout_position_creator_1.LayoutPositionOtherSubDocumentCreator(this.layout, this.subDocument, lowerPosition - 1, this.pageIndex, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character))
                .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true))))
            offset = -1;
        var posWhenStartEquivalents = -1;
        while (charIterator.nextChar()) {
            if (offset == this.templateLength) {
                if (charIterator.char.match(rich_utils_1.RichUtils.isAlphanumeric)) {
                    offset = -1;
                    continue;
                }
                else {
                    this.lastFound = fixed_1.FixedInterval.fromPositions(posWhenStartEquivalents, charIterator.getCurrLogPosition() + 1);
                    this.currentPos = this.lastFound.end;
                    return true;
                }
            }
            if (offset < 0 && !charIterator.char.match(rich_utils_1.RichUtils.isAlphanumeric) ||
                offset >= 0 && this.template[offset] == (this.matchCase ? charIterator.char : charIterator.char.toUpperCase())) {
                offset += 1;
                if (offset == 0)
                    posWhenStartEquivalents = charIterator.getCurrLogPosition();
            }
            else
                offset = -1;
        }
        this.currentPos = greaterPosition;
        this.lastFound = null;
        return false;
    };
    FindReplaceHelper.prototype.findNextUpInternal = function (lowerPosition, greaterPosition) {
        if (greaterPosition < 1) {
            this.currentPos = 0;
            this.lastFound = null;
            return false;
        }
        var charIterator = new BackwardCharacterIterator(this.modelManager, this.layoutFormatterManager, this.pageIndex, this.subDocument, lowerPosition, greaterPosition);
        if (this.wholeWordsOnly)
            return this.findNextUpWholeWordsOnly(charIterator, greaterPosition);
        var offset = 0;
        var templateLastInd = this.templateLength - 1;
        var posWhenStartEquivalents = -1;
        while (charIterator.prevChar()) {
            if (!this.matchCase)
                charIterator.char = charIterator.char.toUpperCase();
            while (offset > 0 && this.template[templateLastInd - offset] != charIterator.char)
                offset = this.supportFunction[offset - 1];
            if (this.template[templateLastInd - offset] == charIterator.char) {
                if (offset == 0)
                    posWhenStartEquivalents = charIterator.getCurrLogPosition() + 1;
                offset++;
            }
            if (offset == this.templateLength) {
                this.lastFound = fixed_1.FixedInterval.fromPositions(charIterator.getCurrLogPosition(), posWhenStartEquivalents);
                this.currentPos = Math.max(0, this.lastFound.start - 1);
                return true;
            }
        }
        this.currentPos = 0;
        this.lastFound = null;
        return false;
    };
    FindReplaceHelper.prototype.findNextUpWholeWordsOnly = function (charIterator, greaterPosition) {
        var offset = 0;
        var layoutPage = this.subDocument.isMain() ? this.layout.getLastValidPage() : this.layout.pages[this.pageIndex];
        var posWhenStartEquivalents = -1;
        if (greaterPosition < layoutPage.getEndPosition() - 1 &&
            this.isCharCanBeInWord((this.subDocument.isMain() ? new layout_position_creator_1.LayoutPositionMainSubDocumentCreator(this.layout, this.subDocument, greaterPosition + 1, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                : new layout_position_creator_1.LayoutPositionOtherSubDocumentCreator(this.layout, this.subDocument, greaterPosition + 1, this.pageIndex, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character))
                .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false))))
            offset = -1;
        else
            posWhenStartEquivalents = charIterator.getCurrLogPosition();
        var templateLastInd = this.templateLength - 1;
        while (charIterator.prevChar()) {
            if (offset == this.templateLength) {
                if (charIterator.char.match(rich_utils_1.RichUtils.isAlphanumeric)) {
                    offset = -1;
                    continue;
                }
                else {
                    this.lastFound = fixed_1.FixedInterval.fromPositions(charIterator.getCurrLogPosition(), posWhenStartEquivalents + 1);
                    this.currentPos = this.lastFound.start;
                    return true;
                }
            }
            if (offset < 0) {
                offset = charIterator.char.match(rich_utils_1.RichUtils.isAlphanumeric) ? -1 : 0;
                if (offset == 0)
                    posWhenStartEquivalents = charIterator.getCurrLogPosition();
            }
            else {
                if (this.template[templateLastInd - offset] == (this.matchCase ? charIterator.char : charIterator.char.toUpperCase()))
                    offset += 1;
            }
        }
        if (offset == this.templateLength) {
            this.currentPos = 0;
            this.lastFound = new fixed_1.FixedInterval(0, posWhenStartEquivalents + 1);
            return true;
        }
        this.currentPos = 0;
        this.lastFound = null;
        return false;
    };
    FindReplaceHelper.prototype.isCharCanBeInWord = function (layoutPos) {
        if (!layoutPos)
            return false;
        layoutPos.switchToStartNextBoxInRow();
        if (layoutPos.box.getType() != layout_box_1.LayoutBoxType.Text && layoutPos.box.getType() != layout_box_1.LayoutBoxType.Dash)
            return false;
        var char;
        if (layoutPos.charOffset >= layoutPos.box.getLength()) {
            if (layoutPos.advanceToNextRow(this.layout))
                char = layoutPos.row.boxes[0].renderGetContent(null)[0];
            else
                return false;
        }
        else
            char = layoutPos.box.renderGetContent(null)[layoutPos.charOffset];
        return char.match(rich_utils_1.RichUtils.isAlphanumeric).length > 0;
    };
    FindReplaceHelper.prototype.getLastFound = function () {
        return this.lastFound ? this.lastFound.clone() : null;
    };
    FindReplaceHelper.prototype.replaceLastFound = function () {
        if (!this.lastFound || this.replaceWith === null || this.replaceWith === undefined)
            return;
        var modelManager = this.modelManager;
        var firstRun = this.subDocument.getRunByPosition(this.lastFound.start);
        var charStyle = firstRun.characterStyle;
        var maskedCharProp = firstRun.maskedCharacterProperties;
        modelManager.history.beginTransaction();
        this.storeSelection(new fixed_1.FixedInterval(this.lastFound.start, this.replaceWith.length));
        modelManager.modelManipulator.range.removeInterval(new sub_document_1.SubDocumentInterval(this.subDocument, this.lastFound.clone()), true, false);
        if (this.replaceWith)
            modelManager.history.addAndRedo(new insert_text_history_item_1.InsertTextHistoryItem(modelManager.modelManipulator, new insert_text_manipulator_params_1.InsertTextManipulatorParams(new sub_document_1.SubDocumentPosition(this.subDocument, this.lastFound.start), new properties_bundle_1.MaskedCharacterPropertiesBundle(maskedCharProp, charStyle), run_type_1.RunType.TextRun, this.replaceWith)));
        modelManager.history.endTransaction();
        var diff = this.replaceWith.length - this.templateLength;
        if (this.searchDirection != SearchDirection.Up)
            this.currentPos += diff;
        if (this.findWithPosition >= this.lastFound.end)
            this.findWithPosition += diff;
        this.lastFound = null;
    };
    FindReplaceHelper.prototype.replaceAll = function (startWithPos) {
        if (this.replaceWith === null || this.replaceWith === undefined)
            return;
        var modelManager = this.modelManager;
        var oldFindWithPosition = this.findWithPosition;
        this.setSearchParams(this.template, this.replaceWith, this.searchDirection, this.matchCase, this.regularExpression, 0, this.wholeWordsOnly);
        var history = modelManager.history;
        var storedPosition = this.subDocument.positionManager.registerPosition(startWithPos);
        this.currentPos = 0;
        var resultIntervals = [];
        while (this.findNextAll() == FindReplaceState.Found)
            resultIntervals.push(this.lastFound);
        if (resultIntervals.length) {
            history.beginTransaction();
            for (var i = resultIntervals.length - 1; i >= 0; i--) {
                var curr = resultIntervals[i];
                var firstRun = this.subDocument.getRunByPosition(curr.start);
                var charStyle = firstRun.characterStyle;
                var maskedCharProp = firstRun.maskedCharacterProperties;
                modelManager.modelManipulator.range.removeInterval(new sub_document_1.SubDocumentInterval(this.subDocument, curr.clone()), true, false);
                if (this.replaceWith.length)
                    modelManager.history.addAndRedo(new insert_text_history_item_1.InsertTextHistoryItem(modelManager.modelManipulator, new insert_text_manipulator_params_1.InsertTextManipulatorParams(new sub_document_1.SubDocumentPosition(this.subDocument, curr.start), new properties_bundle_1.MaskedCharacterPropertiesBundle(maskedCharProp, charStyle), run_type_1.RunType.TextRun, this.replaceWith)));
            }
            this.storeSelection(new fixed_1.FixedInterval(Math.min(storedPosition.value, this.subDocument.getDocumentEndPosition() - 1), 0));
            history.endTransaction();
        }
        this.subDocument.positionManager.unregisterPosition(storedPosition);
        this.setSearchParams(this.template, this.replaceWith, this.searchDirection, this.matchCase, this.regularExpression, oldFindWithPosition, this.wholeWordsOnly);
        return resultIntervals.length;
    };
    FindReplaceHelper.prototype.crateSupportFunction = function () {
        var template;
        if (this.searchDirection == SearchDirection.Up) {
            template = "";
            for (var i = this.templateLength - 1; i >= 0; i--)
                template += this.template[i];
        }
        else
            template = this.template;
        this.supportFunction = [0];
        for (var topBound = 2; topBound < this.templateLength; topBound++)
            this.supportFunction.push(FindReplaceHelper.calcOneElemSuppFunc(template, topBound));
    };
    FindReplaceHelper.calcOneElemSuppFunc = function (template, topBound) {
        for (var offset = 1; offset < topBound; offset++) {
            var lenComparePhrase = topBound - offset;
            var i = 0;
            for (; i < lenComparePhrase; i++) {
                if (template[i] != template[i + offset])
                    break;
            }
            if (i == lenComparePhrase)
                return topBound - offset;
        }
        return 0;
    };
    return FindReplaceHelper;
}());
exports.FindReplaceHelper = FindReplaceHelper;
var CharacterIteratorBase = (function () {
    function CharacterIteratorBase(modelManager, layoutFormatterManager, pageIndex, subDocument, startPosition, endPosition) {
        this.modelManager = modelManager;
        this.subDocument = subDocument;
        this.iterator = subDocument.isMain() ?
            new layout_box_iterator_main_sub_document_1.LayoutBoxIteratorMainSubDocument(subDocument, layoutFormatterManager.layout, startPosition, endPosition) :
            new layout_box_iterator_other_sub_document_1.LayoutBoxIteratorOtherSubDocument(subDocument, layoutFormatterManager.layout, startPosition, endPosition, pageIndex);
        while (!this.iterator.isInitialized())
            layoutFormatterManager.forceFormatPage(layoutFormatterManager.layout.validPageCount);
        this.char = null;
        this.charIndexInBox = -1;
    }
    CharacterIteratorBase.prototype.getCurrLogPosition = function () {
        return this.iterator.position.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box) + this.charIndexInBox;
    };
    CharacterIteratorBase.prototype.getCharInternal = function () {
        switch (this.iterator.position.box.getType()) {
            case layout_box_1.LayoutBoxType.Text:
                this.char = this.iterator.position.box.text[this.charIndexInBox];
                break;
            case layout_box_1.LayoutBoxType.Dash:
                this.char = this.iterator.position.box.text[this.charIndexInBox];
                break;
            case layout_box_1.LayoutBoxType.Space:
                this.char = rich_utils_1.RichUtils.specialCharacters.Space;
                break;
            case layout_box_1.LayoutBoxType.TabSpace:
                this.char = rich_utils_1.RichUtils.specialCharacters.TabMark;
                break;
            case layout_box_1.LayoutBoxType.ParagraphMark:
                this.char = rich_utils_1.RichUtils.specialCharacters.ParagraphMark;
                break;
            default:
                this.char = String.fromCharCode(0);
        }
    };
    return CharacterIteratorBase;
}());
exports.CharacterIteratorBase = CharacterIteratorBase;
var ForwardCharacterIterator = (function (_super) {
    tslib_1.__extends(ForwardCharacterIterator, _super);
    function ForwardCharacterIterator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ForwardCharacterIterator.prototype.getCharInternal = function () {
        if (this.charIndexInBox >= this.iterator.position.box.getLength())
            return false;
        _super.prototype.getCharInternal.call(this);
        return true;
    };
    ForwardCharacterIterator.prototype.nextChar = function () {
        if (this.charIndexInBox == -1) {
            this.iterator.moveNext(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true));
            this.charIndexInBox = this.iterator.position.charOffset;
            if (this.iterator.position.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box) + this.charIndexInBox < this.iterator.intervalEnd && this.getCharInternal())
                return true;
        }
        else
            this.charIndexInBox++;
        this.char = null;
        while (this.char == null && this.iterator.position.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box) + this.charIndexInBox < this.iterator.intervalEnd) {
            if (!this.getCharInternal()) {
                if (this.iterator.moveNext(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true)))
                    this.charIndexInBox = this.iterator.position.charOffset;
                else
                    return false;
            }
        }
        return !!this.char;
    };
    return ForwardCharacterIterator;
}(CharacterIteratorBase));
exports.ForwardCharacterIterator = ForwardCharacterIterator;
var BackwardCharacterIterator = (function (_super) {
    tslib_1.__extends(BackwardCharacterIterator, _super);
    function BackwardCharacterIterator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BackwardCharacterIterator.prototype.getCharInternal = function () {
        if (this.charIndexInBox < 0)
            return false;
        _super.prototype.getCharInternal.call(this);
        return true;
    };
    BackwardCharacterIterator.prototype.prevChar = function () {
        if (this.charIndexInBox == -1) {
            this.iterator.movePrev(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false));
            this.charIndexInBox = this.iterator.position.charOffset - 1;
            if (this.charIndexInBox == -1) {
                if (!this.iterator.movePrev(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false)))
                    return false;
                this.charIndexInBox = this.iterator.position.box.getLength() - 1;
            }
            if (this.getCharInternal())
                return true;
            this.charIndexInBox = -1;
        }
        else
            this.charIndexInBox--;
        this.char = null;
        while (this.char == null && this.iterator.position.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box) + this.charIndexInBox >= this.iterator.intervalStart) {
            if (!this.getCharInternal()) {
                if (this.iterator.position.charOffset > 0)
                    this.iterator.movePrev(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true));
                if (this.iterator.movePrev(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true)))
                    this.charIndexInBox = this.iterator.position.box.getLength() - 1;
                else
                    return false;
            }
        }
        return !!this.char;
    };
    return BackwardCharacterIterator;
}(CharacterIteratorBase));
exports.BackwardCharacterIterator = BackwardCharacterIterator;
