// devexpress-richedit/lib/core/model/sentence-model-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var search_1 = require("@devexpress/utils/lib/utils/search");
var string_1 = require("@devexpress/utils/lib/utils/string");
var layout_box_iterator_main_sub_document_1 = require("../layout-engine/layout-box-iterator/layout-box-iterator-main-sub-document");
var layout_box_iterator_other_sub_document_1 = require("../layout-engine/layout-box-iterator/layout-box-iterator-other-sub-document");
var layout_position_creator_1 = require("../layout-engine/layout-position-creator");
var layout_box_1 = require("../layout/main-structures/layout-boxes/layout-box");
var rich_utils_1 = require("./rich-utils");
var SwitchTextCaseManipulator = (function () {
    function SwitchTextCaseManipulator(dispatcher) {
        this.dispatcher = dispatcher;
    }
    return SwitchTextCaseManipulator;
}());
exports.SwitchTextCaseManipulator = SwitchTextCaseManipulator;
var WordPart = (function () {
    function WordPart(position, text, type) {
        this.position = position;
        this.text = text;
        this.type = type;
    }
    WordPart.prototype.merge = function (pos, text, type) {
        if (this.position + this.text.length == pos && this.type == type && (type == layout_box_1.LayoutBoxType.Text || type == layout_box_1.LayoutBoxType.Dash || type == layout_box_1.LayoutBoxType.Space)) {
            this.text += text;
            return true;
        }
        return false;
    };
    WordPart.prototype.getEndPosition = function () {
        return this.position + this.text.length;
    };
    return WordPart;
}());
exports.WordPart = WordPart;
var SentenceWord = (function () {
    function SentenceWord() {
        this.parts = [];
    }
    SentenceWord.prototype.getLastWordPart = function () {
        return this.parts[this.parts.length - 1];
    };
    return SentenceWord;
}());
exports.SentenceWord = SentenceWord;
var Sentence = (function () {
    function Sentence() {
        this.words = [];
    }
    Sentence.prototype.getLastWord = function () {
        return this.words[this.words.length - 1];
    };
    return Sentence;
}());
exports.Sentence = Sentence;
var SentenceStructureBuilder = (function () {
    function SentenceStructureBuilder(subDocument, layout, selection, interval) {
        this.currSentence = null;
        this.currWord = null;
        this.currWordPart = null;
        this.isSentenceEnd = false;
        this.isWordEnd = false;
        this.findEndLastSentence = false;
        this.subDocument = subDocument;
        this.layout = layout;
        this.interval = interval;
        this.sentences = [];
        this.selection = selection;
    }
    SentenceStructureBuilder.getBuilder = function (layoutFormatterManager, selection, subDocument, interval, splitByInterval) {
        var layout = layoutFormatterManager.layout;
        var builder = new SentenceStructureBuilder(subDocument, layout, selection, interval);
        while (!builder.build())
            layoutFormatterManager.forceFormatPage(layout.validPageCount);
        if (splitByInterval)
            SentenceStructureBuilder.splitPartsByInterval(builder.sentences, interval);
        SentenceStructureBuilder.correctFirstSentence(builder.sentences);
        return builder;
    };
    SentenceStructureBuilder.correctFirstSentence = function (sentences) {
        var firstSentenceWords = sentences[0].words;
        for (var startFirstSentenceWordIndex = 0, word; word = firstSentenceWords[startFirstSentenceWordIndex]; startFirstSentenceWordIndex++) {
            var type = word.parts[0].type;
            if (type != layout_box_1.LayoutBoxType.Space && type != layout_box_1.LayoutBoxType.LineBreak && type != layout_box_1.LayoutBoxType.TabSpace)
                break;
        }
        if (startFirstSentenceWordIndex > 0 && startFirstSentenceWordIndex < firstSentenceWords.length) {
            var newFirstSentence = new Sentence();
            for (; startFirstSentenceWordIndex > 0; startFirstSentenceWordIndex--)
                newFirstSentence.words.push(firstSentenceWords.shift());
            sentences.unshift(newFirstSentence);
        }
    };
    SentenceStructureBuilder.splitPartsByInterval = function (sentences, interval) {
        SentenceStructureBuilder.splitPartsByPosition(sentences, interval.start);
        SentenceStructureBuilder.splitPartsByPosition(sentences, interval.end);
    };
    SentenceStructureBuilder.splitPartsByPosition = function (sentences, position) {
        var sentenceIndex = Math.max(0, search_1.SearchUtils.normedInterpolationIndexOf(sentences, function (s) { return s.words[0].parts[0].position; }, position));
        var sentence = sentences[sentenceIndex];
        var wordIndex = Math.max(0, search_1.SearchUtils.normedInterpolationIndexOf(sentence.words, function (w) { return w.parts[0].position; }, position));
        var sentenceWord = sentence.words[wordIndex];
        for (var wordPartIndex = 0, part; part = sentenceWord.parts[wordPartIndex]; wordPartIndex++) {
            if (part.position > position)
                break;
        }
        wordPartIndex = wordPartIndex > 0 ? wordPartIndex - 1 : 0;
        part = sentenceWord.parts[wordPartIndex];
        if (position > part.position && position < part.position + part.text.length) {
            sentenceWord.parts.splice(wordPartIndex + 1, 0, new WordPart(position, part.text.substr(position - part.position), part.type));
            part.text = part.text.substr(0, position - part.position);
        }
    };
    SentenceStructureBuilder.prototype.build = function () {
        var endPosition = this.subDocument.isMain() ? this.layout.getLastValidPage().getEndPosition() : this.layout.pages[this.selection.pageIndex].getEndPosition();
        if (this.findEndLastSentence) {
            this.interval = fixed_1.FixedInterval.fromPositions(this.interval.end, endPosition);
            this.layoutBoxIterator.resetToInterval(this.interval.start, this.interval.end);
            this.collect();
        }
        else {
            if (this.needCalculateMoreLayout())
                return false;
            var firstSentenceStartPosition = this.findFirstSentenceStartPosition();
            this.layoutBoxIterator = this.subDocument.isMain() ? new layout_box_iterator_main_sub_document_1.LayoutBoxIteratorMainSubDocument(this.subDocument, this.layout, firstSentenceStartPosition, this.interval.end) : new layout_box_iterator_other_sub_document_1.LayoutBoxIteratorOtherSubDocument(this.subDocument, this.layout, firstSentenceStartPosition, this.interval.end, this.selection.pageIndex);
            this.addNewSentence();
            this.addNewWord();
            this.collect();
            this.findEndLastSentence = true;
        }
        if (this.currSentence.words.length == 1 && this.currWord.parts.length == 0) {
            this.sentences.pop();
            return true;
        }
        if (this.interval.end == endPosition)
            return this.layout.isFullyFormatted;
        else
            return this.build();
    };
    SentenceStructureBuilder.prototype.collect = function () {
        while (this.layoutBoxIterator.moveNext(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false))) {
            var layoutPos = this.layoutBoxIterator.position;
            if (this.lastBox && this.lastBox === layoutPos.box)
                continue;
            this.lastBox = layoutPos.box;
            switch (layoutPos.box.getType()) {
                case layout_box_1.LayoutBoxType.Text:
                    var textBox = layoutPos.box;
                    for (var charIndex = layoutPos.charOffset, currAbsolutePosition = layoutPos.getLogPosition() + charIndex, char; char = textBox.text[charIndex]; charIndex++, currAbsolutePosition++) {
                        switch (char) {
                            case ".":
                            case ";":
                                if (this.isSentenceEnd) {
                                    this.addNewSentence();
                                    this.addNewWord();
                                    if (this.findEndLastSentence)
                                        return;
                                }
                                this.addNewWordPart(currAbsolutePosition, char, layout_box_1.LayoutBoxType.Text);
                                this.isSentenceEnd = true;
                                break;
                            case "!":
                                var prevChar = string_1.StringUtils.getSymbolFromEnd(this.currWordPart.text, 1);
                                if (this.isSentenceEnd && prevChar && prevChar != "!") {
                                    this.addNewSentence();
                                    this.addNewWord();
                                    if (this.findEndLastSentence)
                                        return;
                                }
                                this.addNewWordPart(currAbsolutePosition, char, layout_box_1.LayoutBoxType.Text);
                                this.isSentenceEnd = true;
                                break;
                            case "?":
                                var prevChar = string_1.StringUtils.getSymbolFromEnd(this.currWordPart.text, 1);
                                if (this.isSentenceEnd && !(prevChar && prevChar == "!")) {
                                    this.addNewSentence();
                                    this.addNewWord();
                                    if (this.findEndLastSentence)
                                        return;
                                }
                                this.addNewWordPart(currAbsolutePosition, char, layout_box_1.LayoutBoxType.Text);
                                this.isSentenceEnd = true;
                                break;
                            case rich_utils_1.RichUtils.specialCharacters.Dot:
                            case rich_utils_1.RichUtils.specialCharacters.Hyphen:
                            case rich_utils_1.RichUtils.specialCharacters.TrademarkSymbol:
                            case rich_utils_1.RichUtils.specialCharacters.CopyrightSymbol:
                            case rich_utils_1.RichUtils.specialCharacters.RegisteredTrademarkSymbol:
                            case rich_utils_1.RichUtils.specialCharacters.Ellipsis:
                            case rich_utils_1.RichUtils.specialCharacters.LeftDoubleQuote:
                            case rich_utils_1.RichUtils.specialCharacters.LeftSingleQuote:
                            case rich_utils_1.RichUtils.specialCharacters.RightDoubleQuote:
                            case rich_utils_1.RichUtils.specialCharacters.RightSingleQuote:
                            case rich_utils_1.RichUtils.specialCharacters.OpeningDoubleQuotationMark:
                            case rich_utils_1.RichUtils.specialCharacters.OpeningSingleQuotationMark:
                            case rich_utils_1.RichUtils.specialCharacters.ClosingDoubleQuotationMark:
                            case rich_utils_1.RichUtils.specialCharacters.ClosingSingleQuotationMark:
                            case rich_utils_1.RichUtils.specialCharacters.RegisteredTrademarkSymbol:
                            case rich_utils_1.RichUtils.specialCharacters.RegisteredTrademarkSymbol:
                            case ",":
                            case "@":
                            case "#":
                            case "$":
                            case "%":
                            case "^":
                            case "&":
                            case "*":
                            case "(":
                            case ")":
                            case "=":
                            case "+":
                            case "[":
                            case "]":
                            case "{":
                            case "}":
                            case "\\":
                            case "|":
                            case ":":
                            case "\'":
                            case "\"":
                            case "<":
                            case ">":
                            case "/":
                            case "~":
                            case "`":
                                if (this.isWordEnd)
                                    this.addNewWord();
                                this.addNewWordPart(currAbsolutePosition, char, layout_box_1.LayoutBoxType.Text);
                                this.isWordEnd = true;
                                break;
                            default:
                                if (this.isSentenceEnd) {
                                    this.addNewSentence();
                                    this.addNewWord();
                                    if (this.findEndLastSentence)
                                        return;
                                    this.isSentenceEnd = false;
                                    this.isWordEnd = false;
                                }
                                else if (this.isWordEnd) {
                                    this.addNewWord();
                                    this.isWordEnd = false;
                                }
                                this.addNewWordPart(currAbsolutePosition, char, layout_box_1.LayoutBoxType.Text);
                                break;
                        }
                    }
                    break;
                case layout_box_1.LayoutBoxType.SectionMark:
                case layout_box_1.LayoutBoxType.ParagraphMark:
                    this.addNewWordPart(layoutPos.getLogPosition(), " ", layoutPos.box.getType());
                    this.addNewSentence();
                    this.addNewWord();
                    if (this.findEndLastSentence)
                        return;
                    this.isSentenceEnd = false;
                    this.isWordEnd = false;
                    break;
                case layout_box_1.LayoutBoxType.Dash:
                case layout_box_1.LayoutBoxType.Space:
                case layout_box_1.LayoutBoxType.TabSpace:
                    this.addNewWordPart(layoutPos.getLogPosition(), " ", layoutPos.box.getType());
                    this.isWordEnd = true;
                    break;
                default:
                    if (this.currWord.parts.length > 0)
                        this.addNewWord();
                    this.addNewWordPart(layoutPos.getLogPosition(), " ", layoutPos.box.getType());
                    if (!this.isSentenceEnd)
                        this.addNewWord();
                    break;
            }
        }
    };
    SentenceStructureBuilder.prototype.addNewSentence = function () {
        this.currSentence = new Sentence();
        this.sentences.push(this.currSentence);
        this.currWord = null;
        this.currWordPart = null;
    };
    SentenceStructureBuilder.prototype.addNewWord = function () {
        this.currWord = new SentenceWord();
        this.currSentence.words.push(this.currWord);
        this.currWordPart = null;
    };
    SentenceStructureBuilder.prototype.addNewWordPart = function (pos, text, type) {
        var mergeSuccess = this.currWordPart && this.currWordPart.merge(pos, text, type);
        if (!mergeSuccess) {
            this.currWordPart = new WordPart(pos, text, type);
            this.currWord.parts.push(this.currWordPart);
        }
    };
    SentenceStructureBuilder.prototype.needCalculateMoreLayout = function () {
        var lastValidPage = this.subDocument.isMain() ? this.layout.getLastValidPage() : this.layout.pages[this.selection.pageIndex];
        return lastValidPage && (this.subDocument.isMain() ? lastValidPage.getEndPosition() : lastValidPage.otherPageAreas[this.subDocument.id].getEndPosition()) < this.interval.end;
    };
    SentenceStructureBuilder.prototype.findFirstSentenceStartPosition = function () {
        var layoutBoxIterator = this.subDocument.isMain() ? new layout_box_iterator_main_sub_document_1.LayoutBoxIteratorMainSubDocument(this.subDocument, this.layout, 0, this.interval.start) : new layout_box_iterator_other_sub_document_1.LayoutBoxIteratorOtherSubDocument(this.subDocument, this.layout, 0, this.interval.start, this.selection.pageIndex);
        var suspiciousPosition = -1;
        var isFindSentenceStart = false;
        while (layoutBoxIterator.movePrev(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true))) {
            var layoutPos = layoutBoxIterator.position;
            switch (layoutPos.box.getType()) {
                case layout_box_1.LayoutBoxType.Text:
                    var textBox = layoutPos.box;
                    var lastIndexBox = layoutPos.charOffset > 0 || layoutPos.getLogPosition() >= this.interval.start ? layoutPos.charOffset : textBox.text.length - 1;
                    for (var charIndex = lastIndexBox, char; char = textBox.text[charIndex]; charIndex--) {
                        switch (char) {
                            case ".":
                            case "!":
                            case "?":
                            case ";":
                                if (layoutPos.getLogPosition() != this.interval.start)
                                    isFindSentenceStart = true;
                                break;
                            default:
                                suspiciousPosition = layoutPos.getLogPosition();
                                break;
                        }
                        if (isFindSentenceStart)
                            break;
                    }
                    break;
                case layout_box_1.LayoutBoxType.SectionMark:
                case layout_box_1.LayoutBoxType.ParagraphMark:
                    layoutBoxIterator.moveNext(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false));
                    suspiciousPosition = layoutBoxIterator.position.getLogPosition();
                    isFindSentenceStart = true;
                    break;
            }
            if (isFindSentenceStart)
                break;
        }
        if (suspiciousPosition < 0)
            return layoutBoxIterator.position.getLogPosition();
        else
            return suspiciousPosition;
    };
    return SentenceStructureBuilder;
}());
exports.SentenceStructureBuilder = SentenceStructureBuilder;
