// devexpress-richedit/lib/core/model/simple-sentence-model-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var string_1 = require("@devexpress/utils/lib/utils/string");
var rich_utils_1 = require("./rich-utils");
var SimpleSentenceWord = (function () {
    function SimpleSentenceWord(position, text) {
        this.position = position;
        this.text = text;
    }
    return SimpleSentenceWord;
}());
exports.SimpleSentenceWord = SimpleSentenceWord;
var SimpleSentence = (function () {
    function SimpleSentence() {
        this.words = [];
    }
    return SimpleSentence;
}());
exports.SimpleSentence = SimpleSentence;
var SimpleSentenceStructureBuilder = (function () {
    function SimpleSentenceStructureBuilder(text) {
        this.sentences = [];
        this.charIndex = 0;
        this.currSentence = null;
        this.currWord = null;
        this.text = text;
    }
    SimpleSentenceStructureBuilder.prototype.build = function () {
        for (; this.charIndex < this.text.length; this.charIndex++) {
            this.currChar = this.text[this.charIndex];
            switch (this.currChar) {
                case "?":
                case ".":
                case ";":
                case "!":
                case rich_utils_1.RichUtils.specialCharacters.SectionMark:
                case rich_utils_1.RichUtils.specialCharacters.ParagraphMark: {
                    if (this.currSentence) {
                        this.finishSentence();
                    }
                    else {
                        if (this.prevChar == '!' && this.currChar == '?' ||
                            this.prevChar == '"' ||
                            this.prevChar == '`' ||
                            this.prevChar == '?' && this.currChar == '?' ||
                            ((this.prevChar == '!' || this.prevChar == '?' || this.prevChar == '.' || this.prevChar == ';') &&
                                (this.currChar == rich_utils_1.RichUtils.specialCharacters.ParagraphMark || this.currChar == rich_utils_1.RichUtils.specialCharacters.SectionMark))) {
                        }
                        else {
                            this.addSentence();
                            this.finishSentence();
                        }
                    }
                    break;
                }
                case rich_utils_1.RichUtils.specialCharacters.LeftDoubleQuote:
                case rich_utils_1.RichUtils.specialCharacters.RightDoubleQuote:
                case rich_utils_1.RichUtils.specialCharacters.EmSpace:
                case rich_utils_1.RichUtils.specialCharacters.Space:
                case rich_utils_1.RichUtils.specialCharacters.TabMark:
                case rich_utils_1.RichUtils.specialCharacters.EmDash:
                case rich_utils_1.RichUtils.specialCharacters.EnDash:
                case rich_utils_1.RichUtils.specialCharacters.LineBreak:
                case rich_utils_1.RichUtils.specialCharacters.PageBreak:
                case rich_utils_1.RichUtils.specialCharacters.ColumnBreak:
                case rich_utils_1.RichUtils.specialCharacters.NonBreakingSpace:
                case rich_utils_1.RichUtils.specialCharacters.ObjectMark:
                case rich_utils_1.RichUtils.specialCharacters.Dot:
                case rich_utils_1.RichUtils.specialCharacters.TrademarkSymbol:
                case rich_utils_1.RichUtils.specialCharacters.CopyrightSymbol:
                case rich_utils_1.RichUtils.specialCharacters.RegisteredTrademarkSymbol:
                case rich_utils_1.RichUtils.specialCharacters.Ellipsis:
                case String.fromCharCode(160):
                case ",":
                case "#":
                case "$":
                case "%":
                case "^":
                case "&":
                case "*":
                case "(":
                case ")":
                case "=":
                case "+":
                case "[":
                case "]":
                case "{":
                case "}":
                case "\\":
                case "|":
                case ":":
                case "\"":
                case "<":
                case ">":
                case "/":
                case "~":
                    this.finishWord();
                    break;
                default:
                    this.addNewWordPart();
                    break;
            }
            this.prevChar = this.currChar;
        }
        this.postprocessing();
        return this.sentences;
    };
    SimpleSentenceStructureBuilder.prototype.finishSentence = function () {
        this.finishWord();
        this.currSentence = null;
    };
    SimpleSentenceStructureBuilder.prototype.finishWord = function () {
        this.currWord = null;
    };
    SimpleSentenceStructureBuilder.prototype.addSentence = function () {
        this.currSentence = new SimpleSentence();
        this.sentences.push(this.currSentence);
        this.currWord = null;
    };
    SimpleSentenceStructureBuilder.prototype.addWord = function () {
        if (!this.currSentence)
            this.addSentence();
        this.currWord = new SimpleSentenceWord(this.charIndex, "");
        this.currSentence.words.push(this.currWord);
    };
    SimpleSentenceStructureBuilder.prototype.addNewWordPart = function () {
        if (!this.currWord)
            this.addWord();
        this.currWord.text += this.currChar;
    };
    SimpleSentenceStructureBuilder.prototype.postprocessing = function () {
        var separators = [rich_utils_1.RichUtils.specialCharacters.LeftSingleQuote, rich_utils_1.RichUtils.specialCharacters.RightSingleQuote, "'"];
        for (var sentInd = 0, sentence = void 0; sentence = this.sentences[sentInd];) {
            for (var wordInd = 0, word = void 0; word = sentence.words[wordInd];) {
                var trimmedFromStart = string_1.StringUtils.trimStart(word.text, separators);
                word.position -= word.text.length - trimmedFromStart.length;
                var trimmedFromBothSides = string_1.StringUtils.trimEnd(trimmedFromStart, separators);
                if (trimmedFromBothSides.length) {
                    word.text = trimmedFromBothSides;
                    wordInd++;
                }
                else
                    sentence.words.splice(wordInd, 1);
            }
            if (sentence.words.length)
                sentInd++;
            else
                this.sentences.splice(sentInd, 1);
        }
    };
    return SimpleSentenceStructureBuilder;
}());
exports.SimpleSentenceStructureBuilder = SimpleSentenceStructureBuilder;
