// devexpress-richedit/lib/base/commands/header-footer/different-first-odd-even-page-header-footer-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var layout_page_1 = require("../../../core/layout/main-structures/layout-page");
var document_properties_history_items_1 = require("../../../core/model/history/items/document-properties-history-items");
var section_properties_history_items_1 = require("../../../core/model/history/items/section-properties-history-items");
var header_footer_1 = require("../../../core/model/section/header-footer");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var common_1 = require("@devexpress/utils/lib/utils/common");
var client_command_1 = require("../client-command");
var change_active_sub_document_command_1 = require("../sub-document/change-active-sub-document-command");
var header_footer_command_base_1 = require("./header-footer-command-base");
var DifferentFirstPageAndOddEvenHeaderFooterCommandBase = (function (_super) {
    tslib_1.__extends(DifferentFirstPageAndOddEvenHeaderFooterCommandBase, _super);
    function DifferentFirstPageAndOddEvenHeaderFooterCommandBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DifferentFirstPageAndOddEvenHeaderFooterCommandBase.prototype.DEPRECATEDConvertOptionsParameter = function (parameter) {
        return typeof parameter === 'boolean' ? parameter : !this.getValue();
    };
    DifferentFirstPageAndOddEvenHeaderFooterCommandBase.prototype.executeCore = function (state, options) {
        var pageIndex = this.selection.pageIndex;
        var sectionIndex = this.getSectionIndex(pageIndex);
        var section = this.control.modelManager.model.sections[sectionIndex];
        var activeSubDoc = this.selection.activeSubDocument;
        var isHeader = activeSubDoc.isHeader();
        var sectionHeadersFooters = header_footer_command_base_1.HeaderFooterCommandBase.getSectionHeadersFooters(isHeader, section);
        var newValue = options.param;
        if (newValue === state.value)
            return false;
        this.history.beginTransaction();
        var page = this.control.layoutFormatterManager.forceFormatPage(pageIndex);
        var isFirstPageOfSection = page.flags.get(layout_page_1.LayoutPageFlags.IsFirstPageOfSection);
        this.changeSectionProperty(section, newValue);
        var headerFooterType = header_footer_1.SectionHeadersFooters.getActualObjectType(section, isFirstPageOfSection, common_1.isEven(page.layoutPageIndex));
        var info = sectionHeadersFooters.getObject(headerFooterType);
        if (!info) {
            var manipulator = this.control.modelManager.modelManipulator.header.getHeaderFooterManipulator(isHeader);
            manipulator.changeObjectIndex(sectionIndex, headerFooterType, manipulator.createObject(headerFooterType));
            info = sectionHeadersFooters.getObject(headerFooterType);
        }
        if (!this.control.layout.isFullyFormatted)
            this.control.layoutFormatterManager.formatSyncAllDocument();
        if (!this.control.commandManager.getCommand(client_command_1.RichEditClientCommand.ChangeActiveSubDocumentToHeaderFooterByPageIndex).execute(this.control.commandManager.isPublicApiCall, new change_active_sub_document_command_1.ChangeActiveSubDocumentToHeaderFooterByPageIndexCommandParameters(this.control, pageIndex, isHeader)))
            this.control.commandManager.getCommand(client_command_1.RichEditClientCommand.ChangeActiveSubDocumentToHeaderFooterByPageIndex).execute(this.control.commandManager.isPublicApiCall, new change_active_sub_document_command_1.ChangeActiveSubDocumentToHeaderFooterByPageIndexCommandParameters(this.control, this.control.layout.pages.length - 1, isHeader));
        this.history.endTransaction();
        return true;
    };
    return DifferentFirstPageAndOddEvenHeaderFooterCommandBase;
}(header_footer_command_base_1.HeaderFooterCommandBase));
exports.DifferentFirstPageAndOddEvenHeaderFooterCommandBase = DifferentFirstPageAndOddEvenHeaderFooterCommandBase;
var DifferentFirstPageHeaderFooterCommand = (function (_super) {
    tslib_1.__extends(DifferentFirstPageHeaderFooterCommand, _super);
    function DifferentFirstPageHeaderFooterCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DifferentFirstPageHeaderFooterCommand.prototype.getValue = function () {
        return this.control.modelManager.model.sections[this.getSectionIndex(this.selection.pageIndex)].sectionProperties.differentFirstPage;
    };
    DifferentFirstPageHeaderFooterCommand.prototype.changeSectionProperty = function (section, newValue) {
        this.history.addAndRedo(new section_properties_history_items_1.SectionDifferentFirstPageHistoryItem(this.modelManipulator, new fixed_1.FixedInterval(section.startLogPosition.value, section.getLength()), newValue));
    };
    return DifferentFirstPageHeaderFooterCommand;
}(DifferentFirstPageAndOddEvenHeaderFooterCommandBase));
exports.DifferentFirstPageHeaderFooterCommand = DifferentFirstPageHeaderFooterCommand;
var DifferentOddEvenHeaderFooterCommand = (function (_super) {
    tslib_1.__extends(DifferentOddEvenHeaderFooterCommand, _super);
    function DifferentOddEvenHeaderFooterCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DifferentOddEvenHeaderFooterCommand.prototype.getValue = function () {
        return this.control.modelManager.model.differentOddAndEvenPages;
    };
    DifferentOddEvenHeaderFooterCommand.prototype.changeSectionProperty = function (_section, newValue) {
        this.history.addAndRedo(new document_properties_history_items_1.DifferentOddAndEvenPagesHistoryItem(this.modelManipulator, newValue));
    };
    return DifferentOddEvenHeaderFooterCommand;
}(DifferentFirstPageAndOddEvenHeaderFooterCommandBase));
exports.DifferentOddEvenHeaderFooterCommand = DifferentOddEvenHeaderFooterCommand;
