// devexpress-richedit/lib/core/layout/main-structures/layout-boxes/layout-field-code-start-box.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var encode_1 = require("@devexpress/utils/lib/utils/encode");
var measure_info_1 = require("../../../measurer/measure-info");
var layout_box_1 = require("./layout-box");
var LayoutFieldCodeStartBox = (function (_super) {
    tslib_1.__extends(LayoutFieldCodeStartBox, _super);
    function LayoutFieldCodeStartBox() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LayoutFieldCodeStartBox.prototype.clone = function () {
        var newObj = new LayoutFieldCodeStartBox(this.characterProperties, this.colorInfo);
        newObj.copyFrom(this);
        return newObj;
    };
    LayoutFieldCodeStartBox.prototype.copyFrom = function (obj) {
        _super.prototype.copyFrom.call(this, obj);
    };
    LayoutFieldCodeStartBox.prototype.getType = function () {
        return layout_box_1.LayoutBoxType.FieldCodeStart;
    };
    LayoutFieldCodeStartBox.prototype.getBoxChar = function () {
        return "{";
    };
    LayoutFieldCodeStartBox.prototype.pushInfoForMeasure = function (info, _showHiddenSymbols) {
        info.push(new measure_info_1.MeasureInfoNonText(this.getBoxChar(), this.characterProperties));
    };
    LayoutFieldCodeStartBox.prototype.popInfoForMeasure = function (info, _showHiddenSymbols) {
        this.setSize(info.pop().resultSize);
    };
    LayoutFieldCodeStartBox.prototype.isVisibleForRowAlign = function () {
        return true;
    };
    LayoutFieldCodeStartBox.prototype.isVisible = function () {
        return true;
    };
    LayoutFieldCodeStartBox.prototype.renderGetContent = function (_renderer) {
        return encode_1.EncodeUtils.encodeHtml(this.getBoxChar());
    };
    LayoutFieldCodeStartBox.prototype.isWhitespace = function () {
        return false;
    };
    LayoutFieldCodeStartBox.prototype.isLineBreak = function () {
        return false;
    };
    return LayoutFieldCodeStartBox;
}(layout_box_1.LayoutBox));
exports.LayoutFieldCodeStartBox = LayoutFieldCodeStartBox;
