// devexpress-richedit/lib/core/model/history/items/document-properties-history-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var history_item_1 = require("../base/history-item");
var DocumentDefaultTabWidthHistoryItem = (function (_super) {
    tslib_1.__extends(DocumentDefaultTabWidthHistoryItem, _super);
    function DocumentDefaultTabWidthHistoryItem(modelManipulator, newDefaultTabWidth) {
        var _this = _super.call(this, modelManipulator) || this;
        _this.newDefaultTabWidth = newDefaultTabWidth;
        return _this;
    }
    DocumentDefaultTabWidthHistoryItem.prototype.redo = function () {
        this.oldDefaultTabWidth = this.modelManipulator.documentProperties.setDefaultTabWidth(this.modelManipulator.model, this.newDefaultTabWidth);
    };
    DocumentDefaultTabWidthHistoryItem.prototype.undo = function () {
        this.modelManipulator.documentProperties.setDefaultTabWidth(this.modelManipulator.model, this.oldDefaultTabWidth);
    };
    return DocumentDefaultTabWidthHistoryItem;
}(history_item_1.HistoryItem));
exports.DocumentDefaultTabWidthHistoryItem = DocumentDefaultTabWidthHistoryItem;
var PageColorHistoryItem = (function (_super) {
    tslib_1.__extends(PageColorHistoryItem, _super);
    function PageColorHistoryItem(modelManipulator, newPageColor) {
        var _this = _super.call(this, modelManipulator) || this;
        _this.newPageColor = newPageColor;
        return _this;
    }
    PageColorHistoryItem.prototype.redo = function () {
        this.oldPageColor = this.modelManipulator.documentProperties.changePageColor(this.modelManipulator.model, this.newPageColor);
    };
    PageColorHistoryItem.prototype.undo = function () {
        this.modelManipulator.documentProperties.changePageColor(this.modelManipulator.model, this.oldPageColor);
    };
    return PageColorHistoryItem;
}(history_item_1.HistoryItem));
exports.PageColorHistoryItem = PageColorHistoryItem;
var DifferentOddAndEvenPagesHistoryItem = (function (_super) {
    tslib_1.__extends(DifferentOddAndEvenPagesHistoryItem, _super);
    function DifferentOddAndEvenPagesHistoryItem(modelManipulator, newValue) {
        var _this = _super.call(this, modelManipulator) || this;
        _this.newValue = newValue;
        return _this;
    }
    DifferentOddAndEvenPagesHistoryItem.prototype.redo = function () {
        this.oldValue = this.modelManipulator.documentProperties.changeDifferentOddAndEvenPages(this.modelManipulator.model, this.newValue);
    };
    DifferentOddAndEvenPagesHistoryItem.prototype.undo = function () {
        this.modelManipulator.documentProperties.changeDifferentOddAndEvenPages(this.modelManipulator.model, this.oldValue);
    };
    return DifferentOddAndEvenPagesHistoryItem;
}(history_item_1.HistoryItem));
exports.DifferentOddAndEvenPagesHistoryItem = DifferentOddAndEvenPagesHistoryItem;
