// devexpress-richedit/lib/client/public/ribbon/items/sub-menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var base_1 = require("./base");
var button_1 = require("./button");
var RibbonSubMenuItem = (function (_super) {
    tslib_1.__extends(RibbonSubMenuItem, _super);
    function RibbonSubMenuItem(id, text, items, options) {
        if (items === void 0) { items = []; }
        if (options === void 0) { options = {}; }
        var _this = _super.call(this, id, options.beginGroup) || this;
        _this.type = base_1.RibbonItemType.SubMenu;
        _this.text = text;
        _this.items = items;
        _this.icon = options.icon;
        _this.localizationId = options.localizationId;
        return _this;
    }
    RibbonSubMenuItem.prototype.convertToButton = function () {
        return new button_1.RibbonButtonItem(this.id, this.text, {
            showText: true,
            toggleMode: false,
            icon: this.icon,
            beginGroup: false,
            localizationId: this.localizationId,
        });
    };
    return RibbonSubMenuItem;
}(base_1.RibbonItemBase));
exports.RibbonSubMenuItem = RibbonSubMenuItem;
