// devexpress-dashboard/common/extensions/data-inspector-extension/data-inspector-extension.js
/**
* DevExpress Dashboard (data-inspector-extension.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var _default_1 = require("../../../data/localization/_default");
var model_1 = require("../../../model");
var disposable_object_1 = require("../../../model/disposable-object");
var caption_toolbar_options_1 = require("../../../viewer-parts/widgets/caption-toolbar/caption-toolbar-options");
var control_options_1 = require("../../control-options");
var _underlying_data_provider_1 = require("../../data/_underlying-data-provider");
var _options_manager_1 = require("../../internal/_options-manager");
var _utils_1 = require("../../internal/_utils");
var _data_inspector_view_model_1 = require("./_data-inspector-view-model");
var name = 'data-inspector';
var nameAlias = 'dataInspector';
var DataInspectorExtension = (function (_super) {
    __extends(DataInspectorExtension, _super);
    function DataInspectorExtension(dashboardControl, options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this) || this;
        _this._optionsManager = new _options_manager_1.OptionsManager();
        _this._defaultOptions = {
            allowInspectAggregatedData: false,
            allowInspectRawData: false
        };
        _this.name = name;
        _this._addContextToolbarItem = function (toolbarOptions, dashboardItem) {
            if ((_this._optionsManager.get('allowInspectAggregatedData') || _this._optionsManager.get('allowInspectRawData')) && dashboardItem instanceof model_1.DataDashboardItem) {
                toolbarOptions.actionItems.push({
                    hint: _default_1.getLocalizationById('DashboardStringId.ActionShowDataInspector'),
                    icon: 'dx-dashboard-data-inspector',
                    name: caption_toolbar_options_1.dashboardToolbarItemNames.dataInspector,
                    type: 'button',
                    click: function () {
                        _this._viewModel.show(dashboardItem, _this._optionsManager.get('allowInspectAggregatedData'), _this._optionsManager.get('allowInspectRawData'));
                    }
                });
            }
        };
        _this._optionsManager.initialize({
            alias: nameAlias,
            dashboardControl: dashboardControl,
            defaultOptions: _this._defaultOptions,
            eventsHolder: _this,
            initOptions: options,
            optionChanged: function (args) { return _this._optionChanged(args); }
        });
        _this._dashboardControl = dashboardControl;
        _this._viewModel = new _data_inspector_view_model_1.DataInspectorViewModel({
            onGridContentReady: function (e) { _this._optionsManager.raiseEvent('gridContentReady', e); },
            onGridInitialized: function (e) { _this._optionsManager.raiseEvent('gridInitialized', e); },
            onDialogShowing: function (e) { _this._optionsManager.raiseEvent('dialogShowing', e); },
            onDialogShown: function (e) { _this._optionsManager.raiseEvent('dialogShown', e); },
            onDialogHidden: function (e) { _this._optionsManager.raiseEvent('dialogHidden', e); },
        });
        _this._customTemplate = {
            name: 'dx-dashboard-data-inspector-extension',
            data: _this._viewModel
        };
        return _this;
    }
    DataInspectorExtension.prototype._optionChanged = function (args) {
        switch (args.name) {
            case 'allowInspectAggregatedData':
            case 'allowInspectRawData':
                this.hideDataInspector();
                return 'updateItemToolbars';
            default:
                return null;
        }
    };
    DataInspectorExtension.prototype.start = function () {
        var _this = this;
        this._dashboardControl.customTemplates.push(this._customTemplate);
        if (this._dashboardControl._serviceClient()) {
            this._viewModel.setUnderlyingDataProvider(new _underlying_data_provider_1.UnderlyingDataProvider(this._dashboardControl._serviceClient()));
        }
        this.toDispose(this._dashboardControl._serviceClient.subscribe(function (serviceClient) {
            _this._viewModel.setUnderlyingDataProvider(new _underlying_data_provider_1.UnderlyingDataProvider(serviceClient));
        }));
        this._dashboardControl._dashboardContext.addContextToolbarItems.add(this._addContextToolbarItem);
    };
    DataInspectorExtension.prototype.stop = function () {
        this._dashboardControl.customTemplates.remove(this._customTemplate);
        this._dashboardControl._dashboardContext.addContextToolbarItems.remove(this._addContextToolbarItem);
    };
    DataInspectorExtension.prototype.showDataInspector = function (dashboardItemName, inspectedType) {
        var dashboard = this._dashboardControl.dashboard();
        if (dashboard) {
            var dashboardItem = _utils_1.findItemForApi(dashboard, dashboardItemName, model_1.DataDashboardItem);
            this._viewModel.show(dashboardItem, this._optionsManager.get('allowInspectAggregatedData'), this._optionsManager.get('allowInspectRawData'), inspectedType);
        }
    };
    DataInspectorExtension.prototype.currentInspectedType = function () {
        return this._viewModel.visible() && this._viewModel.inspectedDataType() || null;
    };
    DataInspectorExtension.prototype.hideDataInspector = function () {
        this._viewModel.visible(false);
    };
    return DataInspectorExtension;
}(disposable_object_1.DisposableObject));
exports.DataInspectorExtension = DataInspectorExtension;
control_options_1.defaultExtensions[name] = function (dashboardControl, options) { return new DataInspectorExtension(dashboardControl, options); };
control_options_1.extensionNameMap[name] = nameAlias;
