// devexpress-dashboard/viewer-parts/viewer-items/data-grid-item/_data-grid-item.js
/**
* DevExpress Dashboard (_data-grid-item.js)
* Version:  21.2.3
* Build date: Oct 25, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var string_1 = require("devextreme/core/utils/string");
var data_grid_1 = require("devextreme/ui/data_grid");
var _default_1 = require("../../../data/localization/_default");
var _common_1 = require("../../../data/_common");
var _jquery_helpers_1 = require("../../../data/_jquery-helpers");
var _localization_ids_1 = require("../../../data/_localization-ids");
var _localizer_1 = require("../../../data/_localizer");
var _utils_1 = require("../../../data/_utils");
var _style_settings_provider_1 = require("../../conditional-formatting/_style-settings-provider");
var _base_item_1 = require("../_base-item");
var _column_width_calculator_1 = require("./_column-width-calculator");
var _grid_column_painter_1 = require("./_grid-column-painter");
var _grid_filter_parser_1 = require("./_grid-filter-parser");
var MAX_CELL_COUNT = 2000, HEIGHT_DELTA_INDICATOR = 12, DATAGRID_CONTEXT_MENU_ICON = 'dashboard-datagriditem-resetcolumnwidths';
var dataGridItem = (function (_super) {
    __extends(dataGridItem, _super);
    function dataGridItem(container, options) {
        var _this = _super.call(this, container, options) || this;
        _this._updateLocked = false;
        _this._digits_string = '0123456789';
        _this.TextAlignment = {
            Left: 'left',
            Right: 'right',
            Center: 'center'
        };
        _this.DisplayMode = {
            Value: 'Value',
            Delta: 'Delta',
            Bar: 'Bar',
            Sparkline: 'Sparkline',
            Image: 'Image',
            Hyperlink: 'Hyperlink'
        };
        _this.SummaryType = {
            Count: 'Count',
            Min: 'Min',
            Max: 'Max',
            Avg: 'Avg',
            Sum: 'Sum'
        };
        _this.CssClasses = {
            wordWrap: 'dx-dashboard-word-wrap',
            gridAdaptiveCellValue: 'dx-adaptive-item-text'
        };
        _this._bestFitProvider = {
            getBestFit: function (index) {
                var bestFit = 0;
                var columnViewModel = _this.options.ViewModel.Columns[index];
                if (columnViewModel.DisplayMode === _this.DisplayMode.Sparkline) {
                    var columnName = _this.options.Name + columnViewModel.DataId;
                    var startValues = _jquery_helpers_1.$unwrap(_this._dataGrid.element()).querySelectorAll('.' + columnName + '_startValue');
                    var endValues = _jquery_helpers_1.$unwrap(_this._dataGrid.element()).querySelectorAll('.' + columnName + '_endValue');
                    var maxStartWidth = _grid_column_painter_1.GridColumnPainter.calcMaxWidth(startValues);
                    var maxEndWidth = _grid_column_painter_1.GridColumnPainter.calcMaxWidth(endValues);
                    bestFit = Math.round(_this._getDefaultBestCharacterCount(index) * _this.charWidth) + maxStartWidth + maxEndWidth;
                }
                else if (columnViewModel.DisplayMode === _this.DisplayMode.Bar) {
                    bestFit = Math.round(_this._getDefaultBestCharacterCount(index) * _this.charWidth);
                }
                else {
                    bestFit = _this._dataGrid.columnOption(index, 'bestFitWidth');
                }
                return bestFit;
            }
        };
        _this._clientStateUpdateDebounced = _utils_1.debounce(function (widthOptions) {
            _this.gridWidthOptionsChanged && _this.gridWidthOptionsChanged({
                widthOptions: widthOptions
            });
        }, 300);
        _this._customizeViewOptions = function (opts) {
            opts.columnAutoWidth = true;
        };
        _this._calculator = new _column_width_calculator_1.ColumnWidthCalculator();
        _this._styleSettingsProvider = new _style_settings_provider_1.styleSettingsProvider();
        _this._styleSettingsProvider.initialize(_this.options.ConditionalFormattingModel, _this.options.ViewModel ? _this.options.ViewModel.WordWrap : false);
        return _this;
    }
    Object.defineProperty(dataGridItem.prototype, "_captionToolbarSeparatorRequired", {
        get: function () {
            return !this.hasParentContainer() || (this.hasParentContainer() && this.visualMode === 'caption');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(dataGridItem.prototype, "dataController", {
        get: function () { return this._dataController; },
        set: function (dataController) { this._dataController = dataController; },
        enumerable: true,
        configurable: true
    });
    dataGridItem.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this._dataGrid && this._dataGrid.dispose();
        this._styleSettingsProvider.dispose();
    };
    dataGridItem.prototype._clearSelectionUnsafe = function () {
        this._dataGrid.clearSelection();
    };
    dataGridItem.prototype._setSelectionUnsafe = function (values) {
        _super.prototype._setSelectionUnsafe.call(this, values);
        this.clearSelection();
        if (values)
            this._setGridSelection(values);
    };
    dataGridItem.prototype._selectTuplesCore = function (tuples, updateTupleDelegate, state) {
        var that = this, currentSelection = that._dataGrid.getSelectedRowKeys(), values = [], processKeys = function (keys) {
            keys.forEach(function (key) {
                if (state) {
                    currentSelection.push(key);
                }
                else {
                    currentSelection.splice(currentSelection.indexOf(key), 1);
                }
            });
            return currentSelection;
        };
        tuples.forEach(function (tuple) {
            var res = updateTupleDelegate(tuple);
            if (res != null && res[0] != null) {
                values.push(res[0].Value);
            }
        });
        that._setGridSelection(values, processKeys);
    };
    dataGridItem.prototype.renderContentUnsafe = function (element, changeExisting, afterRenderCallback) {
        if (!this.options)
            return false;
        var opts = this._getViewOptions();
        if (this.options.ViewModel && this.options.ViewModel.WordWrap) {
            element.classList.add(this.CssClasses.wordWrap);
        }
        else {
            element.classList.remove(this.CssClasses.wordWrap);
        }
        if (changeExisting && this._dataGrid) {
            if (this.resetClientStateOnUpdate) {
                this._resetColumnWidths();
            }
            _jquery_helpers_1.deepExtend(opts, this._getCommonOptions());
            this._raiseItemWidgetOptionsPrepared(opts);
            this._dataGrid.option(opts);
        }
        else {
            var options = __assign({}, opts, this._getCommonOptions());
            this._raiseItemWidgetOptionsPrepared(options);
            this._dataGrid = new data_grid_1.default(element, options);
            this._updateCharWidth();
        }
        return false;
    };
    dataGridItem.prototype.getInfoUnsafe = function () {
        var that = this, gridScroll = this._dataGrid.getScrollable(), isVScrollbarVisible = that._dataGrid.isScrollbarVisible(), isHScrollbarVisible = gridScroll.scrollWidth() > gridScroll.clientWidth(), scrollRowData = that._dataGrid.getTopVisibleRowData(), combinedFilter = that._dataGrid.getCombinedFilter(), visibleColumns = that._dataGrid.getVisibleColumns(), topPath = [], leftPath = [];
        if (scrollRowData) {
            that.options.ViewModel.RowIdentificatorDataMembers.forEach(function (dataMember) {
                var value = scrollRowData[dataMember];
                topPath.push(value);
            });
            leftPath.push(this._calculator.getLeftPrintingColumnIndex(gridScroll.scrollLeft()));
        }
        return _jquery_helpers_1.deepExtend(_super.prototype.getInfoUnsafe.call(this), {
            scroll: {
                horizontal: isHScrollbarVisible,
                vertical: isVScrollbarVisible,
                topPath: topPath,
                leftPath: leftPath
            },
            sortInfo: that._getSortInfo(),
            combinedFilter: combinedFilter !== undefined ? _grid_filter_parser_1.parseFilter(combinedFilter, visibleColumns, undefined, this.dateToString) : '',
            widthOptions: that._calculator.getClientWidthOptions()
        });
    };
    dataGridItem.prototype._getSortInfo = function () {
        var that = this, sortInfo = [];
        that._dataGrid.getVisibleColumns().forEach(function (info) {
            if (info.sortIndex != undefined && info.sortOrder != undefined) {
                sortInfo.push({
                    dataField: info.dataField,
                    sortIndex: info.sortIndex,
                    sortOrder: info.sortOrder
                });
            }
        });
        sortInfo.sort(function (a, b) {
            return a.sortIndex - b.sortIndex;
        });
        return sortInfo;
    };
    dataGridItem.prototype._initializeData = function (newOptions) {
        _super.prototype._initializeData.call(this, newOptions);
        var isResetColumnWidthsRequired = !!this.options && !!this.options.ViewModel && !!newOptions.ViewModel && this.options.ViewModel.ColumnWidthMode !== newOptions.ViewModel.ColumnWidthMode;
        if (isResetColumnWidthsRequired) {
            this._resetColumnWidths();
        }
        if (this._styleSettingsProvider) {
            this._styleSettingsProvider.initialize(this.options.ConditionalFormattingModel, this.options.ViewModel ? this.options.ViewModel.WordWrap : false);
        }
    };
    dataGridItem.prototype.getValueItem = function (columnName, index) {
        return this._dataController.getValueItem(columnName, index);
    };
    dataGridItem.prototype._resetColumnWidths = function () {
        this._calculator.reset(this.options.ViewModel, this._getColumnWidthMode());
        var columnsWidth = this._calculator.calcColumnsWidth(this._bestFitProvider, (_jquery_helpers_1.$wrap(this._dataGrid.element())).width() - this._dataGrid.getScrollbarWidth(), this.charWidth);
        this._updateColumnsWidth(columnsWidth);
        this._applySelection();
    };
    dataGridItem.prototype._updateColumnsWidth = function (columnWidths) {
        var _this = this;
        this._beginResize();
        columnWidths.forEach(function (width, i) {
            _this._dataGrid.columnOption(i, _this._getColumnWidthProperty(), width);
        });
        this._endResize();
    };
    dataGridItem.prototype._getColumnWidthProperty = function () {
        return 'visibleWidth';
    };
    dataGridItem.prototype._getColumnWidthMode = function () {
        return this.options.ViewModel.ColumnWidthMode;
    };
    dataGridItem.prototype._getDefaultBestCharacterCount = function (index) {
        return this.options.ViewModel.Columns[index].DefaultBestCharacterCount;
    };
    dataGridItem.prototype._beginResize = function () {
        this._updateLocked = true;
        this._dataGrid.beginUpdate();
    };
    dataGridItem.prototype._endResize = function () {
        this._dataGrid.endUpdate();
        this._updateLocked = false;
    };
    dataGridItem.prototype._updateCharWidth = function () {
        var span = document.createElement('span');
        span.classList.add('dx-widget');
        span.style.display = 'inline-block';
        span.innerText = this._digits_string;
        document.body.appendChild(span);
        this.charWidth = _jquery_helpers_1.getWidth(span) / 10;
        document.body.removeChild(span);
    };
    dataGridItem.prototype._onColumnsChanging = function (e) {
        var grid = e.component, columnResized = e.optionNames.width, gridResized = e.optionNames.visibleWidth && grid.columnOption(0, 'visibleWidth') !== undefined, viewModel = this.options.ViewModel;
        if (this._updateLocked || !viewModel || !viewModel.Columns ||
            viewModel.Columns.length == 0 || grid.columnCount() !== viewModel.Columns.length) {
            return;
        }
        if (grid.getController('data').isLoaded() && gridResized && !columnResized) {
            this._calculator.onDataLoaded(viewModel, this._getColumnWidthMode());
        }
        if (columnResized) {
            var leftColumnIndex = grid.getController('columnsResizer')._resizingInfo.currentColumnIndex;
            var leftColumnWidth = this._dataGrid.columnOption(leftColumnIndex, 'width');
            var rightColumnWidth = this._dataGrid.columnOption(leftColumnIndex + 1, 'width');
            var columnWidths = this._calculator.onColumnResized(leftColumnIndex, leftColumnWidth, rightColumnWidth);
            this._updateColumnsWidth(columnWidths);
            this._clientStateUpdateDebounced(this._calculator.widthOptions);
        }
        else if (gridResized) {
            var columnsWidth = this._calculator.calcColumnsWidth(this._bestFitProvider, _jquery_helpers_1.$wrap(this._dataGrid.element()).width() - this._dataGrid.getScrollbarWidth(), this.charWidth);
            this._updateColumnsWidth(columnsWidth);
        }
    };
    dataGridItem.prototype._getViewOptions = function () {
        var that = this;
        var viewModel = that.options.ViewModel;
        if (!viewModel)
            return {};
        var dataSource = that._dataController.getDataSource(), columns = that._getColumns();
        var hasSlowColumns = columns.some(function (c) { return c.displayMode == that.DisplayMode.Bar || c.displayMode == that.DisplayMode.Sparkline || c.displayMode == that.DisplayMode.Image; });
        var viewOptions = {
            dataSource: dataSource,
            keyExpr: 'index',
            columns: columns,
            summary: {
                totalItems: that._getTotals(),
                calculateCustomSummary: function () { return ''; }
            },
            scrolling: {
                newMode: false,
                mode: hasSlowColumns || dataSource.length * (columns.length || 1) >= MAX_CELL_COUNT ? 'virtual' : 'standard'
            },
            rowAlternationEnabled: viewModel.EnableBandedRows,
            showRowLines: viewModel.ShowHorizontalLines,
            showColumnLines: viewModel.ShowVerticalLines,
            wordWrapEnabled: viewModel.WordWrap,
            showColumnHeaders: viewModel.ShowColumnHeaders,
            allowColumnResizing: this._getColumnWidthMode() !== 'AutoFitToContents',
        };
        this._customizeViewOptions(viewOptions);
        return viewOptions;
    };
    dataGridItem.prototype._getCommonOptions = function () {
        var _this = this;
        var that = this, commonOptions = {};
        commonOptions.remoteOperations = { summary: true };
        commonOptions.paging = { enabled: false };
        commonOptions.sorting = {
            mode: 'multiple',
            ascendingText: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.GridSortAscending),
            descendingText: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.GridSortDescending),
            clearText: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.GridClearSorting)
        };
        if (this.manualyResetClientState) {
            commonOptions.onContextMenuPreparing = function (options) {
                if (options.target === 'content' && _this._calculator.columnsResized) {
                    options.items = [{
                            text: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.GridResetColumnWidths),
                            icon: DATAGRID_CONTEXT_MENU_ICON,
                            onItemClick: function () {
                                _this._resetColumnWidths();
                            }
                        }];
                }
            };
        }
        commonOptions.onContentReady = function () {
            that._styleSettingsProvider.draw();
            _this._changeGridSparklineColumnsWidth(_this.options && _this.options.ViewModel && _this.options.ViewModel.Columns);
        };
        commonOptions.noDataText = _localizer_1.localizer.getString(_localization_ids_1.localizationId.MessageGridHasNoData);
        commonOptions.onCellClick = function (data) {
            var columnType = data.column && data.column.type || null;
            if (data.rowType === 'data' && columnType !== 'adaptive')
                that._raiseItemClick(data);
        };
        commonOptions.onCellHoverChanged = function (data) {
            that._raiseItemHover(data, data.eventType == 'mouseover');
        };
        commonOptions['onColumnsChanging'] = function (e) { return _this._onColumnsChanging(e); };
        commonOptions.width = '100%';
        commonOptions.cellHintEnabled = true;
        commonOptions.loadPanel = { enabled: false };
        commonOptions.searchPanel = { visible: false };
        commonOptions.keyboardNavigation = { enabled: false };
        return commonOptions;
    };
    dataGridItem.prototype._getRowsValues = function (data) {
        var that = this, selectionMembers = that.options.ViewModel.SelectionDataMembers, visibleValues = [], value;
        selectionMembers.forEach(function (member) {
            value = that.getValueItem(member, data.index).getUniqueValue();
            if (value !== undefined) {
                visibleValues.push(value);
            }
        });
        return that._dataController.getSelectionValues(visibleValues);
    };
    dataGridItem.prototype._getSelectedRowIndices = function () {
        var selectedValues = this._getSelectedValues();
        return selectedValues ? this._dataController.getSelectedRowKeys(selectedValues) : [];
    };
    dataGridItem.prototype._getTotalCaptionTemplate = function (totalType) {
        if (totalType == 'Auto') {
            return _default_1.getLocalizationById('DashboardStringId.GridTotalAutoTemplate');
        }
        else {
            var id = void 0;
            switch (totalType) {
                case 'Min':
                    id = 'DashboardStringId.GridTotalTypeMin';
                    break;
                case 'Max':
                    id = 'DashboardStringId.GridTotalTypeMax';
                    break;
                case 'Avg':
                    id = 'DashboardStringId.GridTotalTypeAvg';
                    break;
                case 'Sum':
                    id = 'DashboardStringId.GridTotalTypeSum';
                    break;
                default:
                    id = 'DashboardStringId.GridTotalTypeCount';
                    break;
            }
            return string_1.format(_default_1.getLocalizationById('DashboardStringId.GridTotalTemplate'), _default_1.getLocalizationById(id), _default_1.getLocalizationById('DashboardStringId.GridTotalValueTemplate'));
        }
    };
    dataGridItem.prototype._getTotals = function () {
        var _this = this;
        var that = this, res = [], columns = that.options.ViewModel.Columns || [];
        columns.forEach(function (column) {
            var columnName = column.DataId, totalsViewModel = column.Totals || [];
            totalsViewModel.forEach(function (totalModel) {
                res.push({
                    column: columnName,
                    summaryType: 'custom',
                    displayFormat: string_1.format(_this._getTotalCaptionTemplate(totalModel.TotalType), that._dataController.getTotalValue(totalModel.DataId))
                });
            });
        });
        return res;
    };
    dataGridItem.prototype._calculateCustomSummary = function (options) {
        options.totalValue = 0;
    };
    dataGridItem.prototype._getColumnName = function (columnViewModel) {
        var gridName = this.options.Name;
        var fieldName = columnViewModel.DataId;
        return gridName + fieldName;
    };
    dataGridItem.prototype._getColumns = function () {
        var _this = this;
        var that = this, res = [], columns = that.options.ViewModel.Columns || [], popupContainer = this.options.controlContainer;
        columns.forEach(function (column, columnIndex) {
            var fieldName = column.DataId, columnName = _this._getColumnName(column), gridColumn = {
                allowEditing: false,
                dataField: fieldName,
                encodeHtml: that._isEncodeHtml(),
                caption: column.Caption,
                alignment: column.HorzAlignment === 'Right' ? that.TextAlignment.Right : that.TextAlignment.Left
            };
            gridColumn[_common_1.DashboardDataIdField] = column.DataId;
            gridColumn.headerAlignment = that.TextAlignment.Left;
            gridColumn.displayMode = column.DisplayMode;
            switch (column.DisplayMode) {
                case that.DisplayMode.Value:
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index);
                        _grid_column_painter_1.GridColumnPainter.renderValue(container, valueItem.getData().displayText, that._isEncodeHtml());
                        var isDetail = that._isDetail(options.rowType);
                        var colummPrefix = isDetail ? columnName + '_detail' : columnName;
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), false, colummPrefix, isDetail);
                        if (isDetail)
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    gridColumn.resized = function () {
                        var visibleWidthReset = that._dataGrid && (that._dataGrid.columnOption(columnIndex, 'visibleWidth') === undefined);
                        if (!visibleWidthReset) {
                            that._styleSettingsProvider.updateBarWidth(columnName);
                            that._styleSettingsProvider.updateBarWidth(columnName + '_detail');
                        }
                    };
                    break;
                case that.DisplayMode.Delta:
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index);
                        var isDetail = that._isDetail(options.rowType);
                        _grid_column_painter_1.GridColumnPainter.renderDelta(container, valueItem.getData(), isDetail);
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), true);
                        if (isDetail)
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    break;
                case that.DisplayMode.Hyperlink:
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index), uri = fieldName ? string_1.format(!!column.UriPattern ? column.UriPattern : '{0}', valueItem.getUriValue()) : undefined;
                        _grid_column_painter_1.GridColumnPainter.renderHyperlink(container, uri, valueItem.getData().displayText, that._isEncodeHtml());
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), false, columnName);
                        if (that._isDetail(options.rowType))
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    break;
                case that.DisplayMode.Sparkline:
                    gridColumn.alignment = that.TextAlignment.Left;
                    gridColumn.cssClass = 'dx-dashboard-datagrid-column-visible-cell-content';
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index);
                        _grid_column_painter_1.GridColumnPainter.renderSparkline(columnName, column.ShowStartEndValues, container, valueItem.getData());
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), true);
                        if (that._isDetail(options.rowType))
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    gridColumn.resized = function () { return that._changeGridSparklineColumnsWidth([column]); };
                    break;
                case that.DisplayMode.Bar:
                    gridColumn.alignment = that.TextAlignment.Left;
                    gridColumn.cssClass = 'dx-dashboard-datagrid-column-visible-cell-content';
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index);
                        var barData = valueItem.getData();
                        _grid_column_painter_1.GridColumnPainter.renderBar(columnName, container, barData.text, barData.normalizedValue, barData.zeroValue);
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), true);
                        if (that._isDetail(options.rowType))
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    gridColumn.resized = function () {
                        var gridRootElement = that._dataGrid ? _jquery_helpers_1.$unwrap(that._dataGrid.element()) : undefined;
                        if (gridRootElement) {
                            _grid_column_painter_1.GridColumnPainter.changeGridBarColumnsWidth(gridRootElement, columnName);
                        }
                    };
                    break;
                case that.DisplayMode.Image:
                    gridColumn.cellTemplate = function (containerElement, options) {
                        var container = _jquery_helpers_1.$unwrap(containerElement);
                        var valueItem = that.getValueItem(fieldName, options.data.index);
                        _grid_column_painter_1.GridColumnPainter.renderImage(container, valueItem.getData());
                        that._styleSettingsProvider.applyStyleSettings(container, popupContainer, valueItem.getStyleSettingsInfo(), true);
                        if (that._isDetail(options.rowType))
                            container.classList.add(that.CssClasses.gridAdaptiveCellValue);
                    };
                    break;
            }
            res.push(gridColumn);
        });
        return res;
    };
    dataGridItem.prototype._changeGridSparklineColumnsWidth = function (columnViewModels) {
        var _this = this;
        if (columnViewModels) {
            var gridRootElement_1 = this._dataGrid ? _jquery_helpers_1.$unwrap(this._dataGrid.element()) : undefined;
            if (gridRootElement_1) {
                var sparklineViewModels = columnViewModels.filter(function (columnViewModel) { return columnViewModel.DisplayMode === _this.DisplayMode.Sparkline; });
                sparklineViewModels.forEach(function (columnViewModel) {
                    _grid_column_painter_1.GridColumnPainter.changeGridSparklineColumnsWidth(gridRootElement_1, _this._getColumnName(columnViewModel));
                });
            }
        }
    };
    dataGridItem.prototype._isDetail = function (rowType) {
        return rowType === 'detailAdaptive';
    };
    dataGridItem.prototype._applySelectionUnsafe = function () {
        this.setSelection(this.options.SelectedValues);
    };
    dataGridItem.prototype._resizeUnsafe = function () {
        _super.prototype._resizeUnsafe.call(this);
        this._dataGrid.resize();
    };
    dataGridItem.prototype._getDataPoint = function (element) {
        var that = this;
        return {
            getValues: function (name) {
                return (name === 'Default') ? that._dataController.getDimensionValues(element.data.index) : null;
            },
            getDeltaIds: function () {
                return that._getColumnDataIdsByColumnType('Delta');
            },
            getMeasureIds: function () {
                return that._getColumnDataIdsByColumnType('Measure');
            },
            getSelectionValues: function (name) {
                return that._getRowsValues(element.data);
            }
        };
    };
    dataGridItem.prototype._getColumnsByColumnType = function (columnType) {
        var columns = this.options.ViewModel.Columns, foundColumns = [];
        columns.forEach(function (column) {
            if (column.ColumnType === columnType)
                foundColumns.push(column);
        });
        return foundColumns;
    };
    dataGridItem.prototype._getColumnDataIdsByColumnType = function (columnType) {
        var that = this, columns = that._getColumnsByColumnType(columnType), ids = [];
        columns.forEach(function (column) {
            ids.push(column.DataId);
        });
        return ids;
    };
    dataGridItem.prototype._getElementInteractionValue = function (element, viewModel) {
        return this._getRowsValues(element.data);
    };
    dataGridItem.prototype._getWidget = function () {
        return this._dataGrid;
    };
    dataGridItem.prototype._setGridSelection = function (values, keyProcessingDelegate) {
        if (values && values.length > 0) {
            var that = this, selectionKeys = that._dataController.getSelectedRowKeys(values);
            that._selectRows(keyProcessingDelegate ? keyProcessingDelegate(selectionKeys) : selectionKeys);
        }
    };
    dataGridItem.prototype._selectRows = function (data) {
        if (this._dataGrid) {
            this._dataGrid.selectRows(data);
        }
    };
    dataGridItem.prototype._isMultiDataSupported = function () {
        return true;
    };
    return dataGridItem;
}(_base_item_1.baseItem));
exports.dataGridItem = dataGridItem;
