// devexpress-richedit/lib/docx/import/model/comment-list-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CommentListInfo = (function () {
    function CommentListInfo() {
    }
    return CommentListInfo;
}());
exports.CommentListInfo = CommentListInfo;
var CommentListInfoCollection = (function () {
    function CommentListInfoCollection() {
        this.mapParaId = {};
        this.mapId = {};
    }
    CommentListInfoCollection.prototype.add = function (info) {
        this.mapParaId[info.paraId] = info;
        this.mapId[info.id] = info;
    };
    return CommentListInfoCollection;
}());
exports.CommentListInfoCollection = CommentListInfoCollection;
var CommentExListInfo = (function () {
    function CommentExListInfo() {
    }
    return CommentExListInfo;
}());
exports.CommentExListInfo = CommentExListInfo;
var CommentExListInfoCollection = (function () {
    function CommentExListInfoCollection() {
        this.mapParaId = {};
    }
    CommentExListInfoCollection.prototype.add = function (info) {
        this.mapParaId[info.paraId] = info;
    };
    return CommentExListInfoCollection;
}());
exports.CommentExListInfoCollection = CommentExListInfoCollection;
