// devexpress-richedit/lib/core/word-bounds-engine/word-bounds-iterators/layout/layout-word-bounds-iterator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var layout_box_iterator_main_sub_document_1 = require("../../../layout-engine/layout-box-iterator/layout-box-iterator-main-sub-document");
var layout_box_iterator_other_sub_document_1 = require("../../../layout-engine/layout-box-iterator/layout-box-iterator-other-sub-document");
var document_layout_details_level_1 = require("../../../layout/document-layout-details-level");
var layout_box_1 = require("../../../layout/main-structures/layout-boxes/layout-box");
var rich_utils_1 = require("../../../model/rich-utils");
var layout_word_bounds_1 = require("../../layout-word-bounds");
var LayoutWordBoundsIterator = (function () {
    function LayoutWordBoundsIterator(layout, subDocument, selection, startPosition, endPosition) {
        this.isInitOk = true;
        this.isFirstCallNextSymbol = true;
        if (layout.validPageCount < 1) {
            this.startResultPosition = -1;
            this.isInitOk = false;
            return;
        }
        this.boxIterator = subDocument.isMain() ?
            new layout_box_iterator_main_sub_document_1.LayoutBoxIteratorMainSubDocument(subDocument, layout, startPosition, endPosition) :
            new layout_box_iterator_other_sub_document_1.LayoutBoxIteratorOtherSubDocument(subDocument, layout, startPosition, endPosition, selection.pageIndex);
    }
    LayoutWordBoundsIterator.prototype.isSet = function () {
        if (!this.isInitOk)
            return false;
        if (this.boxIterator.isInitialized())
            return true;
        this.startResultPosition = -1;
        return false;
    };
    LayoutWordBoundsIterator.prototype.currSymbolStartPosition = function () {
        return this.boxIterator.position.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Box) + this.charOffset;
    };
    LayoutWordBoundsIterator.prototype.getNextSymbolGroup = function () {
        if (!this.getNextSymbol())
            return false;
        this.groupMask = this.getGroupMask();
        return true;
    };
    LayoutWordBoundsIterator.prototype.getNextSymbol = function () {
        if (this.isFirstCallNextSymbol) {
            this.isFirstCallNextSymbol = false;
            if (!this.moveIterator())
                return false;
            this.charOffset = this.boxIterator.position.charOffset;
            if (this.needExcessMoveBoxIterator())
                this.moveIterator();
            return true;
        }
        if (this.nextCallsSetCharOffset())
            return true;
        if (!this.moveIterator())
            return false;
        this.setCharacterOffsetOnNextCalls();
        return true;
    };
    LayoutWordBoundsIterator.prototype.getGroupMask = function () {
        switch (this.boxIterator.position.box.getType()) {
            case layout_box_1.LayoutBoxType.Text:
                return this.getTextGroupMask();
            case layout_box_1.LayoutBoxType.Space:
                return layout_word_bounds_1.WordGroupMask.Space;
            case layout_box_1.LayoutBoxType.Dash:
                return layout_word_bounds_1.WordGroupMask.PunctuationMark;
            case layout_box_1.LayoutBoxType.Picture:
            case layout_box_1.LayoutBoxType.ParagraphMark:
            case layout_box_1.LayoutBoxType.PageBreak:
            case layout_box_1.LayoutBoxType.ColumnBreak:
            case layout_box_1.LayoutBoxType.SectionMark:
            case layout_box_1.LayoutBoxType.TabSpace:
            case layout_box_1.LayoutBoxType.LineBreak:
            case layout_box_1.LayoutBoxType.NumberingList:
            case layout_box_1.LayoutBoxType.FieldCodeStart:
            case layout_box_1.LayoutBoxType.FieldCodeEnd:
            case layout_box_1.LayoutBoxType.LayoutDependent:
            case layout_box_1.LayoutBoxType.AnchorPicture:
            case layout_box_1.LayoutBoxType.AnchorTextBox:
            default:
                return layout_word_bounds_1.WordGroupMask.DiffersFromAll;
        }
    };
    LayoutWordBoundsIterator.prototype.getTextGroupMask = function () {
        switch (this.boxIterator.position.box.renderGetContent(null)[this.charOffset]) {
            case rich_utils_1.RichUtils.specialCharacters.QmSpace:
            case rich_utils_1.RichUtils.specialCharacters.EmSpace:
            case rich_utils_1.RichUtils.specialCharacters.EnSpace:
            case rich_utils_1.RichUtils.specialCharacters.NonBreakingSpace:
                return layout_word_bounds_1.WordGroupMask.Space;
            case rich_utils_1.RichUtils.specialCharacters.LeftSingleQuote:
                return layout_word_bounds_1.WordGroupMask.LeftSingleQuote;
            case rich_utils_1.RichUtils.specialCharacters.LeftDoubleQuote:
                return layout_word_bounds_1.WordGroupMask.LeftDoubleQuote;
            case rich_utils_1.RichUtils.specialCharacters.RightDoubleQuote:
                return layout_word_bounds_1.WordGroupMask.RightDoubleQuote;
            case '"':
                return layout_word_bounds_1.WordGroupMask.DoubleQuote;
            case "(":
            case ")":
            case "«":
            case "»":
            case "<":
            case ">":
            case "/":
            case "№":
            case "%":
            case "!":
            case ":":
            case "?":
            case ";":
            case "|":
            case "+":
            case ",":
            case ".":
            case "*":
            case "=":
            case "\\":
                return layout_word_bounds_1.WordGroupMask.PunctuationMark;
            default:
                return layout_word_bounds_1.WordGroupMask.Others;
        }
    };
    return LayoutWordBoundsIterator;
}());
exports.LayoutWordBoundsIterator = LayoutWordBoundsIterator;
