// devexpress-richedit/lib/base/bars/context-menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var enums_1 = require("../../core/model/changes/enums");
var flag_1 = require("@devexpress/utils/lib/class/flag");
var base_1 = require("./base");
var ContextMenuEventType;
(function (ContextMenuEventType) {
    ContextMenuEventType[ContextMenuEventType["None"] = 0] = "None";
    ContextMenuEventType[ContextMenuEventType["FullReset"] = 1] = "FullReset";
    ContextMenuEventType[ContextMenuEventType["CheckSelectionChange"] = 2] = "CheckSelectionChange";
})(ContextMenuEventType = exports.ContextMenuEventType || (exports.ContextMenuEventType = {}));
var ContextMenuBarBase = (function (_super) {
    tslib_1.__extends(ContextMenuBarBase, _super);
    function ContextMenuBarBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.selection = null;
        _this.isSpellingMenu = false;
        return _this;
    }
    ContextMenuBarBase.prototype.forceUpdate = function (queryCommands) {
        if (this.isUpdateLocked())
            this.registerOccurredEvent(ContextMenuEventType.FullReset);
        else
            this.updateItemsState(queryCommands);
    };
    ContextMenuBarBase.prototype.getSuggestions = function () {
        var suggestions = undefined;
        if (this.core.spellChecker.settings.isEnabled) {
            var selectedMisspelledInterval = this.core.spellChecker.getSelectedMisspelledInterval(this.core.selection.intervals);
            if (selectedMisspelledInterval)
                suggestions = selectedMisspelledInterval.errorInfo.suggestions;
        }
        return suggestions;
    };
    ContextMenuBarBase.prototype.onUpdateUnlocked = function (occurredEvents) {
        this.applyEvent(occurredEvents);
    };
    ContextMenuBarBase.prototype.modelChanged = function (change) {
        var occurredEvents = this.getModelChangeEvent(change);
        if (this.isUpdateLocked())
            this.registerOccurredEvent(occurredEvents);
        else
            this.applyEvent(occurredEvents);
    };
    ContextMenuBarBase.prototype.NotifySelectionChanged = function (selection) {
        this.selection = selection;
        if (this.isUpdateLocked())
            this.registerOccurredEvent(ContextMenuEventType.CheckSelectionChange);
        else
            this.applyEvent(this.getSelectionChangeEvent());
    };
    ContextMenuBarBase.prototype.applyEvent = function (occurredEvents) {
        if (!this.updateEnabled)
            return;
        var flag = new flag_1.Flag(occurredEvents);
        if (flag.get(ContextMenuEventType.CheckSelectionChange))
            flag.add(this.getSelectionChangeEvent());
        if (flag.get(ContextMenuEventType.FullReset))
            this.updateItemsState();
    };
    ContextMenuBarBase.prototype.getSelectionChangeEvent = function () {
        return this.selection && !this.selection.shouldResetInputPosition() ?
            ContextMenuEventType.None :
            ContextMenuEventType.FullReset;
    };
    ContextMenuBarBase.prototype.getModelChangeEvent = function (change) {
        switch (change.type) {
            case enums_1.ModelChangeType.DifferentOddAndEvenPages:
            case enums_1.ModelChangeType.PageColor:
            case enums_1.ModelChangeType.DefaultTabWidth:
            case enums_1.ModelChangeType.HeaderFooterCreated:
            case enums_1.ModelChangeType.HeaderFooterIndexChanged:
            case enums_1.ModelChangeType.LoadFontInfo:
            case enums_1.ModelChangeType.SectionFormattingChanged:
            case enums_1.ModelChangeType.CreateStyleLink:
            case enums_1.ModelChangeType.DeleteStyleLink:
            case enums_1.ModelChangeType.AbstractNumberingListAdded:
            case enums_1.ModelChangeType.AbstractNumberingListDeleted:
            case enums_1.ModelChangeType.NumberingListAdded:
            case enums_1.ModelChangeType.NumberingListDeleted:
            case enums_1.ModelChangeType.ListLevelPropertyChanged:
            case enums_1.ModelChangeType.ListLevelParagraphPropertyChanged:
            case enums_1.ModelChangeType.ListLevelCharacterPropertyChanged:
            case enums_1.ModelChangeType.IOverrideListLevelChanged:
            case enums_1.ModelChangeType.BookmarkCreated:
            case enums_1.ModelChangeType.BookmarkDeleted:
            case enums_1.ModelChangeType.TabInserted:
            case enums_1.ModelChangeType.TabDeleted:
            case enums_1.ModelChangeType.LoadPicturesInfo:
            case enums_1.ModelChangeType.InlinePicturesUpdated:
            case enums_1.ModelChangeType.InlinePictureInserted:
            case enums_1.ModelChangeType.AnchoredPictureInserted:
            case enums_1.ModelChangeType.AnchoredPictureSizeChanged:
            case enums_1.ModelChangeType.InlineObjectRunPropertyChanged:
            case enums_1.ModelChangeType.FieldInserted:
            case enums_1.ModelChangeType.FieldDeleted:
            case enums_1.ModelChangeType.HyperlinkInfoChanged:
            case enums_1.ModelChangeType.FieldsShowCodeChanged:
            case enums_1.ModelChangeType.SimpleRunInserted:
            case enums_1.ModelChangeType.TextBufferChanged:
            case enums_1.ModelChangeType.ParagraphInserted:
            case enums_1.ModelChangeType.SectionInserted:
            case enums_1.ModelChangeType.AnchoredTextBoxInserted:
            case enums_1.ModelChangeType.AnchorObjectRemoved:
            case enums_1.ModelChangeType.IntervalRemoved:
            case enums_1.ModelChangeType.ParagraphMerged:
            case enums_1.ModelChangeType.SectionMerged:
            case enums_1.ModelChangeType.CharacterFormattingChanged:
            case enums_1.ModelChangeType.CharacterPropertiesChanged:
            case enums_1.ModelChangeType.ParagraphFormattingChanged:
            case enums_1.ModelChangeType.ParagraphPropertiesChanged:
            case enums_1.ModelChangeType.ParagraphAndCharacterMergedPropertiesReset:
            case enums_1.ModelChangeType.AnchoredTextBoxSizeChanged:
            case enums_1.ModelChangeType.AnchoredTextBoxPropertiesChanged:
            case enums_1.ModelChangeType.AnchorInfoPropertyChanged:
            case enums_1.ModelChangeType.ShapeChanged:
            case enums_1.ModelChangeType.ShapePropertyChanged:
            case enums_1.ModelChangeType.CharacterStyleApplied:
            case enums_1.ModelChangeType.ParagraphStyleApplied:
            case enums_1.ModelChangeType.TableStyleChanged:
            case enums_1.ModelChangeType.ParagraphNumberingListChanged:
            case enums_1.ModelChangeType.TableCreated:
            case enums_1.ModelChangeType.TableRemoved:
            case enums_1.ModelChangeType.TableStartPositionShifted:
            case enums_1.ModelChangeType.TableCellPropertyChanged:
            case enums_1.ModelChangeType.TablePropertyChanged:
            case enums_1.ModelChangeType.TableRowPropertyChanged:
            case enums_1.ModelChangeType.TableCellSplittedHorizontally:
            case enums_1.ModelChangeType.TableCellMergedHorizontally:
            case enums_1.ModelChangeType.TableRowInserted:
            case enums_1.ModelChangeType.TableRowRemoved:
            case enums_1.ModelChangeType.TableCellRemoved:
            case enums_1.ModelChangeType.TableCellInserted:
            default:
                return ContextMenuEventType.None;
        }
    };
    return ContextMenuBarBase;
}(base_1.BarBase));
exports.ContextMenuBarBase = ContextMenuBarBase;
