// devexpress-richedit/lib/base/mouse-handler/mouse-handler/mouse-handler-state-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var manipulator_handler_state_base_1 = require("../base/manipulator-handler-state-base");
var MouseHandlerStateBase = (function (_super) {
    tslib_1.__extends(MouseHandlerStateBase, _super);
    function MouseHandlerStateBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MouseHandlerStateBase.prototype.onMouseDoubleClick = function (_evt) { };
    MouseHandlerStateBase.prototype.onMouseDown = function (_evt) { };
    MouseHandlerStateBase.prototype.onMouseUp = function (_evt) { };
    MouseHandlerStateBase.prototype.onMouseMove = function (_evt) { };
    MouseHandlerStateBase.prototype.onMouseWheel = function (_evt) { };
    MouseHandlerStateBase.prototype.onShortcut = function (_shortcutCode) { };
    return MouseHandlerStateBase;
}(manipulator_handler_state_base_1.ManipulatorHandlerStateBase));
exports.MouseHandlerStateBase = MouseHandlerStateBase;
